// @byterover/cipher/dist/src/core/chroma-ZIU3QWAU.js
import {
  VectorDimensionError,
  VectorStoreConnectionError,
  VectorStoreError
} from "./chunk-JOPES53R.js";
import {
  DEFAULTS,
  ERROR_MESSAGES,
  LOG_PREFIXES
} from "./chunk-4JSJAC4K.js";
import "./chunk-7GLOCKG4.js";
import {
  createLogger
} from "./chunk-GPOSIAVD.js";
import "./chunk-7UQA6M3D.js";
import "./chunk-NUVPDHVT.js";
import {
  init_esm_shims
} from "./chunk-KFOQGTVJ.js";

// src/core/vector_storage/backend/chroma.ts
init_esm_shims();
import { ChromaClient } from "chromadb";

// src/core/vector_storage/backend/chroma-payload-adapter.ts
init_esm_shims();
var DEFAULT_CONFIG = {
  defaultStrategy: "json-string",
  autoFlattenNested: true,
  maxNestingDepth: 3,
  fieldConfigs: {
    // Legacy field configurations for backward compatibility
    tags: { strategy: "comma-separated" },
    currentProgress: { strategy: "json-string" },
    bugsEncountered: { strategy: "json-string" },
    workContext: { strategy: "json-string" },
    // Common field patterns
    id: { strategy: "preserve" },
    created_at: { strategy: "preserve" },
    updated_at: { strategy: "preserve" },
    timestamp: { strategy: "preserve" },
    score: { strategy: "preserve" },
    version: { strategy: "preserve" }
  }
};
var DefaultChromaPayloadAdapter = class {
  config;
  logger;
  constructor(config) {
    this.config = { ...DEFAULT_CONFIG, ...config };
    this.logger = createLogger({
      level: process.env.CIPHER_LOG_LEVEL || "info"
    });
    if (config?.fieldConfigs) {
      this.config.fieldConfigs = { ...DEFAULT_CONFIG.fieldConfigs, ...config.fieldConfigs };
    }
  }
  /**
   * Serialize complex payload to ChromaDB-compatible metadata
   */
  serialize(payload) {
    if (!payload || typeof payload !== "object") {
      return {};
    }
    const result = {};
    for (const [key, value] of Object.entries(payload)) {
      if (value === null || value === void 0) {
        continue;
      }
      const fieldConfig = this.config.fieldConfigs[key];
      const strategy = fieldConfig?.strategy || this.getStrategyForValue(value);
      try {
        const transformedEntries = this.serializeField(key, value, strategy, fieldConfig);
        Object.assign(result, transformedEntries);
      } catch (error) {
        this.logger.warn(`${LOG_PREFIXES.CHROMA} Failed to serialize field '${key}'`, {
          key,
          strategy,
          error: error instanceof Error ? error.message : String(error)
        });
        if (strategy !== "json-string") {
          try {
            result[key] = JSON.stringify(value);
          } catch (jsonError) {
            this.logger.error(`${LOG_PREFIXES.CHROMA} Failed JSON fallback for field '${key}'`, {
              key,
              jsonError: jsonError instanceof Error ? jsonError.message : String(jsonError)
            });
          }
        }
      }
    }
    return result;
  }
  /**
   * Deserialize ChromaDB metadata back to original payload format
   */
  deserialize(metadata) {
    if (!metadata || typeof metadata !== "object") {
      return {};
    }
    const result = {};
    const processedKeys = /* @__PURE__ */ new Set();
    for (const [key, value] of Object.entries(metadata)) {
      if (processedKeys.has(key)) {
        continue;
      }
      const fieldConfig = this.config.fieldConfigs[key];
      if (fieldConfig) {
        try {
          result[key] = this.deserializeField(key, value, fieldConfig.strategy, fieldConfig);
          processedKeys.add(key);
        } catch (error) {
          this.logger.warn(`${LOG_PREFIXES.CHROMA} Failed to deserialize field '${key}'`, {
            key,
            strategy: fieldConfig.strategy,
            error: error instanceof Error ? error.message : String(error)
          });
          result[key] = value;
          processedKeys.add(key);
        }
      }
    }
    if (this.config.autoFlattenNested) {
      const nestedGroups = this.groupFlattenedKeys(metadata, processedKeys);
      for (const [rootKey, nestedData] of Object.entries(nestedGroups)) {
        result[rootKey] = this.reconstructNestedObject(nestedData);
        for (const flatKey in metadata) {
          if (flatKey.startsWith(rootKey + "_")) {
            processedKeys.add(flatKey);
          }
        }
      }
    }
    for (const [key, value] of Object.entries(metadata)) {
      if (!processedKeys.has(key)) {
        const strategy = this.inferDeserializationStrategy(value);
        result[key] = this.deserializeField(key, value, strategy);
      }
    }
    return result;
  }
  /**
   * Get transformation configuration
   */
  getConfig() {
    return { ...this.config };
  }
  /**
   * Update transformation configuration
   */
  updateConfig(config) {
    const mergedFieldConfigs = config.fieldConfigs ? { ...this.config.fieldConfigs, ...config.fieldConfigs } : this.config.fieldConfigs;
    this.config = {
      ...this.config,
      ...config,
      fieldConfigs: mergedFieldConfigs
    };
  }
  /**
   * Determine the best strategy for a given value
   */
  getStrategyForValue(value) {
    if (value === null || value === void 0) {
      return "preserve";
    }
    if (typeof value === "string" || typeof value === "number" || typeof value === "boolean") {
      return "preserve";
    }
    if (Array.isArray(value)) {
      if (value.every((item) => typeof item === "string" || typeof item === "number")) {
        return "comma-separated";
      }
      return "json-string";
    }
    if (typeof value === "object") {
      if (Object.keys(value).length === 0) {
        return "json-string";
      }
      if (this.isSimpleObject(value)) {
        return "dot-notation";
      }
      return "json-string";
    }
    return this.config.defaultStrategy;
  }
  /**
   * Infer deserialization strategy based on stored value characteristics
   */
  inferDeserializationStrategy(value) {
    if (typeof value === "string") {
      if (value.includes(",") && !value.startsWith("{") && !value.startsWith("[")) {
        return "comma-separated";
      }
      if (value.startsWith("{") && value.endsWith("}") || value.startsWith("[") && value.endsWith("]")) {
        return "json-string";
      }
    }
    return "preserve";
  }
  /**
   * Check if an object is simple enough for dot notation flattening
   */
  isSimpleObject(obj, depth = 0) {
    if (depth >= this.config.maxNestingDepth) {
      return false;
    }
    if (!obj || typeof obj !== "object" || Array.isArray(obj)) {
      return false;
    }
    const keys = Object.keys(obj);
    if (keys.length > 10) {
      return false;
    }
    return Object.values(obj).every((value) => {
      if (value === null || value === void 0) return true;
      if (typeof value === "string" || typeof value === "number" || typeof value === "boolean") {
        return true;
      }
      if (Array.isArray(value)) {
        return value.length <= 5 && value.every((item) => typeof item === "string" || typeof item === "number");
      }
      if (typeof value === "object") {
        return this.isSimpleObject(value, depth + 1);
      }
      return false;
    });
  }
  /**
   * Serialize a single field based on strategy
   */
  serializeField(key, value, strategy, config) {
    const result = {};
    if (config?.customTransformer) {
      const transformed = config.customTransformer.serialize(value);
      if (typeof transformed === "object" && !Array.isArray(transformed)) {
        return transformed;
      }
      result[key] = transformed;
      return result;
    }
    switch (strategy) {
      case "preserve":
        if (typeof value === "string" || typeof value === "number" || typeof value === "boolean") {
          result[key] = value;
        } else {
          result[key] = String(value);
        }
        break;
      case "json-string":
        result[key] = JSON.stringify(value);
        break;
      case "comma-separated":
        if (Array.isArray(value)) {
          result[key] = value.filter((item) => item != null).join(",");
        } else {
          result[key] = String(value);
        }
        break;
      case "boolean-flags":
        if (Array.isArray(value)) {
          const prefix = config?.prefix || key;
          for (const item of value) {
            if (item != null) {
              result[`${prefix}_${String(item)}`] = true;
            }
          }
        } else {
          result[key] = Boolean(value);
        }
        break;
      case "dot-notation":
        if (typeof value === "object" && value !== null && !Array.isArray(value)) {
          const flattened = this.flattenObject(value, key);
          Object.assign(result, flattened);
        } else {
          result[key] = String(value);
        }
        break;
      default:
        result[key] = JSON.stringify(value);
    }
    return result;
  }
  /**
   * Deserialize a single field based on strategy
   */
  deserializeField(key, value, strategy, config) {
    if (config?.customTransformer) {
      return config.customTransformer.deserialize(value);
    }
    switch (strategy) {
      case "preserve":
        return value;
      case "json-string":
        if (typeof value === "string") {
          try {
            return JSON.parse(value);
          } catch {
            return value;
          }
        }
        return value;
      case "comma-separated":
        if (typeof value === "string") {
          if (value === "") {
            return [];
          }
          return value.split(",").filter((item) => item.trim().length > 0).map((item) => {
            const trimmed = item.trim();
            if (/^-?\d+(\.\d+)?$/.test(trimmed)) {
              const num = Number(trimmed);
              return isNaN(num) ? trimmed : num;
            }
            return trimmed;
          });
        }
        return Array.isArray(value) ? value : [value];
      case "boolean-flags":
        return value;
      case "dot-notation":
        return value;
      default:
        if (typeof value === "string" && (value.startsWith("{") || value.startsWith("["))) {
          try {
            return JSON.parse(value);
          } catch {
            return value;
          }
        }
        return value;
    }
  }
  /**
   * Flatten nested object using dot notation
   */
  flattenObject(obj, prefix = "", depth = 0) {
    const result = {};
    if (depth >= this.config.maxNestingDepth) {
      result[prefix] = JSON.stringify(obj);
      return result;
    }
    for (const [key, value] of Object.entries(obj)) {
      if (value === null || value === void 0) {
        continue;
      }
      const newKey = prefix ? `${prefix}_${key}` : key;
      if (typeof value === "string" || typeof value === "number" || typeof value === "boolean") {
        result[newKey] = value;
      } else if (Array.isArray(value)) {
        if (value.every((item) => typeof item === "string" || typeof item === "number")) {
          result[newKey] = value.join(",");
        } else {
          result[newKey] = JSON.stringify(value);
        }
      } else if (typeof value === "object") {
        Object.assign(result, this.flattenObject(value, newKey, depth + 1));
      } else {
        result[newKey] = String(value);
      }
    }
    return result;
  }
  /**
   * Group flattened keys by their root key
   */
  groupFlattenedKeys(metadata, processedKeys) {
    const groups = {};
    for (const key in metadata) {
      if (processedKeys.has(key)) {
        continue;
      }
      const underscoreIndex = key.indexOf("_");
      if (underscoreIndex > 0) {
        const rootKey = key.substring(0, underscoreIndex);
        const remainingKey = key.substring(underscoreIndex + 1);
        if (!groups[rootKey]) {
          groups[rootKey] = {};
        }
        groups[rootKey][remainingKey] = metadata[key];
      }
    }
    return groups;
  }
  /**
   * Reconstruct nested object from flattened keys
   */
  reconstructNestedObject(flatData) {
    const result = {};
    for (const [key, value] of Object.entries(flatData)) {
      const parts = key.split("_");
      let current = result;
      for (let i = 0; i < parts.length - 1; i++) {
        const part = parts[i];
        if (!part) continue;
        if (!(part in current)) {
          current[part] = {};
        }
        current = current[part];
      }
      const finalKey = parts[parts.length - 1];
      if (!finalKey) continue;
      if (typeof value === "string" && value.includes(",")) {
        const arrayValue = value.split(",").filter((item) => item.trim().length > 0).map((item) => {
          const trimmed = item.trim();
          if (/^-?\d+(\.\d+)?$/.test(trimmed)) {
            const num = Number(trimmed);
            return isNaN(num) ? trimmed : num;
          }
          return trimmed;
        });
        current[finalKey] = arrayValue.length > 1 ? arrayValue : value;
      } else {
        current[finalKey] = value;
      }
    }
    return result;
  }
};

// src/core/vector_storage/backend/chroma.ts
var ChromaBackend = class {
  client;
  collection = null;
  config;
  collectionName;
  dimension;
  logger;
  payloadAdapter;
  connected = false;
  constructor(config, payloadAdapter) {
    this.config = config;
    this.collectionName = config.collectionName;
    this.dimension = config.dimension;
    this.logger = createLogger({
      level: process.env.CIPHER_LOG_LEVEL || "info"
    });
    this.payloadAdapter = payloadAdapter || new DefaultChromaPayloadAdapter();
    const clientConfig = {};
    if (config.url) {
      const url = new URL(config.url);
      clientConfig.host = url.hostname;
      clientConfig.port = parseInt(url.port) || (url.protocol === "https:" ? 443 : 80);
      clientConfig.ssl = url.protocol === "https:";
    } else if (config.host) {
      clientConfig.host = config.host;
      clientConfig.port = config.port || 8e3;
      clientConfig.ssl = config.ssl || false;
    }
    if (config.headers) {
      clientConfig.fetchOptions = {
        headers: config.headers
      };
    }
    this.client = new ChromaClient(clientConfig);
    this.logger.debug(`${LOG_PREFIXES.CHROMA} Initialized`, {
      collection: this.collectionName,
      dimension: this.dimension,
      url: config.url || `${clientConfig.ssl ? "https" : "http"}://${clientConfig.host}:${clientConfig.port}` || "embedded"
    });
  }
  /**
   * Convert search filters to ChromaDB filter format
   */
  createFilter(filters) {
    if (!filters) return void 0;
    const chromaFilter = {};
    for (const [key, value] of Object.entries(filters)) {
      if (value === null || value === void 0) continue;
      if (typeof value === "object" && !Array.isArray(value) && ("gte" in value || "gt" in value || "lte" in value || "lt" in value)) {
        const rangeConditions = {};
        if ("gte" in value && value.gte !== void 0) rangeConditions.$gte = value.gte;
        if ("gt" in value && value.gt !== void 0) rangeConditions.$gt = value.gt;
        if ("lte" in value && value.lte !== void 0) rangeConditions.$lte = value.lte;
        if ("lt" in value && value.lt !== void 0) rangeConditions.$lt = value.lt;
        chromaFilter[key] = rangeConditions;
      } else if (typeof value === "object" && !Array.isArray(value) && ("any" in value || "all" in value)) {
        if ("any" in value && Array.isArray(value.any)) {
          chromaFilter[key] = { $in: value.any };
        } else if ("all" in value && Array.isArray(value.all) && value.all.length > 0) {
          chromaFilter[key] = value.all[0];
        }
      } else {
        chromaFilter[key] = value;
      }
    }
    return Object.keys(chromaFilter).length > 0 ? chromaFilter : void 0;
  }
  /**
   * Validate vector dimension
   */
  validateDimension(vector, operation) {
    if (vector.length !== this.dimension) {
      throw new VectorDimensionError(
        `${ERROR_MESSAGES.INVALID_DIMENSION}: expected ${this.dimension}, got ${vector.length}`,
        this.dimension,
        vector.length
      );
    }
  }
  /**
   * Validate ID (ChromaDB supports string and number IDs, we'll convert numbers to strings)
   */
  validateId(id) {
    if (!Number.isInteger(id) || isNaN(id)) {
      throw new VectorStoreError("ChromaDB point IDs must be valid integers", "id");
    }
  }
  /**
   * Get the payload adapter instance
   */
  getPayloadAdapter() {
    return this.payloadAdapter;
  }
  // VectorStore implementation
  async insert(vectors, ids, payloads) {
    if (!this.connected) {
      throw new VectorStoreError(ERROR_MESSAGES.NOT_CONNECTED, "insert");
    }
    if (!this.collection) {
      this.logger.error(`${LOG_PREFIXES.CHROMA} Collection is null during insert`);
      throw new VectorStoreError("ChromaDB collection not initialized", "insert");
    }
    if (vectors.length !== ids.length || vectors.length !== payloads.length) {
      throw new VectorStoreError("Vectors, IDs, and payloads must have the same length", "insert");
    }
    for (let i = 0; i < vectors.length; i++) {
      const vector = vectors[i];
      const id = ids[i];
      if (!vector) {
        throw new VectorStoreError(`Vector missing at index ${i}`, "insert");
      }
      if (id === void 0 || id === null) {
        throw new VectorStoreError(`ID missing at index ${i}`, "insert");
      }
      this.validateDimension(vector, "insert");
      this.validateId(id);
    }
    this.logger.debug(`${LOG_PREFIXES.CHROMA} Inserting ${vectors.length} vectors`);
    const stringIds = ids.map((id) => id.toString());
    const chromaCompatiblePayloads = payloads.map(
      (payload) => this.payloadAdapter.serialize(payload)
    );
    const addParams = {
      ids: stringIds,
      embeddings: vectors,
      metadatas: chromaCompatiblePayloads
    };
    try {
      await this.collection.upsert(addParams);
      this.logger.debug(`${LOG_PREFIXES.CHROMA} Successfully inserted ${vectors.length} vectors`);
    } catch (error) {
      this.logger.error(`${LOG_PREFIXES.CHROMA} Insert failed`, {
        error: error instanceof Error ? error.message : String(error),
        vectorCount: vectors.length,
        collectionName: this.collectionName
      });
      throw new VectorStoreError("Failed to insert vectors", "insert", error);
    }
  }
  async search(query, limit = DEFAULTS.SEARCH_LIMIT, filters) {
    if (!this.connected) {
      throw new VectorStoreError(ERROR_MESSAGES.NOT_CONNECTED, "search");
    }
    if (!this.collection) {
      this.logger.error(`${LOG_PREFIXES.CHROMA} Collection is null during search`);
      throw new VectorStoreError("ChromaDB collection not initialized", "search");
    }
    this.validateDimension(query, "search");
    this.logger.debug(`${LOG_PREFIXES.CHROMA} Searching with limit ${limit}`, {
      hasFilters: !!filters
    });
    try {
      const chromaFilter = this.createFilter(filters);
      const searchParams = {
        queryEmbeddings: [query],
        nResults: limit
      };
      if (chromaFilter) {
        searchParams.where = chromaFilter;
      }
      const searchResponse = await this.collection.query(searchParams);
      const results = [];
      if (searchResponse.ids && searchResponse.ids[0]) {
        for (let i = 0; i < searchResponse.ids[0].length; i++) {
          const id = searchResponse.ids[0][i];
          const distance = searchResponse.distances?.[0]?.[i] ?? 0;
          const metadata = searchResponse.metadatas?.[0]?.[i] ?? {};
          const score = 1 - distance;
          results.push({
            id: parseInt(id, 10),
            // Convert back to number
            score,
            payload: this.payloadAdapter.deserialize(metadata)
          });
        }
      }
      this.logger.debug(`${LOG_PREFIXES.CHROMA} Found ${results.length} results`);
      return results;
    } catch (error) {
      this.logger.error(`${LOG_PREFIXES.CHROMA} Search failed`, {
        error: error instanceof Error ? error.message : String(error),
        queryLength: query.length,
        limit,
        collectionName: this.collectionName
      });
      throw new VectorStoreError(ERROR_MESSAGES.SEARCH_FAILED, "search", error);
    }
  }
  async get(vectorId) {
    if (!this.connected) {
      throw new VectorStoreError(ERROR_MESSAGES.NOT_CONNECTED, "get");
    }
    this.validateId(vectorId);
    this.logger.debug(`${LOG_PREFIXES.CHROMA} Getting vector ${vectorId}`);
    try {
      const response = await this.collection.get({
        ids: [vectorId.toString()],
        include: ["embeddings", "metadatas"]
      });
      if (!response.ids || response.ids.length === 0) {
        return null;
      }
      const id = response.ids[0];
      const embedding = response.embeddings?.[0];
      const metadata = response.metadatas?.[0] ?? {};
      return {
        id: parseInt(id, 10),
        vector: embedding || [],
        payload: this.payloadAdapter.deserialize(metadata),
        score: 1
      };
    } catch (error) {
      this.logger.error(`${LOG_PREFIXES.CHROMA} Get failed`, { error });
      throw new VectorStoreError("Failed to retrieve vector", "get", error);
    }
  }
  async update(vectorId, vector, payload) {
    if (!this.connected) {
      throw new VectorStoreError(ERROR_MESSAGES.NOT_CONNECTED, "update");
    }
    this.validateDimension(vector, "update");
    this.validateId(vectorId);
    this.logger.debug(`${LOG_PREFIXES.CHROMA} Updating vector ${vectorId}`);
    try {
      const convertedPayload = this.payloadAdapter.serialize(payload);
      await this.collection.upsert({
        ids: [vectorId.toString()],
        embeddings: [vector],
        metadatas: [convertedPayload]
      });
      this.logger.debug(`${LOG_PREFIXES.CHROMA} Successfully updated vector ${vectorId}`);
    } catch (error) {
      this.logger.error(`${LOG_PREFIXES.CHROMA} Update failed`, { error });
      throw new VectorStoreError("Failed to update vector", "update", error);
    }
  }
  async delete(vectorId) {
    if (!this.connected) {
      throw new VectorStoreError(ERROR_MESSAGES.NOT_CONNECTED, "delete");
    }
    this.validateId(vectorId);
    this.logger.debug(`${LOG_PREFIXES.CHROMA} Deleting vector ${vectorId}`);
    try {
      await this.collection.delete({
        ids: [vectorId.toString()]
      });
      this.logger.debug(`${LOG_PREFIXES.CHROMA} Successfully deleted vector ${vectorId}`);
    } catch (error) {
      this.logger.error(`${LOG_PREFIXES.CHROMA} Delete failed`, { error });
      throw new VectorStoreError("Failed to delete vector", "delete", error);
    }
  }
  async deleteCollection() {
    if (!this.connected) {
      throw new VectorStoreError(ERROR_MESSAGES.NOT_CONNECTED, "deleteCollection");
    }
    this.logger.warn(`${LOG_PREFIXES.CHROMA} Deleting collection ${this.collectionName}`);
    try {
      await this.client.deleteCollection({ name: this.collectionName });
      this.logger.info(
        `${LOG_PREFIXES.CHROMA} Successfully deleted collection ${this.collectionName}`
      );
    } catch (error) {
      this.logger.error(`${LOG_PREFIXES.CHROMA} Delete collection failed`, { error });
      throw new VectorStoreError("Failed to delete collection", "deleteCollection", error);
    }
  }
  async list(filters, limit = 1e4) {
    if (!this.connected) {
      throw new VectorStoreError(ERROR_MESSAGES.NOT_CONNECTED, "list");
    }
    this.logger.debug(`${LOG_PREFIXES.CHROMA} Listing vectors`, {
      hasFilters: !!filters,
      limit
    });
    try {
      const chromaFilter = this.createFilter(filters);
      const getParams = {
        include: ["embeddings", "metadatas"],
        limit
      };
      if (chromaFilter) {
        getParams.where = chromaFilter;
      }
      const response = await this.collection.get(getParams);
      const results = [];
      if (response.ids) {
        for (let i = 0; i < response.ids.length; i++) {
          const id = response.ids[i];
          const embedding = response.embeddings?.[i];
          const metadata = response.metadatas?.[i] ?? {};
          results.push({
            id: parseInt(id, 10),
            score: 1,
            // Default score for exact match
            payload: this.payloadAdapter.deserialize(metadata),
            vector: embedding
          });
        }
      }
      this.logger.info(`${LOG_PREFIXES.CHROMA} Listed ${results.length} vectors`);
      return [results, results.length];
    } catch (error) {
      this.logger.error(`${LOG_PREFIXES.CHROMA} List failed`, { error });
      throw new VectorStoreError("Failed to list vectors", "list", error);
    }
  }
  async connect() {
    if (this.connected) {
      this.logger.debug(`${LOG_PREFIXES.CHROMA} Already connected`);
      return;
    }
    this.logger.info(`${LOG_PREFIXES.CHROMA} Connecting to ChromaDB`);
    try {
      try {
        this.collection = await this.client.getCollection({
          name: this.collectionName,
          embeddingFunction: null
          // Use null to indicate no embedding function
        });
        this.logger.debug(
          `${LOG_PREFIXES.CHROMA} Collection already exists: ${this.collectionName}`
        );
      } catch (error) {
        this.logger.debug(`${LOG_PREFIXES.CHROMA} Creating collection ${this.collectionName}`);
        const distance = this.config.distance || "cosine";
        let metricName = "cosine";
        if (distance === "euclidean" || distance === "l2") {
          metricName = "l2";
        } else if (distance === "ip" || distance === "dot") {
          metricName = "ip";
        }
        this.collection = await this.client.createCollection({
          name: this.collectionName,
          metadata: {
            "hnsw:space": metricName
          },
          embeddingFunction: null
          // Use null to indicate no embedding function
        });
      }
      this.connected = true;
      this.logger.info(`${LOG_PREFIXES.CHROMA} Successfully connected`);
    } catch (error) {
      this.logger.error(`${LOG_PREFIXES.CHROMA} Connection failed`, { error });
      if (error instanceof VectorStoreConnectionError) {
        throw error;
      }
      throw new VectorStoreConnectionError(
        ERROR_MESSAGES.CONNECTION_FAILED,
        "chroma",
        error
      );
    }
  }
  async disconnect() {
    if (!this.connected) {
      this.logger.debug(`${LOG_PREFIXES.CHROMA} Already disconnected`);
      return;
    }
    this.logger.info(`${LOG_PREFIXES.CHROMA} Disconnecting from ChromaDB`);
    this.collection = null;
    this.connected = false;
    this.logger.info(`${LOG_PREFIXES.CHROMA} Successfully disconnected`);
  }
  isConnected() {
    return this.connected;
  }
  getBackendType() {
    return "chroma";
  }
  getDimension() {
    return this.dimension;
  }
  getCollectionName() {
    return this.collectionName;
  }
};
export {
  ChromaBackend
};
