// @byterover/cipher/dist/src/core/chunk-DU36577C.js
import {
  logger
} from "./chunk-GPOSIAVD.js";
import {
  init_esm_shims
} from "./chunk-KFOQGTVJ.js";

// src/core/brain/tools/definitions/system/index.ts
init_esm_shims();

// src/core/brain/tools/definitions/system/bash.ts
init_esm_shims();
import { spawn } from "child_process";
import { EventEmitter } from "events";

// src/core/brain/tools/types.ts
init_esm_shims();
var INTERNAL_TOOL_PREFIX = "cipher_";
function createInternalToolName(baseName) {
  return baseName.startsWith(INTERNAL_TOOL_PREFIX) ? baseName : `${INTERNAL_TOOL_PREFIX}${baseName}`;
}
function isInternalToolName(toolName) {
  return toolName.startsWith(INTERNAL_TOOL_PREFIX);
}

// src/core/brain/tools/definitions/system/bash.ts
var BashSession = class extends EventEmitter {
  process = null;
  outputBuffer = "";
  errorBuffer = "";
  isRunning = false;
  currentWorkingDir;
  sessionId;
  constructor(sessionId, workingDir = process.cwd()) {
    super();
    this.sessionId = sessionId;
    this.currentWorkingDir = workingDir;
  }
  /**
   * Initialize the bash session
   */
  async initialize() {
    if (this.process) {
      logger.warn(`BashSession ${this.sessionId}: Already initialized`);
      return;
    }
    try {
      this.process = spawn("/bin/bash", [], {
        cwd: this.currentWorkingDir,
        stdio: ["pipe", "pipe", "pipe"],
        env: { ...process.env }
      });
      this.isRunning = true;
      this.process.stdout?.on("data", (data) => {
        this.outputBuffer += data.toString();
      });
      this.process.stderr?.on("data", (data) => {
        this.errorBuffer += data.toString();
      });
      this.process.on("exit", (code, signal) => {
        this.isRunning = false;
        logger.debug(
          `BashSession ${this.sessionId}: Process exited with code ${code}, signal ${signal}`
        );
        this.emit("exit", { code, signal });
      });
      this.process.on("error", (error) => {
        this.isRunning = false;
        logger.error(`BashSession ${this.sessionId}: Process error`, error);
        this.emit("error", error);
      });
      await new Promise((resolve) => setTimeout(resolve, 100));
      logger.debug(`BashSession ${this.sessionId}: Initialized successfully`);
    } catch (error) {
      logger.error(`BashSession ${this.sessionId}: Failed to initialize`, error);
      throw error;
    }
  }
  /**
   * Execute a command in the session
   */
  async executeCommand(command, timeout = 3e4) {
    if (!this.process || !this.isRunning) {
      throw new Error("Bash session is not running");
    }
    const startTime = Date.now();
    this.outputBuffer = "";
    this.errorBuffer = "";
    return new Promise((resolve, reject) => {
      const timeoutId = setTimeout(() => {
        reject(new Error(`Command timeout after ${timeout}ms`));
      }, timeout);
      const marker = `CMD_COMPLETE_${Date.now()}`;
      const fullCommand = `${command}; echo "${marker}"; echo "EXIT_CODE:$?" `;
      this.process.stdin?.write(`${fullCommand}
`);
      const checkCompletion = () => {
        if (this.outputBuffer.includes(marker)) {
          clearTimeout(timeoutId);
          const parts = this.outputBuffer.split(marker);
          const output = parts[0]?.trim() || "";
          const exitCodeMatch = this.outputBuffer.match(/EXIT_CODE:(\d+)/);
          const exitCode = exitCodeMatch?.[1] ? parseInt(exitCodeMatch[1], 10) : 0;
          const result = {
            output,
            error: this.errorBuffer.trim(),
            exitCode,
            command,
            duration: Date.now() - startTime,
            pid: this.process?.pid,
            workingDir: this.currentWorkingDir || void 0
          };
          resolve(result);
        } else {
          setTimeout(checkCompletion, 100);
        }
      };
      setTimeout(checkCompletion, 50);
    });
  }
  /**
   * Close the bash session
   */
  async close() {
    if (this.process) {
      this.process.stdin?.write("exit\n");
      this.process.kill();
      this.process = null;
      this.isRunning = false;
      logger.debug(`BashSession ${this.sessionId}: Closed`);
    }
  }
  /**
   * Check if session is running
   */
  isActive() {
    return this.isRunning && this.process !== null;
  }
};
var BashSessionManager = class _BashSessionManager {
  sessions = /* @__PURE__ */ new Map();
  static instance;
  static getInstance() {
    if (!_BashSessionManager.instance) {
      _BashSessionManager.instance = new _BashSessionManager();
    }
    return _BashSessionManager.instance;
  }
  /**
   * Get or create a bash session
   */
  async getSession(sessionId, workingDir) {
    let session = this.sessions.get(sessionId);
    if (!session || !session.isActive()) {
      session = new BashSession(sessionId, workingDir);
      await session.initialize();
      this.sessions.set(sessionId, session);
    }
    return session;
  }
  /**
   * Close a specific session
   */
  async closeSession(sessionId) {
    const session = this.sessions.get(sessionId);
    if (session) {
      await session.close();
      this.sessions.delete(sessionId);
    }
  }
  /**
   * Close all sessions
   */
  async closeAllSessions() {
    const closePromises = Array.from(this.sessions.values()).map((session) => session.close());
    await Promise.all(closePromises);
    this.sessions.clear();
  }
};
async function executeBashCommand(options) {
  const { command, timeout = 3e4, workingDir, environment, shell = "/bin/bash" } = options;
  logger.debug("Executing bash command", { command, timeout, workingDir });
  const startTime = Date.now();
  return new Promise((resolve, reject) => {
    const timeoutId = setTimeout(() => {
      reject(new Error(`Command timeout after ${timeout}ms`));
    }, timeout);
    let output = "";
    let error = "";
    const childProcess = spawn(shell, ["-c", command], {
      cwd: workingDir || process.cwd(),
      env: { ...process.env, ...environment },
      stdio: ["pipe", "pipe", "pipe"]
    });
    childProcess.stdout?.on("data", (data) => {
      output += data.toString();
    });
    childProcess.stderr?.on("data", (data) => {
      error += data.toString();
    });
    childProcess.on("close", (code) => {
      clearTimeout(timeoutId);
      const result = {
        output: output.trim(),
        error: error.trim(),
        exitCode: code || 0,
        command,
        duration: Date.now() - startTime,
        pid: childProcess.pid,
        workingDir: workingDir || process.cwd() || void 0
      };
      logger.debug("Bash command completed", {
        command,
        exitCode: result.exitCode,
        duration: result.duration
      });
      resolve(result);
    });
    childProcess.on("error", (err) => {
      clearTimeout(timeoutId);
      logger.error("Bash command failed", { command, error: err.message });
      reject(err);
    });
  });
}
var bashHandler = async (args, context) => {
  const {
    command,
    timeout = 3e4,
    workingDir,
    persistent = false,
    sessionId: customSessionId
  } = args;
  if (!command || typeof command !== "string") {
    return {
      content: "Error: Command is required and must be a string",
      isError: true
    };
  }
  try {
    let result;
    if (persistent) {
      const sessionManager = BashSessionManager.getInstance();
      const sessionId = customSessionId || context?.sessionId || "default";
      const session = await sessionManager.getSession(sessionId, workingDir);
      result = await session.executeCommand(command, timeout);
    } else {
      result = await executeBashCommand({
        command,
        timeout,
        workingDir
      });
    }
    const formattedOutput = [
      `Command: ${result.command}`,
      `Exit Code: ${result.exitCode}`,
      `Duration: ${result.duration}ms`,
      result.workingDir ? `Working Dir: ${result.workingDir}` : "",
      "",
      "Output:",
      result.output || "(no output)",
      result.error ? "\nError:" : "",
      result.error || ""
    ].filter(Boolean).join("\n");
    if (context?.services?.vectorStoreManager && result.output) {
      try {
        logger.debug("Command result could be stored in memory", { command });
      } catch (memoryError) {
        logger.warn("Failed to store command result in memory", memoryError);
      }
    }
    return {
      content: formattedOutput,
      isError: result.exitCode !== 0
    };
  } catch (error) {
    const errorMessage = error instanceof Error ? error.message : String(error);
    logger.error("Bash tool execution failed", { command, error: errorMessage });
    return {
      content: `Error executing command: ${errorMessage}`,
      isError: true
    };
  }
};
var bashTool = {
  name: createInternalToolName("bash"),
  category: "system",
  internal: true,
  agentAccessible: true,
  description: "Execute bash commands in the system shell. Supports both one-off commands and persistent session execution.",
  parameters: {
    type: "object",
    properties: {
      command: {
        type: "string",
        description: "The bash command to execute. Can be a single command or a chain of commands using && or ;"
      },
      timeout: {
        type: "number",
        description: "Timeout in milliseconds for command execution (default: 30000)",
        default: 3e4
      },
      workingDir: {
        type: "string",
        description: "Working directory for command execution (defaults to current directory)"
      },
      persistent: {
        type: "boolean",
        description: "Whether to use a persistent bash session (maintains state between commands)",
        default: false
      },
      sessionId: {
        type: "string",
        description: "Session ID for persistent sessions (defaults to context session ID)"
      }
    },
    required: ["command"]
  },
  handler: bashHandler,
  version: "1.0.0"
};

// src/core/brain/tools/definitions/system/index.ts
function getSystemTools() {
  return {
    [bashTool.name]: bashTool
  };
}

export {
  createInternalToolName,
  isInternalToolName,
  bashTool,
  getSystemTools
};
