// @byterover/cipher/dist/src/core/chunk-6PWZHI5T.js
import {
  external_exports,
  init_zod
} from "./chunk-NUVPDHVT.js";
import {
  init_esm_shims
} from "./chunk-KFOQGTVJ.js";

// src/core/knowledge_graph/config.ts
init_esm_shims();
init_zod();
var BaseBackendSchema = external_exports.object({
  /** Backend type identifier */
  type: external_exports.string(),
  /** Connection timeout in milliseconds */
  timeout: external_exports.number().min(1e3).max(3e5).default(3e4),
  /** Maximum number of retry attempts for failed operations */
  maxRetries: external_exports.number().min(0).max(10).default(3),
  /** Whether to enable connection pooling (if supported by backend) */
  enablePooling: external_exports.boolean().default(true),
  /** Pool size for connections (if pooling is enabled) */
  poolSize: external_exports.number().min(1).max(100).default(10)
});
var InMemoryBackendSchema = BaseBackendSchema.extend({
  type: external_exports.literal("in-memory"),
  /** Maximum number of nodes to store in memory */
  maxNodes: external_exports.number().min(100).max(1e6).default(1e4),
  /** Maximum number of edges to store in memory */
  maxEdges: external_exports.number().min(100).max(5e6).default(5e4),
  /** Whether to enable indexing for faster lookups */
  enableIndexing: external_exports.boolean().default(true),
  /** Whether to enable automatic garbage collection of orphaned nodes */
  enableGarbageCollection: external_exports.boolean().default(false)
});
var Neo4jBackendSchema = BaseBackendSchema.extend({
  type: external_exports.literal("neo4j"),
  /** Neo4j server hostname (alternative to uri) */
  host: external_exports.string().optional(),
  /** Neo4j server port (alternative to uri) */
  port: external_exports.number().min(1).max(65535).optional(),
  /** Complete Neo4j connection URI (alternative to host/port) */
  uri: external_exports.string().url().optional(),
  /** Database username */
  username: external_exports.string().min(1),
  /** Database password */
  password: external_exports.string().min(1),
  /** Database name to connect to */
  database: external_exports.string().min(1).default("neo4j"),
  /** Whether to use encrypted connection */
  encrypted: external_exports.boolean().default(false),
  /** Whether to trust the server certificate (for encrypted connections) */
  trustServerCertificate: external_exports.boolean().default(false),
  /** Maximum transaction retry time in milliseconds */
  maxTransactionRetryTime: external_exports.number().min(1e3).max(3e5).default(3e4),
  /** Connection acquisition timeout in milliseconds */
  connectionAcquisitionTimeout: external_exports.number().min(1e3).max(6e4).default(6e4),
  /** Maximum connection lifetime in milliseconds */
  maxConnectionLifetime: external_exports.number().min(6e4).max(36e5).default(36e5),
  /** Connection liveness check timeout in milliseconds */
  connectionLivenessCheckTimeout: external_exports.number().min(1e3).max(3e4).default(3e4)
});
var BackendConfigSchema = external_exports.discriminatedUnion("type", [
  InMemoryBackendSchema,
  Neo4jBackendSchema
]);
var KnowledgeGraphSchema = external_exports.intersection(
  BackendConfigSchema,
  external_exports.object({
    /** Whether to enable automatic indexing of node properties */
    enableAutoIndexing: external_exports.boolean().default(true),
    /** Whether to enable performance metrics collection */
    enableMetrics: external_exports.boolean().default(false),
    /** Whether to enable query caching (if supported by backend) */
    enableQueryCache: external_exports.boolean().default(true),
    /** Query cache TTL in milliseconds */
    queryCacheTTL: external_exports.number().min(1e3).max(36e5).default(3e5),
    /** Whether to enable schema validation for nodes and edges */
    enableSchemaValidation: external_exports.boolean().default(true),
    /** Default batch size for bulk operations */
    defaultBatchSize: external_exports.number().min(1).max(1e4).default(1e3)
  })
);
var KnowledgeGraphEnvConfigSchema = external_exports.object({
  /** Whether knowledge graph is enabled */
  KNOWLEDGE_GRAPH_ENABLED: external_exports.boolean().default(false),
  /** Backend type */
  KNOWLEDGE_GRAPH_TYPE: external_exports.enum(["neo4j", "in-memory"]).default("in-memory"),
  /** Neo4j host */
  KNOWLEDGE_GRAPH_HOST: external_exports.string().optional(),
  /** Neo4j port */
  KNOWLEDGE_GRAPH_PORT: external_exports.number().optional(),
  /** Neo4j URI */
  KNOWLEDGE_GRAPH_URI: external_exports.string().optional(),
  /** Neo4j username */
  KNOWLEDGE_GRAPH_USERNAME: external_exports.string().optional(),
  /** Neo4j password */
  KNOWLEDGE_GRAPH_PASSWORD: external_exports.string().optional(),
  /** Neo4j database name */
  KNOWLEDGE_GRAPH_DATABASE: external_exports.string().default("neo4j")
});
function parseKnowledgeGraphConfig(config) {
  return KnowledgeGraphSchema.parse(config);
}
function parseKnowledgeGraphConfigFromEnv(env = process.env) {
  const envConfig = KnowledgeGraphEnvConfigSchema.parse({
    KNOWLEDGE_GRAPH_ENABLED: env.KNOWLEDGE_GRAPH_ENABLED === "true",
    KNOWLEDGE_GRAPH_TYPE: env.KNOWLEDGE_GRAPH_TYPE || "in-memory",
    KNOWLEDGE_GRAPH_HOST: env.KNOWLEDGE_GRAPH_HOST,
    KNOWLEDGE_GRAPH_PORT: env.KNOWLEDGE_GRAPH_PORT ? parseInt(env.KNOWLEDGE_GRAPH_PORT, 10) : void 0,
    KNOWLEDGE_GRAPH_URI: env.KNOWLEDGE_GRAPH_URI,
    KNOWLEDGE_GRAPH_USERNAME: env.KNOWLEDGE_GRAPH_USERNAME,
    KNOWLEDGE_GRAPH_PASSWORD: env.KNOWLEDGE_GRAPH_PASSWORD,
    KNOWLEDGE_GRAPH_DATABASE: env.KNOWLEDGE_GRAPH_DATABASE || "neo4j"
  });
  if (!envConfig.KNOWLEDGE_GRAPH_ENABLED) {
    return null;
  }
  if (envConfig.KNOWLEDGE_GRAPH_TYPE === "neo4j") {
    const neo4jConfig = {
      type: "neo4j",
      database: envConfig.KNOWLEDGE_GRAPH_DATABASE
    };
    if (envConfig.KNOWLEDGE_GRAPH_URI) {
      neo4jConfig.uri = envConfig.KNOWLEDGE_GRAPH_URI;
    } else if (envConfig.KNOWLEDGE_GRAPH_HOST && envConfig.KNOWLEDGE_GRAPH_PORT) {
      neo4jConfig.host = envConfig.KNOWLEDGE_GRAPH_HOST;
      neo4jConfig.port = envConfig.KNOWLEDGE_GRAPH_PORT;
    } else {
      throw new Error(
        "For Neo4j backend, either KNOWLEDGE_GRAPH_URI or both KNOWLEDGE_GRAPH_HOST and KNOWLEDGE_GRAPH_PORT must be provided"
      );
    }
    if (!envConfig.KNOWLEDGE_GRAPH_USERNAME || !envConfig.KNOWLEDGE_GRAPH_PASSWORD) {
      throw new Error(
        "For Neo4j backend, both KNOWLEDGE_GRAPH_USERNAME and KNOWLEDGE_GRAPH_PASSWORD must be provided"
      );
    }
    neo4jConfig.username = envConfig.KNOWLEDGE_GRAPH_USERNAME;
    neo4jConfig.password = envConfig.KNOWLEDGE_GRAPH_PASSWORD;
    if (!neo4jConfig.uri && (!neo4jConfig.host || !neo4jConfig.port)) {
      throw new Error(
        "For Neo4j backend, either KNOWLEDGE_GRAPH_URI or both KNOWLEDGE_GRAPH_HOST and KNOWLEDGE_GRAPH_PORT must be provided"
      );
    }
    return parseKnowledgeGraphConfig(neo4jConfig);
  } else {
    return parseKnowledgeGraphConfig({
      type: "in-memory"
    });
  }
}
function validateKnowledgeGraphConfig(config) {
  try {
    const data = parseKnowledgeGraphConfig(config);
    return { success: true, data };
  } catch (error) {
    if (error instanceof external_exports.ZodError) {
      return { success: false, errors: error };
    }
    throw error;
  }
}

export {
  InMemoryBackendSchema,
  Neo4jBackendSchema,
  BackendConfigSchema,
  KnowledgeGraphSchema,
  KnowledgeGraphEnvConfigSchema,
  parseKnowledgeGraphConfig,
  parseKnowledgeGraphConfigFromEnv,
  validateKnowledgeGraphConfig
};
