// @byterover/cipher/dist/src/core/chunk-RNE2AKGD.js
import {
  init_esm_shims
} from "./chunk-KFOQGTVJ.js";

// src/core/brain/systemPrompt/providers/dynamic-provider.ts
init_esm_shims();

// src/core/brain/systemPrompt/interfaces.ts
init_esm_shims();
var ProviderType = /* @__PURE__ */ ((ProviderType2) => {
  ProviderType2["STATIC"] = "static";
  ProviderType2["DYNAMIC"] = "dynamic";
  ProviderType2["FILE_BASED"] = "file-based";
  return ProviderType2;
})(ProviderType || {});

// src/core/brain/systemPrompt/providers/base-provider.ts
init_esm_shims();
var BasePromptProvider = class {
  id;
  name;
  type;
  priority;
  enabled;
  config = {};
  initialized = false;
  constructor(id, name, type, priority, enabled = true) {
    this.id = id;
    this.name = name;
    this.type = type;
    this.priority = priority;
    this.enabled = enabled;
  }
  /**
   * Default validation - can be overridden by concrete providers
   */
  validateConfig(config) {
    return typeof config === "object" && config !== null;
  }
  /**
   * Initialize the provider with configuration
   */
  async initialize(config) {
    if (!this.validateConfig(config)) {
      throw new Error(`Invalid configuration for provider ${this.id}`);
    }
    this.config = { ...config };
    this.initialized = true;
  }
  /**
   * Default cleanup - can be overridden by concrete providers
   */
  async destroy() {
    this.config = {};
    this.initialized = false;
  }
  /**
   * Check if provider is initialized
   */
  ensureInitialized() {
    if (!this.initialized) {
      throw new Error(`Provider ${this.id} is not initialized`);
    }
  }
  /**
   * Check if provider is enabled and initialized
   */
  canGenerate() {
    return this.enabled && this.initialized;
  }
};

// src/core/brain/systemPrompt/providers/dynamic-provider.ts
var DynamicPromptProvider = class _DynamicPromptProvider extends BasePromptProvider {
  generatorName = "";
  generatorConfig = {};
  template = "";
  generatorFunction;
  // Cached content for this provider
  cachedContent = null;
  // Static registry of generator functions
  static generators = /* @__PURE__ */ new Map();
  constructor(id, name, priority, enabled = true) {
    super(id, name, "dynamic" /* DYNAMIC */, priority, enabled);
  }
  /**
   * Register a dynamic content generator function
   */
  static registerGenerator(name, generator) {
    _DynamicPromptProvider.generators.set(name, generator);
  }
  /**
   * Get all registered generator names
   */
  static getRegisteredGenerators() {
    return Array.from(_DynamicPromptProvider.generators.keys());
  }
  /**
   * Check if a generator is registered
   */
  static isGeneratorRegistered(name) {
    return _DynamicPromptProvider.generators.has(name);
  }
  validateConfig(config) {
    if (!super.validateConfig(config)) {
      return false;
    }
    const typedConfig = config;
    if (typeof typedConfig.generator !== "string" || !typedConfig.generator.trim()) {
      return false;
    }
    if (!_DynamicPromptProvider.isGeneratorRegistered(typedConfig.generator)) {
      return false;
    }
    if (typedConfig.generatorConfig !== void 0) {
      if (typeof typedConfig.generatorConfig !== "object" || typedConfig.generatorConfig === null) {
        return false;
      }
    }
    if (typedConfig.template !== void 0) {
      if (typeof typedConfig.template !== "string") {
        return false;
      }
    }
    return true;
  }
  async initialize(config) {
    await super.initialize(config);
    const typedConfig = config;
    this.generatorName = typedConfig.generator;
    this.generatorConfig = typedConfig.generatorConfig || {};
    this.template = typedConfig.template || "{{content}}";
    const generatorFunction = _DynamicPromptProvider.generators.get(this.generatorName);
    if (!generatorFunction) {
      throw new Error(`Generator function '${this.generatorName}' not found`);
    }
    this.generatorFunction = generatorFunction;
  }
  async generateContent(context) {
    this.ensureInitialized();
    if (!this.canGenerate() || !this.generatorFunction) {
      return "";
    }
    if (this.cachedContent !== null) {
      return this.cachedContent;
    }
    try {
      const dynamicContent = await this.generatorFunction(context, this.generatorConfig);
      const result = this.template.replace("{{content}}", dynamicContent);
      return result;
    } catch (error) {
      throw new Error(
        `Failed to generate dynamic content: ${error instanceof Error ? error.message : "Unknown error"}`
      );
    }
  }
  /**
   * Set the cached content for this provider
   */
  setCachedContent(content) {
    this.cachedContent = content;
  }
  /**
   * Get the cached content for this provider
   */
  getCachedContent() {
    return this.cachedContent;
  }
  /**
   * Clear the cached content (for cache invalidation)
   */
  clearCachedContent() {
    this.cachedContent = null;
  }
  async destroy() {
    await super.destroy();
    this.generatorName = "";
    this.generatorConfig = {};
    this.template = "";
    this.generatorFunction = void 0;
    this.cachedContent = null;
  }
};

export {
  ProviderType,
  BasePromptProvider,
  DynamicPromptProvider
};
