// @byterover/cipher/dist/src/core/chunk-VKPE7HMS.js
import {
  init_esm_shims
} from "./chunk-KFOQGTVJ.js";

// src/core/brain/tools/definitions/knowledge_graph/get-neighbors.ts
init_esm_shims();
async function getNeighborsHandler(args, context) {
  try {
    if (!args.nodeId || typeof args.nodeId !== "string") {
      return {
        success: false,
        message: "Node ID must be a non-empty string"
      };
    }
    const direction = args.direction || "both";
    const limit = args.limit || 10;
    const kgManager = context?.services?.knowledgeGraphManager;
    if (!kgManager) {
      return {
        success: false,
        message: "KnowledgeGraphManager not available in context.services"
      };
    }
    const graph = kgManager.getGraph();
    if (!graph) {
      return {
        success: false,
        message: "Knowledge graph backend is not connected"
      };
    }
    const neighbors = await graph.getNeighbors(args.nodeId, direction, args.edgeTypes, limit);
    return {
      success: true,
      message: "Neighbors retrieved",
      neighbors
    };
  } catch (error) {
    return {
      success: false,
      message: `Failed to get neighbors: ${error.message}`
    };
  }
}
var getNeighborsTool = {
  name: "get_neighbors",
  description: "Get neighboring nodes in the knowledge graph with optional filtering",
  category: "knowledge_graph",
  internal: true,
  handler: getNeighborsHandler,
  parameters: {
    type: "object",
    properties: {
      nodeId: {
        type: "string",
        description: "ID of the node to get neighbors for"
      },
      direction: {
        type: "string",
        enum: ["in", "out", "both"],
        description: "Direction of relationships to traverse",
        default: "both"
      },
      edgeTypes: {
        type: "array",
        items: { type: "string" },
        description: 'Optional edge types to filter by (e.g., ["DEPENDS_ON", "CALLS"])'
      },
      limit: {
        type: "number",
        description: "Maximum number of neighbors to return",
        minimum: 1,
        maximum: 100,
        default: 10
      }
    },
    required: ["nodeId"]
  }
};

export {
  getNeighborsTool
};
