// @byterover/cipher/dist/src/core/chunk-IG2VBOT3.js
import {
  BACKEND_TYPES,
  DEFAULTS,
  DISTANCE_METRICS,
  ERROR_MESSAGES,
  LOG_PREFIXES,
  TIMEOUTS
} from "./chunk-4JSJAC4K.js";
import {
  createLogger,
  logger
} from "./chunk-GPOSIAVD.js";
import {
  env,
  init_env
} from "./chunk-7UQA6M3D.js";
import {
  external_exports,
  init_zod
} from "./chunk-NUVPDHVT.js";
import {
  init_esm_shims
} from "./chunk-KFOQGTVJ.js";

// src/core/vector_storage/manager.ts
init_esm_shims();

// src/core/vector_storage/config.ts
init_esm_shims();
init_zod();
var BaseVectorStoreSchema = external_exports.object({
  /** Name of the collection/index to use */
  collectionName: external_exports.string().min(1).describe("Collection name"),
  /** Dimension of vectors (must match embedding model output) */
  dimension: external_exports.number().int().positive().default(DEFAULTS.DIMENSION).describe("Vector dimension"),
  /** Maximum number of concurrent connections */
  maxConnections: external_exports.number().int().positive().optional().describe("Maximum connections"),
  /** Connection timeout in milliseconds */
  connectionTimeoutMillis: external_exports.number().int().positive().optional().describe("Connection timeout"),
  /** Backend-specific options */
  options: external_exports.record(external_exports.any()).optional().describe("Backend-specific options")
});
var InMemoryBackendSchema = BaseVectorStoreSchema.extend({
  type: external_exports.literal("in-memory"),
  /** Maximum number of vectors to store (prevents memory overflow) */
  maxVectors: external_exports.number().int().positive().default(1e4).describe("Maximum vectors to store")
}).strict();
var QdrantBackendSchema = BaseVectorStoreSchema.extend({
  type: external_exports.literal("qdrant"),
  /** Qdrant connection URL (http://...) - overrides individual params if provided */
  url: external_exports.string().url().optional().describe("Qdrant connection URL"),
  /** Qdrant server hostname */
  host: external_exports.string().optional().describe("Qdrant host"),
  /** Qdrant REST API port (default: 6333) */
  port: external_exports.number().int().positive().default(DEFAULTS.QDRANT_PORT).optional().describe("Qdrant port"),
  /** Qdrant API key for authentication */
  apiKey: external_exports.string().optional().describe("Qdrant API key"),
  /** Store vectors on disk (for large datasets) */
  onDisk: external_exports.boolean().optional().describe("Store vectors on disk"),
  /** Path for local Qdrant storage (if not using remote server) */
  path: external_exports.string().optional().describe("Local storage path"),
  /** Distance metric for similarity search */
  distance: external_exports.enum([
    DISTANCE_METRICS.COSINE,
    DISTANCE_METRICS.EUCLIDEAN,
    DISTANCE_METRICS.DOT_PRODUCT,
    DISTANCE_METRICS.MANHATTAN
  ]).default(DEFAULTS.QDRANT_DISTANCE).optional().describe("Distance metric")
}).strict();
var MilvusBackendSchema = BaseVectorStoreSchema.extend({
  type: external_exports.literal("milvus"),
  /** Milvus connection URL (http://...) - overrides individual params if provided */
  url: external_exports.string().url().optional().describe("Milvus connection URL"),
  /** Milvus server hostname */
  host: external_exports.string().optional().describe("Milvus host"),
  /** Milvus REST API port (default: 19530) */
  port: external_exports.number().int().positive().default(19530).optional().describe("Milvus port"),
  /** Milvus username for authentication (Zilliz Cloud) */
  username: external_exports.string().optional().describe("Milvus username"),
  /** Milvus password for authentication (Zilliz Cloud) */
  password: external_exports.string().optional().describe("Milvus password"),
  /** Milvus API token for authentication (Zilliz Cloud) */
  token: external_exports.string().optional().describe("Milvus API token")
}).strict();
var FaissBackendSchema = BaseVectorStoreSchema.extend({
  type: external_exports.literal("faiss"),
  /** Distance metric for similarity search */
  distance: external_exports.enum(["Cosine", "Euclidean", "IP"]).default("Cosine").optional().describe("Distance metric"),
  /** Path to store the FAISS index file (for persistence) */
  baseStoragePath: external_exports.string().optional().describe("Base directory for FAISS collection data")
}).strict();
var ChromaBackendSchema = BaseVectorStoreSchema.extend({
  type: external_exports.literal("chroma"),
  /** ChromaDB connection URL (http://...) - overrides individual params if provided */
  url: external_exports.string().url().optional().describe("ChromaDB connection URL"),
  /** ChromaDB server hostname */
  host: external_exports.string().optional().describe("ChromaDB host"),
  /** ChromaDB HTTP port (default: 8000) */
  port: external_exports.number().int().positive().default(8e3).optional().describe("ChromaDB port"),
  /** Use SSL/TLS for connection (default: false) */
  ssl: external_exports.boolean().default(false).optional().describe("Use SSL/TLS for connection"),
  /** Custom HTTP headers for authentication */
  headers: external_exports.record(external_exports.string()).optional().describe("Custom HTTP headers"),
  /** Distance metric for similarity search */
  distance: external_exports.enum(["cosine", "l2", "euclidean", "ip", "dot"]).default("cosine").optional().describe("Distance metric"),
  /** Custom path for ChromaDB API endpoints */
  path: external_exports.string().optional().describe("Custom API path")
}).strict();
var PineconeBackendSchema = BaseVectorStoreSchema.extend({
  type: external_exports.literal("pinecone"),
  /** Pinecone API key for authentication */
  apiKey: external_exports.string().min(1).describe("Pinecone API key"),
  /** Pinecone provider (optional) */
  provider: external_exports.string().optional().describe("Pinecone provider"),
  /** Pinecone region (default: 'us-west1') */
  region: external_exports.string().default(DEFAULTS.PINECONE_REGION).optional().describe("Pinecone region"),
  /** Distance metric for similarity search */
  metric: external_exports.enum(["cosine", "euclidean", "dotproduct"]).default("cosine").optional().describe("Distance metric"),
  /** Pinecone pod type (for performance tuning) */
  podType: external_exports.string().optional().describe("Pinecone pod type"),
  /** Number of replicas for high availability */
  replicas: external_exports.number().int().positive().optional().describe("Number of replicas"),
  /** Source collection for cloning */
  sourceCollection: external_exports.string().optional().describe("Source collection for cloning")
}).strict();
var PgVectorBackendSchema = BaseVectorStoreSchema.extend({
  type: external_exports.literal("pgvector"),
  /** PostgreSQL connection URL (postgresql://...) - overrides individual params if provided */
  url: external_exports.string().url().optional().describe("PostgreSQL connection URL"),
  /** Use SSL/TLS for connection (default: false) */
  ssl: external_exports.boolean().default(false).optional().describe("Use SSL/TLS for connection"),
  /** Distance metric for similarity search */
  distance: external_exports.enum(["Cosine", "Euclidean", "Dot", "Manhattan"]).default("Cosine").optional().describe("Distance metric for pgvector"),
  /** Connection pool size (default: 10) */
  poolSize: external_exports.number().int().positive().default(10).optional().describe("Connection pool size"),
  /** Index type for vector similarity search */
  indexType: external_exports.enum(["ivfflat", "hnsw"]).default("hnsw").optional().describe("Vector index type"),
  indexMetric: external_exports.enum(["vector_l2_ops", "vector_ip_ops", "vector_cosine_ops"]).default("vector_l2_ops").optional().describe("Vector index metric for pgvector"),
  /** Schema name (default: 'public') */
  schema: external_exports.string().default("public").optional().describe("PostgreSQL schema name")
}).strict();
var RedisBackendSchema = BaseVectorStoreSchema.extend({
  type: external_exports.literal("redis"),
  url: external_exports.string().url().describe("Redis connection URL (redis://...)"),
  host: external_exports.string().optional().describe("Redis host"),
  port: external_exports.number().int().positive().default(6379).optional().describe("Redis port"),
  username: external_exports.string().optional().describe("Redis username"),
  password: external_exports.string().optional().describe("Redis password"),
  distance: external_exports.enum(["COSINE", "L2", "IP"]).default("COSINE").optional().describe("Distance metric for Redis vector similarity search")
}).strict();
var WeaviateBackendSchema = BaseVectorStoreSchema.extend({
  type: external_exports.literal("weaviate"),
  /** Weaviate connection URL (http://...) - overrides individual params if provided */
  url: external_exports.string().optional().describe("Weaviate connection URL"),
  /** Weaviate server hostname */
  host: external_exports.string().optional().describe("Weaviate host"),
  /** Weaviate REST API port (default: 8080) */
  port: external_exports.number().int().positive().default(8080).optional().describe("Weaviate port"),
  /** Weaviate gRPC port (default: 50051) */
  grpcPort: external_exports.number().int().positive().default(50051).optional().describe("Weaviate gRPC port"),
  /** Weaviate API key for authentication */
  apiKey: external_exports.string().optional().describe("Weaviate API key"),
  /** Weaviate username for authentication */
  username: external_exports.string().optional().describe("Weaviate username"),
  /** Weaviate password for authentication */
  password: external_exports.string().optional().describe("Weaviate password"),
  /** Additional headers (e.g., for third-party API keys) */
  headers: external_exports.record(external_exports.string()).optional().describe("Additional headers"),
  /** Use HTTPS/secure connection */
  secure: external_exports.boolean().default(true).optional().describe("Use secure connection"),
  /** Connection timeout in milliseconds */
  timeout: external_exports.number().int().positive().default(3e4).optional().describe("Connection timeout"),
  /** Distance metric for similarity search */
  distance: external_exports.enum(["Cosine", "Euclidean", "IP"]).default("Cosine").optional().describe("Distance metric")
}).strict();
var BackendConfigSchema = external_exports.discriminatedUnion(
  "type",
  [
    InMemoryBackendSchema,
    QdrantBackendSchema,
    MilvusBackendSchema,
    ChromaBackendSchema,
    PineconeBackendSchema,
    PgVectorBackendSchema,
    FaissBackendSchema,
    RedisBackendSchema,
    WeaviateBackendSchema
  ],
  {
    errorMap: (issue, ctx) => {
      if (issue.code === external_exports.ZodIssueCode.invalid_union_discriminator) {
        return {
          message: `Invalid backend type. Expected 'in-memory', 'qdrant', 'milvus', 'chroma', 'pinecone', 'pgvector', 'redis', or 'weaviate'.`
        };
      }
      return { message: ctx.defaultError };
    }
  }
).describe("Backend configuration for vector storage system").superRefine((data, ctx) => {
  if (data.type === "qdrant") {
    if (!data.url && !data.host) {
      ctx.addIssue({
        code: external_exports.ZodIssueCode.custom,
        message: "Qdrant backend requires either 'url' or 'host' to be specified",
        path: ["url"]
      });
    }
  }
  if (data.type === "milvus") {
    if (!data.url && !data.host) {
      ctx.addIssue({
        code: external_exports.ZodIssueCode.custom,
        message: "Milvus backend requires either 'url' or 'host' to be specified",
        path: ["url"]
      });
    }
  }
  if (data.type === "chroma") {
    if (!data.url && !data.host) {
      ctx.addIssue({
        code: external_exports.ZodIssueCode.custom,
        message: "ChromaDB backend requires either 'url' or 'host' to be specified",
        path: ["url"]
      });
    }
  }
  if (data.type === "pinecone") {
    if (!data.apiKey) {
      ctx.addIssue({
        code: external_exports.ZodIssueCode.custom,
        message: "Pinecone backend requires 'apiKey' to be specified",
        path: ["apiKey"]
      });
    }
  }
  if (data.type === "pgvector") {
    if (!data.url) {
      ctx.addIssue({
        code: external_exports.ZodIssueCode.custom,
        message: "PgVector backend requires either 'url' or both 'host' and 'database' to be specified",
        path: ["url"]
      });
    }
  }
  if (data.type === "faiss") {
    if (!data.baseStoragePath) {
      console.log("Faiss backend requires baseStoragePath");
      ctx.addIssue({
        code: external_exports.ZodIssueCode.custom,
        message: "Faiss backend requires 'baseStoragePath' to be specified",
        path: ["baseStoragePath"]
      });
    }
  }
  if (data.type === "redis") {
    if (!data.url && (!data.host || !data.port)) {
      ctx.addIssue({
        code: external_exports.ZodIssueCode.custom,
        message: "Redis backend requires either 'url' or both 'host' and 'port' to be specified",
        path: ["url"]
      });
    }
  }
  if (data.type === "weaviate") {
    if (!data.url && !data.host) {
      ctx.addIssue({
        code: external_exports.ZodIssueCode.custom,
        message: "Weaviate backend requires either 'url' or 'host' to be specified",
        path: ["url"]
      });
    }
  }
  if (!/^[a-zA-Z0-9_-]+$/.test(data.collectionName)) {
    ctx.addIssue({
      code: external_exports.ZodIssueCode.custom,
      message: "Collection name must contain only letters, numbers, underscores, and hyphens",
      path: ["collectionName"]
    });
  }
});
var VectorStoreSchema = BackendConfigSchema;

// src/core/events/event-types.ts
init_esm_shims();
var ServiceEvents = {
  CIPHER_STARTED: "cipher:started",
  CIPHER_STOPPED: "cipher:stopped",
  CIPHER_ERROR: "cipher:error",
  SERVICE_STARTED: "cipher:serviceStarted",
  SERVICE_ERROR: "cipher:serviceError",
  ALL_SERVICES_READY: "cipher:allServicesReady",
  TOOL_REGISTERED: "cipher:toolRegistered",
  TOOL_UNREGISTERED: "cipher:toolUnregistered",
  TOOL_ERROR: "cipher:toolError",
  MCP_CLIENT_CONNECTED: "cipher:mcpClientConnected",
  MCP_CLIENT_DISCONNECTED: "cipher:mcpClientDisconnected",
  MCP_CLIENT_ERROR: "cipher:mcpClientError",
  MEMORY_OPERATION_STARTED: "cipher:memoryOperationStarted",
  MEMORY_OPERATION_COMPLETED: "cipher:memoryOperationCompleted",
  MEMORY_OPERATION_FAILED: "cipher:memoryOperationFailed",
  VECTOR_STORE_CONNECTED: "cipher:vectorStoreConnected",
  VECTOR_STORE_DISCONNECTED: "cipher:vectorStoreDisconnected",
  VECTOR_STORE_ERROR: "cipher:vectorStoreError",
  LLM_PROVIDER_REGISTERED: "cipher:llmProviderRegistered",
  LLM_PROVIDER_ERROR: "cipher:llmProviderError"
};
var SessionEvents = {
  SESSION_CREATED: "session:created",
  SESSION_ACTIVATED: "session:activated",
  SESSION_DEACTIVATED: "session:deactivated",
  SESSION_EXPIRED: "session:expired",
  SESSION_DELETED: "session:deleted",
  TOOL_EXECUTION_STARTED: "tool:executionStarted",
  TOOL_EXECUTION_COMPLETED: "tool:executionCompleted",
  TOOL_EXECUTION_FAILED: "tool:executionFailed",
  LLM_THINKING: "llm:thinking",
  LLM_RESPONSE_STARTED: "llm:responseStarted",
  LLM_RESPONSE_CHUNK: "llm:responseChunk",
  LLM_RESPONSE_COMPLETED: "llm:responseCompleted",
  LLM_RESPONSE_ERROR: "llm:responseError",
  MEMORY_STORED: "memory:stored",
  MEMORY_RETRIEVED: "memory:retrieved",
  MEMORY_SEARCHED: "memory:searched",
  CONVERSATION_MESSAGE_ADDED: "conversation:messageAdded",
  CONVERSATION_MESSAGE_UPDATED: "conversation:messageUpdated",
  CONVERSATION_CLEARED: "conversation:cleared",
  CONTEXT_UPDATED: "context:updated",
  CONTEXT_TRUNCATED: "context:truncated"
};

// src/core/vector_storage/event-aware-store.ts
init_esm_shims();
var EventAwareVectorStore = class {
  store;
  eventManager;
  sessionId;
  constructor(store, eventManager, sessionId) {
    this.store = store;
    this.eventManager = eventManager;
    this.sessionId = sessionId;
  }
  /**
   * Set the session context for event emission
   */
  setSessionContext(sessionId) {
    this.sessionId = sessionId;
  }
  /**
   * Set the event manager for event emission
   */
  setEventManager(eventManager) {
    this.eventManager = eventManager;
  }
  async insert(vectors, ids, payloads) {
    const startTime = Date.now();
    try {
      await this.store.insert(vectors, ids, payloads);
      if (this.eventManager && this.sessionId) {
        this.eventManager.emitSessionEvent(this.sessionId, SessionEvents.MEMORY_STORED, {
          sessionId: this.sessionId,
          type: "knowledge",
          size: vectors.length,
          timestamp: Date.now()
        });
      }
      logger.debug("Memory operation: insert completed", {
        vectorCount: vectors.length,
        duration: Date.now() - startTime,
        sessionId: this.sessionId
      });
    } catch (error) {
      logger.error("Memory operation: insert failed", {
        error: error instanceof Error ? error.message : String(error),
        vectorCount: vectors.length,
        sessionId: this.sessionId
      });
      throw error;
    }
  }
  async search(query, limit, filters) {
    const startTime = Date.now();
    try {
      const results = await this.store.search(query, limit, filters);
      if (this.eventManager && this.sessionId) {
        this.eventManager.emitSessionEvent(this.sessionId, SessionEvents.MEMORY_SEARCHED, {
          sessionId: this.sessionId,
          query: "vector_search",
          resultCount: results.length,
          duration: Date.now() - startTime,
          timestamp: Date.now()
        });
      }
      logger.debug("Memory operation: search completed", {
        resultCount: results.length,
        limit: limit || 10,
        duration: Date.now() - startTime,
        sessionId: this.sessionId
      });
      return results;
    } catch (error) {
      logger.error("Memory operation: search failed", {
        error: error instanceof Error ? error.message : String(error),
        limit: limit || 10,
        sessionId: this.sessionId
      });
      throw error;
    }
  }
  async get(vectorId) {
    const startTime = Date.now();
    try {
      const result = await this.store.get(vectorId);
      if (this.eventManager && this.sessionId) {
        this.eventManager.emitSessionEvent(this.sessionId, SessionEvents.MEMORY_RETRIEVED, {
          sessionId: this.sessionId,
          type: "knowledge",
          count: result ? 1 : 0,
          timestamp: Date.now()
        });
      }
      logger.debug("Memory operation: get completed", {
        vectorId,
        found: !!result,
        duration: Date.now() - startTime,
        sessionId: this.sessionId
      });
      return result;
    } catch (error) {
      logger.error("Memory operation: get failed", {
        error: error instanceof Error ? error.message : String(error),
        vectorId,
        sessionId: this.sessionId
      });
      throw error;
    }
  }
  async update(vectorId, vector, payload) {
    const startTime = Date.now();
    try {
      await this.store.update(vectorId, vector, payload);
      if (this.eventManager && this.sessionId) {
        this.eventManager.emitSessionEvent(this.sessionId, SessionEvents.MEMORY_STORED, {
          sessionId: this.sessionId,
          type: "knowledge",
          size: 1,
          timestamp: Date.now()
        });
      }
      logger.debug("Memory operation: update completed", {
        vectorId,
        duration: Date.now() - startTime,
        sessionId: this.sessionId
      });
    } catch (error) {
      logger.error("Memory operation: update failed", {
        error: error instanceof Error ? error.message : String(error),
        vectorId,
        sessionId: this.sessionId
      });
      throw error;
    }
  }
  async delete(vectorId) {
    const startTime = Date.now();
    try {
      await this.store.delete(vectorId);
      if (this.eventManager && this.sessionId) {
        this.eventManager.emitSessionEvent(this.sessionId, SessionEvents.MEMORY_STORED, {
          sessionId: this.sessionId,
          type: "knowledge",
          size: 1,
          timestamp: Date.now()
        });
      }
      logger.debug("Memory operation: delete completed", {
        vectorId,
        duration: Date.now() - startTime,
        sessionId: this.sessionId
      });
    } catch (error) {
      logger.error("Memory operation: delete failed", {
        error: error instanceof Error ? error.message : String(error),
        vectorId,
        sessionId: this.sessionId
      });
      throw error;
    }
  }
  async deleteCollection() {
    const startTime = Date.now();
    try {
      await this.store.deleteCollection();
      logger.info("Memory operation: collection deleted", {
        duration: Date.now() - startTime,
        sessionId: this.sessionId
      });
    } catch (error) {
      logger.error("Memory operation: deleteCollection failed", {
        error: error instanceof Error ? error.message : String(error),
        sessionId: this.sessionId
      });
      throw error;
    }
  }
  async list(filters, limit) {
    const startTime = Date.now();
    try {
      const [results, total] = await this.store.list(filters, limit);
      logger.debug("Memory operation: list completed", {
        resultCount: results.length,
        totalCount: total,
        limit: limit || "unlimited",
        duration: Date.now() - startTime,
        sessionId: this.sessionId
      });
      return [results, total];
    } catch (error) {
      logger.error("Memory operation: list failed", {
        error: error instanceof Error ? error.message : String(error),
        limit: limit || "unlimited",
        sessionId: this.sessionId
      });
      throw error;
    }
  }
  // Delegate connection management methods
  async connect() {
    return this.store.connect();
  }
  async disconnect() {
    return this.store.disconnect();
  }
  isConnected() {
    return this.store.isConnected();
  }
  // Delegate metadata methods
  getBackendType() {
    return this.store.getBackendType();
  }
  getDimension() {
    return this.store.getDimension();
  }
  getCollectionName() {
    return this.store.getCollectionName();
  }
};

// src/core/vector_storage/manager.ts
var VectorStoreManager = class _VectorStoreManager {
  // Core state
  store;
  connected = false;
  config;
  logger;
  eventManager;
  // Connection tracking
  connectionAttempts = 0;
  lastConnectionError;
  // Backend metadata
  backendMetadata = {
    type: "unknown",
    isFallback: false,
    connectionTime: 0
  };
  // Lazy loading module references (static to share across instances)
  static qdrantModule;
  static inMemoryModule;
  static milvusModule;
  static chromaModule;
  static pgVectorModule;
  static pineconeModule;
  static faissModule;
  static redisModule;
  static weaviateModule;
  // In VectorStoreManager, track if in-memory is used as fallback or primary
  usedFallback = false;
  factoryFallback = false;
  /**
   * Creates a new VectorStoreManager instance
   *
   * @param config - Vector storage configuration
   * @throws {Error} If configuration is invalid
   */
  constructor(config) {
    const configWithFallback = config;
    if (configWithFallback._fallbackFrom) {
      this.factoryFallback = true;
      this.usedFallback = true;
      delete configWithFallback._fallbackFrom;
    }
    const validationResult = VectorStoreSchema.safeParse(config);
    if (!validationResult.success) {
      throw new Error(
        `${ERROR_MESSAGES.INVALID_CONFIG}: ${validationResult.error.errors.map((e) => `${e.path.join(".")}: ${e.message}`).join(", ")}`
      );
    }
    this.config = validationResult.data;
    this.logger = createLogger({
      level: process.env.LOG_LEVEL || "info"
    });
    this.logger.debug(`${LOG_PREFIXES.MANAGER} Initialized with configuration`, {
      type: this.config.type,
      collection: this.config.collectionName,
      dimension: this.config.dimension,
      fallback: this.usedFallback
    });
  }
  /**
   * Set the event manager for emitting memory operation events
   */
  setEventManager(eventManager) {
    this.eventManager = eventManager;
  }
  /**
   * Get the current vector storage configuration
   *
   * @returns The vector storage configuration
   */
  getConfig() {
    return this.config;
  }
  /**
   * Get information about the vector storage system
   *
   * @returns Vector storage system information including connection status and backend type
   */
  getInfo() {
    return {
      connected: this.connected,
      backend: {
        type: this.backendMetadata.type,
        connected: this.store?.isConnected() ?? false,
        fallback: this.usedFallback,
        collectionName: this.config.collectionName,
        dimension: this.config.dimension
      },
      connectionAttempts: this.connectionAttempts,
      lastError: this.lastConnectionError?.message
    };
  }
  /**
   * Get the current vector store if connected
   *
   * @returns The vector store or null if not connected
   */
  getStore() {
    if (!this.connected || !this.store) {
      return null;
    }
    return this.store;
  }
  /**
   * Get an event-aware vector store for a specific session
   *
   * @param sessionId - The session ID to associate with memory operations
   * @returns Event-aware vector store or null if not connected
   */
  getEventAwareStore(sessionId) {
    if (!this.connected || !this.store) {
      return null;
    }
    if (this.eventManager) {
      return new EventAwareVectorStore(this.store, this.eventManager, sessionId);
    }
    return this.store;
  }
  /**
   * Check if the vector storage manager is connected
   *
   * @returns true if backend is connected
   */
  isConnected() {
    return this.connected && this.store?.isConnected() === true;
  }
  /**
   * Connect to vector storage backend
   *
   * @returns The connected vector store
   * @throws {VectorStoreConnectionError} If backend fails to connect
   */
  async connect() {
    if (!this.factoryFallback) {
      this.usedFallback = false;
    }
    if (this.connected && this.store) {
      this.logger.debug(`${LOG_PREFIXES.MANAGER} Already connected`, {
        type: this.backendMetadata.type
      });
      return this.store;
    }
    this.connectionAttempts++;
    this.logger.debug(
      `${LOG_PREFIXES.MANAGER} Starting connection attempt ${this.connectionAttempts} for ${this.config.collectionName}`
    );
    try {
      const startTime = Date.now();
      try {
        this.store = await this.createBackend();
        await this.store.connect();
        if (!this.factoryFallback) {
          this.usedFallback = false;
        }
        this.logger.debug(`${LOG_PREFIXES.MANAGER} Connected to ${this.config.collectionName}`, {
          type: this.backendMetadata.type,
          isFallback: this.backendMetadata.isFallback,
          connectionTime: `${this.backendMetadata.connectionTime}ms`
        });
      } catch (backendError) {
        this.logger.warn(`${LOG_PREFIXES.MANAGER} Backend connection failed, attempting fallback`, {
          error: backendError instanceof Error ? backendError.message : String(backendError),
          originalType: this.config.type
        });
        if (this.config.type !== BACKEND_TYPES.IN_MEMORY) {
          const { InMemoryBackend } = await import("./in-memory-UU6M4DWS.js");
          this.store = new InMemoryBackend({
            type: "in-memory",
            collectionName: this.config.collectionName,
            dimension: this.config.dimension,
            maxVectors: 1e4
          });
          await this.store.connect();
          this.backendMetadata.type = BACKEND_TYPES.IN_MEMORY;
          this.backendMetadata.isFallback = true;
          this.backendMetadata.connectionTime = Date.now() - startTime;
          this.usedFallback = true;
          this.logger.info(`${LOG_PREFIXES.MANAGER} Connected to fallback backend`, {
            type: this.backendMetadata.type,
            originalType: this.config.type
          });
        } else {
          if (!this.factoryFallback) {
            this.usedFallback = false;
          }
          throw backendError;
        }
      }
      this.connected = true;
      if (this.eventManager) {
        this.eventManager.emitServiceEvent(ServiceEvents.VECTOR_STORE_CONNECTED, {
          provider: this.backendMetadata.type,
          timestamp: Date.now()
        });
      }
      this.logger.info(`${LOG_PREFIXES.MANAGER} Vector storage system connected`, {
        backend: this.backendMetadata.type,
        totalConnectionTime: `${this.backendMetadata.connectionTime}ms`
      });
      return this.store;
    } catch (error) {
      this.lastConnectionError = error;
      if (this.store?.isConnected()) {
        await this.store.disconnect().catch(
          (err) => this.logger.error(`${LOG_PREFIXES.MANAGER} Error during cleanup disconnect`, {
            error: err
          })
        );
      }
      this.store = void 0;
      this.connected = false;
      this.usedFallback = false;
      throw error;
    }
  }
  /**
   * Disconnect from vector storage backend
   */
  async disconnect() {
    if (!this.connected) {
      this.logger.debug(`${LOG_PREFIXES.MANAGER} Already disconnected`);
      return;
    }
    this.logger.info(`${LOG_PREFIXES.MANAGER} Disconnecting vector storage backend`);
    try {
      if (this.store?.isConnected()) {
        await Promise.race([
          this.store.disconnect(),
          new Promise(
            (_, reject) => setTimeout(() => reject(new Error("Disconnect timeout")), TIMEOUTS.SHUTDOWN)
          )
        ]);
        if (this.eventManager) {
          this.eventManager.emitServiceEvent(ServiceEvents.VECTOR_STORE_DISCONNECTED, {
            provider: this.backendMetadata.type,
            reason: "Normal disconnection",
            timestamp: Date.now()
          });
        }
        this.logger.info(`${LOG_PREFIXES.MANAGER} Disconnected successfully`);
      }
    } catch (error) {
      if (this.eventManager) {
        this.eventManager.emitServiceEvent(ServiceEvents.VECTOR_STORE_ERROR, {
          provider: this.backendMetadata.type,
          error: error instanceof Error ? error.message : String(error),
          timestamp: Date.now()
        });
      }
      this.logger.error(`${LOG_PREFIXES.MANAGER} Disconnect error`, { error });
      throw error;
    } finally {
      this.store = void 0;
      this.connected = false;
      this.usedFallback = false;
      this.backendMetadata = {
        type: "unknown",
        isFallback: false,
        connectionTime: 0
      };
      this.logger.info(`${LOG_PREFIXES.MANAGER} Vector storage system disconnected`);
    }
  }
  /**
   * Perform health check on the backend
   *
   * @returns Health check results
   */
  async healthCheck() {
    if (!this.connected || !this.store) {
      return {
        backend: false,
        overall: false,
        details: {
          backend: { status: "not_connected" }
        }
      };
    }
    this.logger.debug(`${LOG_PREFIXES.MANAGER} Performing health check`);
    try {
      const startTime = Date.now();
      const isConnected = this.store.isConnected();
      const latency = Date.now() - startTime;
      const result = {
        backend: isConnected,
        overall: isConnected,
        details: {
          backend: {
            status: isConnected ? "healthy" : "unhealthy",
            latency
          }
        }
      };
      this.logger.info(`${LOG_PREFIXES.MANAGER} Health check completed`, {
        overall: result.overall,
        latency: `${latency}ms`
      });
      return result;
    } catch (error) {
      this.logger.error(`${LOG_PREFIXES.MANAGER} Health check failed`, { error });
      return {
        backend: false,
        overall: false,
        details: {
          backend: {
            status: "error",
            error: error instanceof Error ? error.message : String(error)
          }
        }
      };
    }
  }
  // Private helper methods
  /**
   * Create vector store backend based on configuration
   */
  async createBackend() {
    const config = this.config;
    switch (config.type) {
      case BACKEND_TYPES.QDRANT: {
        try {
          if (!_VectorStoreManager.qdrantModule) {
            this.logger.debug(`${LOG_PREFIXES.MANAGER} Lazy loading Qdrant module`);
            const { QdrantBackend: QdrantBackend2 } = await import("./qdrant-7YKPPI77.js");
            _VectorStoreManager.qdrantModule = QdrantBackend2;
          }
          const QdrantBackend = _VectorStoreManager.qdrantModule;
          this.backendMetadata.type = BACKEND_TYPES.QDRANT;
          this.backendMetadata.isFallback = false;
          return new QdrantBackend(config);
        } catch (error) {
          this.logger.debug(`${LOG_PREFIXES.MANAGER} Failed to create Qdrant backend`, {
            error: error instanceof Error ? error.message : String(error)
          });
          throw error;
        }
      }
      case BACKEND_TYPES.MILVUS: {
        try {
          if (!_VectorStoreManager.milvusModule) {
            this.logger.debug(`${LOG_PREFIXES.MANAGER} Lazy loading Milvus module`);
            const { MilvusBackend: MilvusBackend2 } = await import("./milvus-U53ZMOKE.js");
            _VectorStoreManager.milvusModule = MilvusBackend2;
          }
          const MilvusBackend = _VectorStoreManager.milvusModule;
          this.backendMetadata.type = BACKEND_TYPES.MILVUS;
          this.backendMetadata.isFallback = false;
          return new MilvusBackend(config);
        } catch (error) {
          this.logger.info(`${LOG_PREFIXES.MANAGER} Failed to create Milvus backend: ${error}`, {
            error: error instanceof Error ? error.message : String(error)
          });
          throw error;
        }
      }
      case BACKEND_TYPES.CHROMA: {
        try {
          if (!_VectorStoreManager.chromaModule) {
            this.logger.debug(`${LOG_PREFIXES.MANAGER} Lazy loading ChromaDB module`);
            const { ChromaBackend: ChromaBackend2 } = await import("./chroma-ZIU3QWAU.js");
            _VectorStoreManager.chromaModule = ChromaBackend2;
          }
          const ChromaBackend = _VectorStoreManager.chromaModule;
          this.backendMetadata.type = BACKEND_TYPES.CHROMA;
          this.backendMetadata.isFallback = false;
          return new ChromaBackend(config);
        } catch (error) {
          this.logger.info(`${LOG_PREFIXES.MANAGER} Failed to create ChromaDB backend: ${error}`, {
            error: error instanceof Error ? error.message : String(error)
          });
          throw error;
        }
      }
      case BACKEND_TYPES.PINECONE: {
        try {
          if (!_VectorStoreManager.pineconeModule) {
            this.logger.debug(`${LOG_PREFIXES.MANAGER} Lazy loading Pinecone module`);
            const { PineconeBackend: PineconeBackend2 } = await import("./pinecone-DR5K5YYG.js");
            _VectorStoreManager.pineconeModule = PineconeBackend2;
          }
          const PineconeBackend = _VectorStoreManager.pineconeModule;
          this.backendMetadata.type = BACKEND_TYPES.PINECONE;
          this.backendMetadata.isFallback = false;
          return new PineconeBackend(config);
        } catch (error) {
          this.logger.info(`${LOG_PREFIXES.MANAGER} Failed to create Pinecone backend: ${error}`, {
            error: error instanceof Error ? error.message : String(error)
          });
          throw error;
        }
      }
      case BACKEND_TYPES.PGVECTOR: {
        if (!_VectorStoreManager.pgVectorModule) {
          this.logger.debug(`${LOG_PREFIXES.MANAGER} Lazy loading PgVector module`);
          const { PgVectorBackend: PgVectorBackend2 } = await import("./pgvector-H2QZBU2V.js");
          _VectorStoreManager.pgVectorModule = PgVectorBackend2;
        }
        const PgVectorBackend = _VectorStoreManager.pgVectorModule;
        this.backendMetadata.type = BACKEND_TYPES.PGVECTOR;
        this.backendMetadata.isFallback = false;
        return new PgVectorBackend(config);
      }
      case BACKEND_TYPES.FAISS: {
        if (!_VectorStoreManager.faissModule) {
          this.logger.debug(`${LOG_PREFIXES.MANAGER} Lazy loading in-memory module`);
          const { FaissBackend: FaissBackend2 } = await import("./faiss-PC5L364O.js");
          _VectorStoreManager.faissModule = FaissBackend2;
        }
        const FaissBackend = _VectorStoreManager.faissModule;
        this.backendMetadata.type = BACKEND_TYPES.FAISS;
        this.backendMetadata.isFallback = false;
        return new FaissBackend(config);
      }
      case BACKEND_TYPES.WEAVIATE: {
        try {
          if (!_VectorStoreManager.weaviateModule) {
            this.logger.debug(`${LOG_PREFIXES.MANAGER} Lazy loading Weaviate module`);
            const { WeaviateBackend: WeaviateBackend2 } = await import("./weaviate-XCM6BZOQ.js");
            _VectorStoreManager.weaviateModule = WeaviateBackend2;
          }
          const WeaviateBackend = _VectorStoreManager.weaviateModule;
          this.backendMetadata.type = BACKEND_TYPES.WEAVIATE;
          this.backendMetadata.isFallback = false;
          return new WeaviateBackend(config);
        } catch (error) {
          this.logger.info(`${LOG_PREFIXES.MANAGER} Failed to create Weaviate backend: ${error}`, {
            error: error instanceof Error ? error.message : String(error)
          });
          throw error;
        }
      }
      case BACKEND_TYPES.REDIS: {
        if (!_VectorStoreManager.redisModule) {
          this.logger.debug(`${LOG_PREFIXES.MANAGER} Lazy loading Redis module`);
          const { RedisBackend: RedisBackend2 } = await import("./redis-IV6PRUHK.js");
          _VectorStoreManager.redisModule = RedisBackend2;
        }
        const RedisBackend = _VectorStoreManager.redisModule;
        this.backendMetadata.type = BACKEND_TYPES.REDIS;
        this.backendMetadata.isFallback = false;
        return new RedisBackend(config);
      }
      case BACKEND_TYPES.IN_MEMORY:
      default: {
        if (!_VectorStoreManager.inMemoryModule) {
          this.logger.debug(`${LOG_PREFIXES.MANAGER} Lazy loading in-memory module`);
          const { InMemoryBackend: InMemoryBackend2 } = await import("./in-memory-UU6M4DWS.js");
          _VectorStoreManager.inMemoryModule = InMemoryBackend2;
        }
        const InMemoryBackend = _VectorStoreManager.inMemoryModule;
        this.backendMetadata.type = BACKEND_TYPES.IN_MEMORY;
        this.backendMetadata.isFallback = false;
        return new InMemoryBackend(config);
      }
    }
  }
};

// src/core/vector_storage/factory.ts
init_esm_shims();

// src/core/vector_storage/dual-collection-manager.ts
init_esm_shims();
init_env();
var DualCollectionVectorManager = class {
  knowledgeManager;
  reflectionManager;
  logger;
  reflectionEnabled;
  eventManager;
  constructor(baseConfig) {
    this.logger = createLogger({
      level: env.CIPHER_LOG_LEVEL || "info"
    });
    this.reflectionEnabled = !!(env.REFLECTION_VECTOR_STORE_COLLECTION && env.REFLECTION_VECTOR_STORE_COLLECTION.trim());
    this.knowledgeManager = new VectorStoreManager(baseConfig);
    if (this.reflectionEnabled) {
      const reflectionConfig = {
        ...baseConfig,
        collectionName: env.REFLECTION_VECTOR_STORE_COLLECTION
      };
      this.reflectionManager = new VectorStoreManager(reflectionConfig);
      this.logger.info("DualCollectionVectorManager: Initialized with dual collections", {
        backend: baseConfig.type,
        knowledgeCollection: baseConfig.collectionName,
        reflectionCollection: env.REFLECTION_VECTOR_STORE_COLLECTION,
        dimension: baseConfig.dimension
      });
    } else {
      this.reflectionManager = null;
      this.logger.info("DualCollectionVectorManager: Initialized with single collection", {
        backend: baseConfig.type,
        knowledgeCollection: baseConfig.collectionName,
        dimension: baseConfig.dimension
      });
    }
  }
  /**
   * Set the event manager for emitting memory operation events
   */
  setEventManager(eventManager) {
    this.eventManager = eventManager;
    this.knowledgeManager.setEventManager(eventManager);
    if (this.reflectionManager) {
      this.reflectionManager.setEventManager(eventManager);
    }
  }
  /**
   * Connect both collections
   */
  async connect() {
    this.logger.info("DualCollectionVectorManager: Connecting collections...");
    await this.knowledgeManager.connect();
    this.logger.info("DualCollectionVectorManager: Knowledge collection connected");
    if (this.reflectionEnabled && this.reflectionManager) {
      try {
        await this.reflectionManager.connect();
        this.logger.info("DualCollectionVectorManager: Reflection collection connected");
      } catch (error) {
        this.logger.warn("DualCollectionVectorManager: Failed to connect reflection collection", {
          error: error instanceof Error ? error.message : String(error)
        });
      }
    }
  }
  /**
   * Disconnect both collections
   */
  async disconnect() {
    this.logger.info("DualCollectionVectorManager: Disconnecting collections...");
    await Promise.allSettled([
      this.knowledgeManager.disconnect(),
      this.reflectionManager?.disconnect()
    ]);
    this.logger.info("DualCollectionVectorManager: All collections disconnected");
  }
  /**
   * Get a vector store by collection type
   */
  getStore(type) {
    switch (type) {
      case "knowledge":
        return this.knowledgeManager.getStore();
      case "reflection":
        if (!this.reflectionEnabled || !this.reflectionManager) {
          this.logger.debug("DualCollectionVectorManager: Reflection memory is disabled");
          return null;
        }
        return this.reflectionManager.getStore();
      default:
        throw new Error(`Unknown collection type: ${type}`);
    }
  }
  /**
   * Get an event-aware vector store by collection type for a specific session
   */
  getEventAwareStore(type, sessionId) {
    switch (type) {
      case "knowledge":
        return this.knowledgeManager.getEventAwareStore(sessionId);
      case "reflection":
        if (!this.reflectionEnabled || !this.reflectionManager) {
          this.logger.debug("DualCollectionVectorManager: Reflection memory is disabled");
          return null;
        }
        return this.reflectionManager.getEventAwareStore(sessionId);
      default:
        throw new Error(`Unknown collection type: ${type}`);
    }
  }
  /**
   * Check if collections are connected
   */
  isConnected(type) {
    if (type) {
      switch (type) {
        case "knowledge":
          return this.knowledgeManager.isConnected();
        case "reflection":
          if (!this.reflectionEnabled) return false;
          return this.reflectionManager?.isConnected() === true;
        default:
          throw new Error(`Unknown collection type: ${type}`);
      }
    }
    const knowledgeConnected = this.knowledgeManager.isConnected();
    const reflectionConnected = !this.reflectionEnabled || this.reflectionManager?.isConnected() === true;
    return knowledgeConnected && reflectionConnected;
  }
  /**
   * Get information about both collections
   */
  getInfo() {
    return {
      knowledge: {
        connected: this.knowledgeManager.isConnected(),
        collectionName: this.knowledgeManager.getConfig().collectionName,
        manager: this.knowledgeManager
      },
      reflection: {
        connected: this.reflectionManager?.isConnected() === true,
        collectionName: env.REFLECTION_VECTOR_STORE_COLLECTION,
        manager: this.reflectionManager,
        enabled: this.reflectionEnabled
      },
      overallConnected: this.isConnected()
    };
  }
  /**
   * Get a manager by collection type (for advanced usage)
   */
  getManager(type) {
    switch (type) {
      case "knowledge":
        return this.knowledgeManager;
      case "reflection":
        return this.reflectionManager;
      default:
        throw new Error(`Unknown collection type: ${type}`);
    }
  }
  /**
   * Health check for both collections
   */
  async healthCheck() {
    const knowledgeHealth = await this.knowledgeManager.healthCheck();
    const reflectionHealth = this.reflectionManager ? await this.reflectionManager.healthCheck() : { overall: true };
    return {
      knowledge: knowledgeHealth,
      reflection: reflectionHealth,
      overall: knowledgeHealth.overall && reflectionHealth.overall
    };
  }
};

// src/core/vector_storage/factory.ts
init_env();

// src/core/brain/memory/service-cache.ts
init_esm_shims();
var ServiceCache = class _ServiceCache {
  static instance = null;
  cache = /* @__PURE__ */ new Map();
  initPromises = /* @__PURE__ */ new Map();
  static getInstance() {
    if (!_ServiceCache.instance) {
      _ServiceCache.instance = new _ServiceCache();
    }
    return _ServiceCache.instance;
  }
  /**
   * Get or create a cached service
   */
  async getOrCreate(key, factory, options = {}) {
    if (this.cache.has(key)) {
      logger.debug(`ServiceCache: Cache hit for ${key}`);
      return this.cache.get(key);
    }
    if (this.initPromises.has(key)) {
      logger.debug(`ServiceCache: Waiting for in-progress creation of ${key}`);
      return await this.initPromises.get(key);
    }
    logger.debug(`ServiceCache: Cache miss, creating ${key}`, {
      cacheSize: this.cache.size,
      pendingPromises: this.initPromises.size,
      existingKeys: Array.from(this.cache.keys()).slice(0, 5)
      // Show first 5 keys for debugging
    });
    const promise = factory();
    this.initPromises.set(key, promise);
    try {
      const instance = await promise;
      this.cache.set(key, instance);
      this.initPromises.delete(key);
      if (options.ttl) {
        setTimeout(() => {
          this.cache.delete(key);
          logger.debug(`ServiceCache: Expired ${key} after ${options.ttl}ms`);
        }, options.ttl);
      }
      logger.debug(`ServiceCache: Cached ${key}`);
      return instance;
    } catch (error) {
      this.initPromises.delete(key);
      throw error;
    }
  }
  /**
   * Check if a service is cached
   */
  has(key) {
    return this.cache.has(key);
  }
  /**
   * Clear the cache
   */
  clear() {
    this.cache.clear();
    this.initPromises.clear();
    logger.debug("ServiceCache: Cache cleared");
  }
  /**
   * Get cache statistics
   */
  getStats() {
    return {
      cacheSize: this.cache.size,
      pendingPromises: this.initPromises.size,
      cachedKeys: Array.from(this.cache.keys())
    };
  }
};
function getServiceCache() {
  return ServiceCache.getInstance();
}
function createServiceKey(type, config) {
  if (!config) {
    return type;
  }
  const normalizedConfig = normalizeConfigForCache(config);
  const configStr = JSON.stringify(normalizedConfig);
  const hash = Array.from(configStr).reduce((hash2, char) => {
    return (hash2 << 5) - hash2 + char.charCodeAt(0);
  }, 0);
  const cacheKey = `${type}:${Math.abs(hash)}`;
  return cacheKey;
}
function normalizeConfigForCache(config) {
  if (!config || typeof config !== "object") {
    return config;
  }
  const normalized = {};
  for (const [key, value] of Object.entries(config)) {
    if (value === void 0 || value === null) {
      normalized[key] = "";
    } else if (typeof value === "boolean") {
      normalized[key] = value;
    } else if (typeof value === "string") {
      normalized[key] = value.trim().toLowerCase();
    } else if (typeof value !== "object") {
      normalized[key] = value;
    } else {
      normalized[key] = normalizeConfigForCache(value);
    }
  }
  return normalized;
}

// src/core/vector_storage/factory.ts
async function createVectorStore(config) {
  const logger2 = createLogger({ level: env.CIPHER_LOG_LEVEL });
  logger2.debug(`${LOG_PREFIXES.FACTORY} Creating vector storage system`, {
    type: config.type,
    collection: config.collectionName,
    dimension: config.dimension
  });
  const manager = new VectorStoreManager(config);
  try {
    const store = await manager.connect();
    logger2.info(`${LOG_PREFIXES.FACTORY} Vector storage system created successfully`, {
      type: manager.getInfo().backend.type,
      collection: config.collectionName,
      connected: manager.isConnected()
    });
    return { manager, store };
  } catch (error) {
    await manager.disconnect().catch(() => {
    });
    logger2.error(`${LOG_PREFIXES.FACTORY} Failed to create vector storage system`, {
      error: error instanceof Error ? error.message : String(error)
    });
    throw error;
  }
}
async function createDefaultVectorStore(collectionName = "knowledge_memory", dimension = 1536) {
  return createVectorStore({
    type: "in-memory",
    collectionName,
    dimension,
    maxVectors: 1e4
  });
}
async function createVectorStoreFromEnv(agentConfig) {
  const logger2 = createLogger({ level: env.CIPHER_LOG_LEVEL });
  const config = getVectorStoreConfigFromEnv(agentConfig);
  logger2.info(`${LOG_PREFIXES.FACTORY} Creating vector storage from environment`, {
    type: config.type,
    collection: config.collectionName,
    dimension: config.dimension
  });
  return createVectorStore(config);
}
async function createDualCollectionVectorStoreFromEnv(agentConfig) {
  const logger2 = createLogger({ level: env.CIPHER_LOG_LEVEL });
  const config = getVectorStoreConfigFromEnv(agentConfig);
  const serviceCache = getServiceCache();
  const cacheKey = createServiceKey("dualCollectionVectorStore", {
    type: config.type,
    collection: config.collectionName,
    reflectionCollection: env.REFLECTION_VECTOR_STORE_COLLECTION || "",
    // Include dimension for proper cache key differentiation
    dimension: config.dimension
  });
  return await serviceCache.getOrCreate(cacheKey, async () => {
    logger2.debug("Creating new dual collection vector store instance");
    return await createDualCollectionVectorStoreInternal(config, logger2);
  });
}
async function createDualCollectionVectorStoreInternal(config, logger2) {
  const reflectionCollection = (env.REFLECTION_VECTOR_STORE_COLLECTION || "").trim();
  if (!reflectionCollection) {
    logger2.info(
      `${LOG_PREFIXES.FACTORY} Reflection collection not set, creating single collection manager only`,
      {
        type: config.type,
        knowledgeCollection: config.collectionName
      }
    );
    const manager2 = new DualCollectionVectorManager(config);
    try {
      await manager2.connect();
      const knowledgeStore = manager2.getStore("knowledge");
      if (!knowledgeStore) {
        throw new Error("Failed to get knowledge store from dual collection manager");
      }
      return {
        manager: manager2,
        knowledgeStore,
        reflectionStore: null
      };
    } catch (error) {
      logger2.warn(
        `${LOG_PREFIXES.FACTORY} Primary vector store connection failed, using fallback`,
        {
          originalType: config.type,
          error: error instanceof Error ? error.message : String(error)
        }
      );
      await manager2.disconnect().catch(() => {
      });
      const fallbackConfig = {
        type: "in-memory",
        collectionName: config.collectionName,
        dimension: config.dimension,
        maxVectors: 1e4
      };
      const fallbackManager = new DualCollectionVectorManager(fallbackConfig);
      await fallbackManager.connect();
      const knowledgeStore = fallbackManager.getStore("knowledge");
      if (!knowledgeStore) {
        throw new Error("Failed to get knowledge store from fallback dual collection manager");
      }
      logger2.info(`${LOG_PREFIXES.FACTORY} Successfully created fallback in-memory store`, {
        fallbackType: "in-memory",
        originalType: config.type
      });
      return {
        manager: fallbackManager,
        knowledgeStore,
        reflectionStore: null
      };
    }
  }
  logger2.info(`${LOG_PREFIXES.FACTORY} Creating dual collection vector storage from environment`, {
    type: config.type,
    knowledgeCollection: config.collectionName,
    reflectionCollection,
    refectionEnabled: true
  });
  const manager = new DualCollectionVectorManager(config);
  try {
    await manager.connect();
    const knowledgeStore = manager.getStore("knowledge");
    const reflectionStore = manager.getStore("reflection");
    if (!knowledgeStore) {
      throw new Error("Failed to get knowledge store from dual collection manager");
    }
    return {
      manager,
      knowledgeStore,
      reflectionStore
    };
  } catch (error) {
    await manager.disconnect().catch(() => {
    });
    const fallbackConfig = {
      type: "in-memory",
      collectionName: config.collectionName,
      dimension: config.dimension,
      maxVectors: 1e4
    };
    try {
      const fallbackManager = new DualCollectionVectorManager(fallbackConfig);
      await fallbackManager.connect();
      const knowledgeStore = fallbackManager.getStore("knowledge");
      const reflectionStore = fallbackManager.getStore("reflection");
      if (!knowledgeStore) {
        throw new Error("Failed to get knowledge store from fallback dual collection manager");
      }
      logger2.info(
        `${LOG_PREFIXES.FACTORY} Successfully created fallback in-memory dual collection`,
        {
          knowledge: !!knowledgeStore,
          reflection: !!reflectionStore,
          originalType: config.type
        }
      );
      return {
        manager: fallbackManager,
        knowledgeStore,
        reflectionStore
      };
    } catch (fallbackError) {
      logger2.error(
        `${LOG_PREFIXES.FACTORY} Failed to create fallback dual collection vector storage`,
        {
          originalError: error instanceof Error ? error.message : String(error),
          fallbackError: fallbackError instanceof Error ? fallbackError.message : String(fallbackError)
        }
      );
      throw new Error(
        `Failed to create both primary and fallback dual collection vector storage: ${error instanceof Error ? error.message : String(error)}`
      );
    }
  }
}
function getVectorStoreConfigFromEnv(agentConfig) {
  const logger2 = createLogger({ level: env.CIPHER_LOG_LEVEL });
  const storeType = env.VECTOR_STORE_TYPE;
  const collectionName = env.VECTOR_STORE_COLLECTION;
  let dimension = env.VECTOR_STORE_DIMENSION !== void 0 && !Number.isNaN(env.VECTOR_STORE_DIMENSION) ? env.VECTOR_STORE_DIMENSION : 1536;
  const maxVectors = env.VECTOR_STORE_MAX_VECTORS !== void 0 && !Number.isNaN(env.VECTOR_STORE_MAX_VECTORS) ? env.VECTOR_STORE_MAX_VECTORS : 1e4;
  if (agentConfig?.embedding && typeof agentConfig.embedding === "object" && agentConfig.embedding.dimensions) {
    const embeddingDimension = agentConfig.embedding.dimensions;
    if (typeof embeddingDimension === "number" && embeddingDimension > 0) {
      logger2.debug("Overriding vector store dimension from agent config", {
        envDimension: dimension,
        agentDimension: embeddingDimension,
        embeddingType: agentConfig.embedding.type
      });
      dimension = embeddingDimension;
    }
  }
  if (storeType === "qdrant") {
    const host = env.VECTOR_STORE_HOST;
    const url = env.VECTOR_STORE_URL;
    const port = Number.isNaN(env.VECTOR_STORE_PORT) ? void 0 : env.VECTOR_STORE_PORT;
    const apiKey = env.VECTOR_STORE_API_KEY;
    const distance = env.VECTOR_STORE_DISTANCE;
    const onDisk = env.VECTOR_STORE_ON_DISK;
    if (!url && !host) {
      return {
        type: "in-memory",
        collectionName,
        dimension,
        maxVectors,
        // Add a special property to indicate this is a fallback from Qdrant
        _fallbackFrom: "qdrant"
      };
    }
    return {
      type: "qdrant",
      collectionName,
      dimension,
      url,
      host,
      port,
      apiKey,
      distance,
      onDisk
    };
  } else if (storeType === "milvus") {
    const host = env.VECTOR_STORE_HOST;
    const url = env.VECTOR_STORE_URL;
    const port = Number.isNaN(env.VECTOR_STORE_PORT) ? void 0 : env.VECTOR_STORE_PORT;
    const username = env.VECTOR_STORE_USERNAME;
    const password = env.VECTOR_STORE_PASSWORD;
    const token = env.VECTOR_STORE_API_KEY;
    if (!url && !host) {
      return {
        type: "in-memory",
        collectionName,
        dimension,
        maxVectors,
        // Add a special property to indicate this is a fallback from Milvus
        _fallbackFrom: "milvus"
      };
    }
    console.log("building milvus backend", {
      collectionName,
      dimension,
      url,
      host,
      port
    });
    return {
      type: "milvus",
      collectionName,
      dimension,
      url,
      host,
      port,
      username,
      password,
      token
    };
  } else if (storeType === "chroma") {
    const host = env.VECTOR_STORE_HOST;
    const url = env.VECTOR_STORE_URL;
    const port = Number.isNaN(env.VECTOR_STORE_PORT) ? void 0 : env.VECTOR_STORE_PORT;
    const headers = void 0;
    const envDistance = env.VECTOR_STORE_DISTANCE;
    let distance = "cosine";
    if (envDistance) {
      switch (envDistance.toLowerCase()) {
        case "cosine":
          distance = "cosine";
          break;
        case "euclidean":
        case "l2":
          distance = "l2";
          break;
        case "dot":
        case "ip":
          distance = "ip";
          break;
        default:
          distance = "cosine";
      }
    }
    if (!url && !host) {
      return {
        type: "in-memory",
        collectionName,
        dimension,
        maxVectors,
        // Add a special property to indicate this is a fallback from ChromaDB
        _fallbackFrom: "chroma"
      };
    }
    return {
      type: "chroma",
      collectionName,
      dimension,
      url,
      host,
      port,
      headers,
      distance
    };
  } else if (storeType === "pinecone") {
    const apiKey = env.VECTOR_STORE_API_KEY;
    const collectionName2 = env.VECTOR_STORE_COLLECTION || "default";
    const provider = env.PINECONE_PROVIDER || "aws";
    const region = env.PINECONE_REGION || "us-east-1";
    const envmetric = env.VECTOR_STORE_DISTANCE || "cosine";
    const dimension2 = env.VECTOR_STORE_DIMENSION || 1536;
    let metric = "cosine";
    switch (envmetric) {
      case "Cosine":
        metric = "cosine";
        break;
      case "Euclidean":
        metric = "euclidean";
        break;
      case "Dot":
        metric = "dotproduct";
        break;
      default:
        metric = "cosine";
        break;
    }
    if (!apiKey) {
      return {
        type: "in-memory",
        collectionName: collectionName2,
        dimension: dimension2,
        maxVectors,
        // Add a special property to indicate this is a fallback from Pinecone
        _fallbackFrom: "pinecone"
      };
    }
    return {
      type: "pinecone",
      apiKey,
      collectionName: collectionName2,
      provider,
      region,
      metric,
      dimension: dimension2
    };
  } else if (storeType === "pgvector") {
    const url = env.VECTOR_STORE_URL;
    const collectionName2 = env.VECTOR_STORE_COLLECTION || "pgvector_memory";
    const dimension2 = env.VECTOR_STORE_DIMENSION || 1536;
    const envdistance = env.VECTOR_STORE_DISTANCE;
    let distance = "Cosine";
    switch (envdistance) {
      case "Cosine":
        distance = "Cosine";
        break;
      case "Euclidean":
        distance = "Euclidean";
        break;
      case "Dot":
        distance = "Dot";
        break;
      default:
        distance = "Cosine";
    }
    if (!url) {
      return {
        type: "in-memory",
        collectionName: collectionName2,
        dimension: dimension2,
        maxVectors,
        // Add a special property to indicate this is a fallback from PgVector
        _fallbackFrom: "pgvector"
      };
    }
    return {
      type: "pgvector",
      url,
      collectionName: collectionName2,
      dimension: dimension2,
      distance
    };
  } else if (storeType === "redis") {
    const url = env.VECTOR_STORE_URL;
    const host = env.VECTOR_STORE_HOST;
    const port = env.VECTOR_STORE_PORT;
    const username = env.VECTOR_STORE_USERNAME;
    const password = env.VECTOR_STORE_PASSWORD;
    const envdistance = env.WORKSPACE_VECTOR_STORE_DISTANCE || "Cosine";
    let distance = "COSINE";
    switch (envdistance) {
      case "Euclidean":
        distance = "L2";
        break;
      case "Dot":
        distance = "IP";
        break;
      case "Cosine":
        distance = "COSINE";
        break;
      default:
        distance = "COSINE";
    }
    if (!url && !host) {
      return {
        type: "in-memory",
        collectionName,
        dimension,
        maxVectors,
        // Add a special property to indicate this is a fallback from Redis
        _fallbackFrom: "redis"
      };
    }
    return {
      type: "redis",
      url: url || "",
      host,
      port,
      username,
      password,
      collectionName,
      dimension,
      distance
    };
  } else if (storeType === "faiss") {
    const baseStoragePath = env.FAISS_BASE_STORAGE_PATH;
    const envmetric = env.VECTOR_STORE_DISTANCE;
    let distance = "Cosine";
    if (distance) {
      switch (distance.toLowerCase()) {
        case "Cosine":
          distance = "Cosine";
          break;
        case "L2":
          distance = "Euclidean";
          break;
        case "IP":
          distance = "IP";
          break;
        default:
          distance = "Cosine";
      }
    }
    if (!baseStoragePath) {
      logger2.warn("Faiss base storage path not provided, falling back to in-memory storage", {
        hasBaseStoragePath: !!baseStoragePath
      });
      return {
        type: "in-memory",
        collectionName,
        dimension,
        maxVectors,
        _fallbackFrom: "faiss"
      };
    }
    logger2.debug("Creating FAISS configuration", {
      baseStoragePath,
      dimension,
      distance
    });
    return {
      type: "faiss",
      collectionName,
      dimension,
      baseStoragePath,
      distance
    };
  } else if (storeType === "weaviate") {
    const host = env.VECTOR_STORE_HOST;
    const url = env.VECTOR_STORE_URL;
    const port = Number.isNaN(env.VECTOR_STORE_PORT) ? void 0 : env.VECTOR_STORE_PORT;
    const username = env.VECTOR_STORE_USERNAME;
    const password = env.VECTOR_STORE_PASSWORD;
    const apiKey = env.VECTOR_STORE_API_KEY;
    const envdistance = env.VECTOR_STORE_DISTANCE;
    let distance = "Cosine";
    if (envdistance) {
      switch (envdistance.toLowerCase()) {
        case "Cosine":
          distance = "Cosine";
          break;
        case "L2":
          distance = "Euclidean";
          break;
        case "IP":
          distance = "IP";
          break;
        default:
          distance = "Cosine";
      }
    }
    if (!url && !host) {
      logger2.warn("Weaviate URL or host not provided, falling back to in-memory storage", {
        hasUrl: !!url,
        hasHost: !!host
      });
      return {
        type: "in-memory",
        collectionName,
        dimension,
        maxVectors,
        _fallbackFrom: "weaviate"
      };
    }
    logger2.debug("Creating Weaviate configuration", {
      url,
      host,
      port,
      username,
      password,
      apiKey,
      dimension,
      distance
    });
    return {
      type: "weaviate",
      url,
      host,
      port,
      collectionName,
      dimension,
      username,
      password,
      apiKey,
      distance
    };
  } else {
    return {
      type: "in-memory",
      collectionName,
      dimension,
      maxVectors
    };
  }
}
function getWorkspaceVectorStoreConfigFromEnv(agentConfig) {
  const logger2 = createLogger({ level: env.CIPHER_LOG_LEVEL });
  const storeType = env.WORKSPACE_VECTOR_STORE_TYPE || env.VECTOR_STORE_TYPE;
  const collectionName = env.WORKSPACE_VECTOR_STORE_COLLECTION || "workspace_memory";
  let dimension = env.WORKSPACE_VECTOR_STORE_DIMENSION !== void 0 && !Number.isNaN(env.WORKSPACE_VECTOR_STORE_DIMENSION) ? env.WORKSPACE_VECTOR_STORE_DIMENSION : env.VECTOR_STORE_DIMENSION !== void 0 && !Number.isNaN(env.VECTOR_STORE_DIMENSION) ? env.VECTOR_STORE_DIMENSION : 1536;
  const maxVectors = env.WORKSPACE_VECTOR_STORE_MAX_VECTORS !== void 0 && !Number.isNaN(env.WORKSPACE_VECTOR_STORE_MAX_VECTORS) ? env.WORKSPACE_VECTOR_STORE_MAX_VECTORS : env.VECTOR_STORE_MAX_VECTORS !== void 0 && !Number.isNaN(env.VECTOR_STORE_MAX_VECTORS) ? env.VECTOR_STORE_MAX_VECTORS : 1e4;
  if (agentConfig?.embedding && typeof agentConfig.embedding === "object" && agentConfig.embedding.dimensions) {
    const embeddingDimension = agentConfig.embedding.dimensions;
    if (typeof embeddingDimension === "number" && embeddingDimension > 0) {
      logger2.debug("Overriding workspace vector store dimension from agent config", {
        envDimension: dimension,
        agentDimension: embeddingDimension,
        embeddingType: agentConfig.embedding.type
      });
      dimension = embeddingDimension;
    }
  }
  logger2.debug("Creating workspace vector store config", {
    storeType,
    collectionName,
    dimension,
    maxVectors,
    usingWorkspaceSpecificType: !!env.WORKSPACE_VECTOR_STORE_TYPE,
    usingWorkspaceSpecificCollection: !!env.WORKSPACE_VECTOR_STORE_COLLECTION
  });
  console.log("storeType", storeType);
  if (storeType === "qdrant") {
    const host = env.WORKSPACE_VECTOR_STORE_HOST || env.VECTOR_STORE_HOST;
    const url = env.WORKSPACE_VECTOR_STORE_URL || env.VECTOR_STORE_URL;
    const port = Number.isNaN(env.WORKSPACE_VECTOR_STORE_PORT) ? Number.isNaN(env.VECTOR_STORE_PORT) ? void 0 : env.VECTOR_STORE_PORT : env.WORKSPACE_VECTOR_STORE_PORT;
    const apiKey = env.WORKSPACE_VECTOR_STORE_API_KEY || env.VECTOR_STORE_API_KEY;
    const distance = env.WORKSPACE_VECTOR_STORE_DISTANCE || env.VECTOR_STORE_DISTANCE;
    const onDisk = env.WORKSPACE_VECTOR_STORE_ON_DISK;
    if (!url && !host) {
      return {
        type: "in-memory",
        collectionName,
        dimension,
        maxVectors,
        // Add a special property to indicate this is a fallback from Qdrant
        _fallbackFrom: "qdrant-workspace"
      };
    }
    return {
      type: "qdrant",
      collectionName,
      dimension,
      url,
      host,
      port,
      apiKey,
      distance,
      onDisk
    };
  } else if (storeType === "milvus") {
    const host = env.WORKSPACE_VECTOR_STORE_HOST || env.VECTOR_STORE_HOST;
    const url = env.WORKSPACE_VECTOR_STORE_URL || env.VECTOR_STORE_URL;
    const port = Number.isNaN(env.WORKSPACE_VECTOR_STORE_PORT) ? Number.isNaN(env.VECTOR_STORE_PORT) ? void 0 : env.VECTOR_STORE_PORT : env.WORKSPACE_VECTOR_STORE_PORT;
    const username = env.WORKSPACE_VECTOR_STORE_USERNAME || env.VECTOR_STORE_USERNAME;
    const password = env.WORKSPACE_VECTOR_STORE_PASSWORD || env.VECTOR_STORE_PASSWORD;
    const token = env.WORKSPACE_VECTOR_STORE_API_KEY || env.VECTOR_STORE_API_KEY;
    if (!url && !host) {
      console.log("Milvus backend not initialzed, returning in-memory config with fallback marker");
      return {
        type: "in-memory",
        collectionName,
        dimension,
        maxVectors,
        // Add a special property to indicate this is a fallback from Milvus
        _fallbackFrom: "milvus-workspace"
      };
    }
    return {
      type: "milvus",
      collectionName,
      dimension,
      url,
      host,
      port,
      username,
      password,
      token
    };
  } else if (storeType === "chroma") {
    const host = env.WORKSPACE_VECTOR_STORE_HOST || env.VECTOR_STORE_HOST;
    const url = env.WORKSPACE_VECTOR_STORE_URL || env.VECTOR_STORE_URL;
    const port = Number.isNaN(env.WORKSPACE_VECTOR_STORE_PORT) ? Number.isNaN(env.VECTOR_STORE_PORT) ? void 0 : env.VECTOR_STORE_PORT : env.WORKSPACE_VECTOR_STORE_PORT;
    const headers = void 0;
    const envDistance = env.WORKSPACE_VECTOR_STORE_DISTANCE || env.VECTOR_STORE_DISTANCE;
    let distance = "cosine";
    if (envDistance) {
      switch (envDistance.toLowerCase()) {
        case "cosine":
          distance = "cosine";
          break;
        case "euclidean":
        case "l2":
          distance = "l2";
          break;
        case "dot":
        case "ip":
          distance = "ip";
          break;
        default:
          distance = "cosine";
      }
    }
    if (!url && !host) {
      return {
        type: "in-memory",
        collectionName,
        dimension,
        maxVectors,
        // Add a special property to indicate this is a fallback from ChromaDB
        _fallbackFrom: "chroma-workspace"
      };
    }
    return {
      type: "chroma",
      collectionName,
      dimension,
      url,
      host,
      port,
      headers,
      distance
    };
  } else if (storeType === "pinecone") {
    const apiKey = env.WORKSPACE_VECTOR_STORE_API_KEY;
    const collectionName2 = env.WORKSPACE_VECTOR_STORE_COLLECTION || "default";
    const provider = env.WORKSPACE_PINECONE_PROVIDER || "aws";
    const region = env.WORKSPACE_PINECONE_REGION || "us-east-1";
    const envmetric = env.WORKSPACE_VECTOR_STORE_DISTANCE || "cosine";
    const dimension2 = env.WORKSPACE_VECTOR_STORE_DIMENSION || 1536;
    let metric = "cosine";
    switch (envmetric) {
      case "Cosine":
        metric = "cosine";
        break;
      case "Euclidean":
        metric = "euclidean";
        break;
      case "Dot":
        metric = "dotproduct";
        break;
      default:
        metric = "cosine";
        break;
    }
    if (!apiKey) {
      return {
        type: "in-memory",
        collectionName: collectionName2,
        dimension: dimension2,
        maxVectors,
        // Add a special property to indicate this is a fallback from PgVector
        _fallbackFrom: "pinecone"
      };
    }
    return {
      type: "pinecone",
      apiKey,
      collectionName: collectionName2,
      provider,
      region,
      metric,
      dimension: dimension2
    };
  } else if (storeType === "pgvector") {
    const url = env.WORKSPACE_VECTOR_STORE_URL;
    const collectionName2 = env.WORKSPACE_VECTOR_STORE_COLLECTION || "pgvector_memory";
    const dimension2 = env.WORKSPACE_VECTOR_STORE_DIMENSION || 1536;
    const distance = env.WORKSPACE_VECTOR_STORE_DISTANCE || "Cosine";
    if (!url) {
      return {
        type: "in-memory",
        collectionName: collectionName2,
        dimension: dimension2,
        maxVectors,
        // Add a special property to indicate this is a fallback from PgVector
        _fallbackFrom: "pgvector"
      };
    }
    return {
      type: "pgvector",
      url,
      collectionName: collectionName2,
      dimension: dimension2,
      distance
    };
  } else if (storeType === "redis") {
    const url = env.WORKSPACE_VECTOR_STORE_URL;
    const host = env.WORKSPACE_VECTOR_STORE_HOST;
    const port = env.WORKSPACE_VECTOR_STORE_PORT;
    const username = env.WORKSPACE_VECTOR_STORE_USERNAME;
    const password = env.WORKSPACE_VECTOR_STORE_PASSWORD;
    const envdistance = env.WORKSPACE_VECTOR_STORE_DISTANCE || "Cosine";
    let distance = "COSINE";
    switch (envdistance) {
      case "Euclidean":
        distance = "L2";
        break;
      case "Dot":
        distance = "IP";
        break;
      case "Cosine":
        distance = "COSINE";
        break;
      default:
        distance = "COSINE";
    }
    if (!url && !host) {
      return {
        type: "in-memory",
        collectionName,
        dimension,
        maxVectors,
        // Add a special property to indicate this is a fallback from Redis
        _fallbackFrom: "redis"
      };
    }
    return {
      type: "redis",
      url: url || "",
      host,
      port,
      username,
      password,
      collectionName,
      dimension,
      distance
    };
  } else if (storeType === "weaviate") {
    const host = env.WORKSPACE_VECTOR_STORE_HOST;
    const url = env.WORKSPACE_VECTOR_STORE_URL;
    const port = Number.isNaN(env.WORKSPACE_VECTOR_STORE_PORT) ? void 0 : env.WORKSPACE_VECTOR_STORE_PORT;
    const username = env.WORKSPACE_VECTOR_STORE_USERNAME;
    const password = env.WORKSPACE_VECTOR_STORE_PASSWORD;
    const apiKey = env.WORKSPACE_VECTOR_STORE_API_KEY;
    const envdistance = env.WORKSPACE_VECTOR_STORE_DISTANCE;
    let distance = "Cosine";
    if (envdistance) {
      switch (envdistance.toLowerCase()) {
        case "Cosine":
          distance = "Cosine";
          break;
        case "L2":
          distance = "Euclidean";
          break;
        case "IP":
          distance = "IP";
          break;
        default:
          distance = "Cosine";
      }
    }
    if (!url && !host) {
      logger2.warn("Weaviate URL or host not provided, falling back to in-memory storage", {
        hasUrl: !!url,
        hasHost: !!host
      });
      return {
        type: "in-memory",
        collectionName,
        dimension,
        maxVectors,
        _fallbackFrom: "weaviate"
      };
    }
    logger2.debug("Creating Weaviate configuration", {
      url,
      host,
      port,
      username,
      password,
      apiKey,
      dimension,
      distance
    });
    return {
      type: "weaviate",
      url,
      host,
      port,
      collectionName,
      dimension,
      username,
      password,
      apiKey,
      distance
    };
  } else {
    return {
      type: "in-memory",
      collectionName,
      dimension,
      maxVectors
    };
  }
}
async function createMultiCollectionVectorStoreFromEnv(agentConfig) {
  const logger2 = createLogger({ level: env.CIPHER_LOG_LEVEL });
  const config = getVectorStoreConfigFromEnv(agentConfig);
  const serviceCache = getServiceCache();
  const cacheKey = createServiceKey("multiCollectionVectorStore", {
    type: config.type,
    collection: config.collectionName,
    reflectionCollection: env.REFLECTION_VECTOR_STORE_COLLECTION || "",
    workspaceCollection: env.WORKSPACE_VECTOR_STORE_COLLECTION || "workspace_memory",
    workspaceEnabled: !!env.USE_WORKSPACE_MEMORY,
    // Include dimension for proper cache key differentiation
    dimension: config.dimension
  });
  return await serviceCache.getOrCreate(cacheKey, async () => {
    logger2.debug("Creating new multi collection vector store instance");
    return await createMultiCollectionVectorStoreInternal(config, logger2);
  });
}
async function createMultiCollectionVectorStoreInternal(config, logger2) {
  const { MultiCollectionVectorManager } = await import("./multi-collection-manager-7DLF36ER.js");
  logger2.info(`${LOG_PREFIXES.FACTORY} Creating multi collection vector storage from environment`, {
    type: config.type,
    knowledgeCollection: config.collectionName,
    reflectionCollection: env.REFLECTION_VECTOR_STORE_COLLECTION || "disabled",
    workspaceCollection: env.USE_WORKSPACE_MEMORY ? env.WORKSPACE_VECTOR_STORE_COLLECTION || "workspace_memory" : "disabled",
    workspaceEnabled: !!env.USE_WORKSPACE_MEMORY
  });
  const manager = new MultiCollectionVectorManager(config);
  try {
    const connected = await manager.connect();
    if (!connected) {
      throw new Error("Failed to connect multi collection vector manager");
    }
    const knowledgeStore = manager.getStore("knowledge");
    const reflectionStore = manager.getStore("reflection");
    const workspaceStore = manager.getStore("workspace");
    if (!knowledgeStore) {
      throw new Error("Failed to get knowledge store from multi collection manager");
    }
    logger2.info(`${LOG_PREFIXES.FACTORY} Multi collection vector storage created successfully`, {
      knowledge: !!knowledgeStore,
      reflection: !!reflectionStore,
      workspace: !!workspaceStore
    });
    return {
      manager,
      knowledgeStore,
      reflectionStore,
      workspaceStore
    };
  } catch (error) {
    logger2.warn(`${LOG_PREFIXES.FACTORY} Multi collection connection failed, attempting fallback`, {
      originalType: config.type,
      error: error instanceof Error ? error.message : String(error)
    });
    await manager.disconnect().catch(() => {
    });
    const fallbackConfig = {
      type: "in-memory",
      collectionName: config.collectionName,
      dimension: config.dimension,
      maxVectors: 1e4
    };
    try {
      const fallbackManager = new MultiCollectionVectorManager(fallbackConfig);
      const connected = await fallbackManager.connect();
      if (!connected) {
        throw new Error("Failed to connect fallback multi collection vector manager");
      }
      const knowledgeStore = fallbackManager.getStore("knowledge");
      const reflectionStore = fallbackManager.getStore("reflection");
      const workspaceStore = fallbackManager.getStore("workspace");
      if (!knowledgeStore) {
        throw new Error("Failed to get knowledge store from fallback multi collection manager");
      }
      logger2.info(
        `${LOG_PREFIXES.FACTORY} Successfully created fallback in-memory multi collection`,
        {
          knowledge: !!knowledgeStore,
          reflection: !!reflectionStore,
          workspace: !!workspaceStore,
          originalType: config.type
        }
      );
      return {
        manager: fallbackManager,
        knowledgeStore,
        reflectionStore,
        workspaceStore
      };
    } catch (fallbackError) {
      logger2.error(
        `${LOG_PREFIXES.FACTORY} Failed to create fallback multi collection vector storage`,
        {
          originalError: error instanceof Error ? error.message : String(error),
          fallbackError: fallbackError instanceof Error ? fallbackError.message : String(fallbackError)
        }
      );
      throw new Error(
        `Failed to create both primary and fallback multi collection vector storage: ${error instanceof Error ? error.message : String(error)}`
      );
    }
  }
}
async function createWorkspaceVectorStoreFromEnv(agentConfig) {
  const logger2 = createLogger({ level: env.CIPHER_LOG_LEVEL });
  const config = getWorkspaceVectorStoreConfigFromEnv(agentConfig);
  logger2.info(`${LOG_PREFIXES.FACTORY} Creating workspace memory vector storage from environment`, {
    type: config.type,
    collection: config.collectionName,
    dimension: config.dimension,
    workspaceSpecific: config.collectionName !== env.VECTOR_STORE_COLLECTION
  });
  return createVectorStore(config);
}
function isVectorStoreFactory(obj) {
  return typeof obj === "object" && obj !== null && "manager" in obj && "store" in obj && obj.manager instanceof VectorStoreManager;
}

export {
  ServiceEvents,
  SessionEvents,
  getServiceCache,
  createServiceKey,
  VectorStoreManager,
  DualCollectionVectorManager,
  createVectorStore,
  createDefaultVectorStore,
  createVectorStoreFromEnv,
  createDualCollectionVectorStoreFromEnv,
  getVectorStoreConfigFromEnv,
  getWorkspaceVectorStoreConfigFromEnv,
  createMultiCollectionVectorStoreFromEnv,
  createWorkspaceVectorStoreFromEnv,
  isVectorStoreFactory
};
