// @byterover/cipher/dist/src/core/chunk-WZZEPDWF.js
import {
  init_esm_shims
} from "./chunk-KFOQGTVJ.js";

// src/core/storage/constants.ts
init_esm_shims();
var LOG_PREFIXES = {
  MANAGER: "[StorageManager]",
  CACHE: "[StorageManager:Cache]",
  DATABASE: "[StorageManager:Database]",
  HEALTH: "[StorageManager:Health]",
  FACTORY: "[StorageFactory]",
  BACKEND: "[StorageBackend]"
};
var ERROR_MESSAGES = {
  // Connection errors
  CACHE_CONNECTION_FAILED: "Failed to connect to cache backend",
  DATABASE_CONNECTION_FAILED: "Failed to connect to database backend",
  ALREADY_CONNECTED: "Storage manager is already connected",
  NOT_CONNECTED: "Storage manager is not connected",
  // Backend errors
  BACKEND_NOT_FOUND: "Storage backend not found",
  INVALID_BACKEND_TYPE: "Invalid backend type specified",
  MODULE_LOAD_FAILED: "Failed to load backend module",
  // Operation errors
  HEALTH_CHECK_FAILED: "Health check failed",
  OPERATION_TIMEOUT: "Storage operation timed out",
  SERIALIZATION_ERROR: "Failed to serialize/deserialize data",
  // Configuration errors
  INVALID_CONFIG: "Invalid storage configuration",
  MISSING_REQUIRED_CONFIG: "Missing required configuration"
};
var TIMEOUTS = {
  CONNECTION: 1e4,
  // 10 seconds
  HEALTH_CHECK: 5e3,
  // 5 seconds
  OPERATION: 3e4,
  // 30 seconds
  SHUTDOWN: 5e3
  // 5 seconds
};
var HEALTH_CHECK = {
  KEY: "storage_manager_health_check",
  VALUE: "ok",
  TTL_SECONDS: 10
};
var BACKEND_TYPES = {
  // Cache backends
  REDIS: "redis",
  MEMCACHED: "memcached",
  IN_MEMORY: "in-memory",
  // Database backends
  SQLITE: "sqlite",
  POSTGRES: "postgres",
  MYSQL: "mysql"
};
var DEFAULTS = {
  MAX_RETRIES: 3,
  RETRY_DELAY: 1e3,
  // 1 second
  CACHE_TTL: 3600,
  // 1 hour in seconds
  MAX_CONNECTIONS: 10,
  IDLE_TIMEOUT: 3e4
  // 30 seconds
};

export {
  LOG_PREFIXES,
  ERROR_MESSAGES,
  TIMEOUTS,
  HEALTH_CHECK,
  BACKEND_TYPES,
  DEFAULTS
};
