// @byterover/cipher/dist/src/core/chunk-JOPES53R.js
import {
  init_esm_shims
} from "./chunk-KFOQGTVJ.js";

// src/core/vector_storage/backend/types.ts
init_esm_shims();
var VectorStoreError = class extends Error {
  constructor(message, operation, cause) {
    super(message);
    this.message = message;
    this.operation = operation;
    this.cause = cause;
    this.name = "VectorStoreError";
    this.cause = cause;
  }
};
var VectorStoreConnectionError = class extends VectorStoreError {
  constructor(message, backendType, cause) {
    super(message, "connection", cause);
    this.message = message;
    this.backendType = backendType;
    this.cause = cause;
    this.name = "VectorStoreConnectionError";
  }
};
var VectorDimensionError = class extends VectorStoreError {
  constructor(message, expectedDimension, actualDimension, cause) {
    super(message, "dimension_mismatch", cause);
    this.message = message;
    this.expectedDimension = expectedDimension;
    this.actualDimension = actualDimension;
    this.cause = cause;
    this.name = "VectorDimensionError";
  }
};
var CollectionNotFoundError = class extends VectorStoreError {
  constructor(message, collectionName, cause) {
    super(message, "collection_not_found", cause);
    this.message = message;
    this.collectionName = collectionName;
    this.cause = cause;
    this.name = "CollectionNotFoundError";
  }
};

export {
  VectorStoreError,
  VectorStoreConnectionError,
  VectorDimensionError,
  CollectionNotFoundError
};
