// @byterover/cipher/dist/src/core/chunk-MMF5ASL2.js
import {
  BACKEND_TYPES,
  ERROR_MESSAGES,
  HEALTH_CHECK,
  LOG_PREFIXES,
  TIMEOUTS
} from "./chunk-WZZEPDWF.js";
import {
  createLogger
} from "./chunk-GPOSIAVD.js";
import {
  external_exports,
  init_zod
} from "./chunk-NUVPDHVT.js";
import {
  init_esm_shims
} from "./chunk-KFOQGTVJ.js";

// src/core/storage/manager.ts
init_esm_shims();

// src/core/storage/config.ts
init_esm_shims();
init_zod();
var BaseBackendSchema = external_exports.object({
  /** Maximum number of concurrent connections to the backend */
  maxConnections: external_exports.number().int().positive().optional().describe("Maximum connections"),
  /** Time in milliseconds before an idle connection is closed */
  idleTimeoutMillis: external_exports.number().int().positive().optional().describe("Idle timeout in milliseconds"),
  /** Time in milliseconds to wait for a connection to be established */
  connectionTimeoutMillis: external_exports.number().int().positive().optional().describe("Connection timeout in milliseconds"),
  /** Backend-specific options that vary by implementation */
  options: external_exports.record(external_exports.any()).optional().describe("Backend-specific options")
});
var InMemoryBackendSchema = BaseBackendSchema.extend({
  type: external_exports.literal("in-memory")
}).strict();
var RedisBackendSchema = BaseBackendSchema.extend({
  type: external_exports.literal("redis"),
  /** Redis connection URL (redis://...) - overrides individual params if provided */
  url: external_exports.string().optional().describe("Redis connection URL (redis://...)"),
  /** Redis server hostname */
  host: external_exports.string().optional().describe("Redis host"),
  /** Redis server port (default: 6379) */
  port: external_exports.number().int().positive().optional().describe("Redis port"),
  /** Redis authentication username */
  username: external_exports.string().optional().describe("Redis username"),
  /** Redis authentication password */
  password: external_exports.string().optional().describe("Redis password"),
  /** Redis database number (0-15, default: 0) */
  database: external_exports.number().int().nonnegative().optional().describe("Redis database number")
}).strict();
var SqliteBackendSchema = BaseBackendSchema.extend({
  type: external_exports.literal("sqlite"),
  /**
   * SQLite database file path.
   * If not provided, will auto-detect using the path resolver.
   */
  path: external_exports.string().optional().describe(
    "SQLite database file path (optional, will auto-detect using path resolver if not provided)"
  ),
  /** Database filename (default: cipher.db) */
  database: external_exports.string().optional().describe("Database filename (default: cipher.db)")
}).strict();
var PostgresBackendSchema = BaseBackendSchema.extend({
  type: external_exports.literal("postgres"),
  /** PostgreSQL connection URL (postgresql://...) - overrides individual params if provided */
  url: external_exports.string().optional().describe("PostgreSQL connection URL (postgresql://...)"),
  /** PostgreSQL server hostname */
  host: external_exports.string().optional().describe("PostgreSQL host"),
  /** PostgreSQL server port (default: 5432) */
  port: external_exports.number().int().positive().optional().describe("PostgreSQL port"),
  /** Database name */
  database: external_exports.string().optional().describe("Database name"),
  /** Username for authentication */
  user: external_exports.string().optional().describe("Username"),
  /** Password for authentication */
  password: external_exports.string().optional().describe("Password"),
  /** Enable SSL connection (default: false) */
  ssl: external_exports.boolean().optional().describe("Enable SSL connection"),
  /** Connection pool settings */
  pool: external_exports.object({
    /** Minimum number of connections in pool */
    min: external_exports.number().int().nonnegative().optional().describe("Minimum pool size"),
    /** Maximum number of connections in pool */
    max: external_exports.number().int().positive().optional().describe("Maximum pool size"),
    /** Connection idle timeout in ms */
    idleTimeoutMillis: external_exports.number().int().positive().optional().describe("Connection idle timeout"),
    /** Connection acquire timeout in ms */
    acquireTimeoutMillis: external_exports.number().int().positive().optional().describe("Connection acquire timeout")
  }).optional().describe("Connection pool settings")
}).strict();
var BackendConfigSchema = external_exports.discriminatedUnion(
  "type",
  [InMemoryBackendSchema, RedisBackendSchema, SqliteBackendSchema, PostgresBackendSchema],
  {
    errorMap: (issue, ctx) => {
      if (issue.code === external_exports.ZodIssueCode.invalid_union_discriminator) {
        return {
          message: `Invalid backend type. Expected 'in-memory', 'redis', 'sqlite', or 'postgres'.`
        };
      }
      return { message: ctx.defaultError };
    }
  }
).describe("Backend configuration for storage system").superRefine((data, ctx) => {
  if (data.type === "redis") {
    if (!data.url && !data.host) {
      ctx.addIssue({
        code: external_exports.ZodIssueCode.custom,
        message: "Redis backend requires either 'url' or 'host' to be specified",
        path: ["url"]
      });
    }
  }
  if (data.type === "postgres") {
    if (!data.url && (!data.host || !data.database)) {
      ctx.addIssue({
        code: external_exports.ZodIssueCode.custom,
        message: "PostgreSQL backend requires either 'url' or both 'host' and 'database' to be specified",
        path: ["url"]
      });
    }
  }
});
var StorageSchema = external_exports.object({
  /** Cache backend for fast, ephemeral storage (Redis, In-Memory) */
  cache: BackendConfigSchema.describe("Cache backend configuration (fast, ephemeral)"),
  /** Database backend for persistent, reliable storage (SQLite, PostgreSQL) */
  database: BackendConfigSchema.describe("Database backend configuration (persistent, reliable)")
}).strict().describe("Storage configuration with cache and database backends");

// src/core/storage/manager.ts
var StorageManager = class _StorageManager {
  // Core state
  cache;
  database;
  connected = false;
  config;
  logger;
  // Connection tracking
  connectionAttempts = 0;
  lastConnectionError;
  // Backend metadata
  cacheMetadata = {
    type: "unknown",
    isFallback: false,
    connectionTime: 0
  };
  databaseMetadata = {
    type: "unknown",
    isFallback: false,
    connectionTime: 0
  };
  // Lazy loading module references (static to share across instances)
  static redisModule;
  static sqliteModule;
  static postgresModule;
  // Health check configuration
  healthCheckKey = HEALTH_CHECK.KEY;
  healthCheckTimeout = TIMEOUTS.HEALTH_CHECK;
  /**
   * Creates a new StorageManager instance
   *
   * @param config - Storage configuration with cache and database backend configs
   * @throws {Error} If configuration is invalid
   */
  constructor(config) {
    const validationResult = StorageSchema.safeParse(config);
    if (!validationResult.success) {
      throw new Error(
        `${ERROR_MESSAGES.INVALID_CONFIG}: ${validationResult.error.errors.map((e) => `${e.path.join(".")}: ${e.message}`).join(", ")}`
      );
    }
    this.config = validationResult.data;
    this.logger = createLogger({
      level: process.env.LOG_LEVEL || "info"
    });
    this.logger.info(`${LOG_PREFIXES.MANAGER} Initialized with configuration`, {
      cacheType: this.config.cache.type,
      databaseType: this.config.database.type
    });
  }
  /**
   * Get the current storage configuration
   *
   * @returns The storage configuration
   */
  getConfig() {
    return this.config;
  }
  /**
   * Get information about the storage system
   *
   * @returns Storage system information including connection status and backend types
   */
  getInfo() {
    return {
      connected: this.connected,
      backends: {
        cache: {
          type: this.cacheMetadata.type,
          connected: this.cache?.isConnected() ?? false,
          fallback: this.cacheMetadata.isFallback
        },
        database: {
          type: this.databaseMetadata.type,
          connected: this.database?.isConnected() ?? false,
          fallback: this.databaseMetadata.isFallback
        }
      },
      connectionAttempts: this.connectionAttempts,
      lastError: this.lastConnectionError?.message
    };
  }
  /**
   * Get the current storage backends if connected
   *
   * @returns The storage backends or null if not connected
   */
  getBackends() {
    if (!this.connected || !this.cache || !this.database) {
      return null;
    }
    return {
      cache: this.cache,
      database: this.database
    };
  }
  /**
   * Check if the storage manager is connected
   *
   * @returns true if both backends are connected
   */
  isConnected() {
    return this.connected && this.cache?.isConnected() === true && this.database?.isConnected() === true;
  }
  // Placeholder methods for next phases
  /**
   * Connect to storage backends
   *
   * @returns The connected storage backends
   * @throws {StorageConnectionError} If strict backends fail to connect
   */
  async connect() {
    if (this.connected) {
      this.logger.debug(`${LOG_PREFIXES.MANAGER} Already connected`, {
        cacheType: this.cacheMetadata.type,
        databaseType: this.databaseMetadata.type
      });
      return {
        cache: this.cache,
        database: this.database
      };
    }
    this.connectionAttempts++;
    this.logger.info(
      `${LOG_PREFIXES.MANAGER} Starting connection attempt ${this.connectionAttempts}`
    );
    try {
      const cacheStartTime = Date.now();
      try {
        this.cache = await this.createCacheBackend();
        await this.cache.connect();
        this.cacheMetadata.connectionTime = Date.now() - cacheStartTime;
        this.logger.info(`${LOG_PREFIXES.CACHE} Connected successfully`, {
          type: this.cacheMetadata.type,
          isFallback: this.cacheMetadata.isFallback,
          connectionTime: `${this.cacheMetadata.connectionTime}ms`
        });
      } catch (cacheError) {
        this.logger.warn(`${LOG_PREFIXES.CACHE} Connection failed, attempting fallback`, {
          error: cacheError instanceof Error ? cacheError.message : String(cacheError),
          originalType: this.config.cache.type
        });
        if (this.config.cache.type !== BACKEND_TYPES.IN_MEMORY) {
          const { InMemoryBackend } = await import("./in-memory-3HMN4Z6B.js");
          this.cache = new InMemoryBackend();
          await this.cache.connect();
          this.cacheMetadata.type = BACKEND_TYPES.IN_MEMORY;
          this.cacheMetadata.isFallback = true;
          this.cacheMetadata.connectionTime = Date.now() - cacheStartTime;
          this.logger.info(`${LOG_PREFIXES.CACHE} Connected to fallback backend`, {
            type: this.cacheMetadata.type,
            originalType: this.config.cache.type
          });
        } else {
          throw cacheError;
        }
      }
      const dbStartTime = Date.now();
      try {
        this.database = await this.createDatabaseBackend();
        await this.database.connect();
        this.databaseMetadata.connectionTime = Date.now() - dbStartTime;
        this.logger.info(`${LOG_PREFIXES.DATABASE} Connected successfully`, {
          type: this.databaseMetadata.type,
          isFallback: this.databaseMetadata.isFallback,
          connectionTime: `${this.databaseMetadata.connectionTime}ms`
        });
      } catch (err) {
        this.logger.error("Failed to connect to database backend", {
          error: err instanceof Error ? err.message : String(err),
          type: this.config.database.type,
          stack: err instanceof Error ? err.stack : void 0
        });
        if (this.config.database.type !== BACKEND_TYPES.IN_MEMORY) {
          this.logger.warn(
            `${LOG_PREFIXES.DATABASE} Primary backend failed, attempting fallback to in-memory`,
            {
              error: err instanceof Error ? err.message : String(err),
              originalType: this.config.database.type
            }
          );
          try {
            const { InMemoryBackend } = await import("./in-memory-3HMN4Z6B.js");
            this.database = new InMemoryBackend();
            await this.database.connect();
            this.databaseMetadata.type = BACKEND_TYPES.IN_MEMORY;
            this.databaseMetadata.isFallback = true;
            this.databaseMetadata.connectionTime = Date.now() - dbStartTime;
            this.logger.info(
              `${LOG_PREFIXES.DATABASE} Successfully connected to fallback backend`,
              {
                type: this.databaseMetadata.type,
                originalType: this.config.database.type,
                connectionTime: `${this.databaseMetadata.connectionTime}ms`
              }
            );
          } catch (fallbackError) {
            this.logger.error(`${LOG_PREFIXES.DATABASE} Fallback to in-memory also failed`, {
              error: fallbackError instanceof Error ? fallbackError.message : String(fallbackError),
              originalError: err instanceof Error ? err.message : String(err)
            });
            throw new Error(
              `Database connection failed: Primary (${this.config.database.type}) and fallback (in-memory) both failed`
            );
          }
        } else {
          this.logger.error(
            `${LOG_PREFIXES.DATABASE} In-memory backend failed, no fallback available`
          );
          throw err;
        }
      }
      this.connected = true;
      this.logger.info(`${LOG_PREFIXES.MANAGER} Storage system connected`, {
        cacheBackend: this.cacheMetadata.type,
        databaseBackend: this.databaseMetadata.type,
        totalConnectionTime: `${this.cacheMetadata.connectionTime + this.databaseMetadata.connectionTime}ms`
      });
      return {
        cache: this.cache,
        database: this.database
      };
    } catch (error) {
      this.lastConnectionError = error;
      if (this.cache?.isConnected()) {
        await this.cache.disconnect().catch(
          (err) => this.logger.error(`${LOG_PREFIXES.CACHE} Error during cleanup disconnect`, {
            error: err
          })
        );
      }
      if (this.database?.isConnected()) {
        await this.database.disconnect().catch(
          (err) => this.logger.error(`${LOG_PREFIXES.DATABASE} Error during cleanup disconnect`, {
            error: err
          })
        );
      }
      this.cache = void 0;
      this.database = void 0;
      this.connected = false;
      throw error;
    }
  }
  /**
   * Disconnect from all storage backends
   */
  async disconnect() {
    if (!this.connected) {
      this.logger.debug(`${LOG_PREFIXES.MANAGER} Already disconnected`);
      return;
    }
    this.logger.info(`${LOG_PREFIXES.MANAGER} Disconnecting storage backends`);
    const disconnectPromises = [];
    if (this.cache?.isConnected()) {
      disconnectPromises.push(
        this.cache.disconnect().then(() => {
          this.logger.info(`${LOG_PREFIXES.CACHE} Disconnected successfully`);
        }).catch((error) => {
          this.logger.error(`${LOG_PREFIXES.CACHE} Disconnect error`, { error });
          throw error;
        })
      );
    }
    if (this.database?.isConnected()) {
      disconnectPromises.push(
        this.database.disconnect().then(() => {
          this.logger.info(`${LOG_PREFIXES.DATABASE} Disconnected successfully`);
        }).catch((error) => {
          this.logger.error(`${LOG_PREFIXES.DATABASE} Disconnect error`, { error });
          throw error;
        })
      );
    }
    try {
      await Promise.race([
        Promise.all(disconnectPromises),
        new Promise(
          (_, reject) => setTimeout(() => reject(new Error("Disconnect timeout")), TIMEOUTS.SHUTDOWN)
        )
      ]);
    } finally {
      this.cache = void 0;
      this.database = void 0;
      this.connected = false;
      this.cacheMetadata = {
        type: "unknown",
        isFallback: false,
        connectionTime: 0
      };
      this.databaseMetadata = {
        type: "unknown",
        isFallback: false,
        connectionTime: 0
      };
      this.logger.info(`${LOG_PREFIXES.MANAGER} Storage system disconnected`);
    }
  }
  /**
   * Perform health check on all backends
   *
   * @returns Health check results for each backend
   */
  async healthCheck() {
    throw new Error("Not implemented yet - Phase 5");
  }
  // Private helper methods
  /**
   * Create cache backend based on configuration
   */
  async createCacheBackend() {
    const config = this.config.cache;
    this.logger.debug(`${LOG_PREFIXES.CACHE} Creating backend`, { type: config.type });
    switch (config.type) {
      case BACKEND_TYPES.REDIS: {
        try {
          if (!_StorageManager.redisModule) {
            this.logger.debug(`${LOG_PREFIXES.CACHE} Lazy loading Redis module`);
            const { RedisBackend: RedisBackend2 } = await import("./redis-backend-L3FPDS2O.js");
            _StorageManager.redisModule = RedisBackend2;
          }
          const RedisBackend = _StorageManager.redisModule;
          this.cacheMetadata.type = BACKEND_TYPES.REDIS;
          this.cacheMetadata.isFallback = false;
          return new RedisBackend(config);
        } catch (error) {
          this.logger.debug(`${LOG_PREFIXES.CACHE} Failed to create Redis backend`, {
            error: error instanceof Error ? error.message : String(error)
          });
          throw error;
        }
      }
      case BACKEND_TYPES.IN_MEMORY:
      default: {
        const { InMemoryBackend } = await import("./in-memory-3HMN4Z6B.js");
        this.cacheMetadata.type = BACKEND_TYPES.IN_MEMORY;
        this.cacheMetadata.isFallback = false;
        return new InMemoryBackend();
      }
    }
  }
  /**
   * Create database backend based on configuration
   */
  async createDatabaseBackend() {
    const config = this.config.database;
    this.logger.debug(`${LOG_PREFIXES.DATABASE} Creating backend`, { type: config.type });
    switch (config.type) {
      case BACKEND_TYPES.SQLITE: {
        try {
          if (!_StorageManager.sqliteModule) {
            this.logger.debug(`${LOG_PREFIXES.DATABASE} Lazy loading SQLite module`);
            const { SqliteBackend: SqliteBackend2 } = await import("./sqlite-HIBPDF6T.js");
            _StorageManager.sqliteModule = SqliteBackend2;
          }
          const SqliteBackend = _StorageManager.sqliteModule;
          this.databaseMetadata.type = BACKEND_TYPES.SQLITE;
          this.databaseMetadata.isFallback = false;
          return new SqliteBackend(config);
        } catch (error) {
          this.logger.debug(`${LOG_PREFIXES.DATABASE} Failed to create SQLite backend`, {
            error: error instanceof Error ? error.message : String(error)
          });
          throw error;
        }
      }
      case BACKEND_TYPES.POSTGRES: {
        try {
          if (!_StorageManager.postgresModule) {
            this.logger.debug(`${LOG_PREFIXES.DATABASE} Lazy loading PostgreSQL module`);
            const { PostgresBackend: PostgresBackend2 } = await import("./postgresql-ZTEDK4UU.js");
            _StorageManager.postgresModule = PostgresBackend2;
          }
          const PostgresBackend = _StorageManager.postgresModule;
          this.databaseMetadata.type = BACKEND_TYPES.POSTGRES;
          this.databaseMetadata.isFallback = false;
          return new PostgresBackend(config);
        } catch (error) {
          this.logger.debug(`${LOG_PREFIXES.DATABASE} Failed to create PostgreSQL backend`, {
            error: error instanceof Error ? error.message : String(error)
          });
          throw error;
        }
      }
      case BACKEND_TYPES.IN_MEMORY:
      default: {
        const { InMemoryBackend } = await import("./in-memory-3HMN4Z6B.js");
        this.databaseMetadata.type = BACKEND_TYPES.IN_MEMORY;
        this.databaseMetadata.isFallback = false;
        return new InMemoryBackend();
      }
    }
  }
};

export {
  StorageSchema,
  StorageManager
};
