// @byterover/cipher/dist/src/core/chunk-RPNUGONV.js
import {
  StorageManager
} from "./chunk-MMF5ASL2.js";
import {
  LOG_PREFIXES
} from "./chunk-WZZEPDWF.js";
import {
  createLogger
} from "./chunk-GPOSIAVD.js";
import {
  env,
  init_env
} from "./chunk-7UQA6M3D.js";
import {
  init_esm_shims
} from "./chunk-KFOQGTVJ.js";

// src/core/storage/factory.ts
init_esm_shims();
init_env();
async function createStorageBackends(config) {
  const logger = createLogger({ level: env.CIPHER_LOG_LEVEL });
  logger.debug(`${LOG_PREFIXES.FACTORY} Creating storage system`, {
    cacheType: config.cache.type,
    databaseType: config.database.type
  });
  const manager = new StorageManager(config);
  try {
    const backends = await manager.connect();
    logger.info(`${LOG_PREFIXES.FACTORY} Storage system created successfully`, {
      cache: manager.getInfo().backends.cache,
      database: manager.getInfo().backends.database
    });
    return { manager, backends };
  } catch (error) {
    await manager.disconnect().catch(() => {
    });
    logger.error(`${LOG_PREFIXES.FACTORY} Failed to create storage system`, {
      error: error instanceof Error ? error.message : String(error)
    });
    throw error;
  }
}
async function createDefaultStorage() {
  return createStorageBackends({
    cache: { type: "in-memory" },
    database: { type: "in-memory" }
  });
}
async function createStorageFromEnv() {
  const logger = createLogger({ level: env.CIPHER_LOG_LEVEL });
  const cacheType = env.STORAGE_CACHE_TYPE;
  let cacheConfig;
  if (cacheType === "redis") {
    const port = env.STORAGE_CACHE_PORT || 6379;
    const database = env.STORAGE_CACHE_DATABASE;
    cacheConfig = {
      type: "redis",
      host: env.STORAGE_CACHE_HOST || "localhost",
      port,
      username: env.STORAGE_CACHE_USERNAME,
      password: env.STORAGE_CACHE_PASSWORD,
      database
    };
  } else {
    cacheConfig = { type: "in-memory" };
  }
  const dbType = env.STORAGE_DATABASE_TYPE;
  let dbConfig;
  if (dbType === "sqlite") {
    dbConfig = {
      type: "sqlite",
      path: env.STORAGE_DATABASE_PATH,
      database: env.STORAGE_DATABASE_NAME
    };
  } else {
    if (dbType && dbType !== "in-memory") {
      logger.warn(
        `${LOG_PREFIXES.FACTORY} Database type '${dbType}' not yet supported, using in-memory`
      );
    }
    dbConfig = { type: "in-memory" };
  }
  logger.info(`${LOG_PREFIXES.FACTORY} Creating storage from environment`, {
    cacheType,
    databaseType: dbType
  });
  return createStorageBackends({
    cache: cacheConfig,
    database: dbConfig
  });
}
function isStorageFactory(obj) {
  return typeof obj === "object" && obj !== null && "manager" in obj && "backends" in obj && obj.manager instanceof StorageManager;
}

export {
  createStorageBackends,
  createDefaultStorage,
  createStorageFromEnv,
  isStorageFactory
};
