// @byterover/cipher/dist/src/core/chunk-MNM4W6RD.js
import {
  BACKEND_TYPES,
  ERROR_MESSAGES
} from "./chunk-WZZEPDWF.js";
import {
  StorageError
} from "./chunk-C3KGEYH3.js";
import {
  createLogger
} from "./chunk-GPOSIAVD.js";
import {
  init_esm_shims
} from "./chunk-KFOQGTVJ.js";

// src/core/storage/backend/in-memory.ts
init_esm_shims();
var InMemoryBackend = class {
  // Storage maps
  store = /* @__PURE__ */ new Map();
  lists = /* @__PURE__ */ new Map();
  // Connection state
  connected = false;
  // Cleanup interval
  cleanupInterval;
  cleanupIntervalMs = 6e4;
  // 1 minute
  // Logger
  logger;
  // Statistics
  stats = {
    hits: 0,
    misses: 0,
    sets: 0,
    deletes: 0,
    expirations: 0
  };
  constructor() {
    this.logger = createLogger({ level: process.env.LOG_LEVEL || "info" });
  }
  // Connection Management
  /**
   * Connect to the in-memory backend
   *
   * For in-memory backend, this just sets the connected flag and
   * starts the cleanup interval for expired entries.
   */
  async connect() {
    if (this.connected) {
      this.logger.debug("InMemoryBackend already connected");
      return;
    }
    this.connected = true;
    this.startCleanupInterval();
    this.logger.info("InMemoryBackend connected", {
      cleanupInterval: `${this.cleanupIntervalMs}ms`
    });
  }
  /**
   * Disconnect from the in-memory backend
   *
   * Clears all data and stops the cleanup interval.
   */
  async disconnect() {
    if (!this.connected) {
      return;
    }
    this.stopCleanupInterval();
    this.clear();
    this.connected = false;
    this.logger.info("InMemoryBackend disconnected", {
      stats: this.stats
    });
  }
  /**
   * Check if backend is connected
   */
  isConnected() {
    return this.connected;
  }
  /**
   * Get backend type identifier
   */
  getBackendType() {
    return BACKEND_TYPES.IN_MEMORY;
  }
  // Basic Operations (shared by both interfaces)
  /**
   * Get a value by key
   *
   * Checks expiration and removes expired entries.
   */
  async get(key) {
    this.checkConnection();
    const entry = this.store.get(key);
    if (!entry) {
      this.stats.misses++;
      return void 0;
    }
    if (entry.expires && Date.now() > entry.expires) {
      this.store.delete(key);
      this.stats.expirations++;
      this.stats.misses++;
      this.logger.debug("Entry expired", { key });
      return void 0;
    }
    this.stats.hits++;
    return this.cloneValue(entry.value);
  }
  /**
   * Set a value with optional TTL (for CacheBackend)
   */
  async set(key, value, ttlSeconds) {
    this.checkConnection();
    const entry = {
      value: this.cloneValue(value)
    };
    if (ttlSeconds && ttlSeconds > 0) {
      entry.expires = Date.now() + ttlSeconds * 1e3;
    }
    this.store.set(key, entry);
    this.stats.sets++;
    this.logger.debug("Value set", {
      key,
      ttl: ttlSeconds,
      expires: entry.expires
    });
  }
  /**
   * Delete a value by key
   */
  async delete(key) {
    this.checkConnection();
    const deleted = this.store.delete(key);
    this.lists.delete(key);
    if (deleted) {
      this.stats.deletes++;
    }
    this.logger.debug("Value deleted", { key, deleted });
  }
  // DatabaseBackend specific methods
  /**
   * List all keys matching a prefix
   */
  async list(prefix) {
    this.checkConnection();
    const keys = [];
    const now = Date.now();
    for (const [key, entry] of Array.from(this.store.entries())) {
      if (entry.expires && now > entry.expires) {
        continue;
      }
      if (key.startsWith(prefix)) {
        keys.push(key);
      }
    }
    for (const key of Array.from(this.lists.keys())) {
      if (key.startsWith(prefix) && !keys.includes(key)) {
        keys.push(key);
      }
    }
    this.logger.debug("Listed keys", { prefix, count: keys.length });
    return keys;
  }
  /**
   * Append an item to a list
   */
  async append(key, item) {
    this.checkConnection();
    let list = this.lists.get(key);
    if (!list) {
      list = [];
      this.lists.set(key, list);
    }
    list.push(this.cloneValue(item));
    this.logger.debug("Item appended to list", {
      key,
      listSize: list.length
    });
  }
  /**
   * Get a range of items from a list
   */
  async getRange(key, start, count) {
    this.checkConnection();
    const list = this.lists.get(key);
    if (!list) {
      return [];
    }
    if (start < 0) {
      start = Math.max(0, list.length + start);
    }
    const end = Math.min(start + count, list.length);
    const items = list.slice(start, end);
    this.logger.debug("Retrieved range from list", {
      key,
      start,
      count,
      returned: items.length,
      totalSize: list.length
    });
    return items.map((item) => this.cloneValue(item));
  }
  // Helper Methods
  /**
   * Check if backend is connected
   * @throws {StorageError} If not connected
   */
  checkConnection() {
    if (!this.connected) {
      throw new StorageError(
        ERROR_MESSAGES.NOT_CONNECTED,
        "operation",
        new Error("InMemoryBackend is not connected")
      );
    }
  }
  /**
   * Clone a value to prevent reference issues
   *
   * Uses JSON serialization for deep cloning.
   * This also ensures consistency with network-based backends
   * that serialize data.
   */
  cloneValue(value) {
    try {
      return JSON.parse(JSON.stringify(value));
    } catch (error) {
      throw new StorageError(ERROR_MESSAGES.SERIALIZATION_ERROR, "clone", error);
    }
  }
  /**
   * Clear all stored data
   */
  clear() {
    this.store.clear();
    this.lists.clear();
    this.logger.debug("All data cleared");
  }
  /**
   * Start the cleanup interval for expired entries
   */
  startCleanupInterval() {
    if (this.cleanupInterval) {
      return;
    }
    this.cleanupInterval = setInterval(() => {
      this.cleanupExpired();
    }, this.cleanupIntervalMs);
    if (this.cleanupInterval.unref) {
      this.cleanupInterval.unref();
    }
  }
  /**
   * Stop the cleanup interval
   */
  stopCleanupInterval() {
    if (this.cleanupInterval) {
      clearInterval(this.cleanupInterval);
      this.cleanupInterval = void 0;
    }
  }
  /**
   * Remove expired entries from the store
   */
  cleanupExpired() {
    const now = Date.now();
    let cleaned = 0;
    for (const [key, entry] of Array.from(this.store.entries())) {
      if (entry.expires && now > entry.expires) {
        this.store.delete(key);
        cleaned++;
        this.stats.expirations++;
      }
    }
    if (cleaned > 0) {
      this.logger.debug("Cleaned up expired entries", {
        cleaned,
        remaining: this.store.size
      });
    }
  }
  // Additional utility methods
  /**
   * Get storage statistics
   */
  getStats() {
    return { ...this.stats };
  }
  /**
   * Get current storage size
   */
  getSize() {
    return {
      keys: this.store.size,
      lists: this.lists.size,
      total: this.store.size + this.lists.size
    };
  }
  /**
   * Manually trigger cleanup of expired entries
   */
  async cleanup() {
    const before = this.store.size;
    this.cleanupExpired();
    const after = this.store.size;
    return before - after;
  }
};

export {
  InMemoryBackend
};
