// @byterover/cipher/dist/src/core/chunk-4JSJAC4K.js
import {
  init_esm_shims
} from "./chunk-KFOQGTVJ.js";

// src/core/vector_storage/constants.ts
init_esm_shims();
var LOG_PREFIXES = {
  MANAGER: "[VectorStoreManager]",
  BACKEND: "[VectorStoreBackend]",
  FACTORY: "[VectorStoreFactory]",
  SEARCH: "[VectorStore:Search]",
  INDEX: "[VectorStore:Index]",
  QDRANT: "[VectorStore:Qdrant]",
  MILVUS: "[VectorStore:Milvus]",
  CHROMA: "[VectorStore:Chroma]",
  PINECONE: "[VectorStore:Pinecone]",
  PGVECTOR: "[VectorStore:PgVector]",
  FAISS: "[VectorStore:Faiss]",
  REDIS: "[VectorStore:Redis]",
  WEAVIATE: "[VectorStore:Weaviate]",
  MEMORY: "[VectorStore:Memory]"
};
var ERROR_MESSAGES = {
  // Connection errors
  CONNECTION_FAILED: "Failed to connect to vector store backend",
  ALREADY_CONNECTED: "Vector store is already connected",
  NOT_CONNECTED: "Vector store is not connected",
  // Backend errors
  BACKEND_NOT_FOUND: "Vector store backend not found",
  INVALID_BACKEND_TYPE: "Invalid backend type specified",
  MODULE_LOAD_FAILED: "Failed to load backend module",
  // Operation errors
  INVALID_DIMENSION: "Vector dimension mismatch",
  COLLECTION_NOT_FOUND: "Collection does not exist",
  VECTOR_NOT_FOUND: "Vector not found",
  BATCH_SIZE_EXCEEDED: "Batch size exceeds maximum allowed",
  SEARCH_FAILED: "Vector search operation failed",
  // Configuration errors
  INVALID_CONFIG: "Invalid vector store configuration",
  MISSING_REQUIRED_CONFIG: "Missing required configuration",
  INVALID_COLLECTION_NAME: "Invalid collection name"
};
var TIMEOUTS = {
  CONNECTION: 3e4,
  // 30 seconds (vector stores may take longer)
  OPERATION: 6e4,
  // 60 seconds
  SEARCH: 1e4,
  // 10 seconds
  SHUTDOWN: 5e3
  // 5 seconds
};
var BACKEND_TYPES = {
  QDRANT: "qdrant",
  PINECONE: "pinecone",
  WEAVIATE: "weaviate",
  CHROMA: "chroma",
  IN_MEMORY: "in-memory",
  MILVUS: "milvus",
  PGVECTOR: "pgvector",
  FAISS: "faiss",
  REDIS: "redis"
};
var DEFAULTS = {
  // Search defaults
  SEARCH_LIMIT: 10,
  SEARCH_SCORE_THRESHOLD: 0,
  // Vector defaults
  DIMENSION: 1536,
  // OpenAI ada-002 dimension
  MAX_BATCH_SIZE: 100,
  // Connection defaults
  MAX_RETRIES: 3,
  RETRY_DELAY: 1e3,
  // 1 second
  MAX_CONNECTIONS: 10,
  IDLE_TIMEOUT: 3e4,
  // 30 seconds
  // Qdrant defaults
  QDRANT_PORT: 6333,
  QDRANT_GRPC_PORT: 6334,
  QDRANT_DISTANCE: "Cosine",
  // ChromaDB defaults
  CHROMA_PORT: 8e3,
  CHROMA_DISTANCE: "cosine",
  // Pinecone defaults
  PINECONE_REGION: "us-east-1",
  PINECONE_PROVIDER: "aws",
  // PGVector defaults
  PGVECTOR_INDEXTYPE: "hnsw",
  PGVECTOR_INDEXMETRIC: "vector_l2_ops"
};
var DISTANCE_METRICS = {
  COSINE: "Cosine",
  EUCLIDEAN: "Euclidean",
  DOT_PRODUCT: "Dot",
  MANHATTAN: "Manhattan"
};

export {
  LOG_PREFIXES,
  ERROR_MESSAGES,
  TIMEOUTS,
  BACKEND_TYPES,
  DEFAULTS,
  DISTANCE_METRICS
};
