// @byterover/cipher/dist/src/core/chunk-TKUZPR6R.js
import {
  init_esm_shims
} from "./chunk-KFOQGTVJ.js";

// src/core/brain/tools/definitions/knowledge_graph/query-graph.ts
init_esm_shims();
var queryGraphTool = {
  name: "query_graph",
  category: "knowledge_graph",
  internal: true,
  description: "Execute custom queries against the knowledge graph.",
  version: "1.0.0",
  parameters: {
    type: "object",
    properties: {
      query: {
        type: "string",
        description: "Query string (Cypher-like syntax or structured query)"
      },
      queryType: {
        type: "string",
        enum: ["node", "edge", "path", "cypher"],
        description: "Type of query to execute",
        default: "cypher"
      },
      parameters: {
        type: "object",
        description: "Query parameters for parameterized queries",
        additionalProperties: true
      },
      limit: {
        type: "number",
        description: "Maximum number of results to return",
        minimum: 1,
        maximum: 1e3,
        default: 100
      }
    },
    required: ["query"]
  },
  handler: async (args, context) => {
    try {
      const queryType = args.queryType || "cypher";
      const limit = args.limit !== void 0 ? args.limit : 100;
      if (!args.query || typeof args.query !== "string" || args.query.trim().length === 0) {
        throw new Error("Query is required and must be a non-empty string");
      }
      if (limit < 1 || limit > 1e3) {
        return {
          success: false,
          message: "Limit must be between 1 and 1000",
          results: null,
          timestamp: (/* @__PURE__ */ new Date()).toISOString()
        };
      }
      const kgManager = context?.services?.knowledgeGraphManager;
      if (!kgManager) {
        return {
          success: false,
          message: "KnowledgeGraphManager not available in context.services",
          results: null,
          timestamp: (/* @__PURE__ */ new Date()).toISOString()
        };
      }
      const graph = kgManager.getGraph();
      if (!graph) {
        return {
          success: false,
          message: "Knowledge graph backend is not connected",
          results: null,
          timestamp: (/* @__PURE__ */ new Date()).toISOString()
        };
      }
      const graphQuery = {
        type: queryType,
        parameters: args.parameters || {},
        limit
      };
      if (queryType === "cypher") {
        graphQuery.query = args.query.trim();
      } else {
        const pattern = {};
        if (args.query.includes("labels:")) {
          const labelsMatch = args.query.match(/labels:\s*\[([^\]]+)\]/);
          if (labelsMatch && labelsMatch[1]) {
            pattern.labels = labelsMatch[1].split(",").map((l) => l.trim().replace(/['"]/g, ""));
          }
        }
        if (args.query.includes("type:")) {
          const typeMatch = args.query.match(/type:\s*(?:['"]([^'"]+)['"]|([A-Z_][A-Z0-9_]*))/);
          if (typeMatch) {
            pattern.type = typeMatch[1] || typeMatch[2];
          }
        }
        graphQuery.pattern = pattern;
      }
      const result = await graph.query(graphQuery);
      return {
        success: true,
        results: result,
        query: {
          original: args.query,
          type: queryType,
          parameters: args.parameters || {},
          limit
        },
        message: "Query executed",
        timestamp: (/* @__PURE__ */ new Date()).toISOString()
      };
    } catch (error) {
      const errorMessage = error instanceof Error ? error.message : String(error);
      return {
        success: false,
        error: errorMessage,
        results: {
          nodes: [],
          edges: [],
          metadata: {
            totalCount: 0,
            queryType: args.queryType || "cypher"
          }
        },
        timestamp: (/* @__PURE__ */ new Date()).toISOString()
      };
    }
  }
};

export {
  queryGraphTool
};
