// @byterover/cipher/dist/src/core/chunk-RVR2H76A.js
import {
  logger
} from "./chunk-GPOSIAVD.js";
import {
  init_esm_shims
} from "./chunk-KFOQGTVJ.js";

// src/core/brain/tools/definitions/knowledge_graph/search-graph.ts
init_esm_shims();
var SEARCH_GRAPH_TOOL = {
  type: "function",
  function: {
    name: "search_graph",
    description: "Search for nodes and edges in the knowledge graph with optional filters.",
    parameters: {
      type: "object",
      properties: {
        searchType: {
          type: "string",
          enum: ["nodes", "edges", "both"],
          description: "Type of entities to search for",
          default: "both"
        },
        nodeLabels: {
          type: "array",
          description: "Filter by node labels (only applicable when searching nodes)",
          items: {
            type: "string"
          }
        },
        edgeTypes: {
          type: "array",
          description: "Filter by edge types (only applicable when searching edges)",
          items: {
            type: "string"
          }
        },
        // Legacy parameter for backward compatibility
        labels: {
          type: "array",
          description: "Filter by node labels or edge types (deprecated, use nodeLabels/edgeTypes)",
          items: {
            type: "string"
          }
        },
        properties: {
          type: "object",
          description: "Filter by properties (key-value pairs)",
          additionalProperties: true
        },
        textSearch: {
          type: "string",
          description: "Full-text search in properties and names"
        },
        limit: {
          type: "number",
          description: "Maximum number of results to return",
          minimum: 1,
          maximum: 1e3,
          default: 50
        }
      },
      additionalProperties: false
    }
  }
};
var searchGraphTool = {
  name: "search_graph",
  category: "knowledge_graph",
  internal: true,
  description: "Search for nodes and edges in the knowledge graph with optional filters.",
  version: "1.0.0",
  parameters: {
    type: "object",
    properties: {
      searchType: {
        type: "string",
        enum: ["nodes", "edges", "both"],
        description: "Type of entities to search for"
      },
      nodeLabels: {
        type: "array",
        description: "Filter by node labels (only applicable when searching nodes)",
        items: {
          type: "string"
        }
      },
      edgeTypes: {
        type: "array",
        description: "Filter by edge types (only applicable when searching edges)",
        items: {
          type: "string"
        }
      },
      // Legacy parameter for backward compatibility
      labels: {
        type: "array",
        description: "Filter by node labels or edge types (deprecated, use nodeLabels/edgeTypes)",
        items: {
          type: "string"
        }
      },
      properties: {
        type: "object",
        description: "Filter by properties (key-value pairs)",
        additionalProperties: true
      },
      textSearch: {
        type: "string",
        description: "Full-text search in properties and names"
      },
      limit: {
        type: "number",
        description: "Maximum number of results to return",
        minimum: 1,
        maximum: 1e3
      }
    },
    required: []
  },
  handler: async (args, context) => {
    try {
      logger.debug("SearchGraph: Processing search request", {
        searchType: args.searchType,
        hasNodeLabels: !!args.nodeLabels,
        hasEdgeTypes: !!args.edgeTypes,
        hasLegacyLabels: !!args.labels,
        hasProperties: !!args.properties,
        hasTextSearch: !!args.textSearch
      });
      const searchType = args.searchType || "both";
      const limit = args.limit !== void 0 ? args.limit : 50;
      if (limit < 1 || limit > 1e3) {
        return {
          success: false,
          message: "Limit must be between 1 and 1000",
          results: null,
          timestamp: (/* @__PURE__ */ new Date()).toISOString()
        };
      }
      const kgManager = context?.services?.knowledgeGraphManager;
      if (!kgManager) {
        return {
          success: false,
          message: "KnowledgeGraphManager not available in context.services",
          results: null,
          timestamp: (/* @__PURE__ */ new Date()).toISOString()
        };
      }
      const graph = kgManager.getGraph();
      if (!graph) {
        return {
          success: false,
          message: "Knowledge graph backend is not connected",
          results: null,
          timestamp: (/* @__PURE__ */ new Date()).toISOString()
        };
      }
      const nodeLabelsToUse = args.nodeLabels || (searchType === "nodes" ? args.labels : void 0);
      const edgeTypesToUse = args.edgeTypes || (searchType === "edges" ? args.labels : void 0);
      let nodes = [];
      let edges = [];
      const startTime = Date.now();
      if (searchType === "nodes" || searchType === "both") {
        const nodeFilters = buildNodeFilters(args.properties, args.textSearch);
        nodes = await graph.findNodes(nodeFilters, nodeLabelsToUse, limit);
        logger.debug("SearchGraph: Found nodes", {
          count: nodes.length,
          labels: nodeLabelsToUse,
          filters: nodeFilters
        });
      }
      if (searchType === "edges" || searchType === "both") {
        const edgeFilters = buildEdgeFilters(args.properties, args.textSearch);
        if (edgeTypesToUse && edgeTypesToUse.length > 0) {
          const edgeResults = await Promise.all(
            edgeTypesToUse.map(
              (edgeType) => graph.findEdges(edgeFilters, edgeType, Math.ceil(limit / edgeTypesToUse.length))
            )
          );
          edges = edgeResults.flat().slice(0, limit);
        } else {
          edges = await graph.findEdges(edgeFilters, void 0, limit);
        }
        logger.debug("SearchGraph: Found edges", {
          count: edges.length,
          types: edgeTypesToUse,
          filters: edgeFilters
        });
      }
      const executionTime = Date.now() - startTime;
      return {
        success: true,
        results: {
          nodes,
          edges,
          totalCount: nodes.length + edges.length,
          searchMetadata: {
            searchType,
            appliedFilters: {
              nodeLabels: nodeLabelsToUse || [],
              edgeTypes: edgeTypesToUse || [],
              legacyLabels: args.labels || [],
              // For backward compatibility tracking
              properties: args.properties || {},
              textSearch: args.textSearch
            },
            limit,
            executionTime
          }
        },
        message: `Search completed - found ${nodes.length} nodes and ${edges.length} edges`,
        timestamp: (/* @__PURE__ */ new Date()).toISOString()
      };
    } catch (error) {
      const errorMessage = error instanceof Error ? error.message : String(error);
      logger.error("SearchGraph: Search failed", {
        error: errorMessage,
        searchType: args.searchType,
        limit: args.limit
      });
      return {
        success: false,
        message: `Search failed: ${errorMessage}`,
        results: {
          nodes: [],
          edges: [],
          totalCount: 0
        },
        timestamp: (/* @__PURE__ */ new Date()).toISOString()
      };
    }
  }
};
function buildNodeFilters(properties, textSearch) {
  const filters = {};
  if (properties && Object.keys(properties).length > 0) {
    Object.assign(filters, properties);
  }
  if (textSearch && textSearch.trim().length > 0) {
    filters.name = textSearch.trim();
  }
  return Object.keys(filters).length > 0 ? filters : void 0;
}
function buildEdgeFilters(properties, textSearch) {
  const filters = {};
  if (properties && Object.keys(properties).length > 0) {
    Object.assign(filters, properties);
  }
  if (textSearch && textSearch.trim().length > 0) {
    filters.description = textSearch.trim();
  }
  return Object.keys(filters).length > 0 ? filters : void 0;
}

export {
  SEARCH_GRAPH_TOOL,
  searchGraphTool
};
