// @byterover/cipher/dist/src/core/chunk-ZUCXE4PW.js
import {
  logger
} from "./chunk-GPOSIAVD.js";
import {
  init_esm_shims
} from "./chunk-KFOQGTVJ.js";

// src/core/brain/tools/definitions/knowledge_graph/update-node.ts
init_esm_shims();
async function updateNodeHandler(args, context) {
  try {
    logger.info("UpdateNode: Processing update request", {
      nodeId: args.nodeId,
      hasProperties: !!args.properties,
      propertiesCount: args.properties ? Object.keys(args.properties).length : 0,
      hasLabels: !!args.labels,
      labelsCount: args.labels ? args.labels.length : 0
    });
    if (!args.nodeId || typeof args.nodeId !== "string" || args.nodeId.trim().length === 0) {
      return {
        success: false,
        message: "Node ID must be a non-empty string",
        timestamp: (/* @__PURE__ */ new Date()).toISOString()
      };
    }
    if (!args.properties || typeof args.properties !== "object" || Object.keys(args.properties).length === 0) {
      return {
        success: false,
        message: "Properties must be provided as a non-empty object",
        timestamp: (/* @__PURE__ */ new Date()).toISOString()
      };
    }
    if (args.labels && (!Array.isArray(args.labels) || args.labels.length === 0)) {
      return {
        success: false,
        message: "Labels must be a non-empty array when provided",
        timestamp: (/* @__PURE__ */ new Date()).toISOString()
      };
    }
    const kgManager = context?.services?.knowledgeGraphManager;
    if (!kgManager) {
      return {
        success: false,
        message: "KnowledgeGraphManager not available in context.services",
        timestamp: (/* @__PURE__ */ new Date()).toISOString()
      };
    }
    const graph = kgManager.getGraph();
    if (!graph) {
      return {
        success: false,
        message: "Knowledge graph backend is not connected",
        timestamp: (/* @__PURE__ */ new Date()).toISOString()
      };
    }
    await graph.updateNode(args.nodeId.trim(), args.properties, args.labels);
    logger.info("UpdateNode: Node updated successfully", {
      nodeId: args.nodeId,
      updatedProperties: Object.keys(args.properties),
      updatedLabels: args.labels
    });
    return {
      success: true,
      message: `Node '${args.nodeId}' updated in knowledge graph`,
      nodeId: args.nodeId,
      timestamp: (/* @__PURE__ */ new Date()).toISOString()
    };
  } catch (error) {
    const errorMessage = error instanceof Error ? error.message : String(error);
    logger.error("UpdateNode: Failed to update node", {
      error: errorMessage,
      nodeId: args.nodeId,
      properties: args.properties ? Object.keys(args.properties) : []
    });
    return {
      success: false,
      message: `Failed to update node: ${errorMessage}`,
      timestamp: (/* @__PURE__ */ new Date()).toISOString()
    };
  }
}
var updateNodeTool = {
  name: "update_node",
  description: "Update an existing node in the knowledge graph",
  category: "knowledge_graph",
  internal: true,
  version: "1.0.0",
  handler: updateNodeHandler,
  parameters: {
    type: "object",
    properties: {
      nodeId: {
        type: "string",
        description: "ID of the node to update"
      },
      properties: {
        type: "object",
        description: "Properties to update (will be merged with existing properties)",
        additionalProperties: true
      },
      labels: {
        type: "array",
        items: { type: "string" },
        description: "Optional new labels to set for the node"
      }
    },
    required: ["nodeId", "properties"]
  }
};

export {
  updateNodeTool
};
