// @byterover/cipher/dist/src/core/chunk-TCBVDNOG.js
import {
  DEFAULTS,
  ERROR_MESSAGES,
  EdgeNotFoundError,
  GraphValidationError,
  InvalidQueryError,
  KnowledgeGraphConnectionError,
  LOG_PREFIXES,
  NodeNotFoundError
} from "./chunk-Z6KHKTWP.js";
import {
  createLogger
} from "./chunk-GPOSIAVD.js";
import {
  init_esm_shims
} from "./chunk-KFOQGTVJ.js";

// src/core/knowledge_graph/backend/in-memory.ts
init_esm_shims();
var InMemoryBackend = class {
  config;
  logger;
  connected = false;
  // Core storage
  nodes = /* @__PURE__ */ new Map();
  edges = /* @__PURE__ */ new Map();
  // Adjacency lists for graph traversal
  outgoingEdges = /* @__PURE__ */ new Map();
  // node ID -> set of outgoing edge IDs
  incomingEdges = /* @__PURE__ */ new Map();
  // node ID -> set of incoming edge IDs
  // Indexes for performance
  nodeIndex = {
    labels: /* @__PURE__ */ new Map(),
    edgeTypes: /* @__PURE__ */ new Map(),
    properties: /* @__PURE__ */ new Map()
  };
  edgeIndex = {
    labels: /* @__PURE__ */ new Map(),
    edgeTypes: /* @__PURE__ */ new Map(),
    properties: /* @__PURE__ */ new Map()
  };
  // Statistics
  stats = {
    nodeCount: 0,
    edgeCount: 0,
    operations: 0,
    queriesExecuted: 0,
    lastOperation: 0
  };
  constructor(config) {
    this.config = {
      ...config,
      maxNodes: config.maxNodes ?? DEFAULTS.MAX_NODES,
      maxEdges: config.maxEdges ?? DEFAULTS.MAX_EDGES,
      enableIndexing: config.enableIndexing ?? true,
      enableGarbageCollection: config.enableGarbageCollection ?? false
    };
    this.logger = createLogger({
      level: "debug"
    });
    this.logger.debug(`${LOG_PREFIXES.IN_MEMORY} Initialized with config:`, {
      maxNodes: this.config.maxNodes,
      maxEdges: this.config.maxEdges,
      enableIndexing: this.config.enableIndexing,
      enableGarbageCollection: this.config.enableGarbageCollection
    });
  }
  // Connection management
  async connect() {
    if (this.connected) {
      this.logger.debug(`${LOG_PREFIXES.IN_MEMORY} Already connected`);
      return;
    }
    try {
      this.clear();
      this.connected = true;
      this.logger.info(`${LOG_PREFIXES.IN_MEMORY} Connected successfully`);
    } catch (error) {
      const connectionError = new KnowledgeGraphConnectionError(
        ERROR_MESSAGES.CONNECTION_FAILED,
        "in-memory",
        error
      );
      this.logger.error(`${LOG_PREFIXES.IN_MEMORY} Connection failed:`, connectionError);
      throw connectionError;
    }
  }
  async disconnect() {
    if (!this.connected) {
      this.logger.debug(`${LOG_PREFIXES.IN_MEMORY} Already disconnected`);
      return;
    }
    try {
      if (this.config.enableGarbageCollection) {
        this.clear();
      }
      this.connected = false;
      this.logger.info(`${LOG_PREFIXES.IN_MEMORY} Disconnected successfully`);
    } catch (error) {
      this.logger.error(`${LOG_PREFIXES.IN_MEMORY} Error during disconnect:`, error);
      throw error;
    }
  }
  isConnected() {
    return this.connected;
  }
  getBackendType() {
    return "in-memory";
  }
  // Node operations
  async addNode(node) {
    this.ensureConnected();
    this.validateNode(node);
    if (this.nodes.has(node.id)) {
      throw new GraphValidationError(`Node with ID '${node.id}' already exists`, "node");
    }
    if (this.stats.nodeCount >= this.config.maxNodes) {
      throw new GraphValidationError(
        `Maximum number of nodes (${this.config.maxNodes}) exceeded`,
        "node"
      );
    }
    const entry = {
      data: this.deepClone(node),
      createdAt: Date.now(),
      updatedAt: Date.now()
    };
    this.nodes.set(node.id, entry);
    this.stats.nodeCount++;
    this.stats.operations++;
    this.stats.lastOperation = Date.now();
    if (this.config.enableIndexing) {
      this.updateNodeIndex(node.id, node, "add");
    }
    this.outgoingEdges.set(node.id, /* @__PURE__ */ new Set());
    this.incomingEdges.set(node.id, /* @__PURE__ */ new Set());
    this.logger.debug(`${LOG_PREFIXES.IN_MEMORY} Added node:`, {
      id: node.id,
      labels: node.labels
    });
  }
  async addNodes(nodes) {
    this.ensureConnected();
    if (this.stats.nodeCount + nodes.length > this.config.maxNodes) {
      throw new GraphValidationError(
        `Adding ${nodes.length} nodes would exceed maximum (${this.config.maxNodes})`,
        "node"
      );
    }
    for (const node of nodes) {
      this.validateNode(node);
      if (this.nodes.has(node.id)) {
        throw new GraphValidationError(`Node with ID '${node.id}' already exists`, "node");
      }
    }
    for (const node of nodes) {
      await this.addNode(node);
    }
    this.logger.debug(`${LOG_PREFIXES.IN_MEMORY} Added ${nodes.length} nodes`);
  }
  async getNode(nodeId) {
    this.ensureConnected();
    const entry = this.nodes.get(nodeId);
    if (!entry) {
      return null;
    }
    return this.deepClone(entry.data);
  }
  async updateNode(nodeId, properties, labels) {
    this.ensureConnected();
    const entry = this.nodes.get(nodeId);
    if (!entry) {
      throw new NodeNotFoundError(ERROR_MESSAGES.NODE_NOT_FOUND, nodeId);
    }
    if (this.config.enableIndexing) {
      this.updateNodeIndex(nodeId, entry.data, "remove");
    }
    const updatedNode = {
      ...entry.data,
      properties: { ...entry.data.properties, ...properties },
      labels: labels ?? entry.data.labels
    };
    entry.data = updatedNode;
    entry.updatedAt = Date.now();
    this.stats.operations++;
    this.stats.lastOperation = Date.now();
    if (this.config.enableIndexing) {
      this.updateNodeIndex(nodeId, updatedNode, "add");
    }
    this.logger.debug(`${LOG_PREFIXES.IN_MEMORY} Updated node:`, { id: nodeId });
  }
  async deleteNode(nodeId) {
    this.ensureConnected();
    const entry = this.nodes.get(nodeId);
    if (!entry) {
      return;
    }
    const outgoing = this.outgoingEdges.get(nodeId) || /* @__PURE__ */ new Set();
    const incoming = this.incomingEdges.get(nodeId) || /* @__PURE__ */ new Set();
    for (const edgeId of [...outgoing, ...incoming]) {
      await this.deleteEdge(edgeId);
    }
    if (this.config.enableIndexing) {
      this.updateNodeIndex(nodeId, entry.data, "remove");
    }
    this.nodes.delete(nodeId);
    this.outgoingEdges.delete(nodeId);
    this.incomingEdges.delete(nodeId);
    this.stats.nodeCount--;
    this.stats.operations++;
    this.stats.lastOperation = Date.now();
    this.logger.debug(`${LOG_PREFIXES.IN_MEMORY} Deleted node:`, { id: nodeId });
  }
  async findNodes(filters, labels, limit) {
    this.ensureConnected();
    const maxResults = limit ?? DEFAULTS.QUERY_LIMIT;
    const results = [];
    let candidateIds = /* @__PURE__ */ new Set();
    if (labels && labels.length > 0 && this.config.enableIndexing) {
      for (const label of labels) {
        const labelIds = this.nodeIndex.labels.get(label) || /* @__PURE__ */ new Set();
        if (candidateIds.size === 0) {
          candidateIds = new Set(labelIds);
        } else {
          candidateIds = new Set([...candidateIds].filter((id) => labelIds.has(id)));
        }
      }
    } else {
      candidateIds = new Set(this.nodes.keys());
    }
    for (const nodeId of candidateIds) {
      if (results.length >= maxResults) break;
      const entry = this.nodes.get(nodeId);
      if (!entry) continue;
      const node = entry.data;
      if (labels && labels.length > 0) {
        if (!labels.some((label) => node.labels.includes(label))) {
          continue;
        }
      }
      if (filters && !this.matchesNodeFilters(node, filters)) {
        continue;
      }
      results.push(this.deepClone(node));
    }
    this.stats.queriesExecuted++;
    this.logger.debug(`${LOG_PREFIXES.IN_MEMORY} Found ${results.length} nodes`);
    return results;
  }
  // Edge operations
  async addEdge(edge) {
    this.ensureConnected();
    this.validateEdge(edge);
    if (this.edges.has(edge.id)) {
      throw new GraphValidationError(`Edge with ID '${edge.id}' already exists`, "edge");
    }
    if (this.stats.edgeCount >= this.config.maxEdges) {
      throw new GraphValidationError(
        `Maximum number of edges (${this.config.maxEdges}) exceeded`,
        "edge"
      );
    }
    if (!this.nodes.has(edge.startNodeId)) {
      throw new GraphValidationError(`Start node '${edge.startNodeId}' not found`, "edge");
    }
    if (!this.nodes.has(edge.endNodeId)) {
      throw new GraphValidationError(`End node '${edge.endNodeId}' not found`, "edge");
    }
    const entry = {
      data: this.deepClone(edge),
      createdAt: Date.now(),
      updatedAt: Date.now()
    };
    this.edges.set(edge.id, entry);
    this.stats.edgeCount++;
    this.stats.operations++;
    this.stats.lastOperation = Date.now();
    this.outgoingEdges.get(edge.startNodeId).add(edge.id);
    this.incomingEdges.get(edge.endNodeId).add(edge.id);
    if (this.config.enableIndexing) {
      this.updateEdgeIndex(edge.id, edge, "add");
    }
    this.logger.debug(`${LOG_PREFIXES.IN_MEMORY} Added edge:`, {
      id: edge.id,
      type: edge.type,
      from: edge.startNodeId,
      to: edge.endNodeId
    });
  }
  async addEdges(edges) {
    this.ensureConnected();
    if (this.stats.edgeCount + edges.length > this.config.maxEdges) {
      throw new GraphValidationError(
        `Adding ${edges.length} edges would exceed maximum (${this.config.maxEdges})`,
        "edge"
      );
    }
    for (const edge of edges) {
      this.validateEdge(edge);
      if (this.edges.has(edge.id)) {
        throw new GraphValidationError(`Edge with ID '${edge.id}' already exists`, "edge");
      }
    }
    for (const edge of edges) {
      await this.addEdge(edge);
    }
    this.logger.debug(`${LOG_PREFIXES.IN_MEMORY} Added ${edges.length} edges`);
  }
  async getEdge(edgeId) {
    this.ensureConnected();
    const entry = this.edges.get(edgeId);
    if (!entry) {
      return null;
    }
    return this.deepClone(entry.data);
  }
  async updateEdge(edgeId, properties) {
    this.ensureConnected();
    const entry = this.edges.get(edgeId);
    if (!entry) {
      throw new EdgeNotFoundError(ERROR_MESSAGES.EDGE_NOT_FOUND, edgeId);
    }
    if (this.config.enableIndexing) {
      this.updateEdgeIndex(edgeId, entry.data, "remove");
    }
    const updatedEdge = {
      ...entry.data,
      properties: { ...entry.data.properties, ...properties }
    };
    entry.data = updatedEdge;
    entry.updatedAt = Date.now();
    this.stats.operations++;
    this.stats.lastOperation = Date.now();
    if (this.config.enableIndexing) {
      this.updateEdgeIndex(edgeId, updatedEdge, "add");
    }
    this.logger.debug(`${LOG_PREFIXES.IN_MEMORY} Updated edge:`, { id: edgeId });
  }
  async deleteEdge(edgeId) {
    this.ensureConnected();
    const entry = this.edges.get(edgeId);
    if (!entry) {
      return;
    }
    const edge = entry.data;
    this.outgoingEdges.get(edge.startNodeId)?.delete(edgeId);
    this.incomingEdges.get(edge.endNodeId)?.delete(edgeId);
    if (this.config.enableIndexing) {
      this.updateEdgeIndex(edgeId, edge, "remove");
    }
    this.edges.delete(edgeId);
    this.stats.edgeCount--;
    this.stats.operations++;
    this.stats.lastOperation = Date.now();
    this.logger.debug(`${LOG_PREFIXES.IN_MEMORY} Deleted edge:`, { id: edgeId });
  }
  async findEdges(filters, edgeType, limit) {
    this.ensureConnected();
    const maxResults = limit ?? DEFAULTS.QUERY_LIMIT;
    const results = [];
    let candidateIds = /* @__PURE__ */ new Set();
    if (edgeType && this.config.enableIndexing) {
      candidateIds = this.edgeIndex.edgeTypes.get(edgeType) || /* @__PURE__ */ new Set();
    } else {
      candidateIds = new Set(this.edges.keys());
    }
    for (const edgeId of candidateIds) {
      if (results.length >= maxResults) break;
      const entry = this.edges.get(edgeId);
      if (!entry) continue;
      const edge = entry.data;
      if (edgeType && edge.type !== edgeType) {
        continue;
      }
      if (filters && !this.matchesEdgeFilters(edge, filters)) {
        continue;
      }
      results.push(this.deepClone(edge));
    }
    this.stats.queriesExecuted++;
    this.logger.debug(`${LOG_PREFIXES.IN_MEMORY} Found ${results.length} edges`);
    return results;
  }
  // Advanced operations
  async query(query) {
    this.ensureConnected();
    const startTime = Date.now();
    let result;
    try {
      switch (query.type) {
        case "node":
          result = await this.executeNodeQuery(query);
          break;
        case "edge":
          result = await this.executeEdgeQuery(query);
          break;
        case "path":
          result = await this.executePathQuery(query);
          break;
        case "cypher":
          throw new InvalidQueryError("Cypher queries not supported in in-memory backend", query);
        default:
          throw new InvalidQueryError(`Unsupported query type: ${query.type}`, query);
      }
      result.metadata.executionTime = Date.now() - startTime;
      this.stats.queriesExecuted++;
      return result;
    } catch (error) {
      throw new InvalidQueryError(
        `Query execution failed: ${error.message}`,
        query,
        error
      );
    }
  }
  async getNeighbors(nodeId, direction = "both", edgeTypes, limit) {
    this.ensureConnected();
    if (!this.nodes.has(nodeId)) {
      throw new NodeNotFoundError(ERROR_MESSAGES.NODE_NOT_FOUND, nodeId);
    }
    const maxResults = limit ?? DEFAULTS.MAX_NEIGHBORS;
    const results = [];
    const edgeIds = /* @__PURE__ */ new Set();
    if (direction === "out" || direction === "both") {
      const outgoing = this.outgoingEdges.get(nodeId) || /* @__PURE__ */ new Set();
      outgoing.forEach((id) => edgeIds.add(id));
    }
    if (direction === "in" || direction === "both") {
      const incoming = this.incomingEdges.get(nodeId) || /* @__PURE__ */ new Set();
      incoming.forEach((id) => edgeIds.add(id));
    }
    for (const edgeId of edgeIds) {
      if (results.length >= maxResults) break;
      const edgeEntry = this.edges.get(edgeId);
      if (!edgeEntry) continue;
      const edge = edgeEntry.data;
      if (edgeTypes && edgeTypes.length > 0 && !edgeTypes.includes(edge.type)) {
        continue;
      }
      const neighborId = edge.startNodeId === nodeId ? edge.endNodeId : edge.startNodeId;
      const neighborEntry = this.nodes.get(neighborId);
      if (!neighborEntry) continue;
      results.push({
        node: this.deepClone(neighborEntry.data),
        edge: this.deepClone(edge)
      });
    }
    this.logger.debug(
      `${LOG_PREFIXES.IN_MEMORY} Found ${results.length} neighbors for node ${nodeId}`
    );
    return results;
  }
  async findPath(startNodeId, endNodeId, maxDepth = DEFAULTS.MAX_PATH_DEPTH, edgeTypes) {
    this.ensureConnected();
    if (!this.nodes.has(startNodeId)) {
      throw new NodeNotFoundError(`Start node not found`, startNodeId);
    }
    if (!this.nodes.has(endNodeId)) {
      throw new NodeNotFoundError(`End node not found`, endNodeId);
    }
    const queue = [
      { nodeId: startNodeId, path: [startNodeId], edgePath: [] }
    ];
    const visited = /* @__PURE__ */ new Set();
    while (queue.length > 0) {
      const current = queue.shift();
      if (current.nodeId === endNodeId) {
        const nodes = [];
        const edges = [];
        for (const nodeId of current.path) {
          const nodeEntry = this.nodes.get(nodeId);
          if (nodeEntry) {
            nodes.push(this.deepClone(nodeEntry.data));
          }
        }
        for (const edgeId of current.edgePath) {
          const edgeEntry = this.edges.get(edgeId);
          if (edgeEntry) {
            edges.push(this.deepClone(edgeEntry.data));
          }
        }
        this.logger.debug(
          `${LOG_PREFIXES.IN_MEMORY} Found path from ${startNodeId} to ${endNodeId} with ${edges.length} edges`
        );
        return { nodes, edges };
      }
      if (current.path.length >= maxDepth || visited.has(current.nodeId)) {
        continue;
      }
      visited.add(current.nodeId);
      const outgoing = this.outgoingEdges.get(current.nodeId) || /* @__PURE__ */ new Set();
      for (const edgeId of outgoing) {
        const edgeEntry = this.edges.get(edgeId);
        if (!edgeEntry) continue;
        const edge = edgeEntry.data;
        if (edgeTypes && edgeTypes.length > 0 && !edgeTypes.includes(edge.type)) {
          continue;
        }
        const nextNodeId = edge.endNodeId;
        if (!visited.has(nextNodeId)) {
          queue.push({
            nodeId: nextNodeId,
            path: [...current.path, nextNodeId],
            edgePath: [...current.edgePath, edgeId]
          });
        }
      }
    }
    this.logger.debug(
      `${LOG_PREFIXES.IN_MEMORY} No path found from ${startNodeId} to ${endNodeId}`
    );
    return null;
  }
  // Graph management
  async clearGraph() {
    this.ensureConnected();
    this.clear();
    this.logger.info(`${LOG_PREFIXES.IN_MEMORY} Graph cleared`);
  }
  async getStatistics() {
    this.ensureConnected();
    const labelCounts = {};
    const edgeTypeCounts = {};
    for (const entry of this.nodes.values()) {
      for (const label of entry.data.labels) {
        labelCounts[label] = (labelCounts[label] || 0) + 1;
      }
    }
    for (const entry of this.edges.values()) {
      const type = entry.data.type;
      edgeTypeCounts[type] = (edgeTypeCounts[type] || 0) + 1;
    }
    return {
      nodeCount: this.stats.nodeCount,
      edgeCount: this.stats.edgeCount,
      labelCounts,
      edgeTypeCounts
    };
  }
  // Private helper methods
  ensureConnected() {
    if (!this.connected) {
      throw new KnowledgeGraphConnectionError(ERROR_MESSAGES.NOT_CONNECTED, "in-memory");
    }
  }
  validateNode(node) {
    if (!node.id || typeof node.id !== "string") {
      throw new GraphValidationError("Node ID must be a non-empty string", "node");
    }
    if (!Array.isArray(node.labels) || node.labels.length === 0) {
      throw new GraphValidationError("Node must have at least one label", "node");
    }
    if (!node.properties || typeof node.properties !== "object") {
      throw new GraphValidationError("Node properties must be an object", "node");
    }
  }
  validateEdge(edge) {
    if (!edge.id || typeof edge.id !== "string") {
      throw new GraphValidationError("Edge ID must be a non-empty string", "edge");
    }
    if (!edge.type || typeof edge.type !== "string") {
      throw new GraphValidationError("Edge type must be a non-empty string", "edge");
    }
    if (!edge.startNodeId || typeof edge.startNodeId !== "string") {
      throw new GraphValidationError("Edge startNodeId must be a non-empty string", "edge");
    }
    if (!edge.endNodeId || typeof edge.endNodeId !== "string") {
      throw new GraphValidationError("Edge endNodeId must be a non-empty string", "edge");
    }
    if (!edge.properties || typeof edge.properties !== "object") {
      throw new GraphValidationError("Edge properties must be an object", "edge");
    }
  }
  deepClone(obj) {
    return JSON.parse(JSON.stringify(obj));
  }
  clear() {
    this.nodes.clear();
    this.edges.clear();
    this.outgoingEdges.clear();
    this.incomingEdges.clear();
    this.clearIndexes();
    this.stats = {
      nodeCount: 0,
      edgeCount: 0,
      operations: 0,
      queriesExecuted: 0,
      lastOperation: 0
    };
  }
  clearIndexes() {
    this.nodeIndex.labels.clear();
    this.nodeIndex.edgeTypes.clear();
    this.nodeIndex.properties.clear();
    this.edgeIndex.labels.clear();
    this.edgeIndex.edgeTypes.clear();
    this.edgeIndex.properties.clear();
  }
  updateNodeIndex(nodeId, node, operation) {
    if (!this.config.enableIndexing) return;
    const { labels, properties } = this.nodeIndex;
    for (const label of node.labels) {
      if (!labels.has(label)) {
        labels.set(label, /* @__PURE__ */ new Set());
      }
      const labelSet = labels.get(label);
      if (operation === "add") {
        labelSet.add(nodeId);
      } else {
        labelSet.delete(nodeId);
        if (labelSet.size === 0) {
          labels.delete(label);
        }
      }
    }
    for (const [key, value] of Object.entries(node.properties)) {
      if (!properties.has(key)) {
        properties.set(key, /* @__PURE__ */ new Map());
      }
      const propMap = properties.get(key);
      if (!propMap.has(value)) {
        propMap.set(value, /* @__PURE__ */ new Set());
      }
      const valueSet = propMap.get(value);
      if (operation === "add") {
        valueSet.add(nodeId);
      } else {
        valueSet.delete(nodeId);
        if (valueSet.size === 0) {
          propMap.delete(value);
          if (propMap.size === 0) {
            properties.delete(key);
          }
        }
      }
    }
  }
  updateEdgeIndex(edgeId, edge, operation) {
    if (!this.config.enableIndexing) return;
    const { edgeTypes, properties } = this.edgeIndex;
    if (!edgeTypes.has(edge.type)) {
      edgeTypes.set(edge.type, /* @__PURE__ */ new Set());
    }
    const typeSet = edgeTypes.get(edge.type);
    if (operation === "add") {
      typeSet.add(edgeId);
    } else {
      typeSet.delete(edgeId);
      if (typeSet.size === 0) {
        edgeTypes.delete(edge.type);
      }
    }
    for (const [key, value] of Object.entries(edge.properties)) {
      if (!properties.has(key)) {
        properties.set(key, /* @__PURE__ */ new Map());
      }
      const propMap = properties.get(key);
      if (!propMap.has(value)) {
        propMap.set(value, /* @__PURE__ */ new Set());
      }
      const valueSet = propMap.get(value);
      if (operation === "add") {
        valueSet.add(edgeId);
      } else {
        valueSet.delete(edgeId);
        if (valueSet.size === 0) {
          propMap.delete(value);
          if (propMap.size === 0) {
            properties.delete(key);
          }
        }
      }
    }
  }
  matchesNodeFilters(node, filters) {
    for (const [key, filter] of Object.entries(filters)) {
      const value = node.properties[key];
      if (!this.matchesFilter(value, filter)) {
        return false;
      }
    }
    return true;
  }
  matchesEdgeFilters(edge, filters) {
    for (const [key, filter] of Object.entries(filters)) {
      const value = edge.properties[key];
      if (!this.matchesFilter(value, filter)) {
        return false;
      }
    }
    return true;
  }
  matchesFilter(value, filter) {
    if (filter === null || filter === void 0) {
      return value === filter;
    }
    if (typeof filter === "object" && !Array.isArray(filter)) {
      if ("gte" in filter && value < filter.gte) return false;
      if ("gt" in filter && value <= filter.gt) return false;
      if ("lte" in filter && value > filter.lte) return false;
      if ("lt" in filter && value >= filter.lt) return false;
      if ("any" in filter) {
        return Array.isArray(filter.any) && filter.any.includes(value);
      }
      if ("all" in filter) {
        return Array.isArray(filter.all) && Array.isArray(value) && filter.all.every((item) => value.includes(item));
      }
      return true;
    }
    return value === filter;
  }
  async executeNodeQuery(query) {
    const { pattern, limit } = query;
    const labels = pattern?.labels;
    const properties = pattern?.properties;
    const nodes = await this.findNodes(properties, labels, limit);
    return {
      nodes,
      edges: [],
      metadata: {
        totalCount: nodes.length,
        queryType: "node"
      }
    };
  }
  async executeEdgeQuery(query) {
    const { pattern, limit } = query;
    const edgeType = pattern?.type;
    const properties = pattern?.properties;
    const edges = await this.findEdges(properties, edgeType, limit);
    return {
      nodes: [],
      edges,
      metadata: {
        totalCount: edges.length,
        queryType: "edge"
      }
    };
  }
  async executePathQuery(query) {
    const { pattern, parameters } = query;
    const startNode = pattern?.startNode;
    const endNode = pattern?.endNode;
    const maxDepth = parameters?.maxDepth ?? DEFAULTS.MAX_PATH_DEPTH;
    if (!startNode?.id || !endNode?.id) {
      throw new InvalidQueryError("Path query requires startNode.id and endNode.id", query);
    }
    const path = await this.findPath(startNode.id, endNode.id, maxDepth);
    if (path) {
      return {
        nodes: path.nodes,
        edges: path.edges,
        paths: [
          {
            nodes: path.nodes,
            edges: path.edges,
            length: path.edges.length
          }
        ],
        metadata: {
          totalCount: 1,
          queryType: "path"
        }
      };
    }
    return {
      nodes: [],
      edges: [],
      paths: [],
      metadata: {
        totalCount: 0,
        queryType: "path"
      }
    };
  }
  // Debug methods for testing
  getInternalStats() {
    return {
      ...this.stats,
      memoryUsage: {
        nodes: this.nodes.size,
        edges: this.edges.size,
        outgoingEdges: this.outgoingEdges.size,
        incomingEdges: this.incomingEdges.size
      },
      indexSizes: this.config.enableIndexing ? {
        nodeLabels: this.nodeIndex.labels.size,
        nodeProperties: this.nodeIndex.properties.size,
        edgeTypes: this.edgeIndex.edgeTypes.size,
        edgeProperties: this.edgeIndex.properties.size
      } : null
    };
  }
};

export {
  InMemoryBackend
};
