// @byterover/cipher/dist/src/core/chunk-7UQA6M3D.js
import {
  external_exports,
  init_zod
} from "./chunk-NUVPDHVT.js";
import {
  __esm,
  __export,
  init_esm_shims
} from "./chunk-KFOQGTVJ.js";

// src/core/env.ts
var env_exports = {};
__export(env_exports, {
  env: () => env,
  validateEnv: () => validateEnv
});
import { config } from "dotenv";
var isMcpMode, envSchema, env, validateEnv;
var init_env = __esm({
  "src/core/env.ts"() {
    init_esm_shims();
    init_zod();
    isMcpMode = process.argv.includes("--mode") && process.argv[process.argv.indexOf("--mode") + 1] === "mcp";
    if (isMcpMode) {
      config({ override: false });
    } else {
      config();
    }
    envSchema = external_exports.object({
      NODE_ENV: external_exports.enum(["development", "production", "test"]).default("development"),
      CIPHER_LOG_LEVEL: external_exports.enum(["error", "warn", "info", "debug", "silly"]).default("info"),
      REDACT_SECRETS: external_exports.boolean().default(true),
      OPENAI_API_KEY: external_exports.string().optional(),
      ANTHROPIC_API_KEY: external_exports.string().optional(),
      OPENROUTER_API_KEY: external_exports.string().optional(),
      QWEN_API_KEY: external_exports.string().optional(),
      OPENAI_BASE_URL: external_exports.string().optional(),
      OLLAMA_BASE_URL: external_exports.string().optional(),
      LMSTUDIO_BASE_URL: external_exports.string().optional(),
      OPENAI_ORG_ID: external_exports.string().optional(),
      // Embedding Configuration
      EMBEDDING_PROVIDER: external_exports.string().optional(),
      EMBEDDING_MODEL: external_exports.string().optional(),
      EMBEDDING_TIMEOUT: external_exports.number().optional(),
      EMBEDDING_MAX_RETRIES: external_exports.number().optional(),
      EMBEDDING_DIMENSIONS: external_exports.number().optional(),
      DISABLE_EMBEDDINGS: external_exports.boolean().default(false),
      EMBEDDING_DISABLED: external_exports.boolean().default(false),
      GEMINI_API_KEY: external_exports.string().optional(),
      GEMINI_BASE_URL: external_exports.string().optional(),
      // Storage Configuration
      STORAGE_CACHE_TYPE: external_exports.enum(["in-memory", "redis"]).default("in-memory"),
      STORAGE_CACHE_HOST: external_exports.string().optional(),
      STORAGE_CACHE_PORT: external_exports.number().optional(),
      STORAGE_CACHE_USERNAME: external_exports.string().optional(),
      STORAGE_CACHE_PASSWORD: external_exports.string().optional(),
      STORAGE_CACHE_DATABASE: external_exports.number().optional(),
      STORAGE_DATABASE_TYPE: external_exports.enum(["in-memory", "sqlite", "postgres"]).default("in-memory"),
      STORAGE_DATABASE_PATH: external_exports.string().optional(),
      STORAGE_DATABASE_NAME: external_exports.string().optional(),
      // PostgreSQL Configuration
      CIPHER_PG_URL: external_exports.string().optional(),
      STORAGE_DATABASE_HOST: external_exports.string().optional(),
      STORAGE_DATABASE_PORT: external_exports.number().optional(),
      STORAGE_DATABASE_USER: external_exports.string().optional(),
      STORAGE_DATABASE_PASSWORD: external_exports.string().optional(),
      STORAGE_DATABASE_SSL: external_exports.boolean().default(false),
      // Vector Storage Configuration
      VECTOR_STORE_TYPE: external_exports.enum(["qdrant", "milvus", "chroma", "pinecone", "in-memory", "faiss"]).default("in-memory"),
      VECTOR_STORE_HOST: external_exports.string().optional(),
      VECTOR_STORE_PORT: external_exports.number().optional(),
      VECTOR_STORE_URL: external_exports.string().optional(),
      VECTOR_STORE_API_KEY: external_exports.string().optional(),
      VECTOR_STORE_USERNAME: external_exports.string().optional(),
      VECTOR_STORE_PASSWORD: external_exports.string().optional(),
      VECTOR_STORE_COLLECTION: external_exports.string().default("knowledge_memory"),
      VECTOR_STORE_DIMENSION: external_exports.number().default(1536),
      VECTOR_STORE_DISTANCE: external_exports.enum(["Cosine", "Euclidean", "Dot", "Manhattan"]).default("Cosine"),
      VECTOR_STORE_ON_DISK: external_exports.boolean().default(false),
      VECTOR_STORE_MAX_VECTORS: external_exports.number().default(1e4),
      // Pinecone-specific Configuration
      PINECONE_PROVIDER: external_exports.string().default("aws"),
      PINECONE_REGION: external_exports.string().default("us-east-1"),
      // PGVECTOR -specific Configuration
      PGVECTOR_INDEXTYPE: external_exports.enum(["hnsw", "ivfflat"]).default("hnsw"),
      PGVECTOR_INDEXMETRIC: external_exports.enum(["vector_l2_ops", "vector_ip_ops"]).default("vector_l2_ops"),
      PINECONE_NAMESPACE: external_exports.string().default("default"),
      // Faiss-specific Configuration
      FAISS_BASE_STORAGE_PATH: external_exports.string().optional(),
      // Knowledge Graph Configuration
      KNOWLEDGE_GRAPH_ENABLED: external_exports.boolean().default(false),
      KNOWLEDGE_GRAPH_TYPE: external_exports.enum(["neo4j", "in-memory"]).default("in-memory"),
      KNOWLEDGE_GRAPH_HOST: external_exports.string().optional(),
      KNOWLEDGE_GRAPH_PORT: external_exports.number().optional(),
      KNOWLEDGE_GRAPH_URI: external_exports.string().optional(),
      KNOWLEDGE_GRAPH_USERNAME: external_exports.string().optional(),
      KNOWLEDGE_GRAPH_PASSWORD: external_exports.string().optional(),
      KNOWLEDGE_GRAPH_DATABASE: external_exports.string().default("neo4j"),
      // Memory Search Configuration
      SEARCH_MEMORY_TYPE: external_exports.enum(["knowledge", "reflection", "both"]).default("knowledge"),
      // Reflection Memory Configuration
      REFLECTION_VECTOR_STORE_COLLECTION: external_exports.string().default("reflection_memory"),
      DISABLE_REFLECTION_MEMORY: external_exports.boolean().default(true),
      // Event Persistence Configuration
      EVENT_PERSISTENCE_ENABLED: external_exports.boolean().default(false),
      EVENT_PERSISTENCE_PATH: external_exports.string().optional(),
      // Lazy Loading Configuration
      ENABLE_LAZY_LOADING: external_exports.string().optional(),
      LAZY_LOADING_ENABLED: external_exports.string().optional(),
      SKIP_HEAVY_SERVICES: external_exports.string().optional(),
      LAZY_EMBEDDING: external_exports.string().optional(),
      LAZY_VECTOR_STORE: external_exports.string().optional(),
      LAZY_MEMORY_OPERATIONS: external_exports.string().optional(),
      DISABLE_BACKGROUND_PRELOAD: external_exports.string().optional(),
      LAZY_INIT_TIMEOUT: external_exports.string().optional(),
      BACKGROUND_PRELOAD_DELAY: external_exports.string().optional(),
      // Workspace Memory Configuration
      USE_WORKSPACE_MEMORY: external_exports.boolean().default(false),
      WORKSPACE_SEARCH_THRESHOLD: external_exports.number().default(0.4),
      DISABLE_DEFAULT_MEMORY: external_exports.boolean().default(false),
      WORKSPACE_VECTOR_STORE_TYPE: external_exports.enum(["qdrant", "milvus", "chroma", "pinecone", "pgvector", "in-memory"]).optional(),
      WORKSPACE_VECTOR_STORE_HOST: external_exports.string().optional(),
      WORKSPACE_VECTOR_STORE_PORT: external_exports.number().optional(),
      WORKSPACE_VECTOR_STORE_URL: external_exports.string().optional(),
      WORKSPACE_VECTOR_STORE_API_KEY: external_exports.string().optional(),
      WORKSPACE_VECTOR_STORE_USERNAME: external_exports.string().optional(),
      WORKSPACE_VECTOR_STORE_PASSWORD: external_exports.string().optional(),
      WORKSPACE_VECTOR_STORE_COLLECTION: external_exports.string().default("workspace_memory"),
      WORKSPACE_VECTOR_STORE_DIMENSION: external_exports.number().optional(),
      WORKSPACE_VECTOR_STORE_DISTANCE: external_exports.enum(["Cosine", "Euclidean", "Dot", "Manhattan"]).optional(),
      WORKSPACE_VECTOR_STORE_ON_DISK: external_exports.boolean().optional(),
      WORKSPACE_VECTOR_STORE_MAX_VECTORS: external_exports.number().optional(),
      // Redis-specific Configuration
      WORKSPACE_REDIS_DATABASE: external_exports.number().optional(),
      // Pinecone-specific Configuration
      WORKSPACE_PINECONE_PROVIDER: external_exports.string().optional(),
      WORKSPACE_PINECONE_REGION: external_exports.string().optional(),
      WORKSPACE_PINECONE_NAMESPACE: external_exports.string().optional(),
      // Query Refinement Configuration
      ENABLE_QUERY_REFINEMENT: external_exports.boolean().default(true),
      // Cross-Tool Memory Sharing Configuration
      CIPHER_USER_ID: external_exports.string().optional(),
      CIPHER_PROJECT_NAME: external_exports.string().optional(),
      CIPHER_WORKSPACE_MODE: external_exports.enum(["shared", "isolated"]).default("isolated"),
      // MCP Aggregator Configuration
      USE_ASK_CIPHER: external_exports.boolean().default(false)
    });
    env = new Proxy({}, {
      get(target, prop) {
        switch (prop) {
          case "NODE_ENV":
            return process.env.NODE_ENV || "development";
          case "CIPHER_LOG_LEVEL":
            return process.env.CIPHER_LOG_LEVEL || "info";
          case "REDACT_SECRETS":
            return process.env.REDACT_SECRETS === "false" ? false : true;
          case "OPENAI_API_KEY":
            return process.env.OPENAI_API_KEY;
          case "ANTHROPIC_API_KEY":
            return process.env.ANTHROPIC_API_KEY;
          case "OPENROUTER_API_KEY":
            return process.env.OPENROUTER_API_KEY;
          case "QWEN_API_KEY":
            return process.env.QWEN_API_KEY;
          case "OPENAI_BASE_URL":
            return process.env.OPENAI_BASE_URL;
          case "OLLAMA_BASE_URL":
            return process.env.OLLAMA_BASE_URL;
          case "OPENAI_ORG_ID":
            return process.env.OPENAI_ORG_ID;
          // Embedding Configuration
          case "EMBEDDING_PROVIDER":
            return process.env.EMBEDDING_PROVIDER;
          case "EMBEDDING_MODEL":
            return process.env.EMBEDDING_MODEL;
          case "EMBEDDING_TIMEOUT":
            return process.env.EMBEDDING_TIMEOUT ? parseInt(process.env.EMBEDDING_TIMEOUT, 10) : void 0;
          case "EMBEDDING_MAX_RETRIES":
            return process.env.EMBEDDING_MAX_RETRIES ? parseInt(process.env.EMBEDDING_MAX_RETRIES, 10) : void 0;
          case "EMBEDDING_DIMENSIONS":
            return process.env.EMBEDDING_DIMENSIONS ? parseInt(process.env.EMBEDDING_DIMENSIONS, 10) : void 0;
          case "DISABLE_EMBEDDINGS":
            return process.env.DISABLE_EMBEDDINGS === "true";
          case "EMBEDDING_DISABLED":
            return process.env.EMBEDDING_DISABLED === "true";
          case "GEMINI_API_KEY":
            return process.env.GEMINI_API_KEY;
          case "GEMINI_BASE_URL":
            return process.env.GEMINI_BASE_URL;
          // Storage Configuration
          case "STORAGE_CACHE_TYPE":
            return process.env.STORAGE_CACHE_TYPE || "in-memory";
          case "STORAGE_CACHE_HOST":
            return process.env.STORAGE_CACHE_HOST;
          case "STORAGE_CACHE_PORT":
            return process.env.STORAGE_CACHE_PORT ? parseInt(process.env.STORAGE_CACHE_PORT, 10) : void 0;
          case "STORAGE_CACHE_USERNAME":
            return process.env.STORAGE_CACHE_USERNAME;
          case "STORAGE_CACHE_PASSWORD":
            return process.env.STORAGE_CACHE_PASSWORD;
          case "STORAGE_CACHE_DATABASE":
            return process.env.STORAGE_CACHE_DATABASE ? parseInt(process.env.STORAGE_CACHE_DATABASE, 10) : void 0;
          case "STORAGE_DATABASE_TYPE":
            return process.env.STORAGE_DATABASE_TYPE || "in-memory";
          case "STORAGE_DATABASE_PATH":
            return process.env.STORAGE_DATABASE_PATH;
          case "STORAGE_DATABASE_NAME":
            return process.env.STORAGE_DATABASE_NAME;
          // PostgreSQL Configuration
          case "CIPHER_PG_URL":
            return process.env.CIPHER_PG_URL;
          case "STORAGE_DATABASE_HOST":
            return process.env.STORAGE_DATABASE_HOST;
          case "STORAGE_DATABASE_PORT":
            return process.env.STORAGE_DATABASE_PORT ? parseInt(process.env.STORAGE_DATABASE_PORT, 10) : void 0;
          case "STORAGE_DATABASE_USER":
            return process.env.STORAGE_DATABASE_USER;
          case "STORAGE_DATABASE_PASSWORD":
            return process.env.STORAGE_DATABASE_PASSWORD;
          case "STORAGE_DATABASE_SSL":
            return process.env.STORAGE_DATABASE_SSL === "true";
          // Vector Storage Configuration
          case "VECTOR_STORE_TYPE":
            return process.env.VECTOR_STORE_TYPE || "in-memory";
          case "VECTOR_STORE_HOST":
            return process.env.VECTOR_STORE_HOST;
          case "VECTOR_STORE_PORT":
            return process.env.VECTOR_STORE_PORT ? parseInt(process.env.VECTOR_STORE_PORT, 10) : void 0;
          case "VECTOR_STORE_URL":
            return process.env.VECTOR_STORE_URL;
          case "VECTOR_STORE_API_KEY":
            return process.env.VECTOR_STORE_API_KEY;
          case "VECTOR_STORE_USERNAME":
            return process.env.VECTOR_STORE_USERNAME;
          case "VECTOR_STORE_PASSWORD":
            return process.env.VECTOR_STORE_PASSWORD;
          case "VECTOR_STORE_COLLECTION":
            return process.env.VECTOR_STORE_COLLECTION || "knowledge_memory";
          case "VECTOR_STORE_DIMENSION":
            return process.env.VECTOR_STORE_DIMENSION ? parseInt(process.env.VECTOR_STORE_DIMENSION, 10) : 1536;
          case "VECTOR_STORE_DISTANCE":
            return process.env.VECTOR_STORE_DISTANCE || "Cosine";
          case "VECTOR_STORE_ON_DISK":
            return process.env.VECTOR_STORE_ON_DISK === "true";
          case "VECTOR_STORE_MAX_VECTORS":
            return process.env.VECTOR_STORE_MAX_VECTORS ? parseInt(process.env.VECTOR_STORE_MAX_VECTORS, 10) : 1e4;
          // Faiss-specific Configuration
          case "FAISS_BASE_STORAGE_PATH":
            return process.env.FAISS_BASE_STORAGE_PATH;
          // Knowledge Graph Configuration
          case "KNOWLEDGE_GRAPH_ENABLED":
            return process.env.KNOWLEDGE_GRAPH_ENABLED === "true";
          case "KNOWLEDGE_GRAPH_TYPE":
            return process.env.KNOWLEDGE_GRAPH_TYPE || "in-memory";
          case "KNOWLEDGE_GRAPH_HOST":
            return process.env.KNOWLEDGE_GRAPH_HOST;
          case "KNOWLEDGE_GRAPH_PORT":
            return process.env.KNOWLEDGE_GRAPH_PORT ? parseInt(process.env.KNOWLEDGE_GRAPH_PORT, 10) : void 0;
          case "KNOWLEDGE_GRAPH_URI":
            return process.env.KNOWLEDGE_GRAPH_URI;
          case "KNOWLEDGE_GRAPH_USERNAME":
            return process.env.KNOWLEDGE_GRAPH_USERNAME;
          case "KNOWLEDGE_GRAPH_PASSWORD":
            return process.env.KNOWLEDGE_GRAPH_PASSWORD;
          case "KNOWLEDGE_GRAPH_DATABASE":
            return process.env.KNOWLEDGE_GRAPH_DATABASE || "neo4j";
          // Memory Search Configuration
          case "SEARCH_MEMORY_TYPE":
            return process.env.SEARCH_MEMORY_TYPE || "knowledge";
          // Reflection Memory Configuration
          case "REFLECTION_VECTOR_STORE_COLLECTION": {
            const value = process.env.REFLECTION_VECTOR_STORE_COLLECTION || "reflection_memory";
            if (value === "true") return true;
            if (value === "false") return false;
            return value;
          }
          case "DISABLE_REFLECTION_MEMORY":
            return process.env.DISABLE_REFLECTION_MEMORY === "true";
          // Event Persistence Configuration
          case "EVENT_PERSISTENCE_ENABLED":
            return process.env.EVENT_PERSISTENCE_ENABLED === "true";
          case "EVENT_PERSISTENCE_PATH":
            return process.env.EVENT_PERSISTENCE_PATH;
          // Lazy Loading Configuration
          case "ENABLE_LAZY_LOADING":
            return process.env.ENABLE_LAZY_LOADING;
          case "LAZY_LOADING_ENABLED":
            return process.env.LAZY_LOADING_ENABLED;
          case "SKIP_HEAVY_SERVICES":
            return process.env.SKIP_HEAVY_SERVICES;
          case "LAZY_EMBEDDING":
            return process.env.LAZY_EMBEDDING;
          case "LAZY_VECTOR_STORE":
            return process.env.LAZY_VECTOR_STORE;
          case "LAZY_MEMORY_OPERATIONS":
            return process.env.LAZY_MEMORY_OPERATIONS;
          case "DISABLE_BACKGROUND_PRELOAD":
            return process.env.DISABLE_BACKGROUND_PRELOAD;
          case "LAZY_INIT_TIMEOUT":
            return process.env.LAZY_INIT_TIMEOUT;
          case "BACKGROUND_PRELOAD_DELAY":
            return process.env.BACKGROUND_PRELOAD_DELAY;
          // Workspace Memory Configuration
          case "USE_WORKSPACE_MEMORY":
            return process.env.USE_WORKSPACE_MEMORY === "true";
          case "WORKSPACE_SEARCH_THRESHOLD":
            return process.env.WORKSPACE_SEARCH_THRESHOLD ? parseFloat(process.env.WORKSPACE_SEARCH_THRESHOLD) : 0.4;
          case "DISABLE_DEFAULT_MEMORY":
            return process.env.DISABLE_DEFAULT_MEMORY === "true";
          case "WORKSPACE_VECTOR_STORE_TYPE":
            return process.env.WORKSPACE_VECTOR_STORE_TYPE;
          case "WORKSPACE_VECTOR_STORE_HOST":
            return process.env.WORKSPACE_VECTOR_STORE_HOST;
          case "WORKSPACE_VECTOR_STORE_PORT":
            return process.env.WORKSPACE_VECTOR_STORE_PORT ? parseInt(process.env.WORKSPACE_VECTOR_STORE_PORT, 10) : void 0;
          case "WORKSPACE_VECTOR_STORE_URL":
            return process.env.WORKSPACE_VECTOR_STORE_URL;
          case "WORKSPACE_VECTOR_STORE_API_KEY":
            return process.env.WORKSPACE_VECTOR_STORE_API_KEY;
          case "WORKSPACE_VECTOR_STORE_USERNAME":
            return process.env.WORKSPACE_VECTOR_STORE_USERNAME;
          case "WORKSPACE_VECTOR_STORE_PASSWORD":
            return process.env.WORKSPACE_VECTOR_STORE_PASSWORD;
          case "WORKSPACE_VECTOR_STORE_COLLECTION":
            return process.env.WORKSPACE_VECTOR_STORE_COLLECTION || "workspace_memory";
          case "WORKSPACE_VECTOR_STORE_DIMENSION":
            return process.env.WORKSPACE_VECTOR_STORE_DIMENSION ? parseInt(process.env.WORKSPACE_VECTOR_STORE_DIMENSION, 10) : process.env.VECTOR_STORE_DIMENSION ? parseInt(process.env.VECTOR_STORE_DIMENSION, 10) : 1536;
          case "WORKSPACE_VECTOR_STORE_DISTANCE":
            return process.env.WORKSPACE_VECTOR_STORE_DISTANCE || process.env.VECTOR_STORE_DISTANCE || "Cosine";
          case "WORKSPACE_VECTOR_STORE_ON_DISK":
            return process.env.WORKSPACE_VECTOR_STORE_ON_DISK !== void 0 ? process.env.WORKSPACE_VECTOR_STORE_ON_DISK === "true" : process.env.VECTOR_STORE_ON_DISK !== void 0 ? process.env.VECTOR_STORE_ON_DISK === "true" : false;
          case "WORKSPACE_VECTOR_STORE_MAX_VECTORS":
            return process.env.WORKSPACE_VECTOR_STORE_MAX_VECTORS ? parseInt(process.env.WORKSPACE_VECTOR_STORE_MAX_VECTORS, 10) : process.env.VECTOR_STORE_MAX_VECTORS ? parseInt(process.env.VECTOR_STORE_MAX_VECTORS, 10) : 1e4;
          // Pinecone-specific Configuration
          case "WORKSPACE_PINECONE_PROVIDER":
            return process.env.WORKSPACE_PINECONE_PROVIDER;
          case "WORKSPACE_PINECONE_REGION":
            return process.env.WORKSPACE_PINECONE_REGION;
          case "WORKSPACE_PINECONE_NAMESPACE":
            return process.env.WORKSPACE_PINECONE_NAMESPACE;
          // Cross-Tool Memory Sharing Configuration
          case "CIPHER_USER_ID":
            return process.env.CIPHER_USER_ID;
          case "CIPHER_PROJECT_NAME":
            return process.env.CIPHER_PROJECT_NAME;
          case "CIPHER_WORKSPACE_MODE":
            return process.env.CIPHER_WORKSPACE_MODE || "isolated";
          case "USE_ASK_CIPHER":
            return process.env.USE_ASK_CIPHER === "true";
          default:
            return process.env[prop];
        }
      }
    });
    validateEnv = () => {
      const embeddingsDisabled = process.env.DISABLE_EMBEDDINGS === "true" || process.env.EMBEDDING_DISABLED === "true";
      if (!embeddingsDisabled) {
        const hasOpenAI = !!process.env.OPENAI_API_KEY;
        const hasGemini = !!process.env.GEMINI_API_KEY;
        const hasOpenRouter = !!process.env.OPENROUTER_API_KEY;
        const hasOllama = !!process.env.OLLAMA_BASE_URL;
        const hasAnyEmbeddingProvider = hasOpenAI || hasGemini || hasOpenRouter || hasOllama;
        if (!hasAnyEmbeddingProvider) {
          const errorMsg = "No embedding provider configured. Set one of: OPENAI_API_KEY, GEMINI_API_KEY, OPENROUTER_API_KEY, OLLAMA_BASE_URL, or set DISABLE_EMBEDDINGS=true to run without memory capabilities";
          if (isMcpMode) {
            process.stderr.write(`[CIPHER-MCP] WARNING: ${errorMsg}
`);
          } else {
            console.warn(errorMsg);
          }
        }
      } else {
        const infoMsg = "Embeddings are disabled - Cipher will run without memory capabilities";
        if (isMcpMode) {
          process.stderr.write(`[CIPHER-MCP] INFO: ${infoMsg}
`);
        } else {
          console.info(infoMsg);
        }
      }
      const envToValidate = {
        NODE_ENV: process.env.NODE_ENV,
        CIPHER_LOG_LEVEL: process.env.CIPHER_LOG_LEVEL,
        REDACT_SECRETS: process.env.REDACT_SECRETS === "false" ? false : true,
        OPENAI_API_KEY: process.env.OPENAI_API_KEY,
        ANTHROPIC_API_KEY: process.env.ANTHROPIC_API_KEY,
        OPENROUTER_API_KEY: process.env.OPENROUTER_API_KEY,
        QWEN_API_KEY: process.env.QWEN_API_KEY,
        OPENAI_BASE_URL: process.env.OPENAI_BASE_URL,
        OLLAMA_BASE_URL: process.env.OLLAMA_BASE_URL,
        OPENAI_ORG_ID: process.env.OPENAI_ORG_ID,
        // Embedding Configuration
        EMBEDDING_PROVIDER: process.env.EMBEDDING_PROVIDER,
        EMBEDDING_MODEL: process.env.EMBEDDING_MODEL,
        EMBEDDING_TIMEOUT: process.env.EMBEDDING_TIMEOUT ? parseInt(process.env.EMBEDDING_TIMEOUT, 10) : void 0,
        EMBEDDING_MAX_RETRIES: process.env.EMBEDDING_MAX_RETRIES ? parseInt(process.env.EMBEDDING_MAX_RETRIES, 10) : void 0,
        EMBEDDING_DIMENSIONS: process.env.EMBEDDING_DIMENSIONS ? parseInt(process.env.EMBEDDING_DIMENSIONS, 10) : void 0,
        DISABLE_EMBEDDINGS: process.env.DISABLE_EMBEDDINGS === "true",
        EMBEDDING_DISABLED: process.env.EMBEDDING_DISABLED === "true",
        GEMINI_API_KEY: process.env.GEMINI_API_KEY,
        GEMINI_BASE_URL: process.env.GEMINI_BASE_URL,
        // Storage Configuration
        STORAGE_CACHE_TYPE: process.env.STORAGE_CACHE_TYPE || "in-memory",
        STORAGE_CACHE_HOST: process.env.STORAGE_CACHE_HOST,
        STORAGE_CACHE_PORT: process.env.STORAGE_CACHE_PORT ? parseInt(process.env.STORAGE_CACHE_PORT, 10) : void 0,
        STORAGE_CACHE_USERNAME: process.env.STORAGE_CACHE_USERNAME,
        STORAGE_CACHE_PASSWORD: process.env.STORAGE_CACHE_PASSWORD,
        STORAGE_CACHE_DATABASE: process.env.STORAGE_CACHE_DATABASE ? parseInt(process.env.STORAGE_CACHE_DATABASE, 10) : void 0,
        STORAGE_DATABASE_TYPE: process.env.STORAGE_DATABASE_TYPE || "in-memory",
        STORAGE_DATABASE_PATH: process.env.STORAGE_DATABASE_PATH,
        STORAGE_DATABASE_NAME: process.env.STORAGE_DATABASE_NAME,
        // PostgreSQL Configuration
        CIPHER_PG_URL: process.env.CIPHER_PG_URL,
        STORAGE_DATABASE_HOST: process.env.STORAGE_DATABASE_HOST,
        STORAGE_DATABASE_PORT: process.env.STORAGE_DATABASE_PORT ? parseInt(process.env.STORAGE_DATABASE_PORT, 10) : void 0,
        STORAGE_DATABASE_USER: process.env.STORAGE_DATABASE_USER,
        STORAGE_DATABASE_PASSWORD: process.env.STORAGE_DATABASE_PASSWORD,
        STORAGE_DATABASE_SSL: process.env.STORAGE_DATABASE_SSL === "true",
        // Vector Storage Configuration
        VECTOR_STORE_TYPE: process.env.VECTOR_STORE_TYPE || "in-memory",
        VECTOR_STORE_HOST: process.env.VECTOR_STORE_HOST,
        VECTOR_STORE_PORT: process.env.VECTOR_STORE_PORT ? parseInt(process.env.VECTOR_STORE_PORT, 10) : void 0,
        VECTOR_STORE_URL: process.env.VECTOR_STORE_URL,
        VECTOR_STORE_API_KEY: process.env.VECTOR_STORE_API_KEY,
        VECTOR_STORE_USERNAME: process.env.VECTOR_STORE_USERNAME,
        VECTOR_STORE_PASSWORD: process.env.VECTOR_STORE_PASSWORD,
        VECTOR_STORE_COLLECTION: process.env.VECTOR_STORE_COLLECTION || "knowledge_memory",
        VECTOR_STORE_DIMENSION: process.env.VECTOR_STORE_DIMENSION ? parseInt(process.env.VECTOR_STORE_DIMENSION, 10) : 1536,
        VECTOR_STORE_DISTANCE: process.env.VECTOR_STORE_DISTANCE || "Cosine",
        VECTOR_STORE_ON_DISK: process.env.VECTOR_STORE_ON_DISK === "true",
        VECTOR_STORE_MAX_VECTORS: process.env.VECTOR_STORE_MAX_VECTORS ? parseInt(process.env.VECTOR_STORE_MAX_VECTORS, 10) : 1e4,
        // Faiss-specific Configuration
        FAISS_BASE_STORAGE_PATH: process.env.FAISS_BASE_STORAGE_PATH,
        // Knowledge Graph Configuration
        KNOWLEDGE_GRAPH_ENABLED: process.env.KNOWLEDGE_GRAPH_ENABLED === "true",
        KNOWLEDGE_GRAPH_TYPE: process.env.KNOWLEDGE_GRAPH_TYPE || "in-memory",
        KNOWLEDGE_GRAPH_HOST: process.env.KNOWLEDGE_GRAPH_HOST,
        KNOWLEDGE_GRAPH_PORT: process.env.KNOWLEDGE_GRAPH_PORT ? parseInt(process.env.KNOWLEDGE_GRAPH_PORT, 10) : void 0,
        KNOWLEDGE_GRAPH_URI: process.env.KNOWLEDGE_GRAPH_URI,
        KNOWLEDGE_GRAPH_USERNAME: process.env.KNOWLEDGE_GRAPH_USERNAME,
        KNOWLEDGE_GRAPH_PASSWORD: process.env.KNOWLEDGE_GRAPH_PASSWORD,
        KNOWLEDGE_GRAPH_DATABASE: process.env.KNOWLEDGE_GRAPH_DATABASE || "neo4j",
        // Memory Search Configuration
        SEARCH_MEMORY_TYPE: process.env.SEARCH_MEMORY_TYPE || "knowledge",
        // Reflection Memory Configuration
        REFLECTION_VECTOR_STORE_COLLECTION: process.env.REFLECTION_VECTOR_STORE_COLLECTION || "reflection_memory",
        DISABLE_REFLECTION_MEMORY: process.env.DISABLE_REFLECTION_MEMORY === "true",
        // Workspace Memory Configuration
        USE_WORKSPACE_MEMORY: process.env.USE_WORKSPACE_MEMORY === "true",
        WORKSPACE_SEARCH_THRESHOLD: process.env.WORKSPACE_SEARCH_THRESHOLD ? parseFloat(process.env.WORKSPACE_SEARCH_THRESHOLD) : 0.4,
        DISABLE_DEFAULT_MEMORY: process.env.DISABLE_DEFAULT_MEMORY === "true",
        WORKSPACE_VECTOR_STORE_TYPE: process.env.WORKSPACE_VECTOR_STORE_TYPE,
        WORKSPACE_VECTOR_STORE_HOST: process.env.WORKSPACE_VECTOR_STORE_HOST,
        WORKSPACE_VECTOR_STORE_PORT: process.env.WORKSPACE_VECTOR_STORE_PORT ? parseInt(process.env.WORKSPACE_VECTOR_STORE_PORT, 10) : void 0,
        WORKSPACE_VECTOR_STORE_URL: process.env.WORKSPACE_VECTOR_STORE_URL,
        WORKSPACE_VECTOR_STORE_API_KEY: process.env.WORKSPACE_VECTOR_STORE_API_KEY,
        WORKSPACE_VECTOR_STORE_USERNAME: process.env.WORKSPACE_VECTOR_STORE_USERNAME,
        WORKSPACE_VECTOR_STORE_PASSWORD: process.env.WORKSPACE_VECTOR_STORE_PASSWORD,
        WORKSPACE_VECTOR_STORE_COLLECTION: process.env.WORKSPACE_VECTOR_STORE_COLLECTION || "workspace_memory",
        WORKSPACE_VECTOR_STORE_DIMENSION: process.env.WORKSPACE_VECTOR_STORE_DIMENSION ? parseInt(process.env.WORKSPACE_VECTOR_STORE_DIMENSION, 10) : 1536,
        WORKSPACE_VECTOR_STORE_DISTANCE: process.env.WORKSPACE_VECTOR_STORE_DISTANCE || "Cosine",
        WORKSPACE_VECTOR_STORE_ON_DISK: process.env.WORKSPACE_VECTOR_STORE_ON_DISK === "true",
        WORKSPACE_VECTOR_STORE_MAX_VECTORS: process.env.WORKSPACE_VECTOR_STORE_MAX_VECTORS ? parseInt(process.env.WORKSPACE_VECTOR_STORE_MAX_VECTORS, 10) : 1e4,
        // Cross-Tool Memory Sharing Configuration
        CIPHER_USER_ID: process.env.CIPHER_USER_ID,
        CIPHER_PROJECT_NAME: process.env.CIPHER_PROJECT_NAME,
        CIPHER_WORKSPACE_MODE: process.env.CIPHER_WORKSPACE_MODE || "isolated"
      };
      const result = envSchema.safeParse(envToValidate);
      if (!result.success) {
        const errorMsg = `Environment validation failed: ${JSON.stringify(result.error.issues)}`;
        if (isMcpMode) {
          process.stderr.write(`[CIPHER-MCP] ERROR: ${errorMsg}
`);
        } else {
          console.error("Environment validation failed:", result.error.issues);
        }
        return false;
      }
      return result.success;
    };
  }
});

export {
  env,
  validateEnv,
  env_exports,
  init_env
};
