// @byterover/cipher/dist/src/core/chunk-THTEMDGP.js
import {
  logger
} from "./chunk-GPOSIAVD.js";
import {
  init_esm_shims
} from "./chunk-KFOQGTVJ.js";

// src/core/brain/tools/definitions/knowledge_graph/relationship-manager.ts
init_esm_shims();
var relationshipManagerTool = {
  name: "relationship_manager",
  category: "knowledge_graph",
  internal: true,
  description: "Intelligently manage entity relationships with support for complex operations like entity replacement, merging, and bulk updates.",
  version: "1.0.0",
  parameters: {
    type: "object",
    properties: {
      instruction: {
        type: "string",
        description: "Natural language instruction for the relationship operation"
      },
      operation: {
        type: "string",
        enum: [
          "replace_entity",
          "update_relationship",
          "merge_entities",
          "delete_relationships",
          "bulk_update",
          "conditional_update",
          "auto"
        ],
        description: "Type of relationship operation (auto will determine from instruction)",
        default: "auto"
      },
      targets: {
        type: "object",
        description: "Target entities or relationships for the operation",
        properties: {
          sourceEntity: {
            type: "string",
            description: "Source entity name or ID"
          },
          targetEntity: {
            type: "string",
            description: "Target entity name or ID"
          },
          relationshipType: {
            type: "string",
            description: "Type of relationship to operate on"
          },
          criteria: {
            type: "object",
            description: "Criteria for bulk operations",
            additionalProperties: true
          }
        }
      },
      options: {
        type: "object",
        description: "Operation configuration options",
        properties: {
          preserveHistory: {
            type: "boolean",
            description: "Whether to preserve historical data",
            default: true
          },
          confidenceThreshold: {
            type: "number",
            description: "Confidence threshold for automatic operations",
            minimum: 0,
            maximum: 1,
            default: 0.8
          },
          autoResolveConflicts: {
            type: "boolean",
            description: "Whether to automatically resolve conflicts",
            default: true
          },
          bulkLimit: {
            type: "number",
            description: "Maximum entities to affect in bulk operations",
            minimum: 1,
            maximum: 1e3,
            default: 100
          },
          validateRelationships: {
            type: "boolean",
            description: "Whether to validate relationships before operations",
            default: true
          }
        }
      }
    },
    required: ["instruction"]
  },
  handler: async (args, context) => {
    try {
      logger.info("RelationshipManager: Processing relationship operation", {
        instructionLength: args.instruction?.length || 0,
        operation: args.operation,
        hasTargets: !!args.targets
      });
      if (!args.instruction || typeof args.instruction !== "string" || args.instruction.trim().length === 0) {
        throw new Error("Instruction is required and must be a non-empty string");
      }
      const instruction = args.instruction.trim();
      const operation = args.operation || "auto";
      const options = {
        preserveHistory: true,
        confidenceThreshold: 0.8,
        autoResolveConflicts: true,
        bulkLimit: 100,
        validateRelationships: true,
        ...args.options
      };
      const kgManager = context?.services?.knowledgeGraphManager;
      const llmService = context?.services?.llmService;
      if (!kgManager) {
        throw new Error("KnowledgeGraphManager not available in context.services");
      }
      if (!llmService) {
        throw new Error("LLM service is required for intelligent relationship management");
      }
      const graph = kgManager.getGraph();
      if (!graph) {
        throw new Error("Knowledge graph backend is not connected");
      }
      const result = {
        success: false,
        message: "",
        timestamp: (/* @__PURE__ */ new Date()).toISOString(),
        operation: {
          type: "replace_entity",
          description: instruction,
          affectedEntities: 0,
          affectedRelationships: 0,
          confidence: 0
        },
        changes: {
          entitiesCreated: [],
          entitiesUpdated: [],
          entitiesDeleted: [],
          entitiesMerged: [],
          relationshipsCreated: [],
          relationshipsUpdated: [],
          relationshipsDeleted: []
        },
        conflicts: []
      };
      const operationAnalysis = await analyzeRelationshipInstruction(
        instruction,
        args.targets,
        operation,
        llmService
      );
      result.operation = operationAnalysis;
      switch (operationAnalysis.type) {
        case "replace_entity":
          await executeEntityReplacement(operationAnalysis, graph, options, result);
          break;
        case "update_relationship":
          await executeRelationshipUpdate(operationAnalysis, graph, options, result);
          break;
        case "merge_entities":
          await executeEntityMerge(operationAnalysis, graph, options, result);
          break;
        case "delete_relationships":
          await executeRelationshipDeletion(operationAnalysis, graph, options, result);
          break;
        case "bulk_update":
          await executeBulkUpdate(operationAnalysis, graph, options, result);
          break;
        case "conditional_update":
          await executeConditionalUpdate(operationAnalysis, graph, options, result);
          break;
        default:
          throw new Error(`Unsupported operation type: ${operationAnalysis.type}`);
      }
      if (result.conflicts.length > 0 && options.autoResolveConflicts) {
        await resolveOperationConflicts(result.conflicts, graph, llmService);
      }
      result.success = true;
      result.message = `Relationship operation completed: ${result.operation.description}. Affected ${result.operation.affectedEntities} entities and ${result.operation.affectedRelationships} relationships.`;
      logger.info("RelationshipManager: Operation completed successfully", {
        operationType: result.operation.type,
        confidence: result.operation.confidence,
        entitiesAffected: result.operation.affectedEntities,
        relationshipsAffected: result.operation.affectedRelationships,
        conflicts: result.conflicts.length
      });
      return result;
    } catch (error) {
      const errorMessage = error instanceof Error ? error.message : String(error);
      logger.error("RelationshipManager: Operation failed", {
        error: errorMessage,
        instructionLength: args.instruction?.length || 0
      });
      return {
        success: false,
        message: `Relationship operation failed: ${errorMessage}`,
        timestamp: (/* @__PURE__ */ new Date()).toISOString(),
        operation: {
          type: "replace_entity",
          description: args.instruction || "",
          affectedEntities: 0,
          affectedRelationships: 0,
          confidence: 0
        },
        changes: {
          entitiesCreated: [],
          entitiesUpdated: [],
          entitiesDeleted: [],
          entitiesMerged: [],
          relationshipsCreated: [],
          relationshipsUpdated: [],
          relationshipsDeleted: []
        },
        conflicts: [],
        error: errorMessage
      };
    }
  }
};
async function analyzeRelationshipInstruction(instruction, targets, operation, llmService) {
  if (operation !== "auto") {
    return {
      type: operation,
      description: instruction,
      affectedEntities: 0,
      affectedRelationships: 0,
      confidence: 0.9,
      params: targets
    };
  }
  try {
    const analysisPrompt = `
You are an expert at analyzing relationship management instructions for a knowledge graph. Your job is to understand what operation the user wants to perform and extract the relevant parameters.

Instruction: "${instruction}"

Possible operations:
1. replace_entity: Replace one entity with another in all relationships (e.g., "not Long but Nam", "replace Google with Alphabet")
2. update_relationship: Update properties of specific relationships (e.g., "update John's role to senior engineer")
3. merge_entities: Merge two entities and combine their relationships (e.g., "merge Google and Alphabet")
4. delete_relationships: Remove relationships matching criteria (e.g., "delete all connections to deprecated services")
5. bulk_update: Update multiple relationships at once (e.g., "update all employees' status to active")
6. conditional_update: Update relationships based on conditions (e.g., "promote all junior developers to mid-level")

Analyze the instruction and provide:
1. The operation type
2. Entities involved (source, target, etc.)
3. Relationship types or criteria
4. Confidence level (0.0-1.0)
5. Specific parameters for the operation

Examples:
- "not Long but Nam and Trang" -> replace_entity: replace "Long" with "Nam", keep "Trang"
- "John works at Microsoft now, not Google" -> replace_entity in WORKS_AT relationship
- "merge Google and Alphabet" -> merge_entities: combine all relationships
- "update John's role to senior engineer" -> update_relationship: modify role property

Respond with a JSON object:
{
  "operationType": "replace_entity" | "update_relationship" | "merge_entities" | "delete_relationships" | "bulk_update" | "conditional_update",
  "entities": {
    "source": "source entity name",
    "target": "target entity name", 
    "toReplace": "entity to replace",
    "replacement": "replacement entity"
  },
  "relationships": {
    "type": "relationship type",
    "properties": {"property": "value"}
  },
  "criteria": {
    "conditions": "matching criteria"
  },
  "confidence": 0.0-1.0,
  "description": "human-readable description of the operation"
}

Respond ONLY with the JSON object:`;
    const response = await llmService.generate(analysisPrompt);
    const cleanResponse = response.replace(/```json|```/g, "").trim();
    const analysis = JSON.parse(cleanResponse);
    return {
      type: analysis.operationType || "replace_entity",
      description: analysis.description || instruction,
      affectedEntities: 0,
      // Will be calculated during execution
      affectedRelationships: 0,
      // Will be calculated during execution
      confidence: Math.max(0.1, Math.min(1, analysis.confidence || 0.8)),
      params: {
        entities: analysis.entities || {},
        relationships: analysis.relationships || {},
        criteria: analysis.criteria || {}
      }
    };
  } catch (error) {
    logger.warn("RelationshipManager: Failed to analyze instruction with LLM, using fallback", {
      error: error instanceof Error ? error.message : String(error)
    });
    const instructionLower = instruction.toLowerCase();
    if (instructionLower.includes("not") && instructionLower.includes("but")) {
      return {
        type: "replace_entity",
        description: "Entity replacement based on pattern matching",
        affectedEntities: 0,
        affectedRelationships: 0,
        confidence: 0.6
      };
    } else if (instructionLower.includes("merge")) {
      return {
        type: "merge_entities",
        description: "Entity merge based on pattern matching",
        affectedEntities: 0,
        affectedRelationships: 0,
        confidence: 0.6
      };
    } else if (instructionLower.includes("update") || instructionLower.includes("change")) {
      return {
        type: "update_relationship",
        description: "Relationship update based on pattern matching",
        affectedEntities: 0,
        affectedRelationships: 0,
        confidence: 0.6
      };
    } else if (instructionLower.includes("delete") || instructionLower.includes("remove")) {
      return {
        type: "delete_relationships",
        description: "Relationship deletion based on pattern matching",
        affectedEntities: 0,
        affectedRelationships: 0,
        confidence: 0.6
      };
    }
    return {
      type: "replace_entity",
      description: "Default operation",
      affectedEntities: 0,
      affectedRelationships: 0,
      confidence: 0.5
    };
  }
}
async function executeEntityReplacement(operation, graph, options, result) {
  try {
    const params = operation.params || {};
    const entities = params.entities || {};
    const toReplace = entities.toReplace || entities.source;
    const replacement = entities.replacement || entities.target;
    if (!toReplace || !replacement) {
      throw new Error("Both source and target entities must be specified for replacement");
    }
    const sourceEntities = await graph.findNodes({ name: toReplace }, void 0, 10);
    if (sourceEntities.length === 0) {
      result.conflicts.push({
        type: "entity_not_found",
        description: `Entity to replace "${toReplace}" not found in graph`
      });
      return;
    }
    let replacementEntities = await graph.findNodes({ name: replacement }, void 0, 10);
    let replacementEntity;
    if (replacementEntities.length === 0) {
      const entityId = `person_${replacement.replace(/[^a-zA-Z0-9]/g, "_")}_${Date.now()}`;
      replacementEntity = {
        id: entityId,
        labels: ["Person"],
        // Default type, could be inferred
        properties: {
          name: replacement,
          createdAt: (/* @__PURE__ */ new Date()).toISOString(),
          createdBy: "relationship_manager"
        }
      };
      await graph.addNode(replacementEntity);
      result.changes.entitiesCreated.push({
        id: entityId,
        name: replacement,
        type: "Person"
      });
    } else {
      replacementEntity = replacementEntities[0];
    }
    for (const sourceEntity of sourceEntities) {
      const neighbors = await graph.getNeighbors(sourceEntity.id, "both", void 0, 1e3);
      for (const neighbor of neighbors) {
        const edge = neighbor.edge;
        const newEdgeId = `rel_${edge.startNodeId === sourceEntity.id ? replacementEntity.id : edge.startNodeId}_${edge.endNodeId === sourceEntity.id ? replacementEntity.id : edge.endNodeId}_${edge.type}_${Date.now()}`;
        const newEdge = {
          id: newEdgeId,
          type: edge.type,
          startNodeId: edge.startNodeId === sourceEntity.id ? replacementEntity.id : edge.startNodeId,
          endNodeId: edge.endNodeId === sourceEntity.id ? replacementEntity.id : edge.endNodeId,
          properties: {
            ...edge.properties,
            replacedFrom: sourceEntity.id,
            replacedAt: (/* @__PURE__ */ new Date()).toISOString()
          }
        };
        await graph.addEdge(newEdge);
        result.changes.relationshipsCreated.push({
          id: newEdgeId,
          type: edge.type,
          description: `Replaced relationship from ${sourceEntity.properties.name}`
        });
        await graph.deleteEdge(edge.id);
        result.changes.relationshipsDeleted.push({
          id: edge.id,
          reason: "Entity replacement"
        });
        result.operation.affectedRelationships++;
      }
      await graph.deleteNode(sourceEntity.id);
      result.changes.entitiesDeleted.push({
        id: sourceEntity.id,
        name: sourceEntity.properties.name || "Unknown",
        reason: "Replaced by another entity"
      });
      result.operation.affectedEntities++;
    }
  } catch (error) {
    logger.error("RelationshipManager: Error in entity replacement", {
      error: error instanceof Error ? error.message : String(error)
    });
    result.conflicts.push({
      type: "operation_error",
      description: `Entity replacement failed: ${error instanceof Error ? error.message : String(error)}`
    });
  }
}
async function executeRelationshipUpdate(operation, graph, options, result) {
  result.conflicts.push({
    type: "not_implemented",
    description: "Relationship update operation not yet implemented"
  });
}
async function executeEntityMerge(operation, graph, options, result) {
  result.conflicts.push({
    type: "not_implemented",
    description: "Entity merge operation not yet implemented"
  });
}
async function executeRelationshipDeletion(operation, graph, options, result) {
  result.conflicts.push({
    type: "not_implemented",
    description: "Relationship deletion operation not yet implemented"
  });
}
async function executeBulkUpdate(operation, graph, options, result) {
  result.conflicts.push({
    type: "not_implemented",
    description: "Bulk update operation not yet implemented"
  });
}
async function executeConditionalUpdate(operation, graph, options, result) {
  result.conflicts.push({
    type: "not_implemented",
    description: "Conditional update operation not yet implemented"
  });
}
async function resolveOperationConflicts(conflicts, _graph, _llmService) {
  for (const conflict of conflicts) {
    try {
      if (conflict.type === "entity_not_found") {
        conflict.resolution = "Entity will be created if needed";
      } else if (conflict.type === "operation_error") {
        conflict.resolution = "Operation skipped due to error";
      } else if (conflict.type === "not_implemented") {
        conflict.resolution = "Feature will be implemented in future version";
      }
    } catch (error) {
      logger.warn("RelationshipManager: Error resolving conflict", {
        conflictType: conflict.type,
        error: error instanceof Error ? error.message : String(error)
      });
    }
  }
}

export {
  relationshipManagerTool
};
