// @byterover/cipher/dist/src/core/chunk-X6TFDOOF.js
import {
  logger
} from "./chunk-GPOSIAVD.js";
import {
  init_esm_shims
} from "./chunk-KFOQGTVJ.js";

// src/core/brain/llm/messages/history/factory.ts
init_esm_shims();

// src/core/brain/llm/messages/history/database.ts
init_esm_shims();
var MESSAGE_LIMIT = 1e3;
var STORAGE_KEY_PREFIX = "messages:";
var DatabaseHistoryProvider = class {
  storageManager;
  constructor(storageManager) {
    this.storageManager = storageManager;
    const backends = this.storageManager.getBackends();
    if (!backends || typeof backends.database?.get !== "function" || typeof backends.database?.set !== "function") {
      throw new Error(
        "DatabaseHistoryProvider: StorageManager database backend is not compatible."
      );
    }
  }
  getKey(sessionId) {
    return `${STORAGE_KEY_PREFIX}${sessionId}`;
  }
  /**
   * Retrieve up to the most recent `limit` messages for a session, in chronological order (oldest to newest).
   */
  async getHistory(sessionId, limit = MESSAGE_LIMIT) {
    const key = this.getKey(sessionId);
    try {
      const backends = this.storageManager.getBackends();
      const messages = await backends?.database.get(key) || [];
      const result = messages.slice(-limit);
      return result;
    } catch (err) {
      logger.error(
        `DatabaseHistoryProvider.getHistory failed for session ${sessionId}: ${err.message}`
      );
      return [];
    }
  }
  /**
   * Save a message to the session's history, enforcing message limit and order.
   */
  async saveMessage(sessionId, message) {
    const key = this.getKey(sessionId);
    try {
      const backends = this.storageManager.getBackends();
      const messages = await backends?.database.get(key) || [];
      messages.push(message);
      const trimmed = messages.slice(-MESSAGE_LIMIT);
      await backends?.database.set(key, trimmed);
    } catch (err) {
      logger.error(
        `DatabaseHistoryProvider.saveMessage failed for session ${sessionId}: ${err.message}`
      );
    }
  }
  /**
   * Clear all conversation history for a session.
   */
  async clearHistory(sessionId) {
    const key = this.getKey(sessionId);
    try {
      const backends = this.storageManager.getBackends();
      await backends?.database.delete(key);
      logger.info(`DatabaseHistoryProvider: Cleared conversation history for session ${sessionId}`);
    } catch (err) {
      logger.error(
        `DatabaseHistoryProvider.clearHistory failed for session ${sessionId}: ${err.message}`
      );
    }
  }
};

// src/core/brain/llm/messages/history/multi-backend.ts
init_esm_shims();
var MultiBackendHistoryProvider = class {
  constructor(primary, backup, wal, flushIntervalMs = 5e3) {
    this.primary = primary;
    this.backup = backup;
    this.wal = wal;
    this.flushIntervalMs = flushIntervalMs;
    this.startWALFlushWorker();
  }
  walFlushInterval = null;
  async getHistory(sessionId, limit) {
    try {
      return await this.primary.getHistory(sessionId, limit);
    } catch (err) {
      console.error("[MultiBackend] Primary getHistory failed, falling back to backup:", err);
      return await this.backup.getHistory(sessionId, limit);
    }
  }
  async saveMessage(sessionId, message) {
    try {
      await this.primary.saveMessage(sessionId, message);
    } catch (err) {
      console.error("[MultiBackend] Primary saveMessage failed:", err);
      throw err;
    }
    try {
      await this.wal.saveMessage(sessionId, message);
    } catch (err) {
      console.error("[MultiBackend] WAL saveMessage failed:", err);
      throw err;
    }
  }
  async clearHistory(sessionId) {
    try {
      await this.primary.clearHistory(sessionId);
    } catch (err) {
      console.error("[MultiBackend] Primary clearHistory failed:", err);
    }
    try {
      await this.backup.clearHistory(sessionId);
    } catch (err) {
      console.error("[MultiBackend] Backup clearHistory failed:", err);
    }
    try {
      await this.wal.clearHistory(sessionId);
    } catch (err) {
      console.error("[MultiBackend] WAL clearHistory failed:", err);
    }
  }
  startWALFlushWorker() {
    if (this.walFlushInterval) return;
    this.walFlushInterval = setInterval(async () => {
      try {
        const pending = await this.wal.getPendingEntries();
        for (const { sessionId, message } of pending) {
          try {
            await this.backup.saveMessage(sessionId, message);
            await this.wal.markFlushed(sessionId, message);
          } catch (err) {
            console.error("[MultiBackend] Backup saveMessage failed during WAL flush:", err);
          }
        }
      } catch (err) {
        console.error("[MultiBackend] WAL flush worker error:", err);
      }
    }, this.flushIntervalMs);
  }
  async disconnect() {
    if (this.walFlushInterval) {
      clearInterval(this.walFlushInterval);
      this.walFlushInterval = null;
    }
    if (typeof this.primary.disconnect === "function") {
      await this.primary.disconnect();
    }
    if (typeof this.backup.disconnect === "function") {
      await this.backup.disconnect();
    }
    if (typeof this.wal.disconnect === "function") {
      await this.wal.disconnect();
    }
  }
};

// src/core/brain/llm/messages/history/wal.ts
init_esm_shims();
var WALHistoryProvider = class {
  log = [];
  async getHistory(sessionId, _limit) {
    return this.log.filter((e) => e.sessionId === sessionId).map((e) => e.message);
  }
  async saveMessage(sessionId, message) {
    this.log.push({ sessionId, message, flushed: false });
  }
  async clearHistory(sessionId) {
    this.log = this.log.filter((e) => e.sessionId !== sessionId);
  }
  async getPendingEntries() {
    return this.log.filter((e) => !e.flushed).map((e) => ({ sessionId: e.sessionId, message: e.message }));
  }
  async markFlushed(sessionId, message) {
    const entry = this.log.find(
      (e) => e.sessionId === sessionId && e.message === message && !e.flushed
    );
    if (entry) entry.flushed = true;
  }
  async disconnect() {
  }
};

// src/core/brain/llm/messages/history/factory.ts
function createDatabaseHistoryProvider(storage) {
  return new DatabaseHistoryProvider(storage);
}
function createMultiBackendHistoryProvider(primary, backup, wal, flushIntervalMs = 5e3) {
  if (!wal) wal = new WALHistoryProvider();
  return new MultiBackendHistoryProvider(primary, backup, wal, flushIntervalMs);
}

export {
  WALHistoryProvider,
  createDatabaseHistoryProvider,
  createMultiBackendHistoryProvider
};
