// @byterover/cipher/dist/src/core/faiss-PC5L364O.js
import {
  VectorDimensionError,
  VectorStoreConnectionError,
  VectorStoreError
} from "./chunk-JOPES53R.js";
import {
  ERROR_MESSAGES,
  LOG_PREFIXES
} from "./chunk-4JSJAC4K.js";
import "./chunk-7GLOCKG4.js";
import {
  createLogger
} from "./chunk-GPOSIAVD.js";
import {
  env,
  init_env
} from "./chunk-7UQA6M3D.js";
import "./chunk-NUVPDHVT.js";
import {
  init_esm_shims
} from "./chunk-KFOQGTVJ.js";

// src/core/vector_storage/backend/faiss.ts
init_esm_shims();
init_env();
import * as fs from "fs/promises";
import * as path from "path";
import { IndexFlatL2, IndexFlatIP } from "faiss-node";
var FaissBackend = class {
  config;
  collectionName;
  dimension;
  logger;
  connected = false;
  faissIndex;
  // Native FAISS index
  payloads = /* @__PURE__ */ new Map();
  // Store all entry data
  collectionFilePath;
  needsSave = false;
  // Track if data needs to be saved
  constructor(config) {
    this.config = config;
    this.collectionName = config.collectionName;
    this.dimension = config.dimension;
    this.logger = createLogger({
      level: env.CIPHER_LOG_LEVEL || "info"
    });
    this.logger.debug(`${LOG_PREFIXES.FAISS} Backend initialized`, {
      collection: this.collectionName,
      dimension: this.dimension
    });
    this.collectionFilePath = path.join(
      this.config.baseStoragePath || "./faiss-data",
      `${this.collectionName}.json`
    );
    this.config.normalize = false;
    switch (this.config.distance) {
      case "Cosine":
        this.faissIndex = new IndexFlatIP(this.dimension);
        this.config.normalize = true;
        break;
      case "Euclidean":
        this.faissIndex = new IndexFlatL2(this.dimension);
        break;
      case "IP":
        this.faissIndex = new IndexFlatIP(this.dimension);
        break;
      default:
        this.faissIndex = new IndexFlatL2(this.dimension);
        this.logger.warn(
          `${LOG_PREFIXES.FAISS} No distance metric specified or unsupported, defaulting to Euclidean (L2).`
        );
    }
  }
  /**
   * Validate the dimension of a given vector.
   */
  validateDimension(vector, _operation) {
    if (vector.length !== this.dimension) {
      throw new VectorDimensionError(
        `${ERROR_MESSAGES.INVALID_DIMENSION}: expected ${this.dimension}, got ${vector.length}`,
        this.dimension,
        vector.length
      );
    }
  }
  normalizeVector(vector) {
    const magnitude = Math.sqrt(vector.reduce((sum, val) => sum + val * val, 0));
    if (magnitude === 0) {
      this.logger.warn(`${LOG_PREFIXES.FAISS} Zero magnitude vector encountered`);
      return [...vector];
    }
    return vector.map((val) => val / magnitude);
  }
  async connect() {
    if (this.connected) {
      this.logger.debug(`${LOG_PREFIXES.FAISS} Already connected`);
      return;
    }
    this.logger.debug(`${LOG_PREFIXES.FAISS} Connecting to in-memory FAISS store`);
    try {
      this.logger.debug(`${LOG_PREFIXES.FAISS} Successfully connected`);
      await this.loadData();
      this.connected = true;
      this.logger.debug(`${LOG_PREFIXES.FAISS} Data loaded from ${this.config.baseStoragePath}`);
    } catch (error) {
      this.logger.error(`${LOG_PREFIXES.FAISS} Connection failed`, { error });
      throw new VectorStoreConnectionError(
        ERROR_MESSAGES.CONNECTION_FAILED,
        "faiss",
        error
      );
    }
  }
  async insert(vectors, ids, payloads) {
    if (!this.connected) {
      throw new VectorStoreError(ERROR_MESSAGES.NOT_CONNECTED, "insert");
    }
    if (vectors.length !== ids.length || vectors.length !== payloads.length) {
      throw new VectorStoreError("Vectors, IDs, and payloads must have the same length", "insert");
    }
    const newEntries = [];
    for (let i = 0; i < vectors.length; i++) {
      let vector = vectors[i];
      const id = ids[i];
      const payload = payloads[i];
      if (this.config.normalize) {
        vector = this.normalizeVector(vector);
      }
      this.validateDimension(vector, "insert");
      const entry = { id, vector, payload };
      this.payloads.set(id, entry);
      newEntries.push(entry);
    }
    await this.saveData(newEntries);
    this.logger.debug(`${LOG_PREFIXES.FAISS} Inserted ${vectors.length} vectors`);
  }
  async search(query, limit = 2, filters) {
    if (!this.connected) {
      throw new VectorStoreError(ERROR_MESSAGES.NOT_CONNECTED, "search");
    }
    this.validateDimension(query, "search");
    if (this.config.normalize) {
      query = this.normalizeVector(query);
    }
    if (!this.faissIndex) {
      throw new VectorStoreError("FAISS index not initialized", "search");
    }
    limit = Math.min(limit, this.faissIndex.ntotal());
    const { distances, labels } = this.faissIndex.search(query, limit);
    const results = [];
    for (let i = 0; i < labels.length; i++) {
      const idx = labels[i];
      if (idx === void 0) {
        this.logger.warn(`${LOG_PREFIXES.FAISS} Search returned undefined label for index ${i}`);
        continue;
      }
      if (idx < 0) continue;
      const entry = this.payloads.get(idx);
      if (!entry) continue;
      const distanceRow = distances[0];
      if (!distanceRow || typeof distanceRow[i] !== "number") continue;
      const distance = distanceRow[i];
      results.push({
        id: entry.id,
        vector: entry.vector,
        payload: entry.payload,
        score: this.config.distance === "Cosine" || this.config.distance === "IP" ? distance : 1 / (1 + distance)
      });
    }
    return results;
  }
  async get(vectorId) {
    if (!this.connected) {
      throw new VectorStoreError(ERROR_MESSAGES.NOT_CONNECTED, "get");
    }
    try {
      const payload = this.payloads.get(vectorId);
      if (!payload) return null;
      return {
        id: vectorId,
        vector: payload.vector,
        payload: payload.payload,
        // Corrected to return only the user-defined payload
        score: 1
        // For direct retrieval, score is 1.0
      };
    } catch (error) {
      this.logger.error(`${LOG_PREFIXES.FAISS} Get failed`, { error });
      throw new VectorStoreError("Failed to retrieve vector", "get", error);
    }
  }
  async update(vectorId, vector, payload) {
    if (!this.connected) {
      throw new VectorStoreError(ERROR_MESSAGES.NOT_CONNECTED, "update");
    }
    this.validateDimension(vector, "update");
    try {
      if (!this.payloads.has(vectorId)) {
        throw new VectorStoreError(`Vector with ID ${vectorId} not found for update`, "update");
      }
      const oldEntry = this.payloads.get(vectorId);
      if (oldEntry && this.faissIndex) {
      }
      const updatedEntry = { ...this.payloads.get(vectorId), vector, payload };
      this.payloads.set(vectorId, updatedEntry);
      this.logger.debug(`${LOG_PREFIXES.FAISS} Updated vector ${vectorId}`);
      await this.saveData([{ id: vectorId, vector, payload }]);
    } catch (error) {
      this.logger.error(`${LOG_PREFIXES.FAISS} Update failed`, { error });
      throw new VectorStoreError("Failed to update vector", "update", error);
    }
  }
  async delete(vectorId) {
    if (!this.connected) {
      throw new VectorStoreError(ERROR_MESSAGES.NOT_CONNECTED, "delete");
    }
    try {
      const deleted = this.payloads.delete(vectorId);
      if (!deleted) {
        this.logger.warn(`${LOG_PREFIXES.FAISS} Vector with ID ${vectorId} not found for deletion`);
      } else {
        this.logger.debug(`${LOG_PREFIXES.FAISS} Deleted vector ${vectorId}`);
        await this.removeFromFile(vectorId);
      }
    } catch (error) {
      this.logger.error(`${LOG_PREFIXES.FAISS} Delete failed`, { error });
      throw new VectorStoreError("Failed to delete vector", "delete", error);
    }
  }
  async deleteCollection() {
    if (!this.connected) {
      throw new VectorStoreError(ERROR_MESSAGES.NOT_CONNECTED, "deleteCollection");
    }
    try {
      this.payloads.clear();
      this.logger.info(`${LOG_PREFIXES.FAISS} Cleared collection ${this.collectionName}`);
      if (this.config.baseStoragePath) {
        try {
          await fs.unlink(this.collectionFilePath);
          this.logger.info(
            `${LOG_PREFIXES.FAISS} Deleted collection file ${this.collectionFilePath}`
          );
        } catch (fileError) {
          if (fileError.code !== "ENOENT") {
            this.logger.warn(
              `${LOG_PREFIXES.FAISS} Failed to delete collection file: ${fileError}`
            );
          }
        }
      }
    } catch (error) {
      this.logger.error(`${LOG_PREFIXES.FAISS} Delete collection failed`, { error });
      throw new VectorStoreError("Failed to delete collection", "deleteCollection", error);
    }
  }
  async list(filters, limit = 1e4) {
    if (!this.connected) {
      throw new VectorStoreError(ERROR_MESSAGES.NOT_CONNECTED, "list");
    }
    try {
      let results = Array.from(this.payloads.entries()).map((entry) => ({
        id: entry[0],
        vector: entry[1].vector,
        payload: entry[1].payload,
        // Corrected to return only the user-defined payload
        score: 1
        // Score is 1.0 for listing all
      }));
      if (filters) {
        results = results.filter(
          (entry) => Object.entries(filters).every(([key, value]) => {
            const payloadValue = entry.payload[key];
            if (typeof value === "object" && value !== null) {
              if ("any" in value && Array.isArray(value.any)) {
                return value.any.includes(payloadValue);
              }
              if (typeof payloadValue === "number") {
                if ("gte" in value && payloadValue < value.gte) return false;
                if ("lte" in value && payloadValue > value.lte) return false;
                if ("gt" in value && payloadValue <= value.gt) return false;
                if ("lt" in value && payloadValue >= value.lt) return false;
              }
              return true;
            }
            return payloadValue === value;
          })
        );
      }
      const paginatedResults = results.slice(0, limit);
      return [paginatedResults, paginatedResults.length];
    } catch (error) {
      this.logger.error(`${LOG_PREFIXES.FAISS} List failed`, { error });
      throw new VectorStoreError("Failed to list vectors", "list", error);
    }
  }
  async disconnect() {
    if (this.connected) {
      this.connected = false;
      this.logger.info(`${LOG_PREFIXES.FAISS} Disconnected`);
    }
  }
  isConnected() {
    return this.connected;
  }
  getBackendType() {
    return "faiss";
  }
  getDimension() {
    return this.dimension;
  }
  getCollectionName() {
    return this.collectionName;
  }
  async listCollections() {
    if (!this.connected) {
      throw new VectorStoreError(ERROR_MESSAGES.NOT_CONNECTED, "listCollections");
    }
    return [this.collectionName];
  }
  /**
   * Saves or updates specific vector data in the JSON file.
   */
  async saveData(specificEntries) {
    if (!this.config.baseStoragePath) {
      this.logger.debug(`${LOG_PREFIXES.FAISS} No storage path configured, skipping save.`);
      return;
    }
    try {
      const baseDir = path.dirname(this.collectionFilePath);
      await fs.mkdir(baseDir, { recursive: true });
      let existingData = [];
      try {
        const fileContent = await fs.readFile(this.collectionFilePath, "utf8");
        existingData = JSON.parse(fileContent);
      } catch (error) {
        if (error.code !== "ENOENT") {
          throw error;
        }
      }
      if (specificEntries) {
        for (const newEntry of specificEntries) {
          const existingIndex = existingData.findIndex((entry) => entry.id === newEntry.id);
          if (existingIndex >= 0) {
            existingData[existingIndex] = newEntry;
          } else {
            existingData.push(newEntry);
          }
        }
      } else {
        existingData = Array.from(this.payloads.entries()).map(([id, entry]) => ({
          id,
          vector: entry.vector,
          payload: entry.payload
        }));
      }
      await fs.writeFile(this.collectionFilePath, JSON.stringify(existingData, null, 2), "utf8");
      this.needsSave = false;
      if (specificEntries) {
        this.logger.debug(
          `${LOG_PREFIXES.FAISS} Updated ${specificEntries.length} specific entries in ${this.collectionFilePath}`
        );
      } else {
        this.logger.debug(
          `${LOG_PREFIXES.FAISS} Full save of ${existingData.length} vectors to ${this.collectionFilePath}`
        );
      }
    } catch (error) {
      this.logger.error(`${LOG_PREFIXES.FAISS} Failed to save data to ${this.collectionFilePath}`, {
        error
      });
      throw error;
    }
  }
  /**
   * Removes specific vector data from the JSON file.
   */
  async removeFromFile(vectorId) {
    if (!this.config.baseStoragePath) {
      return;
    }
    try {
      let existingData = [];
      try {
        const fileContent = await fs.readFile(this.collectionFilePath, "utf8");
        existingData = JSON.parse(fileContent);
      } catch (error) {
        if (error.code === "ENOENT") {
          return;
        }
        throw error;
      }
      const filteredData = existingData.filter((entry) => entry.id !== vectorId);
      if (filteredData.length !== existingData.length) {
        await fs.writeFile(this.collectionFilePath, JSON.stringify(filteredData, null, 2), "utf8");
        this.logger.debug(
          `${LOG_PREFIXES.FAISS} Removed vector ID ${vectorId} from ${this.collectionFilePath}`
        );
      }
    } catch (error) {
      this.logger.error(
        `${LOG_PREFIXES.FAISS} Failed to remove vector ID ${vectorId} from ${this.collectionFilePath}`,
        {
          error
        }
      );
      throw error;
    }
  }
  /**
   * Loads vector data from a single JSON file into memory.
   */
  async loadData() {
    if (!this.config.baseStoragePath) {
      this.logger.debug(`${LOG_PREFIXES.FAISS} No storage path configured, skipping load.`);
      return;
    }
    try {
      this.payloads.clear();
      const fileContent = await fs.readFile(this.collectionFilePath, "utf8");
      const entries = JSON.parse(fileContent);
      for (const entry of entries) {
        this.payloads.set(entry.id, entry);
        this.faissIndex?.add(entry.vector);
        this.logger.debug(`${LOG_PREFIXES.FAISS} Loaded vector ID ${entry.id}`);
      }
      this.needsSave = false;
      this.logger.debug(
        `${LOG_PREFIXES.FAISS} Loaded ${entries.length} vectors from ${this.collectionFilePath}`
      );
    } catch (error) {
      if (error.code === "ENOENT") {
        this.logger.debug(
          `${LOG_PREFIXES.FAISS} Collection file not found at ${this.collectionFilePath}. Starting fresh.`
        );
      } else {
        this.logger.error(
          `${LOG_PREFIXES.FAISS} Failed to load data from ${this.collectionFilePath}`,
          {
            error
          }
        );
        throw error;
      }
    }
  }
};
export {
  FaissBackend
};
