// @byterover/cipher/dist/src/core/in-memory-UU6M4DWS.js
import {
  VectorDimensionError,
  VectorStoreError
} from "./chunk-JOPES53R.js";
import {
  DEFAULTS,
  ERROR_MESSAGES,
  LOG_PREFIXES
} from "./chunk-4JSJAC4K.js";
import "./chunk-7GLOCKG4.js";
import {
  createLogger
} from "./chunk-GPOSIAVD.js";
import "./chunk-7UQA6M3D.js";
import "./chunk-NUVPDHVT.js";
import {
  init_esm_shims
} from "./chunk-KFOQGTVJ.js";

// src/core/vector_storage/backend/in-memory.ts
init_esm_shims();
var InMemoryBackend = class {
  config;
  collectionName;
  dimension;
  maxVectors;
  logger;
  connected = false;
  // In-memory storage
  vectors = /* @__PURE__ */ new Map();
  constructor(config) {
    this.config = config;
    this.collectionName = config.collectionName;
    this.dimension = config.dimension;
    this.maxVectors = config.maxVectors || 1e4;
    this.logger = createLogger({
      level: process.env.LOG_LEVEL || "info"
    });
    this.logger.debug(`${LOG_PREFIXES.MEMORY} Initialized`, {
      collection: this.collectionName,
      dimension: this.dimension,
      maxVectors: this.maxVectors
    });
  }
  /**
   * Calculate cosine similarity between two vectors
   */
  cosineSimilarity(a, b) {
    let dotProduct = 0;
    let normA = 0;
    let normB = 0;
    const minLength = Math.min(a.length, b.length);
    for (let i = 0; i < minLength; i++) {
      const aVal = a[i];
      const bVal = b[i];
      dotProduct += aVal * bVal;
      normA += aVal * aVal;
      normB += bVal * bVal;
    }
    normA = Math.sqrt(normA);
    normB = Math.sqrt(normB);
    if (normA === 0 || normB === 0) {
      return 0;
    }
    return dotProduct / (normA * normB);
  }
  /**
   * Deep clone an object to prevent reference issues
   */
  deepClone(obj) {
    if (obj === null || typeof obj !== "object") {
      return obj;
    }
    if (Array.isArray(obj)) {
      return obj.map((item) => this.deepClone(item));
    }
    const cloned = {};
    for (const key in obj) {
      if (Object.prototype.hasOwnProperty.call(obj, key)) {
        cloned[key] = this.deepClone(obj[key]);
      }
    }
    return cloned;
  }
  /**
   * Check if a vector entry matches the given filters
   */
  matchesFilters(entry, filters) {
    if (!filters) return true;
    for (const [key, value] of Object.entries(filters)) {
      const payloadValue = entry.payload[key];
      if (value === null || value === void 0) continue;
      if (typeof value === "object" && !Array.isArray(value) && ("gte" in value || "gt" in value || "lte" in value || "lt" in value)) {
        if (typeof payloadValue !== "number") return false;
        if ("gte" in value && payloadValue < value.gte) return false;
        if ("gt" in value && payloadValue <= value.gt) return false;
        if ("lte" in value && payloadValue > value.lte) return false;
        if ("lt" in value && payloadValue >= value.lt) return false;
      } else if (typeof value === "object" && !Array.isArray(value) && ("any" in value || "all" in value)) {
        if ("any" in value && Array.isArray(value.any)) {
          if (!value.any.includes(payloadValue)) return false;
        }
      } else {
        if (payloadValue !== value) return false;
      }
    }
    return true;
  }
  /**
   * Validate vector dimension
   */
  validateDimension(vector, _operation) {
    if (vector.length !== this.dimension) {
      throw new VectorDimensionError(
        `${ERROR_MESSAGES.INVALID_DIMENSION}: expected ${this.dimension}, got ${vector.length}`,
        this.dimension,
        vector.length
      );
    }
  }
  // VectorStore implementation
  async insert(vectors, ids, payloads) {
    if (!this.connected) {
      throw new VectorStoreError(ERROR_MESSAGES.NOT_CONNECTED, "insert");
    }
    if (vectors.length !== ids.length || vectors.length !== payloads.length) {
      throw new VectorStoreError("Vectors, IDs, and payloads must have the same length", "insert");
    }
    if (this.vectors.size + vectors.length > this.maxVectors) {
      throw new VectorStoreError(
        `Cannot insert ${vectors.length} vectors. Would exceed max limit of ${this.maxVectors}`,
        "insert"
      );
    }
    for (let i = 0; i < vectors.length; i++) {
      const vector = vectors[i];
      const id = ids[i];
      const payload = payloads[i];
      if (!vector || typeof id !== "number" || !Number.isInteger(id) || !payload) {
        throw new VectorStoreError(
          `Invalid input at index ${i}: vector, integer id, and payload are required`,
          "insert"
        );
      }
      this.validateDimension(vector, "insert");
      this.vectors.set(id, {
        id,
        vector: [...vector],
        // Clone to prevent external modification
        payload: this.deepClone(payload)
        // Deep clone payload
      });
    }
    this.logger.debug(`${LOG_PREFIXES.INDEX} Inserted ${vectors.length} vectors`);
  }
  async search(query, limit = DEFAULTS.SEARCH_LIMIT, filters) {
    if (!this.connected) {
      throw new VectorStoreError(ERROR_MESSAGES.NOT_CONNECTED, "search");
    }
    this.validateDimension(query, "search");
    const results = [];
    for (const entry of this.vectors.values()) {
      if (!this.matchesFilters(entry, filters)) {
        continue;
      }
      const score = this.cosineSimilarity(query, entry.vector);
      results.push({ entry, score });
    }
    results.sort((a, b) => b.score - a.score);
    const topResults = results.slice(0, limit);
    const formattedResults = topResults.map(({ entry, score }) => ({
      id: entry.id,
      vector: [...entry.vector],
      payload: this.deepClone(entry.payload),
      score
    }));
    this.logger.debug(`${LOG_PREFIXES.SEARCH} Found ${formattedResults.length} results`);
    return formattedResults;
  }
  async get(vectorId) {
    if (!this.connected) {
      throw new VectorStoreError(ERROR_MESSAGES.NOT_CONNECTED, "get");
    }
    const entry = this.vectors.get(vectorId);
    if (!entry) {
      return null;
    }
    return {
      id: entry.id,
      vector: [...entry.vector],
      // Clone vector
      payload: this.deepClone(entry.payload),
      // Deep clone payload
      score: 1
      // Perfect match for direct retrieval
    };
  }
  async update(vectorId, vector, payload) {
    if (!this.connected) {
      throw new VectorStoreError(ERROR_MESSAGES.NOT_CONNECTED, "update");
    }
    this.validateDimension(vector, "update");
    if (!this.vectors.has(vectorId)) {
      throw new VectorStoreError(`Vector with ID ${vectorId} not found`, "update");
    }
    this.vectors.set(vectorId, {
      id: vectorId,
      vector: [...vector],
      payload: this.deepClone(payload)
    });
    this.logger.debug(`${LOG_PREFIXES.BACKEND} Updated vector ${vectorId}`);
  }
  async delete(vectorId) {
    if (!this.connected) {
      throw new VectorStoreError(ERROR_MESSAGES.NOT_CONNECTED, "delete");
    }
    if (!this.vectors.has(vectorId)) {
      this.logger.warn(`${LOG_PREFIXES.BACKEND} Vector ${vectorId} not found for deletion`);
      return;
    }
    this.vectors.delete(vectorId);
    this.logger.debug(`${LOG_PREFIXES.BACKEND} Deleted vector ${vectorId}`);
  }
  async deleteCollection() {
    if (!this.connected) {
      throw new VectorStoreError(ERROR_MESSAGES.NOT_CONNECTED, "deleteCollection");
    }
    const count = this.vectors.size;
    this.vectors.clear();
    this.logger.info(
      `${LOG_PREFIXES.BACKEND} Deleted collection ${this.collectionName} with ${count} vectors`
    );
  }
  async list(filters, limit = 100) {
    if (!this.connected) {
      throw new VectorStoreError(ERROR_MESSAGES.NOT_CONNECTED, "list");
    }
    const results = [];
    let count = 0;
    for (const entry of this.vectors.values()) {
      if (this.matchesFilters(entry, filters)) {
        count++;
        if (results.length < limit) {
          results.push({
            id: entry.id,
            vector: [...entry.vector],
            payload: this.deepClone(entry.payload),
            score: 1
            // Default score for list operations
          });
        }
      }
    }
    this.logger.info(`${LOG_PREFIXES.BACKEND} Listed ${results.length} of ${count} vectors`);
    return [results, count];
  }
  async connect() {
    if (this.connected) {
      this.logger.debug(`${LOG_PREFIXES.MEMORY} Already connected`);
      return;
    }
    this.connected = true;
    this.logger.debug(`${LOG_PREFIXES.MEMORY} Connected (in-memory)`);
  }
  async disconnect() {
    if (!this.connected) {
      this.logger.debug(`${LOG_PREFIXES.MEMORY} Already disconnected`);
      return;
    }
    this.connected = false;
    this.vectors.clear();
    this.logger.info(`${LOG_PREFIXES.MEMORY} Disconnected and cleared data`);
  }
  isConnected() {
    return this.connected;
  }
  getBackendType() {
    return "in-memory";
  }
  getDimension() {
    return this.dimension;
  }
  getCollectionName() {
    return this.collectionName;
  }
};
export {
  InMemoryBackend
};
