// @byterover/cipher/dist/src/core/index.js
import {
  Neo4jBackend
} from "./chunk-UDUOGZVQ.js";
import {
  InMemoryBackend as InMemoryBackend2
} from "./chunk-TCBVDNOG.js";
import {
  createDefaultStorage,
  createStorageBackends,
  createStorageFromEnv,
  isStorageFactory
} from "./chunk-RPNUGONV.js";
import {
  AwsService,
  AzureService,
  CONNECTION_MODES,
  CantInferProviderError,
  ContextManager,
  ConversationSession,
  DEFAULT_CONNECTION_MODE,
  DEFAULT_TIMEOUT_MS,
  ENV_VARS,
  ERROR_MESSAGES,
  EmbeddingConfigSchema,
  KnowledgeGraphManager,
  LLMConfigSchema,
  LMStudioService,
  LOG_PREFIXES,
  MAX_TIMEOUT_MS,
  MCPClient,
  MCPManager,
  MIN_TIMEOUT_MS,
  OllamaService,
  OpenRouterService,
  ReasoningContentDetector,
  SearchContextManager,
  ServerConfigsSchema,
  SessionManager,
  TRANSPORT_TYPES,
  createAgentServices,
  createDefaultKnowledgeGraph,
  createEnhancedAgentServices,
  createInMemoryKnowledgeGraph,
  createKnowledgeGraph,
  createKnowledgeGraphFromEnv,
  createNeo4jKnowledgeGraph,
  getKnowledgeGraphConfigFromEnv,
  isKnowledgeGraphEnabled,
  isKnowledgeGraphFactory,
  isNeo4jConfigAvailable,
  shouldEnableLazyLoading
} from "./chunk-4TOIWRCS.js";
import {
  BACKEND_TYPES as BACKEND_TYPES3,
  DEFAULTS as DEFAULTS3,
  ERROR_MESSAGES as ERROR_MESSAGES3,
  EdgeNotFoundError,
  GraphValidationError,
  INDEX_TEMPLATES,
  InvalidQueryError,
  KnowledgeGraphConnectionError,
  KnowledgeGraphError,
  LOG_PREFIXES as LOG_PREFIXES3,
  METRICS_EVENTS,
  NodeNotFoundError,
  QUERY_TEMPLATES,
  SCHEMA,
  TIMEOUTS as TIMEOUTS2
} from "./chunk-Z6KHKTWP.js";
import {
  BackendConfigSchema,
  InMemoryBackendSchema,
  KnowledgeGraphEnvConfigSchema,
  KnowledgeGraphSchema,
  Neo4jBackendSchema,
  parseKnowledgeGraphConfig,
  parseKnowledgeGraphConfigFromEnv,
  validateKnowledgeGraphConfig
} from "./chunk-6PWZHI5T.js";
import {
  StorageManager,
  StorageSchema
} from "./chunk-MMF5ASL2.js";
import "./chunk-JVU3SARR.js";
import "./chunk-ZUCXE4PW.js";
import "./chunk-35RZTXDH.js";
import "./chunk-TKUZPR6R.js";
import "./chunk-77KNNZZ6.js";
import "./chunk-JU4Q2IV6.js";
import "./chunk-THTEMDGP.js";
import "./chunk-DU36577C.js";
import {
  DualCollectionVectorManager,
  VectorStoreManager,
  createDefaultVectorStore,
  createDualCollectionVectorStoreFromEnv,
  createMultiCollectionVectorStoreFromEnv,
  createVectorStore,
  createVectorStoreFromEnv,
  createWorkspaceVectorStoreFromEnv,
  getVectorStoreConfigFromEnv,
  getWorkspaceVectorStoreConfigFromEnv,
  isVectorStoreFactory
} from "./chunk-IG2VBOT3.js";
import "./chunk-SUYRJ7QN.js";
import "./chunk-F6PMSUBO.js";
import "./chunk-RVR2H76A.js";
import "./chunk-VKPE7HMS.js";
import "./chunk-36DMOB2T.js";
import {
  CollectionNotFoundError,
  VectorDimensionError,
  VectorStoreConnectionError,
  VectorStoreError
} from "./chunk-JOPES53R.js";
import {
  BACKEND_TYPES as BACKEND_TYPES2,
  DEFAULTS as DEFAULTS2,
  DISTANCE_METRICS
} from "./chunk-4JSJAC4K.js";
import "./chunk-RNE2AKGD.js";
import "./chunk-X6TFDOOF.js";
import {
  InMemoryBackend
} from "./chunk-MNM4W6RD.js";
import {
  BACKEND_TYPES,
  DEFAULTS,
  ERROR_MESSAGES as ERROR_MESSAGES2,
  HEALTH_CHECK,
  LOG_PREFIXES as LOG_PREFIXES2,
  TIMEOUTS
} from "./chunk-WZZEPDWF.js";
import {
  StorageConnectionError,
  StorageError,
  StorageNotFoundError
} from "./chunk-C3KGEYH3.js";
import "./chunk-7GLOCKG4.js";
import {
  Logger,
  createLogger,
  getGlobalLogLevel,
  logger,
  setGlobalLogLevel
} from "./chunk-GPOSIAVD.js";
import {
  env,
  init_env,
  validateEnv
} from "./chunk-7UQA6M3D.js";
import {
  external_exports,
  init_zod
} from "./chunk-NUVPDHVT.js";
import {
  __export,
  init_esm_shims
} from "./chunk-KFOQGTVJ.js";

// src/core/index.ts
init_esm_shims();

// src/core/mcp/index.ts
init_esm_shims();

// src/core/brain/index.ts
init_esm_shims();

// src/core/brain/memAgent/index.ts
init_esm_shims();

// src/core/brain/memAgent/agent.ts
init_esm_shims();
var requiredServices = [
  "mcpManager",
  "promptManager",
  "stateManager",
  "sessionManager",
  "internalToolManager",
  "unifiedToolManager"
];
var MemAgent = class {
  mcpManager;
  promptManager;
  stateManager;
  sessionManager;
  internalToolManager;
  // Will be properly typed later
  unifiedToolManager;
  // Will be properly typed later
  services;
  defaultSession = null;
  currentDefaultSessionId = "default";
  currentActiveSessionId = "default";
  // Will be set properly in constructor
  isStarted = false;
  isStopped = false;
  config;
  appMode = null;
  constructor(config, appMode) {
    this.config = config;
    this.appMode = appMode || null;
    if (appMode === "cli") {
      this.currentActiveSessionId = this.currentDefaultSessionId;
    } else {
      this.currentActiveSessionId = this.generateUniqueSessionId();
    }
    if (appMode !== "cli") {
      logger.debug("MemAgent created");
    }
  }
  /**
   * Generate a unique session ID for API/MCP modes
   * CLI mode uses the default session for persistence
   */
  generateUniqueSessionId() {
    const timestamp = Date.now();
    const random = Math.random().toString(36).substring(2, 8);
    return `session-${timestamp}-${random}`;
  }
  /**
   * Start the MemAgent
   */
  async start() {
    if (this.isStarted) {
      throw new Error("MemAgent is already started");
    }
    try {
      if (this.appMode !== "cli") {
        logger.debug("Starting MemAgent...");
      }
      const useLazyLoading = shouldEnableLazyLoading({
        appMode: this.appMode || void 0
      });
      let services;
      if (useLazyLoading) {
        logger.debug("MemAgent: Using enhanced services with lazy loading");
        services = await createEnhancedAgentServices(this.config, {
          enableLazyLoading: true,
          ...this.appMode && { appMode: this.appMode }
        });
      } else {
        logger.debug("MemAgent: Using standard services");
        services = await createAgentServices(this.config, this.appMode || void 0);
      }
      for (const service of requiredServices) {
        if (!services[service]) {
          throw new Error(`Required service ${service} is missing during agent start`);
        }
      }
      Object.assign(this, {
        mcpManager: services.mcpManager,
        promptManager: services.promptManager,
        stateManager: services.stateManager,
        sessionManager: services.sessionManager,
        internalToolManager: services.internalToolManager,
        unifiedToolManager: services.unifiedToolManager,
        services
      });
      this.isStarted = true;
      if (this.appMode !== "cli") {
        logger.debug("MemAgent started successfully");
      }
    } catch (error) {
      logger.error("Failed to start MemAgent:", error);
      throw error;
    }
  }
  /**
   * Stop the MemAgent
   */
  async stop() {
    if (this.isStopped) {
      logger.warn("MemAgent is already stopped");
      return;
    }
    if (!this.isStarted) {
      throw new Error("MemAgent must be started before stopping");
    }
    try {
      logger.info("Stopping MemAgent...");
      const shutdownErrors = [];
      try {
        if (this.sessionManager) {
          logger.info("Saving all sessions before shutdown...");
          await this.sessionManager.shutdown();
          logger.debug("SessionManager shutdown completed");
        }
      } catch (error) {
        const err = error instanceof Error ? error : new Error(String(error));
        shutdownErrors.push(new Error(`SessionManager shutdown failed: ${err.message}`));
      }
      try {
        if (this.mcpManager) {
          await this.mcpManager.disconnectAll();
          logger.debug("MCPManager disconnected all clients successfully");
        }
      } catch (error) {
        const err = error instanceof Error ? error : new Error(String(error));
        shutdownErrors.push(new Error(`MCPManager disconnect failed: ${err.message}`));
      }
      this.isStopped = true;
      this.isStarted = false;
      if (shutdownErrors.length > 0) {
        const errorMessages = shutdownErrors.map((e) => e.message).join("; ");
        logger.warn(`MemAgent stopped with some errors: ${errorMessages}`);
      } else {
        logger.info("MemAgent stopped successfully.");
      }
    } catch (error) {
      logger.error("Failed to stop MemAgent:", error);
      throw error;
    }
  }
  /**
   * Get the status of the MemAgent
   */
  getIsStarted() {
    return this.isStarted;
  }
  /**
   * Get the status of the MemAgent
   */
  getIsStopped() {
    return this.isStopped;
  }
  ensureStarted() {
    if (this.isStopped) {
      throw new Error("MemAgent has been stopped and cannot be used");
    }
    if (!this.isStarted) {
      throw new Error("MemAgent must be started before use. Call agent.start() first.");
    }
  }
  /**
   * Run the MemAgent
   */
  async run(userInput, imageDataInput, sessionId, stream = false, options) {
    this.ensureStarted();
    try {
      let session;
      if (sessionId) {
        session = await this.sessionManager.getSession(sessionId) ?? await this.sessionManager.createSession(sessionId);
        this.currentActiveSessionId = sessionId;
      } else {
        session = await this.sessionManager.getSession(this.currentActiveSessionId) ?? await this.sessionManager.createSession(this.currentActiveSessionId);
      }
      logger.debug(`MemAgent.run: using session ${session.id}`);
      const { response, backgroundOperations } = await session.run(
        userInput,
        imageDataInput,
        stream,
        {
          ...options?.memoryMetadata !== void 0 && { memoryMetadata: options.memoryMetadata },
          ...options?.sessionOptions !== void 0 && {
            contextOverrides: options.sessionOptions
          }
        }
      );
      const finalResponse = response && response.trim() !== "" ? response : null;
      return { response: finalResponse, backgroundOperations };
    } catch (error) {
      logger.error("MemAgent.run: error", error);
      throw error;
    }
  }
  async createSession(sessionId) {
    this.ensureStarted();
    logger.debug(`MemAgent: Creating session with ID: ${sessionId || "auto-generated"}`);
    const session = await this.sessionManager.createSession(sessionId);
    logger.info(`MemAgent: Created session: ${session.id}`);
    return session;
  }
  async getSession(sessionId) {
    this.ensureStarted();
    let session = await this.sessionManager.getSession(sessionId);
    if (!session && this.appMode === "cli" && sessionId === "default") {
      logger.debug(
        "MemAgent: Default session not found, creating new default session for CLI mode"
      );
      try {
        session = await this.sessionManager.createSession("default");
        logger.info("MemAgent: Created default session for CLI mode");
      } catch (error) {
        logger.error("MemAgent: Failed to create default session:", error);
        return null;
      }
    }
    return session;
  }
  /**
   * Get the current active session ID
   */
  getCurrentSessionId() {
    this.ensureStarted();
    return this.currentActiveSessionId;
  }
  /**
   * Load conversation history for a specific session
   */
  async loadSessionHistory(sessionId) {
    this.ensureStarted();
    const session = await this.sessionManager.getSession(sessionId);
    if (!session) {
      throw new Error(`Session not found: ${sessionId}`);
    }
    try {
      await session.refreshConversationHistory();
    } catch (error) {
      logger.warn(`MemAgent: Failed to load conversation history for session ${sessionId}:`, error);
      throw error;
    }
  }
  /**
   * Load (switch to) a specific session
   */
  async loadSession(sessionId) {
    this.ensureStarted();
    let session = await this.sessionManager.getSession(sessionId);
    if (!session) {
      throw new Error(`Session not found: ${sessionId}`);
    }
    try {
      logger.debug(`MemAgent: Loading session ${sessionId}...`);
      await session.init();
      await session.refreshConversationHistory();
      logger.info(`MemAgent: Successfully loaded session ${sessionId} with conversation history`);
    } catch (error) {
      logger.warn(`MemAgent: Failed to initialize session ${sessionId}:`, error);
    }
    this.currentActiveSessionId = sessionId;
    logger.debug(`MemAgent: Switched to session ${sessionId}`);
    return session;
  }
  /**
   * Get all active session IDs
   */
  async listSessions() {
    this.ensureStarted();
    return await this.sessionManager.getActiveSessionIds();
  }
  /**
   * Remove a session
   */
  async removeSession(sessionId) {
    this.ensureStarted();
    if (sessionId === this.currentActiveSessionId) {
      throw new Error(
        "Cannot remove the currently active session. Switch to another session first."
      );
    }
    return await this.sessionManager.removeSession(sessionId);
  }
  /**
   * Get session metadata including message count
   */
  async getSessionMetadata(sessionId) {
    this.ensureStarted();
    const session = await this.sessionManager.getSession(sessionId);
    if (!session) {
      return null;
    }
    let messageCount = 0;
    try {
      if (!session.getContextManager) {
        await session.init();
      }
      const history = await session.getConversationHistory();
      messageCount = history.length;
    } catch (error) {
      logger.warn(`Failed to get message count for session ${sessionId}:`, error);
      try {
        const contextManager = session.getContextManager();
        if (contextManager) {
          const rawMessages = contextManager.getRawMessages();
          messageCount = rawMessages.length;
        }
      } catch (fallbackError) {
        logger.warn(
          `Failed to get message count via fallback for session ${sessionId}:`,
          fallbackError
        );
      }
    }
    return {
      id: sessionId,
      createdAt: Date.now(),
      // Placeholder - actual creation time would need to be tracked
      lastActivity: Date.now(),
      // Placeholder - actual last activity would need to be tracked
      messageCount
    };
  }
  /**
   * Get conversation history for the current session
   */
  async getCurrentSessionHistory() {
    this.ensureStarted();
    const session = await this.sessionManager.getSession(this.currentActiveSessionId);
    if (!session) {
      return [];
    }
    return await session.getConversationHistory();
  }
  /**
   * Get conversation history for a specific session
   */
  async getSessionHistory(sessionId) {
    this.ensureStarted();
    const session = await this.sessionManager.getSession(sessionId);
    if (!session) {
      throw new Error(`Session not found: ${sessionId}`);
    }
    const contextManager = session.contextManager;
    if (!contextManager) {
      return [];
    }
    const rawMessages = contextManager.getRawMessages();
    return rawMessages.map((msg, index) => ({
      id: index + 1,
      role: msg.role,
      content: msg.content,
      timestamp: (/* @__PURE__ */ new Date()).toISOString(),
      // Placeholder - actual timestamps would need to be tracked
      ...msg.toolCalls && { toolCalls: msg.toolCalls },
      ...msg.toolCallId && { toolCallId: msg.toolCallId },
      ...msg.name && { name: msg.name }
    }));
  }
  getCurrentLLMConfig() {
    this.ensureStarted();
    return structuredClone(this.stateManager.getLLMConfig());
  }
  async connectMcpServer(name, config) {
    this.ensureStarted();
    try {
      const validation = this.stateManager.addMcpServer(name, config);
      if (!validation.isValid) {
        const errorMessages = validation.errors.map((e) => e.message).join(", ");
        throw new Error(`Invalid MCP server configuration: ${errorMessages}`);
      }
      await this.mcpManager.connectServer(name, config);
      logger.info(`MemAgent: Successfully added and connected to MCP server '${name}'.`);
    } catch (error) {
      const errorMessage = error instanceof Error ? error.message : String(error);
      logger.error(`MemAgent: Failed to add MCP server '${name}': ${errorMessage}`);
      this.stateManager.removeMcpServer(name);
      throw error;
    }
  }
  async removeMcpServer(name) {
    this.ensureStarted();
    await this.mcpManager.removeClient(name);
    this.stateManager.removeMcpServer(name);
  }
  async executeMcpTool(toolName, args) {
    this.ensureStarted();
    return await this.mcpManager.executeTool(toolName, args);
  }
  async getAllMcpTools() {
    this.ensureStarted();
    return await this.mcpManager.getAllTools();
  }
  getMcpClients() {
    this.ensureStarted();
    return this.mcpManager.getClients();
  }
  getMcpFailedConnections() {
    this.ensureStarted();
    return this.mcpManager.getFailedConnections();
  }
  getAllMcpServers() {
    this.ensureStarted();
    const clients = this.mcpManager.getClients();
    const failedConnections = this.mcpManager.getFailedConnections();
    const servers = [];
    for (const [name, client] of clients.entries()) {
      try {
        client.getServerInfo();
        servers.push({
          id: name,
          // Use client name as id for API compatibility
          name,
          status: "connected"
        });
      } catch (error) {
        servers.push({
          id: name,
          // Use client name as id for API compatibility
          name,
          status: "error",
          error: error instanceof Error ? error.message : String(error)
        });
      }
    }
    for (const [name, error] of Object.entries(failedConnections)) {
      servers.push({
        id: name,
        // Use client name as id for API compatibility
        name,
        status: "error",
        error
      });
    }
    return servers;
  }
  getEffectiveConfig(sessionId) {
    this.ensureStarted();
    return sessionId ? this.stateManager.getRuntimeConfig(sessionId) : this.stateManager.getRuntimeConfig();
  }
  getCurrentActiveSessionId() {
    return this.currentActiveSessionId;
  }
  /**
   * Manually save all sessions to persistent storage
   */
  async saveAllSessions() {
    this.ensureStarted();
    const stats = await this.sessionManager.saveAllSessions();
    return {
      saved: stats.savedSessions,
      failed: stats.failedSessions,
      total: stats.totalSessions
    };
  }
  /**
   * Manually load all sessions from persistent storage
   */
  async loadAllSessions() {
    this.ensureStarted();
    const stats = await this.sessionManager.loadAllSessions();
    return {
      restored: stats.restoredSessions,
      failed: stats.failedSessions,
      total: stats.totalSessions
    };
  }
};

// src/core/brain/memAgent/config.ts
init_esm_shims();
init_zod();
var AgentCardSchema = external_exports.object({
  name: external_exports.string().default("cipher"),
  description: external_exports.string().default(
    "cipher is an AI assistant capable of store valuable software development knowledge for your  vibe coding agents"
  ),
  provider: external_exports.object({
    organization: external_exports.string().default("byterover-inc"),
    url: external_exports.string().url().default("https://byterover.dev")
  }).optional(),
  version: external_exports.string().default("1.0.0"),
  defaultInputModes: external_exports.array(external_exports.string()).default(["application/json", "text/plain"]),
  defaultOutputModes: external_exports.array(external_exports.string()).default(["application/json", "text/event-stream", "text/plain"]),
  skills: external_exports.array(
    external_exports.object({
      id: external_exports.string(),
      name: external_exports.string(),
      description: external_exports.string(),
      tags: external_exports.array(external_exports.string()),
      examples: external_exports.array(external_exports.string()).optional(),
      inputModes: external_exports.array(external_exports.string()).optional().default(["text/plain"]),
      outputModes: external_exports.array(external_exports.string()).optional().default(["text/plain"])
    })
  ).default([
    {
      id: "chat_with_agent",
      name: "chat_with_agent",
      description: "Allows you to chat with an AI agent. Send a message to interact.",
      tags: ["chat", "AI", "assistant", "mcp", "natural language"],
      examples: [
        `Send a JSON-RPC request to /mcp with method: "chat_with_agent" and params: {"message":"Your query..."}`,
        "Alternatively, use a compatible MCP client library."
      ]
    }
  ])
}).strict();
var EventPersistenceConfigSchema = external_exports.object({
  enabled: external_exports.boolean().default(true),
  storageType: external_exports.enum(["file", "memory", "database"]).default("file"),
  maxEvents: external_exports.number().optional(),
  rotationSize: external_exports.number().optional(),
  retentionDays: external_exports.number().optional(),
  filePath: external_exports.string().optional()
});
var AgentConfigSchema = external_exports.object({
  agentCard: AgentCardSchema.describe("Configuration for the agent card").optional(),
  systemPrompt: external_exports.string().describe(
    "The system prompt content as a string, or a structured system prompt configuration"
  ),
  mcpServers: ServerConfigsSchema.default({}).describe(
    "Configurations for MCP (Model Context Protocol) servers used by the agent"
  ),
  llm: LLMConfigSchema.describe("Core LLM configuration for the agent"),
  evalLlm: LLMConfigSchema.optional().describe(
    "Evaluation LLM configuration for non-thinking tasks (optional, falls back to main LLM if not provided)"
  ),
  embedding: external_exports.union([EmbeddingConfigSchema, external_exports.object({ disabled: external_exports.boolean() }), external_exports.boolean(), external_exports.null()]).optional().describe(
    "Embedding configuration for the agent (optional, falls back to environment auto-detection if not provided). Set to false, null, or {disabled: true} to disable embeddings."
  ),
  sessions: external_exports.object({
    maxSessions: external_exports.number().int().positive().default(100).describe("Maximum number of concurrent sessions allowed, defaults to 100"),
    sessionTTL: external_exports.number().int().positive().default(36e5).describe("Session time-to-live in milliseconds, defaults to 3600000ms (1 hour)")
  }).default({
    maxSessions: 100,
    sessionTTL: 36e5
  }).describe("Session management configuration"),
  eventPersistence: EventPersistenceConfigSchema.optional()
}).strict().describe("Main configuration for an agent, including its LLM and server connections");

// src/core/session/index.ts
init_esm_shims();

// src/core/brain/reasoning/index.ts
init_esm_shims();

// src/core/utils/index.ts
init_esm_shims();

// src/core/utils/path.ts
init_esm_shims();
import * as path from "path";
import { fileURLToPath } from "url";
var DEFAULT_CONFIG_PATH = "memAgent/cipher.yml";
function resolveConfigPath(configPath) {
  if (path.isAbsolute(configPath)) {
    return configPath;
  }
  if (configPath === DEFAULT_CONFIG_PATH) {
    const currentFileUrl = import.meta.url;
    const currentFilePath = fileURLToPath(currentFileUrl);
    const packageRoot = path.resolve(path.dirname(currentFilePath), "../../..");
    return path.resolve(packageRoot, configPath);
  }
  return path.resolve(process.cwd(), configPath);
}

// src/core/index.ts
init_env();

// src/core/storage/index.ts
var storage_exports = {};
__export(storage_exports, {
  BACKEND_TYPES: () => BACKEND_TYPES,
  DEFAULTS: () => DEFAULTS,
  ERROR_MESSAGES: () => ERROR_MESSAGES2,
  HEALTH_CHECK: () => HEALTH_CHECK,
  InMemoryBackend: () => InMemoryBackend,
  LOG_PREFIXES: () => LOG_PREFIXES2,
  MemoryHistoryStorageService: () => MemoryHistoryStorageService,
  StorageConnectionError: () => StorageConnectionError,
  StorageError: () => StorageError,
  StorageManager: () => StorageManager,
  StorageNotFoundError: () => StorageNotFoundError,
  StorageSchema: () => StorageSchema,
  TIMEOUTS: () => TIMEOUTS,
  createDefaultStorage: () => createDefaultStorage,
  createMemoryHistoryEntry: () => createMemoryHistoryEntry,
  createMemoryHistoryService: () => createMemoryHistoryService,
  createStorageBackends: () => createStorageBackends,
  createStorageFromEnv: () => createStorageFromEnv,
  isStorageFactory: () => isStorageFactory
});
init_esm_shims();

// src/core/storage/memory-history/index.ts
init_esm_shims();

// src/core/storage/memory-history/service.ts
init_esm_shims();
init_env();
var MemoryHistoryStorageService = class {
  logger;
  storageManager;
  connected = false;
  schemaInitialized = false;
  constructor() {
    this.logger = createLogger({ level: env.CIPHER_LOG_LEVEL || "info" });
  }
  /**
   * Initialize connection to storage backend
   */
  async connect() {
    if (this.connected) {
      this.logger.debug(`${LOG_PREFIXES2.BACKEND} Memory history service already connected`);
      return;
    }
    try {
      this.logger.info(`${LOG_PREFIXES2.BACKEND} Connecting memory history service`);
      const { createStorageFromEnv: createStorageFromEnv2 } = await import("./factory-RNOSUZME.js");
      const storageFactory = await createStorageFromEnv2();
      this.storageManager = storageFactory.manager;
      await this.initializeSchema();
      this.connected = true;
      this.logger.info(`${LOG_PREFIXES2.BACKEND} Memory history service connected successfully`);
    } catch (error) {
      this.logger.error(`${LOG_PREFIXES2.BACKEND} Failed to connect memory history service`, {
        error: error instanceof Error ? error.message : String(error)
      });
      throw new StorageConnectionError(
        `Failed to connect memory history service: ${error instanceof Error ? error.message : String(error)}`,
        "memory-history"
      );
    }
  }
  /**
   * Disconnect from storage backend
   */
  async disconnect() {
    if (!this.connected) {
      return;
    }
    try {
      if (this.storageManager) {
        await this.storageManager.disconnect();
        this.storageManager = void 0;
      }
      this.connected = false;
      this.schemaInitialized = false;
      this.logger.info(`${LOG_PREFIXES2.BACKEND} Memory history service disconnected`);
    } catch (error) {
      this.logger.error(`${LOG_PREFIXES2.BACKEND} Error disconnecting memory history service`, {
        error: error instanceof Error ? error.message : String(error)
      });
      throw error;
    }
  }
  /**
   * Check if service is connected
   */
  isConnected() {
    return this.connected && this.storageManager?.isConnected() === true;
  }
  /**
   * Record a memory operation in history
   */
  async recordOperation(entry) {
    if (!this.isConnected()) {
      throw new StorageError("Memory history service not connected", "recordOperation");
    }
    try {
      const startTime = Date.now();
      await this.ensureSchemaInitialized();
      this.validateEntry(entry);
      const key = `memory_history:${entry.id}`;
      const backends = await this.storageManager.getBackends();
      if (!backends) {
        throw new Error("Storage backends not available");
      }
      await backends.database.set(key, entry);
      await this.updateRecentHistoryCache(entry);
      const duration = Date.now() - startTime;
      this.logger.debug(`${LOG_PREFIXES2.BACKEND} Recorded memory operation`, {
        id: entry.id,
        operation: entry.operation,
        projectId: entry.projectId,
        success: entry.success,
        duration: `${duration}ms`
      });
    } catch (error) {
      this.logger.error(`${LOG_PREFIXES2.BACKEND} Failed to record memory operation`, {
        id: entry.id,
        operation: entry.operation,
        error: error instanceof Error ? error.message : String(error)
      });
      throw new StorageError(
        `Failed to record memory operation: ${error instanceof Error ? error.message : String(error)}`,
        "recordOperation"
      );
    }
  }
  /**
   * Get memory operation history with filters
   */
  async getHistory(filters) {
    if (!this.isConnected()) {
      throw new StorageError("Memory history service not connected", "getHistory");
    }
    try {
      await this.ensureSchemaInitialized();
      const results = await this.queryHistory(filters);
      this.logger.debug(`${LOG_PREFIXES2.BACKEND} Retrieved memory history`, {
        count: results.length,
        filters: this.sanitizeFilters(filters)
      });
      return results;
    } catch (error) {
      this.logger.error(`${LOG_PREFIXES2.BACKEND} Failed to get memory history`, {
        filters: this.sanitizeFilters(filters),
        error: error instanceof Error ? error.message : String(error)
      });
      throw new StorageError(
        `Failed to get memory history: ${error instanceof Error ? error.message : String(error)}`,
        "getHistory"
      );
    }
  }
  /**
   * Get history by project ID
   */
  async getByProjectId(projectId, options = {}) {
    return this.getHistory({ projectId, options });
  }
  /**
   * Get history by user ID
   */
  async getByUserId(userId, options = {}) {
    return this.getHistory({ userId, options });
  }
  /**
   * Get history by tags
   */
  async getByTags(tags, options = {}) {
    return this.getHistory({ tags, options });
  }
  /**
   * Get history by time range
   */
  async getByTimeRange(startTime, endTime, options = {}) {
    return this.getHistory({ startTime, endTime, options });
  }
  /**
   * Get operation statistics
   */
  async getOperationStats(projectId, userId) {
    if (!this.isConnected()) {
      throw new StorageError("Memory history service not connected", "getOperationStats");
    }
    try {
      await this.ensureSchemaInitialized();
      const filters = {};
      if (projectId) filters.projectId = projectId;
      if (userId) filters.userId = userId;
      const entries = await this.queryHistory(filters);
      const stats = this.calculateStats(entries);
      this.logger.debug(`${LOG_PREFIXES2.BACKEND} Retrieved operation statistics`, {
        projectId,
        userId,
        totalOperations: stats.totalOperations
      });
      return stats;
    } catch (error) {
      this.logger.error(`${LOG_PREFIXES2.BACKEND} Failed to get operation statistics`, {
        projectId,
        userId,
        error: error instanceof Error ? error.message : String(error)
      });
      throw new StorageError(
        `Failed to get operation statistics: ${error instanceof Error ? error.message : String(error)}`,
        "getOperationStats"
      );
    }
  }
  /**
   * Get success rate for operations
   */
  async getSuccessRate(projectId, userId) {
    const stats = await this.getOperationStats(projectId, userId);
    if (stats.totalOperations === 0) {
      return 0;
    }
    return stats.successCount / stats.totalOperations;
  }
  // Private helper methods
  /**
   * Initialize database schema
   */
  async initializeSchema() {
    if (this.schemaInitialized) {
      return;
    }
    try {
      const backends = await this.storageManager.getBackends();
      if (!backends) {
        throw new Error("Storage backends not available");
      }
      const backendType = backends.database.getBackendType();
      if (backendType === "in-memory") {
        this.schemaInitialized = true;
        return;
      }
      this.schemaInitialized = true;
      this.logger.info(
        `${LOG_PREFIXES2.BACKEND} Memory history schema initialized for ${backendType}`
      );
    } catch (error) {
      this.logger.error(`${LOG_PREFIXES2.BACKEND} Failed to initialize schema`, {
        error: error instanceof Error ? error.message : String(error)
      });
      throw error;
    }
  }
  /**
   * Ensure schema is initialized
   */
  async ensureSchemaInitialized() {
    if (!this.schemaInitialized) {
      await this.initializeSchema();
    }
  }
  /**
   * Validate memory history entry
   */
  validateEntry(entry) {
    const required = ["id", "projectId", "memoryId", "name", "operation", "timestamp", "success"];
    for (const field of required) {
      if (!(field in entry) || entry[field] === void 0) {
        throw new StorageError(`Missing required field: ${field}`, "validateEntry");
      }
    }
    const validOperations = ["ADD", "UPDATE", "DELETE", "SEARCH", "RETRIEVE"];
    if (!validOperations.includes(entry.operation)) {
      throw new StorageError(`Invalid operation type: ${entry.operation}`, "validateEntry");
    }
    if (isNaN(Date.parse(entry.timestamp))) {
      throw new StorageError(`Invalid timestamp format: ${entry.timestamp}`, "validateEntry");
    }
    if (!Array.isArray(entry.tags)) {
      throw new StorageError("Tags must be an array", "validateEntry");
    }
  }
  /**
   * Query history with filters
   */
  async queryHistory(filters) {
    const backends = await this.storageManager.getBackends();
    if (!backends) {
      throw new Error("Storage backends not available");
    }
    const keyPrefix = "memory_history:";
    const keys = await backends.database.list(keyPrefix);
    const entries = [];
    for (const key of keys) {
      const entry = await backends.database.get(key);
      if (entry) {
        entries.push(entry);
      }
    }
    let filteredEntries = entries;
    if (filters.projectId) {
      filteredEntries = filteredEntries.filter((entry) => entry.projectId === filters.projectId);
    }
    if (filters.userId) {
      filteredEntries = filteredEntries.filter((entry) => entry.userId === filters.userId);
    }
    if (filters.memoryId) {
      filteredEntries = filteredEntries.filter((entry) => entry.memoryId === filters.memoryId);
    }
    if (filters.operation) {
      const operations = Array.isArray(filters.operation) ? filters.operation : [filters.operation];
      filteredEntries = filteredEntries.filter((entry) => operations.includes(entry.operation));
    }
    if (filters.tags && filters.tags.length > 0) {
      filteredEntries = filteredEntries.filter(
        (entry) => filters.tags.every((tag) => entry.tags.includes(tag))
      );
    }
    if (filters.sessionId) {
      filteredEntries = filteredEntries.filter((entry) => entry.sessionId === filters.sessionId);
    }
    if (filters.success !== void 0) {
      filteredEntries = filteredEntries.filter((entry) => entry.success === filters.success);
    }
    if (filters.startTime) {
      filteredEntries = filteredEntries.filter((entry) => entry.timestamp >= filters.startTime);
    }
    if (filters.endTime) {
      filteredEntries = filteredEntries.filter((entry) => entry.timestamp <= filters.endTime);
    }
    const sortBy = filters.options?.sortBy || "timestamp";
    const sortOrder = filters.options?.sortOrder || "desc";
    filteredEntries.sort((a, b) => {
      const aValue = a[sortBy];
      const bValue = b[sortBy];
      if (aValue === void 0 && bValue === void 0) return 0;
      if (aValue === void 0) return sortOrder === "asc" ? -1 : 1;
      if (bValue === void 0) return sortOrder === "asc" ? 1 : -1;
      let comparison = 0;
      if (aValue < bValue) comparison = -1;
      if (aValue > bValue) comparison = 1;
      return sortOrder === "asc" ? comparison : -comparison;
    });
    const offset = filters.options?.offset || 0;
    const limit = filters.options?.limit;
    if (limit) {
      filteredEntries = filteredEntries.slice(offset, offset + limit);
    } else if (offset > 0) {
      filteredEntries = filteredEntries.slice(offset);
    }
    return filteredEntries;
  }
  /**
   * Update recent history cache
   */
  async updateRecentHistoryCache(entry) {
    try {
      const backends = await this.storageManager.getBackends();
      if (!backends) {
        throw new Error("Storage backends not available");
      }
      const recentKey = "memory_history:recent";
      const recentEntries = await backends.cache.get(recentKey) || [];
      recentEntries.unshift(entry);
      const trimmedEntries = recentEntries.slice(0, 1e3);
      await backends.cache.set(recentKey, trimmedEntries, 3600);
    } catch (error) {
      this.logger.warn(`${LOG_PREFIXES2.BACKEND} Failed to update recent history cache`, {
        error: error instanceof Error ? error.message : String(error)
      });
    }
  }
  /**
   * Calculate statistics from entries
   */
  calculateStats(entries) {
    const operationCounts = {
      ADD: 0,
      UPDATE: 0,
      DELETE: 0,
      SEARCH: 0,
      RETRIEVE: 0
    };
    let successCount = 0;
    let errorCount = 0;
    let totalDuration = 0;
    let durationCount = 0;
    const tagCounts = {};
    let earliest = entries[0]?.timestamp;
    let latest = entries[0]?.timestamp;
    for (const entry of entries) {
      operationCounts[entry.operation]++;
      if (entry.success) {
        successCount++;
      } else {
        errorCount++;
      }
      if (entry.duration !== void 0) {
        totalDuration += entry.duration;
        durationCount++;
      }
      for (const tag of entry.tags) {
        tagCounts[tag] = (tagCounts[tag] || 0) + 1;
      }
      if (earliest === void 0 || entry.timestamp < earliest) earliest = entry.timestamp;
      if (latest === void 0 || entry.timestamp > latest) latest = entry.timestamp;
    }
    const topTags = Object.entries(tagCounts).sort((a, b) => b[1] - a[1]).slice(0, 10).map(([tag, count]) => ({ tag, count }));
    return {
      totalOperations: entries.length,
      operationCounts,
      successCount,
      errorCount,
      averageDuration: durationCount > 0 ? Math.round(totalDuration / durationCount) : 0,
      topTags,
      dateRange: {
        earliest: earliest || (/* @__PURE__ */ new Date()).toISOString(),
        latest: latest || (/* @__PURE__ */ new Date()).toISOString()
      }
    };
  }
  /**
   * Sanitize filters for logging (remove sensitive data)
   */
  sanitizeFilters(filters) {
    const sanitized = { ...filters };
    return sanitized;
  }
};

// src/core/storage/memory-history/schema.ts
init_esm_shims();

// src/core/storage/memory-history/index.ts
function createMemoryHistoryService() {
  return new MemoryHistoryStorageService();
}
function createMemoryHistoryEntry(params) {
  return {
    id: params.id || `mh_${Date.now()}_${Math.random().toString(36).substr(2, 9)}`,
    timestamp: params.timestamp || (/* @__PURE__ */ new Date()).toISOString(),
    ...params
  };
}

// src/core/vector_storage/index.ts
var vector_storage_exports = {};
__export(vector_storage_exports, {
  BACKEND_TYPES: () => BACKEND_TYPES2,
  CollectionNotFoundError: () => CollectionNotFoundError,
  DEFAULTS: () => DEFAULTS2,
  DISTANCE_METRICS: () => DISTANCE_METRICS,
  DualCollectionVectorManager: () => DualCollectionVectorManager,
  VectorDimensionError: () => VectorDimensionError,
  VectorStoreConnectionError: () => VectorStoreConnectionError,
  VectorStoreError: () => VectorStoreError,
  VectorStoreManager: () => VectorStoreManager,
  createDefaultVectorStore: () => createDefaultVectorStore,
  createDualCollectionVectorStoreFromEnv: () => createDualCollectionVectorStoreFromEnv,
  createMultiCollectionVectorStoreFromEnv: () => createMultiCollectionVectorStoreFromEnv,
  createVectorStore: () => createVectorStore,
  createVectorStoreFromEnv: () => createVectorStoreFromEnv,
  createWorkspaceVectorStoreFromEnv: () => createWorkspaceVectorStoreFromEnv,
  getVectorStoreConfigFromEnv: () => getVectorStoreConfigFromEnv,
  getWorkspaceVectorStoreConfigFromEnv: () => getWorkspaceVectorStoreConfigFromEnv,
  isVectorStoreFactory: () => isVectorStoreFactory
});
init_esm_shims();

// src/core/knowledge_graph/index.ts
var knowledge_graph_exports = {};
__export(knowledge_graph_exports, {
  BACKEND_TYPES: () => BACKEND_TYPES3,
  BackendConfigSchema: () => BackendConfigSchema,
  DEFAULTS: () => DEFAULTS3,
  ERROR_MESSAGES: () => ERROR_MESSAGES3,
  EdgeNotFoundError: () => EdgeNotFoundError,
  GraphValidationError: () => GraphValidationError,
  INDEX_TEMPLATES: () => INDEX_TEMPLATES,
  InMemoryBackend: () => InMemoryBackend2,
  InMemoryBackendSchema: () => InMemoryBackendSchema,
  InvalidQueryError: () => InvalidQueryError,
  KnowledgeGraphConnectionError: () => KnowledgeGraphConnectionError,
  KnowledgeGraphEnvConfigSchema: () => KnowledgeGraphEnvConfigSchema,
  KnowledgeGraphError: () => KnowledgeGraphError,
  KnowledgeGraphManager: () => KnowledgeGraphManager,
  KnowledgeGraphSchema: () => KnowledgeGraphSchema,
  LOG_PREFIXES: () => LOG_PREFIXES3,
  METRICS_EVENTS: () => METRICS_EVENTS,
  Neo4jBackend: () => Neo4jBackend,
  Neo4jBackendSchema: () => Neo4jBackendSchema,
  NodeNotFoundError: () => NodeNotFoundError,
  QUERY_TEMPLATES: () => QUERY_TEMPLATES,
  SCHEMA: () => SCHEMA,
  TIMEOUTS: () => TIMEOUTS2,
  createDefaultKnowledgeGraph: () => createDefaultKnowledgeGraph,
  createInMemoryKnowledgeGraph: () => createInMemoryKnowledgeGraph,
  createKnowledgeGraph: () => createKnowledgeGraph,
  createKnowledgeGraphFromEnv: () => createKnowledgeGraphFromEnv,
  createNeo4jKnowledgeGraph: () => createNeo4jKnowledgeGraph,
  getKnowledgeGraphConfigFromEnv: () => getKnowledgeGraphConfigFromEnv,
  isKnowledgeGraphEnabled: () => isKnowledgeGraphEnabled,
  isKnowledgeGraphFactory: () => isKnowledgeGraphFactory,
  isNeo4jConfigAvailable: () => isNeo4jConfigAvailable,
  parseKnowledgeGraphConfig: () => parseKnowledgeGraphConfig,
  parseKnowledgeGraphConfigFromEnv: () => parseKnowledgeGraphConfigFromEnv,
  validateKnowledgeGraphConfig: () => validateKnowledgeGraphConfig
});
init_esm_shims();

// src/core/knowledge_graph/types.ts
init_esm_shims();

// src/core/knowledge_graph/backend/knowledge-graph.ts
init_esm_shims();
export {
  AgentConfigSchema,
  AwsService,
  AzureService,
  CONNECTION_MODES,
  CantInferProviderError,
  ContextManager,
  ConversationSession,
  DEFAULT_CONFIG_PATH,
  DEFAULT_CONNECTION_MODE,
  DEFAULT_TIMEOUT_MS,
  ENV_VARS,
  ERROR_MESSAGES,
  knowledge_graph_exports as KnowledgeGraph,
  LLMConfigSchema,
  LMStudioService,
  LOG_PREFIXES,
  Logger,
  MAX_TIMEOUT_MS,
  MCPClient,
  MCPManager,
  MIN_TIMEOUT_MS,
  MemAgent,
  OllamaService,
  OpenRouterService,
  ReasoningContentDetector,
  SearchContextManager,
  SessionManager,
  storage_exports as Storage,
  TRANSPORT_TYPES,
  vector_storage_exports as VectorStorage,
  createAgentServices,
  createLogger,
  env,
  getGlobalLogLevel,
  logger,
  resolveConfigPath,
  setGlobalLogLevel,
  validateEnv
};
