// @byterover/cipher/dist/src/core/milvus-U53ZMOKE.js
import {
  VectorDimensionError,
  VectorStoreConnectionError,
  VectorStoreError
} from "./chunk-JOPES53R.js";
import {
  ERROR_MESSAGES,
  LOG_PREFIXES
} from "./chunk-4JSJAC4K.js";
import "./chunk-7GLOCKG4.js";
import {
  createLogger,
  logger
} from "./chunk-GPOSIAVD.js";
import {
  env,
  init_env
} from "./chunk-7UQA6M3D.js";
import "./chunk-NUVPDHVT.js";
import {
  init_esm_shims
} from "./chunk-KFOQGTVJ.js";

// src/core/vector_storage/backend/milvus.ts
init_esm_shims();
import { DataType } from "@zilliz/milvus2-sdk-node";
init_env();

// src/core/vector_storage/connection-pool.ts
init_esm_shims();
import { MilvusClient } from "@zilliz/milvus2-sdk-node";
init_env();
var MilvusConnectionPool = class _MilvusConnectionPool {
  static instance = null;
  connections = /* @__PURE__ */ new Map();
  maxConnections = 10;
  // Prevent connection bloat
  connectionTtl = 3e5;
  // 5 minutes TTL
  healthCheckInterval = 6e4;
  // 1 minute health checks
  healthCheckTimer;
  constructor() {
    this.startHealthCheckTimer();
  }
  /**
   * Get the singleton instance of MilvusConnectionPool
   */
  static getInstance() {
    if (!_MilvusConnectionPool.instance) {
      _MilvusConnectionPool.instance = new _MilvusConnectionPool();
    }
    return _MilvusConnectionPool.instance;
  }
  /**
   * Generate a unique connection key based on configuration
   */
  generateConnectionKey(config) {
    const address = config.url || (config.host && config.port ? `http://${config.host}:${config.port}` : env.VECTOR_STORE_URL);
    const username = config.username || env.VECTOR_STORE_USERNAME || "";
    const password = config.password ? "[REDACTED]" : "";
    return `${address}:${username}:${!!password}`;
  }
  /**
   * Get or create a Milvus client from the pool
   */
  async getClient(config) {
    const connectionKey = this.generateConnectionKey(config);
    const existingConnection = this.connections.get(connectionKey);
    if (existingConnection && existingConnection.isHealthy) {
      existingConnection.lastUsed = Date.now();
      existingConnection.refCount++;
      logger.debug("MilvusConnectionPool: Reusing existing connection", {
        connectionKey,
        refCount: existingConnection.refCount,
        poolSize: this.connections.size
      });
      return existingConnection.client;
    }
    if (existingConnection && !existingConnection.isHealthy) {
      logger.debug("MilvusConnectionPool: Removing unhealthy connection", { connectionKey });
      this.connections.delete(connectionKey);
    }
    if (this.connections.size >= this.maxConnections) {
      await this.evictLeastRecentlyUsedConnection();
    }
    logger.debug("MilvusConnectionPool: Creating new connection", {
      connectionKey,
      poolSize: this.connections.size
    });
    const client = await this.createMilvusClient(config);
    const pooledConnection = {
      client,
      createdAt: Date.now(),
      lastUsed: Date.now(),
      isHealthy: true,
      refCount: 1
    };
    this.connections.set(connectionKey, pooledConnection);
    logger.debug("MilvusConnectionPool: Connection created and pooled", {
      connectionKey,
      poolSize: this.connections.size
    });
    return client;
  }
  /**
   * Release a client reference (decrement ref count)
   */
  releaseClient(config) {
    const connectionKey = this.generateConnectionKey(config);
    const connection = this.connections.get(connectionKey);
    if (connection) {
      connection.refCount = Math.max(0, connection.refCount - 1);
      logger.debug("MilvusConnectionPool: Client reference released", {
        connectionKey,
        refCount: connection.refCount
      });
    }
  }
  /**
   * Create a new Milvus client with the given configuration
   */
  async createMilvusClient(config) {
    const address = config.url || (config.host && config.port ? `http://${config.host}:${config.port}` : env.VECTOR_STORE_URL || "");
    const username = config.username || env.VECTOR_STORE_USERNAME || "";
    const password = config.password || env.VECTOR_STORE_PASSWORD || "";
    try {
      const client = new MilvusClient({
        address,
        username,
        password
      });
      await client.showCollections();
      return client;
    } catch (error) {
      logger.error("MilvusConnectionPool: Failed to create client", {
        error: error instanceof Error ? error.message : String(error),
        address
      });
      throw error;
    }
  }
  /**
   * Evict the least recently used connection to make room for new ones
   */
  async evictLeastRecentlyUsedConnection() {
    let oldestKey = null;
    let oldestTime = Date.now();
    for (const [key, connection] of this.connections) {
      if (connection.refCount === 0 && connection.lastUsed < oldestTime) {
        oldestTime = connection.lastUsed;
        oldestKey = key;
      }
    }
    if (!oldestKey) {
      for (const [key, connection] of this.connections) {
        if (connection.lastUsed < oldestTime) {
          oldestTime = connection.lastUsed;
          oldestKey = key;
        }
      }
    }
    if (oldestKey) {
      logger.debug("MilvusConnectionPool: Evicting LRU connection", {
        evictedKey: oldestKey,
        poolSize: this.connections.size
      });
      await this.closeConnection(oldestKey);
    }
  }
  /**
   * Start periodic health check timer
   */
  startHealthCheckTimer() {
    this.healthCheckTimer = setInterval(() => {
      this.performHealthChecks();
    }, this.healthCheckInterval);
  }
  /**
   * Perform health checks on all pooled connections
   */
  async performHealthChecks() {
    const now = Date.now();
    const connectionsToRemove = [];
    for (const [key, connection] of this.connections) {
      if (now - connection.createdAt > this.connectionTtl) {
        logger.debug("MilvusConnectionPool: Connection expired", { key });
        connectionsToRemove.push(key);
        continue;
      }
      try {
        await connection.client.showCollections();
        connection.isHealthy = true;
      } catch (error) {
        logger.warn("MilvusConnectionPool: Connection health check failed", {
          key,
          error: error instanceof Error ? error.message : String(error)
        });
        connection.isHealthy = false;
        if (connection.refCount === 0) {
          connectionsToRemove.push(key);
        }
      }
    }
    for (const key of connectionsToRemove) {
      await this.closeConnection(key);
    }
    if (connectionsToRemove.length > 0) {
      logger.debug("MilvusConnectionPool: Health check completed", {
        removedConnections: connectionsToRemove.length,
        remainingConnections: this.connections.size
      });
    }
  }
  /**
   * Close and remove a connection from the pool
   */
  async closeConnection(key) {
    const connection = this.connections.get(key);
    if (connection) {
      try {
        this.connections.delete(key);
        logger.debug("MilvusConnectionPool: Connection closed and removed", { key });
      } catch (error) {
        logger.warn("MilvusConnectionPool: Error closing connection", {
          key,
          error: error instanceof Error ? error.message : String(error)
        });
        this.connections.delete(key);
      }
    }
  }
  /**
   * Get pool statistics for monitoring
   */
  getStats() {
    const now = Date.now();
    const connectionDetails = Array.from(this.connections.entries()).map(([key, connection]) => ({
      key,
      createdAt: connection.createdAt,
      lastUsed: connection.lastUsed,
      refCount: connection.refCount,
      isHealthy: connection.isHealthy,
      age: now - connection.createdAt
    }));
    return {
      totalConnections: this.connections.size,
      maxConnections: this.maxConnections,
      connectionDetails
    };
  }
  /**
   * Force close all connections and cleanup
   */
  async shutdown() {
    if (this.healthCheckTimer) {
      clearInterval(this.healthCheckTimer);
      this.healthCheckTimer = void 0;
    }
    const connectionKeys = Array.from(this.connections.keys());
    await Promise.allSettled(connectionKeys.map((key) => this.closeConnection(key)));
    logger.debug("MilvusConnectionPool: Shutdown completed", {
      closedConnections: connectionKeys.length
    });
  }
  /**
   * Get current pool size
   */
  size() {
    return this.connections.size;
  }
  /**
   * Check if a connection exists for the given config
   */
  hasConnection(config) {
    const connectionKey = this.generateConnectionKey(config);
    const connection = this.connections.get(connectionKey);
    return connection !== void 0 && connection.isHealthy;
  }
};
var getMilvusConnectionPool = () => {
  return MilvusConnectionPool.getInstance();
};

// src/core/vector_storage/backend/milvus.ts
var MilvusBackend = class {
  client = null;
  config;
  collectionName;
  dimension;
  logger;
  connected = false;
  connectionConfig;
  // Milvus collection configuration
  MILVUS_COLLECTION_CONFIG = {
    schema: [
      {
        name: "id",
        description: "Primary key",
        data_type: DataType.Int64,
        is_primary_key: true
      },
      {
        name: "vector",
        description: "Vector field",
        data_type: DataType.FloatVector
      },
      {
        name: "payload",
        description: "Payload",
        data_type: DataType.JSON
      }
    ]
  };
  constructor(config) {
    this.config = config;
    this.collectionName = config.collectionName;
    this.dimension = config.dimension;
    this.logger = createLogger({
      level: env.CIPHER_LOG_LEVEL || "info"
    });
    this.connectionConfig = {
      url: config.url || void 0,
      host: config.host || void 0,
      port: config.port || void 0,
      username: config.username || env.VECTOR_STORE_USERNAME || void 0,
      password: config.password || env.VECTOR_STORE_PASSWORD || void 0
    };
    this.logger.debug(`${LOG_PREFIXES.MILVUS} Backend initialized with connection pooling`, {
      collection: this.collectionName,
      dimension: this.dimension,
      connectionKey: this.generateConnectionKey()
    });
  }
  /**
   * Generate a connection key for logging (without sensitive data)
   */
  generateConnectionKey() {
    const address = this.connectionConfig.url || (this.connectionConfig.host && this.connectionConfig.port ? `${this.connectionConfig.host}:${this.connectionConfig.port}` : env.VECTOR_STORE_URL);
    return `${address}:${this.connectionConfig.username || "anonymous"}`;
  }
  /**
   * Ensure the client is available and connected
   */
  ensureClient() {
    if (!this.client) {
      throw new VectorStoreError(ERROR_MESSAGES.NOT_CONNECTED, "operation");
    }
    return this.client;
  }
  validateDimension(vector, _operation) {
    if (vector.length !== this.dimension) {
      throw new VectorDimensionError(
        `${ERROR_MESSAGES.INVALID_DIMENSION}: expected ${this.dimension}, got ${vector.length}`,
        this.dimension,
        vector.length
      );
    }
  }
  async connect() {
    if (this.connected) {
      this.logger.debug(`${LOG_PREFIXES.MILVUS} Already connected`);
      return;
    }
    this.logger.debug(`${LOG_PREFIXES.MILVUS} Connecting to Milvus via connection pool`);
    try {
      const connectionPool = getMilvusConnectionPool();
      this.client = await connectionPool.getClient(this.connectionConfig);
      const client = this.ensureClient();
      const collections = await client.showCollections();
      const exists = collections.data.some((c) => c.name === this.collectionName);
      if (!exists) {
        this.logger.debug(`${LOG_PREFIXES.MILVUS} Creating collection ${this.collectionName}`);
        const schema = this.MILVUS_COLLECTION_CONFIG.schema.map(
          (field) => field.name === "vector" ? { ...field, dim: this.dimension } : { ...field }
        );
        await client.createCollection({
          collection_name: this.collectionName,
          fields: schema,
          // Include index parameters during collection creation
          index_params: [
            {
              field_name: "vector",
              index_name: "vector_index",
              index_type: "AUTOINDEX",
              metric_type: "COSINE"
            }
          ]
        });
        this.logger.debug(
          `${LOG_PREFIXES.MILVUS} Collection created with indexes: ${this.collectionName}`
        );
      } else {
        this.logger.debug(
          `${LOG_PREFIXES.MILVUS} Collection already exists: ${this.collectionName}`
        );
        try {
          const indexes = await client.describeIndex({
            collection_name: this.collectionName,
            field_name: "vector"
          });
          if (!indexes || !indexes.index_descriptions || indexes.index_descriptions.length === 0) {
            this.logger.debug(`${LOG_PREFIXES.MILVUS} Creating missing vector index`);
            await this.createMissingIndexes();
          }
        } catch {
          this.logger.debug(`${LOG_PREFIXES.MILVUS} Creating missing vector index`);
          await this.createMissingIndexes();
        }
      }
      const clientForLoad = this.ensureClient();
      await clientForLoad.loadCollection({ collection_name: this.collectionName });
      this.connected = true;
      this.logger.debug(`${LOG_PREFIXES.MILVUS} Successfully connected`);
    } catch (error) {
      this.logger.error(`${LOG_PREFIXES.MILVUS} Connection failed`, { error });
      if (error instanceof VectorStoreConnectionError) {
        throw error;
      }
      throw new VectorStoreConnectionError(
        ERROR_MESSAGES.CONNECTION_FAILED,
        "milvus",
        error
      );
    }
  }
  /**
   * Create missing indexes for existing collections
   * This is a fallback method for collections that were created without indexes
   */
  async createMissingIndexes() {
    try {
      const clientForIndex = this.ensureClient();
      await clientForIndex.createIndex({
        collection_name: this.collectionName,
        field_name: "vector",
        index_name: "vector_index",
        index_type: "AUTOINDEX",
        metric_type: "COSINE"
      });
      await this.createPayloadFieldIndexes();
      this.logger.debug(`${LOG_PREFIXES.MILVUS} Reloading collection after index creation`);
      const clientForReload = this.ensureClient();
      await clientForReload.loadCollection({ collection_name: this.collectionName });
    } catch (error) {
      this.logger.error(`${LOG_PREFIXES.MILVUS} Failed to create indexes and reload collection`);
      throw error;
    }
  }
  /**
   * Create indexes for payload fields to improve filtering performance
   */
  async createPayloadFieldIndexes() {
    const payloadFields = ["type", "category", "sessionId", "traceId", "timestamp"];
    for (const fieldName of payloadFields) {
      try {
        this.logger.debug(`${LOG_PREFIXES.MILVUS} Creating index for field: ${fieldName}`);
        const clientForFieldIndex = this.ensureClient();
        await clientForFieldIndex.createIndex({
          collection_name: this.collectionName,
          field_name: `payload.${fieldName}`,
          index_type: "AUTOINDEX"
        });
      } catch (error) {
        this.logger.warn(
          `${LOG_PREFIXES.MILVUS} Failed to create index for field ${fieldName}:`,
          error
        );
      }
    }
  }
  async insert(vectors, ids, payloads) {
    if (!this.connected)
      return Promise.reject(new VectorStoreError(ERROR_MESSAGES.NOT_CONNECTED, "insert"));
    if (vectors.length !== ids.length || vectors.length !== payloads.length) {
      return Promise.reject(
        new VectorStoreError("Vectors, IDs, and payloads must have the same length", "insert")
      );
    }
    for (const vector of vectors) this.validateDimension(vector, "insert");
    const data = vectors.map((vector, idx) => ({
      id: ids[idx],
      vector,
      payload: payloads[idx]
    }));
    try {
      const clientForInsert = this.ensureClient();
      await clientForInsert.insert({
        collection_name: this.collectionName,
        data
      });
      this.logger.debug(`Inserted ${vectors.length} vectors`);
    } catch (error) {
      this.logger.error(`Insert failed`, { error });
      throw new VectorStoreError("Failed to insert vectors", "insert", error);
    }
  }
  async search(query, limit = 10, filters) {
    if (!this.connected) throw new VectorStoreError(ERROR_MESSAGES.NOT_CONNECTED, "search");
    this.validateDimension(query, "search");
    try {
      const expr = filtersToExpr(filters);
      const searchParams = {
        collection_name: this.collectionName,
        data: [query],
        anns_field: "vector",
        params: { nprobe: 10 },
        limit,
        output_fields: ["id", "payload"],
        ...expr ? { filter: expr } : {}
      };
      const clientForSearch = this.ensureClient();
      const res = await clientForSearch.search(searchParams);
      return res.results.map((hit) => ({
        id: hit.id,
        score: hit.score,
        payload: hit.payload
      }));
    } catch (error) {
      this.logger.error(`Search failed`, { error });
      throw new VectorStoreError(ERROR_MESSAGES.SEARCH_FAILED, "search", error);
    }
  }
  async get(vectorId) {
    if (!this.connected)
      return Promise.reject(new VectorStoreError(ERROR_MESSAGES.NOT_CONNECTED, "get"));
    try {
      const clientForQuery = this.ensureClient();
      const res = await clientForQuery.query({
        collection_name: this.collectionName,
        output_fields: ["id", "vector", "payload"],
        filter: `id == ${vectorId}`
      });
      if (!res.data.length || !res.data[0]) return null;
      const doc = res.data[0];
      if (!doc) return null;
      return {
        id: vectorId,
        vector: doc.vector,
        payload: doc.payload,
        score: 1
      };
    } catch (error) {
      this.logger.error(`Get failed`, { error });
      throw new VectorStoreError("Failed to retrieve vector", "get", error);
    }
  }
  async update(vectorId, vector, payload) {
    if (!this.connected)
      return Promise.reject(new VectorStoreError(ERROR_MESSAGES.NOT_CONNECTED, "update"));
    this.validateDimension(vector, "update");
    try {
      const clientForUpsert = this.ensureClient();
      await clientForUpsert.upsert({
        collection_name: this.collectionName,
        data: [{ id: vectorId, vector, payload }]
      });
      this.logger.debug(`${LOG_PREFIXES.MILVUS} Updated vector ${vectorId}`);
    } catch (error) {
      this.logger.error(`Update failed`, { error });
      throw new VectorStoreError("Failed to update vector", "update", error);
    }
  }
  async delete(vectorId) {
    if (!this.connected)
      return Promise.reject(new VectorStoreError(ERROR_MESSAGES.NOT_CONNECTED, "delete"));
    try {
      const clientForDelete = this.ensureClient();
      await clientForDelete.deleteEntities({
        collection_name: this.collectionName,
        expr: `id == ${vectorId}`
      });
      this.logger.debug(`${LOG_PREFIXES.MILVUS} Deleted vector ${vectorId}`);
    } catch (error) {
      this.logger.error(`Delete failed`, { error });
      throw new VectorStoreError("Failed to delete vector", "delete", error);
    }
  }
  async deleteCollection() {
    if (!this.connected)
      throw new VectorStoreError(ERROR_MESSAGES.NOT_CONNECTED, "deleteCollection");
    try {
      const clientForDrop = this.ensureClient();
      await clientForDrop.dropCollection({ collection_name: this.collectionName });
      this.logger.info(`Deleted collection ${this.collectionName}`);
    } catch (error) {
      this.logger.error(`Delete collection failed`, { error });
      throw new VectorStoreError("Failed to delete collection", "deleteCollection", error);
    }
  }
  async list(filters, limit = 1e4) {
    if (!this.connected) throw new VectorStoreError(ERROR_MESSAGES.NOT_CONNECTED, "list");
    try {
      const expr = filtersToExpr(filters);
      const queryParams = {
        collection_name: this.collectionName,
        output_fields: ["id", "vector", "payload"],
        limit,
        ...expr ? { filter: expr } : {}
      };
      const clientForList = this.ensureClient();
      const res = await clientForList.query(queryParams);
      const results = (res.data || []).map(
        (doc) => doc && {
          id: doc.id,
          vector: doc.vector,
          payload: doc.payload,
          score: 1
        }
      ).filter(Boolean);
      return [results, results.length];
    } catch (error) {
      this.logger.error(`List failed`, { error });
      throw new VectorStoreError("Failed to list vectors", "list", error);
    }
  }
  async disconnect() {
    if (this.client) {
      const connectionPool = getMilvusConnectionPool();
      connectionPool.releaseClient(this.connectionConfig);
      this.client = null;
    }
    this.connected = false;
    this.logger.info(`${LOG_PREFIXES.MILVUS} Disconnected and released connection to pool`);
  }
  isConnected() {
    return this.connected;
  }
  getBackendType() {
    return "milvus";
  }
  getDimension() {
    return this.dimension;
  }
  getCollectionName() {
    return this.collectionName;
  }
  async listCollections() {
    if (!this.connected)
      return Promise.reject(new VectorStoreError(ERROR_MESSAGES.NOT_CONNECTED, "listCollections"));
    const clientForListCollections = this.ensureClient();
    const collections = await clientForListCollections.showCollections();
    return collections.data.map((c) => c.name);
  }
};
function filtersToExpr(filters) {
  if (!filters) return void 0;
  return Object.entries(filters).map(([key, value]) => {
    if (typeof value === "string" || typeof value === "number" || typeof value === "boolean") {
      return `${key} == "${value}"`;
    }
    if (value && typeof value === "object") {
      if ("any" in value && Array.isArray(value.any)) {
        const arr = value.any.map((v) => `"${v}"`).join(", ");
        return `${key} in [${arr}]`;
      }
      const ops = { gte: ">=", lte: "<=", gt: ">", lt: "<" };
      return Object.entries(ops).filter(([op]) => op in value).map(([op, symbol]) => `${key} ${symbol} ${value[op]}`).join(" && ");
    }
    return "";
  }).filter(Boolean).join(" && ");
}
export {
  MilvusBackend
};
