// @byterover/cipher/dist/src/core/multi-collection-manager-7DLF36ER.js
import {
  VectorStoreManager,
  getWorkspaceVectorStoreConfigFromEnv
} from "./chunk-IG2VBOT3.js";
import "./chunk-4JSJAC4K.js";
import "./chunk-7GLOCKG4.js";
import {
  createLogger
} from "./chunk-GPOSIAVD.js";
import {
  env,
  init_env
} from "./chunk-7UQA6M3D.js";
import "./chunk-NUVPDHVT.js";
import {
  init_esm_shims
} from "./chunk-KFOQGTVJ.js";

// src/core/vector_storage/multi-collection-manager.ts
init_esm_shims();
init_env();
var MultiCollectionVectorManager = class {
  knowledgeManager;
  reflectionManager;
  workspaceManager;
  logger;
  reflectionEnabled;
  workspaceEnabled;
  eventManager;
  constructor(baseConfig) {
    this.logger = createLogger({
      level: env.CIPHER_LOG_LEVEL || "info"
    });
    this.reflectionEnabled = !!(env.REFLECTION_VECTOR_STORE_COLLECTION && env.REFLECTION_VECTOR_STORE_COLLECTION.trim());
    this.workspaceEnabled = !!(env.USE_WORKSPACE_MEMORY && env.WORKSPACE_VECTOR_STORE_COLLECTION && env.WORKSPACE_VECTOR_STORE_COLLECTION.trim());
    this.knowledgeManager = new VectorStoreManager(baseConfig);
    if (this.reflectionEnabled) {
      const reflectionConfig = {
        ...baseConfig,
        collectionName: env.REFLECTION_VECTOR_STORE_COLLECTION
      };
      this.reflectionManager = new VectorStoreManager(reflectionConfig);
    } else {
      this.reflectionManager = null;
    }
    if (this.workspaceEnabled) {
      const workspaceConfig = getWorkspaceVectorStoreConfigFromEnv(baseConfig);
      this.workspaceManager = new VectorStoreManager(workspaceConfig);
      this.logger.info("MultiCollectionVectorManager: Initialized with multiple collections", {
        backend: baseConfig.type,
        knowledgeCollection: baseConfig.collectionName,
        reflectionCollection: this.reflectionEnabled ? env.REFLECTION_VECTOR_STORE_COLLECTION : "disabled",
        workspaceCollection: env.WORKSPACE_VECTOR_STORE_COLLECTION,
        dimension: workspaceConfig.dimension
        // Use dimension from workspaceConfig
      });
    } else {
      this.workspaceManager = null;
      this.logger.info("MultiCollectionVectorManager: Initialized without workspace memory", {
        backend: baseConfig.type,
        knowledgeCollection: baseConfig.collectionName,
        reflectionCollection: this.reflectionEnabled ? env.REFLECTION_VECTOR_STORE_COLLECTION : "disabled",
        workspaceCollection: "disabled",
        dimension: baseConfig.dimension
      });
    }
  }
  /**
   * Set event manager for event-aware operations
   */
  setEventManager(eventManager) {
    this.eventManager = eventManager;
    this.knowledgeManager.setEventManager(eventManager);
    if (this.reflectionManager) {
      this.reflectionManager.setEventManager(eventManager);
    }
    if (this.workspaceManager) {
      this.workspaceManager.setEventManager(eventManager);
    }
  }
  /**
   * Connect to all enabled vector stores
   */
  async connect() {
    const promises = [];
    promises.push(this.knowledgeManager.connect());
    if (this.reflectionManager) {
      promises.push(this.reflectionManager.connect());
    }
    if (this.workspaceManager) {
      promises.push(this.workspaceManager.connect());
    }
    try {
      const results = await Promise.all(promises);
      const knowledgeConnected = !!results[0];
      if (!knowledgeConnected) {
        this.logger.error("MultiCollectionVectorManager: Failed to connect knowledge manager");
        return false;
      }
      this.logger.info("MultiCollectionVectorManager: Connection completed", {
        knowledge: knowledgeConnected,
        reflection: this.reflectionEnabled ? !!results[1] : "disabled",
        workspace: this.workspaceEnabled ? !!results[results.length - 1] : "disabled"
      });
      return true;
    } catch (error) {
      this.logger.error("MultiCollectionVectorManager: Connection failed", {
        error: error instanceof Error ? error.message : String(error)
      });
      return false;
    }
  }
  /**
   * Disconnect from all vector stores
   */
  async disconnect() {
    const promises = [];
    promises.push(this.knowledgeManager.disconnect());
    if (this.reflectionManager) {
      promises.push(this.reflectionManager.disconnect());
    }
    if (this.workspaceManager) {
      promises.push(this.workspaceManager.disconnect());
    }
    await Promise.all(promises);
  }
  /**
   * Get vector store by collection type
   */
  getStore(type) {
    switch (type) {
      case "knowledge":
        return this.knowledgeManager.getStore();
      case "reflection":
        if (!this.reflectionEnabled || !this.reflectionManager) {
          this.logger.debug("MultiCollectionVectorManager: Reflection memory is disabled");
          return null;
        }
        return this.reflectionManager.getStore();
      case "workspace":
        if (!this.workspaceEnabled || !this.workspaceManager) {
          this.logger.debug("MultiCollectionVectorManager: Workspace memory is disabled");
          return null;
        }
        return this.workspaceManager.getStore();
      default:
        throw new Error(`Unknown collection type: ${type}`);
    }
  }
  /**
   * Get vector store by named collection (for backward compatibility)
   */
  getNamedStore(collectionName) {
    if (collectionName === env.WORKSPACE_VECTOR_STORE_COLLECTION) {
      return this.getStore("workspace");
    }
    if (collectionName === env.REFLECTION_VECTOR_STORE_COLLECTION) {
      return this.getStore("reflection");
    }
    return this.getStore("knowledge");
  }
  /**
   * Get an event-aware vector store by collection type for a specific session
   */
  getEventAwareStore(type, sessionId) {
    switch (type) {
      case "knowledge":
        return this.knowledgeManager.getEventAwareStore(sessionId);
      case "reflection":
        if (!this.reflectionEnabled || !this.reflectionManager) {
          this.logger.debug("MultiCollectionVectorManager: Reflection memory is disabled");
          return null;
        }
        return this.reflectionManager.getEventAwareStore(sessionId);
      case "workspace":
        if (!this.workspaceEnabled || !this.workspaceManager) {
          this.logger.debug("MultiCollectionVectorManager: Workspace memory is disabled");
          return null;
        }
        return this.workspaceManager.getEventAwareStore(sessionId);
      default:
        throw new Error(`Unknown collection type: ${type}`);
    }
  }
  /**
   * Health check for all collections
   */
  async healthCheck() {
    const knowledgeHealthy = (await this.knowledgeManager.healthCheck()).overall;
    let reflectionHealthy = true;
    if (this.reflectionEnabled && this.reflectionManager) {
      reflectionHealthy = (await this.reflectionManager.healthCheck()).overall;
    }
    let workspaceHealthy = true;
    if (this.workspaceEnabled && this.workspaceManager) {
      workspaceHealthy = (await this.workspaceManager.healthCheck()).overall;
    }
    return {
      knowledge: knowledgeHealthy,
      reflection: reflectionHealthy,
      workspace: workspaceHealthy,
      overall: knowledgeHealthy && reflectionHealthy && workspaceHealthy
    };
  }
  /**
   * Get vector store manager by collection type
   */
  getManager(type) {
    switch (type) {
      case "knowledge":
        return this.knowledgeManager;
      case "reflection":
        return this.reflectionManager;
      case "workspace":
        return this.workspaceManager;
      default:
        throw new Error(`Unknown collection type: ${type}`);
    }
  }
  /**
   * Get information about all collections
   */
  getInfo() {
    return {
      knowledge: {
        connected: this.knowledgeManager.isConnected(),
        collectionName: this.knowledgeManager.getInfo().backend.collectionName,
        manager: this.knowledgeManager
      },
      reflection: {
        connected: this.reflectionManager?.isConnected() ?? false,
        collectionName: env.REFLECTION_VECTOR_STORE_COLLECTION || "",
        manager: this.reflectionManager,
        enabled: this.reflectionEnabled
      },
      workspace: {
        connected: this.workspaceManager?.isConnected() ?? false,
        collectionName: env.WORKSPACE_VECTOR_STORE_COLLECTION || "",
        manager: this.workspaceManager,
        enabled: this.workspaceEnabled
      },
      overallConnected: this.knowledgeManager.isConnected() && (!this.reflectionEnabled || this.reflectionManager?.isConnected() === true) && (!this.workspaceEnabled || this.workspaceManager?.isConnected() === true)
    };
  }
  /**
   * Check if any collection is connected
   */
  isConnected() {
    return this.knowledgeManager.isConnected();
  }
};
export {
  MultiCollectionVectorManager
};
