// @byterover/cipher/dist/src/core/pgvector-H2QZBU2V.js
import {
  VectorDimensionError,
  VectorStoreConnectionError,
  VectorStoreError
} from "./chunk-JOPES53R.js";
import {
  DEFAULTS,
  DISTANCE_METRICS,
  ERROR_MESSAGES,
  LOG_PREFIXES
} from "./chunk-4JSJAC4K.js";
import "./chunk-7GLOCKG4.js";
import {
  createLogger
} from "./chunk-GPOSIAVD.js";
import "./chunk-7UQA6M3D.js";
import "./chunk-NUVPDHVT.js";
import {
  init_esm_shims
} from "./chunk-KFOQGTVJ.js";

// src/core/vector_storage/backend/pgvector.ts
init_esm_shims();
import { Pool, types } from "pg";
var VECTOR_OID = 16405;
var parseVector = (value) => {
  if (value.startsWith("[") && value.endsWith("]")) {
    return value.substring(1, value.length - 1).split(",").map(parseFloat);
  }
  return [];
};
types.setTypeParser(VECTOR_OID, parseVector);
var PgVectorBackend = class {
  pool;
  config;
  collectionName;
  dimension;
  logger;
  connected = false;
  vectorOid;
  constructor(config) {
    if (config.type !== "pgvector") {
      throw new VectorStoreError("Invalid config type for PgVectorBackend", "initialization");
    }
    this.config = config;
    this.collectionName = config.collectionName;
    this.dimension = config.dimension;
    this.logger = createLogger({
      level: process.env.LOG_LEVEL || "info"
    });
    const Poolconfig = {
      max: config.poolSize || 10
    };
    if (config.url) {
      Poolconfig.connectionString = config.url;
      Poolconfig.ssl = config.ssl || false;
    } else {
      throw new VectorStoreError(
        "Missing connection configuration for PgVectorBackend",
        "initialization"
      );
    }
    this.pool = new Pool(Poolconfig);
    this.logger.debug(`${LOG_PREFIXES.PGVECTOR} Initialized`, {
      collection: this.collectionName,
      dimension: this.dimension
    });
  }
  // @override
  async connect() {
    if (this.connected) {
      this.logger.debug(`${LOG_PREFIXES.PGVECTOR} Already connected`);
      return;
    }
    this.logger.info(`${LOG_PREFIXES.PGVECTOR} Connecting to PostgreSQL`);
    try {
      const client = await this.pool.connect();
      try {
        await client.query("CREATE EXTENSION IF NOT EXISTS vector");
        const { rows } = await client.query("SELECT oid FROM pg_type WHERE typname = 'vector'");
        this.vectorOid = rows[0]?.oid;
        if (this.vectorOid) types.setTypeParser(this.vectorOid, parseVector);
        const tableExists = await client.query(
          `SELECT EXISTS (
            SELECT FROM information_schema.tables
            WHERE table_name = $1
          )`,
          [this.collectionName]
        );
        if (!tableExists.rows[0].exists) {
          this.logger.info(`${LOG_PREFIXES.PGVECTOR} Creating table ${this.collectionName}`);
          await client.query(
            `CREATE TABLE "${this.collectionName}" (
              id BIGSERIAL PRIMARY KEY,
              vector vector(${this.dimension}),
              payload JSONB
            )`
          );
          const indexType = this.config.indexType || "ivfflat";
          const indexMetric = this.config.indexMetric || "vector_l2_ops";
          await client.query(
            `CREATE INDEX IF NOT EXISTS ${this.collectionName}_vector_idx
             ON "${this.collectionName}"
             USING ${indexType} (vector ${indexMetric})
             WITH (lists = 100)`
          );
        }
        const payloadFields = ["type", "category", "sessionId", "traceId", "timestamp"];
        for (const field of payloadFields) {
          await client.query(
            `CREATE INDEX IF NOT EXISTS ${this.collectionName}_payload_${field}_idx ON "${this.collectionName}" ((payload->>'${field}'))`
          );
        }
        await client.query(
          `CREATE INDEX IF NOT EXISTS ${this.collectionName}_tags_gin_idx ON "${this.collectionName}" USING GIN ((payload->'tags'));`
        );
      } finally {
        client.release();
      }
      this.connected = true;
      this.logger.info(`${LOG_PREFIXES.PGVECTOR} Successfully connected`);
    } catch (error) {
      this.logger.error(`${LOG_PREFIXES.PGVECTOR} Connection failed: ${error}`);
      throw new VectorStoreConnectionError(
        ERROR_MESSAGES.CONNECTION_FAILED,
        "pgvector",
        error
      );
    }
  }
  // @override
  async disconnect() {
    if (!this.connected) {
      this.logger.debug(`${LOG_PREFIXES.PGVECTOR} Already disconnected`);
      return;
    }
    this.logger.info(`${LOG_PREFIXES.PGVECTOR} Disconnecting from PostgreSQL`);
    await this.pool.end();
    this.connected = false;
    this.logger.info(`${LOG_PREFIXES.PGVECTOR} Successfully disconnected`);
  }
  // @override
  isConnected() {
    return this.connected;
  }
  // @override
  getBackendType() {
    return "pgvector";
  }
  // @override
  getDimension() {
    return this.dimension;
  }
  // @override
  getCollectionName() {
    return this.collectionName;
  }
  // @override
  async insert(vectors, ids, payloads) {
    if (!this.connected) {
      throw new VectorStoreError(ERROR_MESSAGES.NOT_CONNECTED, "insert");
    }
    const client = await this.pool.connect();
    try {
      await client.query("BEGIN");
      const values = [];
      const placeholders = [];
      for (let i = 0; i < vectors.length; i++) {
        const vector = vectors[i];
        const payload = payloads[i];
        if (!vector || !Array.isArray(vector)) {
          throw new VectorStoreError("Invalid vector", "insert");
        }
        if (!payload || typeof payload !== "object") {
          throw new VectorStoreError("Invalid payload", "insert");
        }
        this.validateDimension(vector, "insert");
        values.push(ids[i], `[${vector.join(",")}]`, payload);
        placeholders.push(`($${i * 3 + 1}, $${i * 3 + 2}, $${i * 3 + 3})`);
      }
      await client.query(
        `INSERT INTO ${this.collectionName} (id, vector, payload) VALUES ${placeholders.join(",")}`,
        values
      );
      await client.query("COMMIT");
    } catch (error) {
      await client.query("ROLLBACK");
      this.logger.error(`${LOG_PREFIXES.PGVECTOR} Insert failed`, { error });
      if (error instanceof VectorDimensionError) {
        throw error;
      }
      throw new VectorStoreError("Failed to insert vectors", "insert", error);
    } finally {
      client.release();
    }
  }
  // @override
  async search(query, limit = DEFAULTS.SEARCH_LIMIT, filters) {
    if (!this.connected) {
      throw new VectorStoreError(ERROR_MESSAGES.NOT_CONNECTED, "search");
    }
    this.validateDimension(query, "search");
    const metricOp = this.resolveDistanceOp(this.config.distance || "Cosine");
    const filterClauses = filters ? this.buildFilterClauses(filters) : "";
    const queryString = `
      SELECT id, payload, vector ${metricOp} $1 AS score
      FROM ${this.collectionName}
      ${filterClauses}
      ORDER BY score
      LIMIT $2
    `;
    const queryParams = [`[${query.join(",")}]`, limit];
    try {
      const result = await this.pool.query(queryString, queryParams);
      return result.rows.map((row) => ({
        id: parseInt(row.id, 10),
        score: row.score,
        payload: row.payload
      }));
    } catch (error) {
      this.logger.error(`${LOG_PREFIXES.PGVECTOR} Search failed`, { error });
      throw new VectorStoreError(ERROR_MESSAGES.SEARCH_FAILED, "search", error);
    }
  }
  // @override
  async get(vectorId) {
    if (!this.connected) {
      throw new VectorStoreError(ERROR_MESSAGES.NOT_CONNECTED, "get");
    }
    const result = await this.pool.query(
      `SELECT id, vector, payload FROM ${this.collectionName} WHERE id = $1`,
      [vectorId]
    );
    if (result.rows.length === 0) {
      return null;
    }
    const row = result.rows[0];
    return {
      id: parseInt(row.id, 10),
      vector: row.vector,
      payload: row.payload,
      score: 1
    };
  }
  // @override
  async update(vectorId, vector, payload) {
    if (!this.connected) {
      throw new VectorStoreError(ERROR_MESSAGES.NOT_CONNECTED, "update");
    }
    this.validateDimension(vector, "update");
    await this.pool.query(
      `UPDATE ${this.collectionName} SET vector = $1, payload = $2 WHERE id = $3`,
      [`[${vector.join(",")}]`, payload, vectorId]
    );
  }
  // @override
  async delete(vectorId) {
    if (!this.connected) {
      throw new VectorStoreError(ERROR_MESSAGES.NOT_CONNECTED, "delete");
    }
    await this.pool.query(`DELETE FROM ${this.collectionName} WHERE id = $1`, [vectorId]);
  }
  // @override
  async deleteCollection() {
    if (!this.connected) {
      throw new VectorStoreError(ERROR_MESSAGES.NOT_CONNECTED, "deleteCollection");
    }
    await this.pool.query(`DROP TABLE IF EXISTS ${this.collectionName}`);
  }
  // @override
  async list(filters, limit = 100) {
    if (!this.connected) {
      throw new VectorStoreError(ERROR_MESSAGES.NOT_CONNECTED, "list");
    }
    let filterClauses = "";
    if (filters && filters.category) {
      filterClauses = `WHERE category = '${filters.category}'`;
    } else if (filters) {
      filterClauses = this.buildFilterClauses(filters);
    }
    const results = await this.pool.query(
      `SELECT id, vector, payload FROM ${this.collectionName} ${filterClauses} LIMIT ${limit}`
    );
    const countResult = await this.pool.query(
      `SELECT COUNT(*) FROM ${this.collectionName} ${filterClauses}`
    );
    const formattedResults = results.rows.map((row) => ({
      id: parseInt(row.id, 10),
      vector: row.vector,
      payload: row.payload,
      score: 1
    }));
    return [formattedResults, parseInt(countResult.rows[0].count, 10)];
  }
  validateDimension(vector, _operation) {
    if (vector.length !== this.dimension) {
      throw new VectorDimensionError(
        `${ERROR_MESSAGES.INVALID_DIMENSION}: expected ${this.dimension}, got ${vector.length}`,
        this.dimension,
        vector.length
      );
    }
  }
  resolveDistanceOp(distance) {
    switch (distance) {
      case DISTANCE_METRICS.EUCLIDEAN:
        return "<->";
      case DISTANCE_METRICS.COSINE:
        return "<=>";
      case DISTANCE_METRICS.DOT_PRODUCT:
        return "<#>";
      case DISTANCE_METRICS.MANHATTAN:
        return "<+>";
      default:
        throw new Error(`Unknown distance metric: ${distance}`);
    }
  }
  buildFilterClauses(filters) {
    const conditions = Object.entries(filters).map(([key, value]) => {
      if (typeof value === "object" && value !== null) {
        if ("gte" in value) return `(payload->>'${key}')::numeric >= ${value.gte}`;
        if ("gt" in value) return `(payload->>'${key}')::numeric > ${value.gt}`;
        if ("lte" in value) return `(payload->>'${key}')::numeric <= ${value.lte}`;
        if ("lt" in value) return `(payload->>'${key}')::numeric < ${value.lt}`;
        if ("any" in value)
          return `payload->'${key}' @> '[${value.any.map((v) => `"${v}"`).join(",")}]'::jsonb`;
      }
      return `payload->>'${key}' = '${value}'`;
    });
    return conditions.length > 0 ? `WHERE ${conditions.join(" AND ")}` : "";
  }
};
export {
  PgVectorBackend
};
