// @byterover/cipher/dist/src/core/pinecone-DR5K5YYG.js
import {
  VectorDimensionError,
  VectorStoreConnectionError,
  VectorStoreError
} from "./chunk-JOPES53R.js";
import {
  DEFAULTS,
  ERROR_MESSAGES,
  LOG_PREFIXES
} from "./chunk-4JSJAC4K.js";
import "./chunk-7GLOCKG4.js";
import {
  createLogger
} from "./chunk-GPOSIAVD.js";
import "./chunk-7UQA6M3D.js";
import "./chunk-NUVPDHVT.js";
import {
  init_esm_shims
} from "./chunk-KFOQGTVJ.js";

// src/core/vector_storage/backend/pinecone.ts
init_esm_shims();
import { Pinecone } from "@pinecone-database/pinecone";
var PineconeBackend = class {
  client;
  index = null;
  config;
  indexName;
  dimension;
  provider;
  region;
  logger;
  connected = false;
  constructor(config) {
    this.config = config;
    this.indexName = config.collectionName;
    this.dimension = config.dimension;
    this.provider = config.provider || DEFAULTS.PINECONE_PROVIDER;
    this.region = config.region || DEFAULTS.PINECONE_REGION;
    this.logger = createLogger({
      level: process.env.CIPHER_LOG_LEVEL || "info"
    });
    this.client = new Pinecone({
      apiKey: this.config.apiKey
    });
    this.logger.debug(`${LOG_PREFIXES.PINECONE} Initialized`, {
      indexName: this.indexName,
      dimension: this.dimension
    });
    console.log("PineconeBackend initialized", {
      indexName: this.indexName,
      dimension: this.dimension
    });
  }
  /**
   * Convert search filters to Pinecone filter format
   */
  createFilter(filters) {
    if (!filters) return void 0;
    const pineconeFilter = {};
    for (const [key, value] of Object.entries(filters)) {
      if (value === null || value === void 0) continue;
      if (typeof value === "object" && !Array.isArray(value) && ("gte" in value || "gt" in value || "lte" in value || "lt" in value)) {
        const rangeConditions = {};
        if ("gte" in value && value.gte !== void 0) rangeConditions.$gte = value.gte;
        if ("gt" in value && value.gt !== void 0) rangeConditions.$gt = value.gt;
        if ("lte" in value && value.lte !== void 0) rangeConditions.$lte = value.lte;
        if ("lt" in value && value.lt !== void 0) rangeConditions.$lt = value.lt;
        pineconeFilter[key] = rangeConditions;
      } else if (typeof value === "object" && !Array.isArray(value) && ("any" in value || "all" in value)) {
        if ("any" in value && Array.isArray(value.any)) {
          pineconeFilter[key] = { $in: value.any };
        } else if ("all" in value && Array.isArray(value.all) && value.all.length > 0) {
          pineconeFilter[key] = value.all[0];
        }
      } else {
        pineconeFilter[key] = { $eq: value };
      }
    }
    return Object.keys(pineconeFilter).length > 0 ? pineconeFilter : void 0;
  }
  /**
   * Validate vector dimension
   */
  validateDimension(vector, operation) {
    if (vector.length !== this.dimension) {
      throw new VectorDimensionError(
        `${ERROR_MESSAGES.INVALID_DIMENSION}: expected ${this.dimension}, got ${vector.length}`,
        this.dimension,
        vector.length
      );
    }
  }
  /**
   * Validate ID (Pinecone supports string IDs)
   */
  validateId(id) {
    if (id === void 0 || id === null || !Number.isInteger(id)) {
      throw new VectorStoreError("Pinecone point IDs must be valid integers", "id");
    }
  }
  /**
   * Map provider string to valid Pinecone cloud values
   */
  getCloudprovider(provider) {
    switch (provider.toLowerCase()) {
      case "aws":
      case "amazon":
        return "aws";
      case "gcp":
      case "google":
      case "gcloud":
        return "gcp";
      case "azure":
      case "microsoft":
        return "azure";
      default:
        this.logger.warn(
          `${LOG_PREFIXES.PINECONE} Unknown provider '${provider}', defaulting to 'aws'`
        );
        return "aws";
    }
  }
  /**
   * Create index in Pinecone
   */
  async createIndex() {
    try {
      this.logger.info(`${LOG_PREFIXES.PINECONE} Creating index '${this.indexName}'`, {
        dimension: this.dimension,
        metric: this.config.metric || "cosine"
      });
      const Cloudprovider = this.getCloudprovider(this.provider);
      const result = await this.client.createIndex({
        name: this.indexName,
        dimension: this.dimension,
        metric: this.config.metric || "cosine",
        spec: {
          serverless: {
            cloud: Cloudprovider,
            region: this.region
          }
        }
      });
    } catch (error) {
      this.logger.error(`${LOG_PREFIXES.PINECONE} Failed to create index - DETAILED ERROR`, {
        error: error instanceof Error ? error.message : String(error),
        indexName: this.indexName,
        errorType: error?.constructor?.name,
        errorCode: error?.code,
        errorStatus: error?.status,
        errorResponse: error?.response?.data,
        fullError: JSON.stringify(error, null, 2)
      });
      throw error;
    }
  }
  async connect() {
    if (this.connected) {
      this.logger.debug(`${LOG_PREFIXES.PINECONE} Already connected`);
      return;
    }
    this.logger.info(`${LOG_PREFIXES.PINECONE} Connecting to Pinecone`);
    try {
      console.log("Trying to list indexes");
      const indexList = await this.client.listIndexes();
      const indexExists = indexList.indexes?.some((index) => index.name === this.indexName);
      console.log("indexExists", indexExists);
      if (!indexExists) {
        this.logger.info(
          `${LOG_PREFIXES.PINECONE} Index '${this.indexName}' does not exist, creating...`
        );
        await this.createIndex();
      }
      this.index = this.client.index(this.indexName);
      this.connected = true;
      this.logger.info(`${LOG_PREFIXES.PINECONE} Successfully connected`, {
        indexName: this.indexName,
        dimension: this.dimension
      });
    } catch (error) {
      this.logger.error(`${LOG_PREFIXES.PINECONE} Connection failed`, {
        error: error instanceof Error ? error.message : String(error),
        indexName: this.indexName
      });
      if (error instanceof Error) {
        if (error.message.includes("404") || error.message.includes("not found") || error.message.includes("does not exist")) {
          throw new VectorStoreConnectionError(
            `Pinecone index '${this.indexName}' not found (HTTP 404). Please create the index in your Pinecone dashboard first or set PINECONE_AUTO_CREATE_INDEX=true.`,
            "pinecone",
            error
          );
        }
        if (error.message.includes("401") || error.message.includes("403") || error.message.includes("Unauthorized") || error.message.includes("Forbidden")) {
          throw new VectorStoreConnectionError(
            `Pinecone authentication failed. Please check your API key.`,
            "pinecone",
            error
          );
        }
        if (error.message.includes("429") || error.message.includes("rate limit")) {
          throw new VectorStoreConnectionError(
            `Pinecone rate limit exceeded. Please try again later.`,
            "pinecone",
            error
          );
        }
      }
      if (error instanceof VectorStoreConnectionError) {
        throw error;
      }
      throw new VectorStoreConnectionError(
        ERROR_MESSAGES.CONNECTION_FAILED,
        "pinecone",
        error
      );
    }
  }
  async disconnect() {
    if (!this.connected) {
      this.logger.debug(`${LOG_PREFIXES.PINECONE} Already disconnected`);
      return;
    }
    this.logger.info(`${LOG_PREFIXES.PINECONE} Disconnecting from Pinecone`);
    this.index = null;
    this.connected = false;
    this.logger.info(`${LOG_PREFIXES.PINECONE} Successfully disconnected`);
  }
  async insert(vectors, ids, payloads) {
    if (!this.connected || !this.index) {
      throw new VectorStoreError(ERROR_MESSAGES.NOT_CONNECTED, "insert");
    }
    if (vectors.length !== ids.length || vectors.length !== payloads.length) {
      throw new VectorStoreError("Vectors, IDs, and payloads must have the same length", "insert");
    }
    if (vectors.length === 0) {
      this.logger.debug(`${LOG_PREFIXES.PINECONE} No vectors to insert`);
      return;
    }
    for (let i = 0; i < vectors.length; i++) {
      const vector = vectors[i];
      const id = ids[i];
      if (!vector) {
        throw new Error(`Vector at index ${i} is missing`);
      }
      if (id == null) {
        throw new Error(`ID at index ${i} is missing`);
      }
      this.validateDimension(vector, "insert");
      this.validateId(id);
    }
    this.logger.debug(`${LOG_PREFIXES.PINECONE} Inserting ${vectors.length} vectors`);
    try {
      const upsertData = vectors.map((vector, idx) => {
        const id = ids[idx];
        if (id == null) {
          throw new Error(`ID at index ${idx} is missing`);
        }
        return {
          id: id.toString(),
          values: vector,
          metadata: payloads[idx] || {}
        };
      });
      await this.index.upsert(upsertData);
      this.logger.debug(`${LOG_PREFIXES.PINECONE} Successfully inserted ${vectors.length} vectors`);
    } catch (error) {
      this.logger.error(`${LOG_PREFIXES.PINECONE} Insert failed`, {
        error: error instanceof Error ? error.message : String(error),
        vectorCount: vectors.length
      });
      throw new VectorStoreError("Failed to insert vectors", "insert", error);
    }
  }
  async search(query, limit = DEFAULTS.SEARCH_LIMIT, filters) {
    if (!this.connected || !this.index) {
      throw new VectorStoreError(ERROR_MESSAGES.NOT_CONNECTED, "search");
    }
    this.validateDimension(query, "search");
    if (limit <= 0) {
      throw new VectorStoreError("Search limit must be positive", "search");
    }
    this.logger.debug(`${LOG_PREFIXES.PINECONE} Searching with limit ${limit}`, {
      hasFilters: !!filters
    });
    try {
      const pineconeFilter = this.createFilter(filters);
      const queryRequest = {
        vector: query,
        topK: Math.min(limit, 1e4),
        // Pinecone limit
        includeMetadata: true,
        includeValues: false
        // Usually not needed for search results
      };
      if (pineconeFilter) {
        queryRequest.filter = pineconeFilter;
      }
      const searchResponse = await this.index.query(queryRequest);
      const results = [];
      if (searchResponse.matches) {
        for (const match of searchResponse.matches) {
          results.push({
            id: parseInt(match.id, 10),
            score: match.score || 0,
            payload: match.metadata || {},
            vector: match.values || []
          });
        }
      }
      this.logger.debug(`${LOG_PREFIXES.PINECONE} Found ${results.length} results`);
      return results;
    } catch (error) {
      this.logger.error(`${LOG_PREFIXES.PINECONE} Search failed`, {
        error: error instanceof Error ? error.message : String(error),
        limit
      });
      throw new VectorStoreError(ERROR_MESSAGES.SEARCH_FAILED, "search", error);
    }
  }
  async get(vectorId) {
    if (!this.connected || !this.index) {
      throw new VectorStoreError(ERROR_MESSAGES.NOT_CONNECTED, "get");
    }
    this.validateId(vectorId);
    this.logger.debug(`${LOG_PREFIXES.PINECONE} Getting vector ${vectorId}`);
    try {
      const fetchRequest = {
        ids: [vectorId.toString()]
      };
      const response = await this.index.fetch(fetchRequest);
      if (!response.vectors || !response.vectors[vectorId.toString()]) {
        return null;
      }
      const vector = response.vectors[vectorId.toString()];
      return {
        id: vectorId,
        vector: vector.values || [],
        payload: vector.metadata || {},
        score: 1
      };
    } catch (error) {
      this.logger.error(`${LOG_PREFIXES.PINECONE} Get failed`, {
        error: error instanceof Error ? error.message : String(error),
        vectorId
      });
      throw new VectorStoreError("Failed to retrieve vector", "get", error);
    }
  }
  async update(vectorId, vector, payload) {
    if (!this.connected || !this.index) {
      throw new VectorStoreError(ERROR_MESSAGES.NOT_CONNECTED, "update");
    }
    this.validateDimension(vector, "update");
    this.validateId(vectorId);
    this.logger.debug(`${LOG_PREFIXES.PINECONE} Updating vector ${vectorId}`);
    try {
      const upsertRequest = [
        {
          id: vectorId.toString(),
          values: vector,
          metadata: payload || {}
        }
      ];
      await this.index.upsert(upsertRequest);
      this.logger.debug(`${LOG_PREFIXES.PINECONE} Successfully updated vector ${vectorId}`);
    } catch (error) {
      this.logger.error(`${LOG_PREFIXES.PINECONE} Update failed`, {
        error: error instanceof Error ? error.message : String(error),
        vectorId
      });
      throw new VectorStoreError("Failed to update vector", "update", error);
    }
  }
  async delete(vectorId) {
    if (!this.connected || !this.index) {
      throw new VectorStoreError(ERROR_MESSAGES.NOT_CONNECTED, "delete");
    }
    this.validateId(vectorId);
    this.logger.debug(`${LOG_PREFIXES.PINECONE} Deleting vector ${vectorId}`);
    try {
      const deleteRequest = {
        ids: [vectorId.toString()]
      };
      await this.index.delete(deleteRequest);
      this.logger.debug(`${LOG_PREFIXES.PINECONE} Successfully deleted vector ${vectorId}`);
    } catch (error) {
      this.logger.error(`${LOG_PREFIXES.PINECONE} Delete failed`, {
        error: error instanceof Error ? error.message : String(error),
        vectorId
      });
      throw new VectorStoreError("Failed to delete vector", "delete", error);
    }
  }
  async deleteCollection() {
    if (!this.connected || !this.index) {
      throw new VectorStoreError(ERROR_MESSAGES.NOT_CONNECTED, "deleteCollection");
    }
    try {
      const deleteRequest = {
        deleteAll: true
      };
      await this.index.delete(deleteRequest);
    } catch (error) {
      this.logger.error(`${LOG_PREFIXES.PINECONE} Delete collection failed`, {
        error: error instanceof Error ? error.message : String(error)
      });
      throw new VectorStoreError("Failed to delete collection", "deleteCollection", error);
    }
  }
  async list(filters, limit = 1e4) {
    throw new VectorStoreError(
      "Pinecone does not support listing all vectors directly. Use search with appropriate filters instead.",
      "list"
    );
  }
  // Getter methods
  isConnected() {
    return this.connected;
  }
  getBackendType() {
    return "pinecone";
  }
  getDimension() {
    return this.dimension;
  }
  getCollectionName() {
    return this.indexName;
  }
};
export {
  PineconeBackend
};
