// @byterover/cipher/dist/src/core/qdrant-7YKPPI77.js
import {
  VectorDimensionError,
  VectorStoreConnectionError,
  VectorStoreError
} from "./chunk-JOPES53R.js";
import {
  DEFAULTS,
  ERROR_MESSAGES,
  LOG_PREFIXES
} from "./chunk-4JSJAC4K.js";
import "./chunk-7GLOCKG4.js";
import {
  createLogger
} from "./chunk-GPOSIAVD.js";
import "./chunk-7UQA6M3D.js";
import "./chunk-NUVPDHVT.js";
import {
  init_esm_shims
} from "./chunk-KFOQGTVJ.js";

// src/core/vector_storage/backend/qdrant.ts
init_esm_shims();
import { QdrantClient } from "@qdrant/js-client-rest";
var QdrantBackend = class {
  client;
  config;
  collectionName;
  dimension;
  logger;
  connected = false;
  constructor(config) {
    this.config = config;
    this.collectionName = config.collectionName;
    this.dimension = config.dimension;
    this.logger = createLogger({
      level: process.env.LOG_LEVEL || "info"
    });
    if (config.url) {
      const clientParams = {
        url: config.url
      };
      if (config.apiKey) {
        clientParams.apiKey = config.apiKey;
      }
      this.client = new QdrantClient(clientParams);
    } else {
      const params = {
        host: config.host || "localhost",
        port: config.port || DEFAULTS.QDRANT_PORT
      };
      if (config.apiKey) {
        params.apiKey = config.apiKey;
      }
      this.client = new QdrantClient(params);
    }
    this.logger.debug(`${LOG_PREFIXES.QDRANT} Initialized`, {
      collection: this.collectionName,
      dimension: this.dimension,
      host: config.host || config.url || "local"
    });
  }
  /**
   * Convert search filters to Qdrant filter format
   */
  createFilter(filters) {
    if (!filters) return void 0;
    const conditions = [];
    for (const [key, value] of Object.entries(filters)) {
      if (value === null || value === void 0) continue;
      if (typeof value === "object" && !Array.isArray(value) && ("gte" in value || "gt" in value || "lte" in value || "lt" in value)) {
        conditions.push({
          key,
          range: value
        });
      } else if (typeof value === "object" && !Array.isArray(value) && ("any" in value || "all" in value)) {
        if ("any" in value && Array.isArray(value.any)) {
          for (const item of value.any) {
            conditions.push({
              key,
              match: { value: item }
            });
          }
        }
      } else {
        conditions.push({
          key,
          match: { value }
        });
      }
    }
    return conditions.length ? { must: conditions } : void 0;
  }
  /**
   * Validate vector dimension
   */
  validateDimension(vector, _operation) {
    if (vector.length !== this.dimension) {
      throw new VectorDimensionError(
        `${ERROR_MESSAGES.INVALID_DIMENSION}: expected ${this.dimension}, got ${vector.length}`,
        this.dimension,
        vector.length
      );
    }
  }
  // Qdrant only supports integer IDs for points in this deployment
  validateId(id) {
    if (!Number.isInteger(id) || isNaN(id)) {
      throw new VectorStoreError("Qdrant point IDs must be valid integers", "id");
    }
  }
  // VectorStore implementation
  async insert(vectors, ids, payloads) {
    if (!this.connected) {
      throw new VectorStoreError(ERROR_MESSAGES.NOT_CONNECTED, "insert");
    }
    if (vectors.length !== ids.length || vectors.length !== payloads.length) {
      throw new VectorStoreError("Vectors, IDs, and payloads must have the same length", "insert");
    }
    for (let i = 0; i < vectors.length; i++) {
      const vector = vectors[i];
      const id = ids[i];
      if (!vector) {
        throw new VectorStoreError(`Vector missing at index ${i}`, "insert");
      }
      if (id === void 0 || id === null) {
        throw new VectorStoreError(`ID missing at index ${i}`, "insert");
      }
      this.validateDimension(vector, "insert");
      this.validateId(id);
    }
    this.logger.debug(`${LOG_PREFIXES.INDEX} Inserting ${vectors.length} vectors`);
    try {
      const points = vectors.map((vector, idx) => {
        const payload = payloads[idx];
        const id = ids[idx];
        if (!payload) throw new VectorStoreError(`Payload missing at index ${idx}`, "insert");
        if (!vector) throw new VectorStoreError(`Vector missing at index ${idx}`, "insert");
        if (id === void 0 || id === null)
          throw new VectorStoreError(`ID missing at index ${idx}`, "insert");
        return {
          id,
          // Always use integer IDs
          vector,
          payload
        };
      });
      await this.client.upsert(this.collectionName, { points });
      this.logger.debug(`${LOG_PREFIXES.INDEX} Successfully inserted ${vectors.length} vectors`);
    } catch (error) {
      this.logger.error(`${LOG_PREFIXES.INDEX} Insert failed`, { error });
      throw new VectorStoreError("Failed to insert vectors", "insert", error);
    }
  }
  async search(query, limit = DEFAULTS.SEARCH_LIMIT, filters) {
    if (!this.connected) {
      throw new VectorStoreError(ERROR_MESSAGES.NOT_CONNECTED, "search");
    }
    this.validateDimension(query, "search");
    this.logger.debug(`${LOG_PREFIXES.SEARCH} Searching with limit ${limit}`, {
      hasFilters: !!filters
    });
    try {
      const queryFilter = this.createFilter(filters);
      const searchParams = {
        vector: query,
        limit,
        with_vector: true,
        with_payload: true
      };
      if (queryFilter) {
        searchParams.filter = queryFilter;
      }
      const searchResponse = await this.client.search(this.collectionName, searchParams);
      const results = searchResponse.result || searchResponse;
      const formattedResults = results.map((hit) => ({
        id: hit.id,
        score: hit.score,
        payload: hit.payload
      }));
      this.logger.debug(`${LOG_PREFIXES.SEARCH} Found ${formattedResults.length} results`);
      return formattedResults;
    } catch (error) {
      this.logger.error(`${LOG_PREFIXES.SEARCH} Search failed`, { error });
      throw new VectorStoreError(ERROR_MESSAGES.SEARCH_FAILED, "search", error);
    }
  }
  async get(vectorId) {
    if (!this.connected) {
      throw new VectorStoreError(ERROR_MESSAGES.NOT_CONNECTED, "get");
    }
    this.validateId(vectorId);
    this.logger.debug(`${LOG_PREFIXES.BACKEND} Getting vector ${vectorId}`);
    try {
      const response = await this.client.retrieve(this.collectionName, {
        ids: [vectorId],
        with_vector: true,
        with_payload: true
      });
      const results = response.result || response;
      if (!results.length) {
        return null;
      }
      const firstResult = results[0];
      if (!firstResult) {
        return null;
      }
      return {
        id: vectorId,
        // Qdrant uses integer IDs
        vector: firstResult.vector || [],
        payload: firstResult.payload || {},
        score: 1
      };
    } catch (error) {
      this.logger.error(`${LOG_PREFIXES.BACKEND} Get failed`, { error });
      throw new VectorStoreError("Failed to retrieve vector", "get", error);
    }
  }
  async update(vectorId, vector, payload) {
    if (!this.connected) {
      throw new VectorStoreError(ERROR_MESSAGES.NOT_CONNECTED, "update");
    }
    this.validateDimension(vector, "update");
    this.validateId(vectorId);
    this.logger.debug(`${LOG_PREFIXES.BACKEND} Updating vector ${vectorId}`);
    try {
      const point = {
        id: vectorId,
        vector,
        payload
      };
      await this.client.upsert(this.collectionName, {
        points: [point]
      });
      this.logger.debug(`${LOG_PREFIXES.BACKEND} Successfully updated vector ${vectorId}`);
    } catch (error) {
      this.logger.error(`${LOG_PREFIXES.BACKEND} Update failed`, { error });
      throw new VectorStoreError("Failed to update vector", "update", error);
    }
  }
  async delete(vectorId) {
    if (!this.connected) {
      throw new VectorStoreError(ERROR_MESSAGES.NOT_CONNECTED, "delete");
    }
    this.validateId(vectorId);
    this.logger.debug(`${LOG_PREFIXES.BACKEND} Deleting vector ${vectorId}`);
    try {
      await this.client.delete(this.collectionName, {
        points: [vectorId]
      });
      this.logger.debug(`${LOG_PREFIXES.BACKEND} Successfully deleted vector ${vectorId}`);
    } catch (error) {
      this.logger.error(`${LOG_PREFIXES.BACKEND} Delete failed`, { error });
      throw new VectorStoreError("Failed to delete vector", "delete", error);
    }
  }
  async deleteCollection() {
    if (!this.connected) {
      throw new VectorStoreError(ERROR_MESSAGES.NOT_CONNECTED, "deleteCollection");
    }
    this.logger.warn(`${LOG_PREFIXES.BACKEND} Deleting collection ${this.collectionName}`);
    try {
      await this.client.deleteCollection(this.collectionName);
      this.logger.info(
        `${LOG_PREFIXES.BACKEND} Successfully deleted collection ${this.collectionName}`
      );
    } catch (error) {
      this.logger.error(`${LOG_PREFIXES.BACKEND} Delete collection failed`, { error });
      throw new VectorStoreError("Failed to delete collection", "deleteCollection", error);
    }
  }
  async list(filters, limit = 1e4) {
    if (!this.connected) {
      throw new VectorStoreError(ERROR_MESSAGES.NOT_CONNECTED, "list");
    }
    this.logger.debug(`${LOG_PREFIXES.BACKEND} Listing vectors`, {
      hasFilters: !!filters,
      limit
    });
    try {
      const queryFilter = this.createFilter(filters);
      const scrollRequest = {
        limit,
        with_payload: true,
        with_vector: true
      };
      if (queryFilter) {
        scrollRequest.filter = queryFilter;
      }
      const response = await this.client.scroll(this.collectionName, scrollRequest);
      const results = (response.result?.points || response.points || []).map(
        (point) => ({
          id: point.id,
          score: 1,
          // Default score for exact match
          payload: point.payload
        })
      );
      const countRequest = {};
      if (queryFilter) {
        countRequest.filter = queryFilter;
      }
      const countResponse = await this.client.count(this.collectionName, countRequest);
      this.logger.info(`${LOG_PREFIXES.BACKEND} Listed ${results.length} vectors`);
      return [results, countResponse.count];
    } catch (error) {
      this.logger.error(`${LOG_PREFIXES.BACKEND} List failed`, { error });
      throw new VectorStoreError("Failed to list vectors", "list", error);
    }
  }
  async connect() {
    if (this.connected) {
      this.logger.debug(`${LOG_PREFIXES.QDRANT} Already connected`);
      return;
    }
    this.logger.info(`${LOG_PREFIXES.QDRANT} Connecting to Qdrant`);
    try {
      const collections = await this.client.getCollections();
      const exists = collections.collections.some((c) => c.name === this.collectionName);
      if (!exists) {
        this.logger.info(`${LOG_PREFIXES.QDRANT} Creating collection ${this.collectionName}`);
        const distanceMap = {
          Cosine: "Cosine",
          Euclidean: "Euclid",
          Dot: "Dot",
          Manhattan: "Manhattan"
        };
        const distance = this.config.distance || DEFAULTS.QDRANT_DISTANCE;
        const qdrantDistance = distanceMap[distance] || "Cosine";
        await this.client.createCollection(this.collectionName, {
          vectors: {
            size: this.dimension,
            distance: qdrantDistance
          }
        });
      } else {
        const collectionInfo = await this.client.getCollection(this.collectionName);
        const vectorConfig = collectionInfo?.config?.params?.vectors;
        if (vectorConfig && "size" in vectorConfig && vectorConfig.size !== this.dimension) {
          throw new VectorStoreConnectionError(
            `Collection ${this.collectionName} exists with different dimension. Expected: ${this.dimension}, got: ${vectorConfig.size}`,
            "qdrant"
          );
        }
      }
      this.connected = true;
      this.logger.info(`${LOG_PREFIXES.QDRANT} Successfully connected`);
    } catch (error) {
      this.logger.error(`${LOG_PREFIXES.QDRANT} Connection failed`, { error });
      if (error instanceof VectorStoreConnectionError) {
        throw error;
      }
      throw new VectorStoreConnectionError(
        ERROR_MESSAGES.CONNECTION_FAILED,
        "qdrant",
        error
      );
    }
  }
  async disconnect() {
    if (!this.connected) {
      this.logger.debug(`${LOG_PREFIXES.QDRANT} Already disconnected`);
      return;
    }
    this.logger.info(`${LOG_PREFIXES.QDRANT} Disconnecting from Qdrant`);
    this.connected = false;
    this.logger.info(`${LOG_PREFIXES.QDRANT} Successfully disconnected`);
  }
  isConnected() {
    return this.connected;
  }
  getBackendType() {
    return "qdrant";
  }
  getDimension() {
    return this.dimension;
  }
  getCollectionName() {
    return this.collectionName;
  }
};
export {
  QdrantBackend
};
