// @byterover/cipher/dist/src/core/redis-backend-L3FPDS2O.js
import {
  init_esm_shims
} from "./chunk-KFOQGTVJ.js";

// src/core/storage/backend/redis-backend.ts
init_esm_shims();
import { Redis } from "ioredis";
var RedisBackend = class {
  /**
   * Creates a new Redis backend instance
   *
   * @param config - Redis configuration options
   */
  constructor(config) {
    this.config = config;
  }
  /** The Redis client instance */
  redis = null;
  /** Connection status flag */
  connected = false;
  /**
   * Establishes connection to Redis server
   *
   * Sets up connection with retry logic, error handling, and event listeners.
   * Configuration supports both individual options and connection URL.
   *
   * @throws {StorageConnectionError} If connection fails after retries
   */
  async connect() {
    if (this.connected) return;
    this.redis = new Redis({
      // Host and port configuration
      ...this.config.host && { host: this.config.host },
      ...this.config.port && { port: this.config.port },
      ...this.config.username && { username: this.config.username },
      ...this.config.password && { password: this.config.password },
      // Database selection (default: 0)
      db: this.config.database || 0,
      // Force IPv4 to avoid IPv6 issues
      family: 4,
      // Timeout configuration
      ...this.config.connectionTimeoutMillis && {
        connectTimeout: this.config.connectionTimeoutMillis
      },
      ...this.config.connectionTimeoutMillis && {
        commandTimeout: this.config.connectionTimeoutMillis
      },
      // Retry configuration
      maxRetriesPerRequest: 3,
      // Lazy connect for explicit connection control
      lazyConnect: true,
      // Merge any additional options from config
      ...this.config.options
    });
    let connectionError = null;
    this.redis.on("error", (err) => {
      connectionError = err;
    });
    await new Promise((resolve, reject) => {
      const onReady = () => {
        this.connected = true;
        cleanup();
        resolve();
      };
      const onError = (err) => {
        cleanup();
        reject(new Error("Redis connection failed: " + err.message));
      };
      const cleanup = () => {
        this.redis?.off("ready", onReady);
        this.redis?.off("error", onError);
      };
      this.redis?.once("ready", onReady);
      this.redis?.once("error", onError);
      const timeoutMs = this.config.connectionTimeoutMillis || 1e4;
      setTimeout(() => {
        cleanup();
        if (connectionError) {
          reject(new Error("Redis connection failed: " + connectionError.message));
        } else {
          reject(new Error("Redis connection failed: timeout"));
        }
      }, timeoutMs);
      this.redis?.connect().catch(onError);
    });
  }
  /**
   * Gracefully disconnects from Redis
   *
   * Ensures all pending commands are completed before closing the connection.
   */
  async disconnect() {
    if (this.redis) {
      await this.redis.quit();
      this.redis = null;
    }
    this.connected = false;
  }
  /**
   * Checks if the Redis connection is active and ready
   *
   * @returns true if connected and Redis status is 'ready'
   */
  isConnected() {
    return this.connected && this.redis?.status === "ready";
  }
  /**
   * Returns the backend type identifier
   *
   * @returns 'redis'
   */
  getBackendType() {
    return "redis";
  }
  // Core operations
  /**
   * Retrieves and deserializes a value from Redis
   *
   * @template T - The expected type of the cached value
   * @param key - The cache key
   * @returns The cached value or undefined if not found
   * @throws {StorageError} If deserialization fails
   */
  async get(key) {
    this.checkConnection();
    const value = await this.redis.get(key);
    return value ? JSON.parse(value) : void 0;
  }
  /**
   * Serializes and stores a value in Redis with optional TTL
   *
   * @template T - The type of the value to cache
   * @param key - The cache key
   * @param value - The value to cache
   * @param ttlSeconds - Optional TTL in seconds
   */
  async set(key, value, ttlSeconds) {
    this.checkConnection();
    const serialized = JSON.stringify(value);
    if (ttlSeconds) {
      await this.redis.setex(key, ttlSeconds, serialized);
    } else {
      await this.redis.set(key, serialized);
    }
  }
  /**
   * Deletes a key from Redis
   *
   * @param key - The cache key to delete
   */
  async delete(key) {
    this.checkConnection();
    await this.redis.del(key);
  }
  /**
   * Batch retrieval of multiple keys
   *
   * More efficient than multiple individual get() calls.
   *
   * @template T - The expected type of the cached values
   * @param keys - Array of cache keys
   * @returns Array of values (undefined for missing keys)
   */
  async mget(keys) {
    this.checkConnection();
    if (keys.length === 0) return [];
    const values = await this.redis.mget(...keys);
    return values.map((value) => value ? JSON.parse(value) : void 0);
  }
  /**
   * Batch storage of multiple key-value pairs
   *
   * Uses pipeline for atomic batch operation.
   * Note: Does not support TTL for batch operations.
   *
   * @template T - The type of values to cache
   * @param entries - Array of [key, value] tuples
   */
  async mset(entries) {
    this.checkConnection();
    if (entries.length === 0) return;
    const pipeline = this.redis.pipeline();
    for (const [key, value] of entries) {
      pipeline.set(key, JSON.stringify(value));
    }
    await pipeline.exec();
  }
  /**
   * Checks if a key exists in Redis
   *
   * @param key - The cache key to check
   * @returns true if key exists, false otherwise
   */
  async exists(key) {
    this.checkConnection();
    const result = await this.redis.exists(key);
    return result === 1;
  }
  /**
   * Sets expiration time for an existing key
   *
   * @param key - The cache key
   * @param ttlSeconds - TTL in seconds
   */
  async expire(key, ttlSeconds) {
    this.checkConnection();
    await this.redis.expire(key, ttlSeconds);
  }
  // Cache-specific operations
  /**
   * Atomically increments a numeric value
   *
   * Creates the key with value 0 if it doesn't exist.
   *
   * @param key - The cache key
   * @param by - Increment amount (default: 1)
   * @returns The new value after increment
   */
  async increment(key, by = 1) {
    this.checkConnection();
    return await this.redis.incrby(key, by);
  }
  /**
   * Atomically decrements a numeric value
   *
   * @param key - The cache key
   * @param by - Decrement amount (default: 1)
   * @returns The new value after decrement
   */
  async decrement(key, by = 1) {
    this.checkConnection();
    return await this.redis.decrby(key, by);
  }
  // List operations (for compatibility with DatabaseBackend patterns)
  /**
   * Appends an item to a Redis list
   *
   * Creates the list if it doesn't exist.
   * Items are added to the end of the list (RPUSH).
   *
   * @template T - The type of list items
   * @param key - The list key
   * @param item - The item to append
   */
  async append(key, item) {
    this.checkConnection();
    await this.redis.rpush(key, JSON.stringify(item));
  }
  /**
   * Retrieves a range of items from a Redis list
   *
   * @template T - The type of list items
   * @param key - The list key
   * @param start - Starting index (0-based)
   * @param count - Number of items to retrieve
   * @returns Array of items in the specified range
   */
  async getRange(key, start, count) {
    this.checkConnection();
    const items = await this.redis.lrange(key, start, start + count - 1);
    return items.map((item) => JSON.parse(item));
  }
  /**
   * Lists all keys matching a pattern
   *
   * Warning: KEYS command can be slow on large databases.
   * Consider using SCAN for production use.
   *
   * @param prefix - Key prefix to search for
   * @returns Array of matching keys
   */
  async list(prefix) {
    this.checkConnection();
    return await this.redis.keys(`${prefix}*`);
  }
  /**
   * Validates connection status before operations
   *
   * @throws {Error} If not connected or Redis is not ready
   * @private
   */
  checkConnection() {
    if (!this.connected || !this.redis || this.redis.status !== "ready") {
      throw new Error("RedisBackend not connected");
    }
  }
  /**
   * Flushes the current Redis database
   *
   * Warning: This will delete ALL data in the current database.
   * Use with caution, typically only in development/testing.
   */
  async flushdb() {
    this.checkConnection();
    await this.redis.flushdb();
  }
  /**
   * Retrieves Redis server information
   *
   * Useful for monitoring and debugging.
   *
   * @returns Raw Redis INFO output
   */
  async info() {
    this.checkConnection();
    return await this.redis.info();
  }
};
export {
  RedisBackend
};
