// @byterover/cipher/dist/src/core/redis-IV6PRUHK.js
import {
  VectorDimensionError,
  VectorStoreConnectionError,
  VectorStoreError
} from "./chunk-JOPES53R.js";
import {
  DEFAULTS,
  ERROR_MESSAGES,
  LOG_PREFIXES
} from "./chunk-4JSJAC4K.js";
import "./chunk-7GLOCKG4.js";
import {
  createLogger
} from "./chunk-GPOSIAVD.js";
import "./chunk-7UQA6M3D.js";
import "./chunk-NUVPDHVT.js";
import {
  init_esm_shims
} from "./chunk-KFOQGTVJ.js";

// src/core/vector_storage/backend/redis.ts
init_esm_shims();
import { Redis } from "ioredis";
var RedisBackend = class {
  client;
  config;
  dimension;
  logger;
  collectionName;
  indexName;
  distance;
  connected = false;
  indexCreated = false;
  constructor(config) {
    this.config = config;
    this.dimension = config.dimension || DEFAULTS.DIMENSION;
    this.collectionName = config.collectionName || "vectors";
    this.indexName = `${this.collectionName}_idx`;
    this.distance = config.distance || "COSINE";
    this.logger = createLogger({
      level: process.env.CIPHER_LOG_LEVEL || "info"
    });
    if (this.config.url) {
      this.client = new Redis(this.config.url);
    } else {
      this.client = new Redis({
        host: this.config.host || "localhost",
        port: this.config.port ?? 6379,
        password: this.config.password || "",
        username: this.config.username || "",
        enableOfflineQueue: false,
        maxRetriesPerRequest: 3
      });
    }
    this.client.on("error", (error) => {
      this.logger.error(`${LOG_PREFIXES.REDIS} Connection error:`, error);
      this.connected = false;
    });
    this.client.on("connect", () => {
      this.logger.info(`${LOG_PREFIXES.REDIS} Connected successfully`);
      this.connected = true;
    });
    this.client.on("close", () => {
      this.logger.info(`${LOG_PREFIXES.REDIS} Connection closed`);
      this.connected = false;
    });
    this.client.on("ready", () => {
      this.logger.info(`${LOG_PREFIXES.REDIS} Client ready`);
      this.connected = true;
    });
    this.logger.debug(`${LOG_PREFIXES.REDIS} Initialized`, {
      collectionName: this.collectionName,
      indexName: this.indexName,
      dimension: this.dimension,
      distance: this.distance
    });
  }
  /**
   * Generate Redis key for vector document
   */
  getDocumentKey(id) {
    return `${this.collectionName}:${id}`;
  }
  /**
   * Validate vector dimension
   */
  validateDimension(vector, operation) {
    if (vector.length !== this.dimension) {
      throw new VectorDimensionError(
        `${ERROR_MESSAGES.INVALID_DIMENSION}: expected ${this.dimension}, got ${vector.length}`,
        this.dimension,
        vector.length
      );
    }
  }
  /**
   * Validate ID (must be a valid integer)
   */
  validateId(id) {
    if (id === void 0 || id === null || !Number.isInteger(id)) {
      throw new VectorStoreError("Redis point IDs must be valid integers", "id");
    }
  }
  /**
   * Convert vector to buffer for Redis storage
   */
  vectorToBuffer(vector) {
    const buffer = Buffer.allocUnsafe(vector.length * 4);
    for (let i = 0; i < vector.length; i++) {
      buffer.writeFloatLE(vector[i] ?? 0, i * 4);
    }
    return buffer;
  }
  /**
   * Convert buffer back to vector
   */
  bufferToVector(buffer) {
    const vector = [];
    for (let i = 0; i < buffer.length; i += 4) {
      vector.push(buffer.readFloatLE(i));
    }
    return vector;
  }
  /**
   * Create vector search index in Redis using raw commands
   */
  async createIndex() {
    if (this.indexCreated) return;
    try {
      try {
        await this.client.call("FT.INFO", this.indexName);
        this.indexCreated = true;
        this.logger.debug(`${LOG_PREFIXES.REDIS} Index ${this.indexName} already exists`);
        return;
      } catch (error) {
      }
      const createArgs = [
        this.indexName,
        "ON",
        "JSON",
        "PREFIX",
        "1",
        `${this.collectionName}:`,
        "SCHEMA",
        "$.vector",
        "AS",
        "vector",
        "VECTOR",
        "FLAT",
        "6",
        "TYPE",
        "FLOAT32",
        "DIM",
        this.dimension.toString(),
        "DISTANCE_METRIC",
        this.distance
      ];
      await this.client.call("FT.CREATE", ...createArgs);
      this.indexCreated = true;
      this.logger.info(`${LOG_PREFIXES.REDIS} Created vector index: ${this.indexName}`);
    } catch (error) {
      this.logger.error(`${LOG_PREFIXES.REDIS} Failed to create index`, {
        indexName: this.indexName,
        error: error instanceof Error ? error.message : String(error)
      });
      throw new VectorStoreError("Failed to create vector index", "createIndex", error);
    }
  }
  /**
   * Build filter query string for FT.SEARCH
   */
  buildFilterQuery(filters) {
    if (!filters || Object.keys(filters).length === 0) return "*";
    const conditions = [];
    for (const [key, value] of Object.entries(filters)) {
      if (value === null || value === void 0) continue;
      if (typeof value === "object" && !Array.isArray(value) && ("gte" in value || "gt" in value || "lte" in value || "lt" in value)) {
        let rangeQuery = `@${key}:[`;
        if ("gte" in value && value.gte !== void 0) {
          rangeQuery += value.gte;
        } else if ("gt" in value && value.gt !== void 0) {
          rangeQuery += `(${value.gt}`;
        } else {
          rangeQuery += "-inf";
        }
        rangeQuery += " ";
        if ("lte" in value && value.lte !== void 0) {
          rangeQuery += value.lte;
        } else if ("lt" in value && value.lt !== void 0) {
          rangeQuery += `${value.lt})`;
        } else {
          rangeQuery += "+inf";
        }
        rangeQuery += "]";
        conditions.push(rangeQuery);
      } else if (typeof value === "object" && !Array.isArray(value) && "any" in value && Array.isArray(value.any)) {
        const values = value.any.map((v) => `"${v}"`).join("|");
        conditions.push(`@${key}:(${values})`);
      } else {
        conditions.push(`@${key}:"${value}"`);
      }
    }
    return conditions.length > 0 ? conditions.join(" ") : "*";
  }
  async connect() {
    if (this.connected && this.client.status === "ready") {
      this.logger.debug(`${LOG_PREFIXES.REDIS} Already connected`);
      return;
    }
    this.logger.info(`${LOG_PREFIXES.REDIS} Connecting to Redis Stack`);
    try {
      if (this.client.status !== "ready") {
        await new Promise((resolve, reject) => {
          let settled = false;
          const timeout = setTimeout(() => {
            if (!settled) {
              settled = true;
              reject(new Error("Connection timeout"));
            }
          }, 1e4);
          const onReady = () => {
            if (!settled) {
              settled = true;
              clearTimeout(timeout);
              resolve(void 0);
            }
          };
          const onError = (error) => {
            if (!settled) {
              settled = true;
              clearTimeout(timeout);
              reject(error);
            }
          };
          this.client.once("ready", onReady);
          this.client.once("error", onError);
        });
      }
      this.connected = true;
      await this.createIndex();
      this.logger.info(`${LOG_PREFIXES.REDIS} Successfully connected to Redis Stack`, {
        collectionName: this.collectionName,
        indexName: this.indexName,
        dimension: this.dimension
      });
    } catch (error) {
      this.logger.error(`${LOG_PREFIXES.REDIS} Connection failed`, {
        error: error instanceof Error ? error.message : String(error)
      });
      throw new VectorStoreConnectionError(
        ERROR_MESSAGES.CONNECTION_FAILED,
        "redis",
        error
      );
    }
  }
  async disconnect() {
    if (!this.connected) {
      this.logger.debug(`${LOG_PREFIXES.REDIS} Already disconnected`);
      return;
    }
    this.logger.info(`${LOG_PREFIXES.REDIS} Disconnecting from Redis`);
    try {
      await this.client.quit();
      this.connected = false;
      this.logger.info(`${LOG_PREFIXES.REDIS} Successfully disconnected`);
    } catch (error) {
      this.logger.error(`${LOG_PREFIXES.REDIS} Error during disconnect:`, error);
    }
  }
  async insert(vectors, ids, payloads) {
    if (!this.connected) {
      throw new VectorStoreError(ERROR_MESSAGES.NOT_CONNECTED, "insert");
    }
    if (vectors.length !== ids.length || vectors.length !== payloads.length) {
      throw new VectorStoreError("Vectors, IDs, and payloads must have the same length", "insert");
    }
    if (vectors.length === 0) {
      this.logger.debug(`${LOG_PREFIXES.REDIS} No vectors to insert`);
      return;
    }
    for (let i = 0; i < vectors.length; i++) {
      this.validateDimension(vectors[i], "insert");
      this.validateId(ids[i]);
    }
    this.logger.debug(`${LOG_PREFIXES.REDIS} Inserting ${vectors.length} vectors`);
    try {
      const pipeline = this.client.pipeline();
      for (let i = 0; i < vectors.length; i++) {
        const id = ids[i];
        const vector = vectors[i];
        const payload = payloads[i] || {};
        const documentKey = this.getDocumentKey(id ?? 0);
        const vectorBuffer = this.vectorToBuffer(vector ?? []);
        const document = {
          id,
          vector: Array.from(vectorBuffer),
          ...payload
        };
        pipeline.call("JSON.SET", documentKey, "$", JSON.stringify(document));
      }
      await pipeline.exec();
      this.logger.debug(`${LOG_PREFIXES.REDIS} Successfully inserted ${vectors.length} vectors`);
    } catch (error) {
      this.logger.error(`${LOG_PREFIXES.REDIS} Insert failed`, {
        error: error instanceof Error ? error.message : String(error),
        vectorCount: vectors.length
      });
      throw new VectorStoreError("Failed to insert vectors", "insert", error);
    }
  }
  async search(query, limit = DEFAULTS.SEARCH_LIMIT, filters) {
    if (!this.connected) {
      throw new VectorStoreError(ERROR_MESSAGES.NOT_CONNECTED, "search");
    }
    this.validateDimension(query, "search");
    if (limit <= 0) {
      throw new VectorStoreError("Search limit must be positive", "search");
    }
    this.logger.debug(`${LOG_PREFIXES.REDIS} Searching with limit ${limit}`, {
      hasFilters: !!filters
    });
    try {
      const queryBuffer = this.vectorToBuffer(query);
      const filterQuery = this.buildFilterQuery(filters);
      const searchQuery = `${filterQuery} => [KNN ${limit} @vector $query_vec AS score]`;
      const searchArgs = [
        this.indexName,
        searchQuery,
        "PARAMS",
        "2",
        "query_vec",
        queryBuffer,
        "RETURN",
        "2",
        "$",
        "score",
        "SORTBY",
        "score",
        "DIALECT",
        "2"
      ];
      const searchResult = await this.client.call("FT.SEARCH", ...searchArgs);
      const results = [];
      const totalResults = searchResult[0];
      for (let i = 1; i < searchResult.length; i += 2) {
        const docKey = searchResult[i];
        const docFields = searchResult[i + 1];
        let docData = {};
        let score = 0;
        for (let j = 0; j < docFields.length; j += 2) {
          const fieldName = docFields[j];
          const fieldValue = docFields[j + 1];
          if (fieldName === "$") {
            docData = JSON.parse(fieldValue);
          } else if (fieldName === "score") {
            score = parseFloat(fieldValue);
          }
        }
        const id = parseInt(docData.id);
        const similarityScore = 1 - score;
        const vectorArray = docData.vector;
        const vectorBuffer = Buffer.from(vectorArray);
        const vector = this.bufferToVector(vectorBuffer);
        const payload = { ...docData };
        delete payload.id;
        delete payload.vector;
        results.push({
          id,
          score: similarityScore,
          payload,
          vector
        });
      }
      this.logger.debug(`${LOG_PREFIXES.REDIS} Found ${results.length} results`);
      return results;
    } catch (error) {
      this.logger.error(`${LOG_PREFIXES.REDIS} Search failed`, {
        error: error instanceof Error ? error.message : String(error),
        limit
      });
      throw new VectorStoreError(ERROR_MESSAGES.SEARCH_FAILED, "search", error);
    }
  }
  async get(vectorId) {
    if (!this.connected) {
      throw new VectorStoreError(ERROR_MESSAGES.NOT_CONNECTED, "get");
    }
    this.validateId(vectorId);
    this.logger.debug(`${LOG_PREFIXES.REDIS} Getting vector ${vectorId}`);
    try {
      const documentKey = this.getDocumentKey(vectorId);
      const documentJson = await this.client.call("JSON.GET", documentKey, "$");
      if (!documentJson) {
        return null;
      }
      const docData = JSON.parse(documentJson)[0];
      const vectorArray = docData.vector;
      const vectorBuffer = Buffer.from(vectorArray);
      const vector = this.bufferToVector(vectorBuffer);
      const payload = { ...docData };
      delete payload.id;
      delete payload.vector;
      return {
        id: vectorId,
        vector,
        payload,
        score: 1
        // Perfect match for exact retrieval
      };
    } catch (error) {
      this.logger.error(`${LOG_PREFIXES.REDIS} Get failed`, {
        error: error instanceof Error ? error.message : String(error),
        vectorId
      });
      throw new VectorStoreError("Failed to retrieve vector", "get", error);
    }
  }
  async update(vectorId, vector, payload) {
    if (!this.connected) {
      throw new VectorStoreError(ERROR_MESSAGES.NOT_CONNECTED, "update");
    }
    this.validateDimension(vector, "update");
    this.validateId(vectorId);
    this.logger.debug(`${LOG_PREFIXES.REDIS} Updating vector ${vectorId}`);
    try {
      const documentKey = this.getDocumentKey(vectorId);
      const existing = await this.client.call("JSON.GET", documentKey, "$");
      if (!existing) {
        throw new VectorStoreError(`Vector ${vectorId} does not exist`, "update");
      }
      const vectorBuffer = this.vectorToBuffer(vector);
      const document = {
        id: vectorId,
        vector: Array.from(vectorBuffer),
        ...payload
      };
      await this.client.call("JSON.SET", documentKey, "$", JSON.stringify(document));
      this.logger.debug(`${LOG_PREFIXES.REDIS} Successfully updated vector ${vectorId}`);
    } catch (error) {
      this.logger.error(`${LOG_PREFIXES.REDIS} Update failed`, {
        error: error instanceof Error ? error.message : String(error),
        vectorId
      });
      throw new VectorStoreError("Failed to update vector", "update", error);
    }
  }
  async delete(vectorId) {
    if (!this.connected) {
      throw new VectorStoreError(ERROR_MESSAGES.NOT_CONNECTED, "delete");
    }
    this.validateId(vectorId);
    this.logger.debug(`${LOG_PREFIXES.REDIS} Deleting vector ${vectorId}`);
    try {
      const documentKey = this.getDocumentKey(vectorId);
      await this.client.del(documentKey);
      this.logger.debug(`${LOG_PREFIXES.REDIS} Successfully deleted vector ${vectorId}`);
    } catch (error) {
      this.logger.error(`${LOG_PREFIXES.REDIS} Delete failed`, {
        error: error instanceof Error ? error.message : String(error),
        vectorId
      });
      throw new VectorStoreError("Failed to delete vector", "delete", error);
    }
  }
  async deleteCollection() {
    if (!this.connected) {
      throw new VectorStoreError(ERROR_MESSAGES.NOT_CONNECTED, "deleteCollection");
    }
    this.logger.info(`${LOG_PREFIXES.REDIS} Deleting collection: ${this.collectionName}`);
    try {
      try {
        await this.client.call("FT.DROPINDEX", this.indexName);
        this.indexCreated = false;
        this.logger.info(`${LOG_PREFIXES.REDIS} Dropped index: ${this.indexName}`);
      } catch (error) {
        this.logger.debug(
          `${LOG_PREFIXES.REDIS} Index ${this.indexName} does not exist or already dropped`
        );
      }
      const pattern = `${this.collectionName}:*`;
      let cursor = "0";
      let deletedCount = 0;
      do {
        const result = await this.client.scan(cursor, "MATCH", pattern, "COUNT", 1e3);
        cursor = result[0];
        const keys = result[1];
        if (keys.length > 0) {
          await this.client.del(...keys);
          deletedCount += keys.length;
        }
      } while (cursor !== "0");
      this.logger.info(
        `${LOG_PREFIXES.REDIS} Deleted ${deletedCount} documents from collection: ${this.collectionName}`
      );
    } catch (error) {
      this.logger.error(`${LOG_PREFIXES.REDIS} Delete collection failed`, {
        error: error instanceof Error ? error.message : String(error)
      });
      throw new VectorStoreError("Failed to delete collection", "deleteCollection", error);
    }
  }
  async list(filters, limit = 1e4) {
    if (!this.connected) {
      throw new VectorStoreError(ERROR_MESSAGES.NOT_CONNECTED, "list");
    }
    this.logger.debug(`${LOG_PREFIXES.REDIS} Listing vectors with limit ${limit}`, {
      hasFilters: !!filters
    });
    try {
      const filterQuery = this.buildFilterQuery(filters);
      const searchArgs = [
        this.indexName,
        filterQuery,
        "LIMIT",
        "0",
        limit.toString(),
        "RETURN",
        "1",
        "$"
      ];
      const searchResult = await this.client.call("FT.SEARCH", ...searchArgs);
      const results = [];
      const totalCount = searchResult[0];
      for (let i = 1; i < searchResult.length; i += 2) {
        const docKey = searchResult[i];
        const docFields = searchResult[i + 1];
        let docData = {};
        for (let j = 0; j < docFields.length; j += 2) {
          const fieldName = docFields[j];
          const fieldValue = docFields[j + 1];
          if (fieldName === "$") {
            docData = JSON.parse(fieldValue);
            break;
          }
        }
        const id = parseInt(docData.id);
        const vectorArray = docData.vector;
        const vectorBuffer = Buffer.from(vectorArray);
        const vector = this.bufferToVector(vectorBuffer);
        const payload = { ...docData };
        delete payload.id;
        delete payload.vector;
        results.push({
          id,
          vector,
          payload,
          score: 1
        });
      }
      this.logger.debug(
        `${LOG_PREFIXES.REDIS} Listed ${results.length} vectors (total: ${totalCount})`
      );
      return [results, totalCount];
    } catch (error) {
      this.logger.error(`${LOG_PREFIXES.REDIS} List failed`, {
        error: error instanceof Error ? error.message : String(error),
        limit
      });
      throw new VectorStoreError("Failed to list vectors", "list", error);
    }
  }
  // Getter methods
  isConnected() {
    return this.connected && this.client.status === "ready";
  }
  getBackendType() {
    return "redis";
  }
  getDimension() {
    return this.dimension;
  }
  getCollectionName() {
    return this.collectionName;
  }
};
export {
  RedisBackend
};
