// @byterover/cipher/dist/src/core/sqlite-HIBPDF6T.js
import {
  BACKEND_TYPES,
  ERROR_MESSAGES
} from "./chunk-WZZEPDWF.js";
import {
  StorageConnectionError,
  StorageError
} from "./chunk-C3KGEYH3.js";
import "./chunk-7GLOCKG4.js";
import {
  createLogger
} from "./chunk-GPOSIAVD.js";
import "./chunk-7UQA6M3D.js";
import "./chunk-NUVPDHVT.js";
import {
  init_esm_shims
} from "./chunk-KFOQGTVJ.js";

// src/core/storage/backend/sqlite.ts
init_esm_shims();
import Database from "better-sqlite3";
import { mkdir } from "fs/promises";
import { dirname, join } from "path";
import { existsSync } from "fs";
var SqliteBackend = class {
  constructor(config) {
    this.config = config;
    this.logger = createLogger({ level: process.env.LOG_LEVEL || "info" });
    const dbPath = config.path || "./data";
    const dbName = config.database || "cipher.db";
    this.dbPath = join(dbPath, dbName);
    this.logger.debug("SQLite backend initialized", {
      path: this.dbPath,
      config: { ...config, path: dbPath, database: dbName }
    });
  }
  logger;
  connected = false;
  db;
  dbPath;
  // Prepared statements for performance
  statements = {};
  async connect() {
    if (this.connected) {
      return;
    }
    try {
      const dir = dirname(this.dbPath);
      this.logger.debug("Ensuring database directory exists", { dir, dbPath: this.dbPath });
      if (!existsSync(dir)) {
        await mkdir(dir, { recursive: true });
        this.logger.debug("Created database directory", { dir });
      }
      this.logger.debug("Opening SQLite database", { path: this.dbPath });
      this.db = new Database(this.dbPath);
      if (!this.db) {
        throw new Error("Failed to create SQLite database connection - database instance is null");
      }
      this.logger.debug("Testing database connection");
      this.db.pragma("journal_mode = WAL");
      this.db.pragma("synchronous = NORMAL");
      this.db.pragma("foreign_keys = ON");
      this.logger.debug("Creating database tables");
      this.createTables();
      this.logger.debug("Preparing SQL statements");
      this.prepareStatements();
      this.connected = true;
      this.logger.info("SQLite backend connected successfully", {
        path: this.dbPath,
        dbSize: this.getDbInfo()
      });
    } catch (error) {
      const errorDetails = {
        path: this.dbPath,
        directory: dirname(this.dbPath),
        directoryExists: existsSync(dirname(this.dbPath)),
        error: error instanceof Error ? error.message : String(error),
        stack: error instanceof Error ? error.stack : void 0,
        nodeVersion: process.version,
        platform: process.platform,
        arch: process.arch
      };
      console.log("Failed to connect to SQLite database", errorDetails);
      throw new StorageConnectionError(
        `Failed to connect to SQLite database: ${error instanceof Error ? error.message : String(error)}`,
        BACKEND_TYPES.SQLITE,
        error
      );
    }
  }
  async disconnect() {
    if (!this.connected || !this.db) {
      return;
    }
    try {
      Object.values(this.statements).forEach((_stmt) => {
        try {
        } catch (error) {
          this.logger.warn("Error finalizing statement", { error });
        }
      });
      this.statements = {};
      this.db.close();
      this.db = void 0;
      this.connected = false;
      this.logger.info("SQLite backend disconnected");
    } catch (error) {
      this.logger.error("Error disconnecting from SQLite database", {
        error: error instanceof Error ? error.message : String(error)
      });
      throw error;
    }
  }
  isConnected() {
    return this.connected;
  }
  getBackendType() {
    return BACKEND_TYPES.SQLITE;
  }
  // DatabaseBackend implementation
  async get(key) {
    this.checkConnection();
    try {
      const row = this.statements.get.get(key);
      if (!row) {
        return void 0;
      }
      return JSON.parse(row.value);
    } catch (error) {
      this.logger.error("Error getting value from SQLite", {
        key,
        error: error instanceof Error ? error.message : String(error)
      });
      throw new StorageError("Failed to get value from SQLite", "get", error);
    }
  }
  async set(key, value) {
    this.checkConnection();
    try {
      const serialized = JSON.stringify(value);
      const now = Date.now();
      this.statements.set.run(key, serialized, now, now);
    } catch (error) {
      this.logger.error("Error setting value in SQLite", {
        key,
        error: error instanceof Error ? error.message : String(error)
      });
      throw new StorageError("Failed to set value in SQLite", "set", error);
    }
  }
  async delete(key) {
    this.checkConnection();
    try {
      const transaction = this.db.transaction(() => {
        this.statements.delete.run(key);
        this.db.prepare("DELETE FROM lists WHERE key = ?").run(key);
      });
      transaction();
    } catch (error) {
      this.logger.error("Error deleting value from SQLite", {
        key,
        error: error instanceof Error ? error.message : String(error)
      });
      throw new StorageError("Failed to delete value from SQLite", "delete", error);
    }
  }
  async list(prefix) {
    this.checkConnection();
    try {
      const rows = this.statements.list.all(prefix + "%");
      return rows.map((row) => row.key);
    } catch (error) {
      this.logger.error("Error listing keys from SQLite", {
        prefix,
        error: error instanceof Error ? error.message : String(error)
      });
      throw new StorageError("Failed to list keys from SQLite", "list", error);
    }
  }
  async append(key, item) {
    this.checkConnection();
    try {
      const serialized = JSON.stringify(item);
      const now = Date.now();
      const maxPosResult = this.db.prepare(
        "SELECT COALESCE(MAX(position), -1) + 1 as next_pos FROM lists WHERE key = ?"
      ).get(key);
      const nextPosition = maxPosResult.next_pos;
      this.db.prepare(
        "INSERT INTO lists (key, value, position, created_at) VALUES (?, ?, ?, ?)"
      ).run(key, serialized, nextPosition, now);
    } catch (error) {
      this.logger.error("Error appending to list in SQLite", {
        key,
        error: error instanceof Error ? error.message : String(error)
      });
      throw new StorageError("Failed to append to list in SQLite", "append", error);
    }
  }
  async getRange(key, start, count) {
    this.checkConnection();
    try {
      const rows = this.statements.getRange.all(key, count, start);
      return rows.map((row) => JSON.parse(row.value));
    } catch (error) {
      this.logger.error("Error getting range from SQLite", {
        key,
        start,
        count,
        error: error instanceof Error ? error.message : String(error)
      });
      throw new StorageError("Failed to get range from SQLite", "getRange", error);
    }
  }
  // Private helper methods
  checkConnection() {
    if (!this.connected || !this.db) {
      throw new StorageError(ERROR_MESSAGES.NOT_CONNECTED, "operation");
    }
  }
  createTables() {
    if (!this.db) {
      throw new StorageError("Database not initialized", "createTables");
    }
    this.db.exec(`
			CREATE TABLE IF NOT EXISTS store (
				key TEXT PRIMARY KEY,
				value TEXT NOT NULL,
				created_at INTEGER NOT NULL,
				updated_at INTEGER NOT NULL
			)
		`);
    this.db.exec(`
			CREATE TABLE IF NOT EXISTS lists (
				key TEXT NOT NULL,
				value TEXT NOT NULL,
				position INTEGER NOT NULL,
				created_at INTEGER NOT NULL,
				PRIMARY KEY (key, position)
			)
		`);
    this.db.exec(`
			CREATE TABLE IF NOT EXISTS list_metadata (
				key TEXT PRIMARY KEY,
				count INTEGER NOT NULL DEFAULT 0,
				created_at INTEGER NOT NULL,
				updated_at INTEGER NOT NULL
			)
		`);
    this.db.exec("CREATE INDEX IF NOT EXISTS idx_store_updated_at ON store(updated_at)");
    this.db.exec("CREATE INDEX IF NOT EXISTS idx_lists_key ON lists(key)");
    this.db.exec("CREATE INDEX IF NOT EXISTS idx_lists_created_at ON lists(created_at)");
    this.logger.debug("SQLite tables and indexes created");
  }
  prepareStatements() {
    if (!this.db) {
      throw new StorageError("Database not initialized", "prepareStatements");
    }
    this.statements = {
      get: this.db.prepare("SELECT value FROM store WHERE key = ?"),
      set: this.db.prepare(`
				INSERT OR REPLACE INTO store (key, value, created_at, updated_at)
				VALUES (?, ?, ?, ?)
			`),
      delete: this.db.prepare("DELETE FROM store WHERE key = ?"),
      list: this.db.prepare("SELECT key FROM store WHERE key LIKE ? ORDER BY key"),
      // Range query for lists
      getRange: this.db.prepare(`
				SELECT value FROM lists 
				WHERE key = ? 
				ORDER BY position 
				LIMIT ? OFFSET ?
			`),
      listCount: this.db.prepare("SELECT COUNT(*) as count FROM lists WHERE key = ?")
    };
    this.logger.debug("SQLite prepared statements created");
  }
  /**
   * Get database file size information
   */
  getDbInfo() {
    const info = { path: this.dbPath };
    if (this.connected && this.db) {
      try {
        const pragmaResult = this.db.pragma("page_count");
        const pageCount = Array.isArray(pragmaResult) ? pragmaResult[0] : pragmaResult;
        const numPageCount = typeof pageCount === "number" ? pageCount : Number(pageCount);
        if (!isNaN(numPageCount)) {
          info.pageCount = numPageCount;
        }
        const pageSizeResult = this.db.pragma("page_size");
        const pageSize = Array.isArray(pageSizeResult) ? pageSizeResult[0] : pageSizeResult;
        const numPageSize = typeof pageSize === "number" ? pageSize : Number(pageSize);
        if (!isNaN(numPageSize)) {
          info.pageSize = numPageSize;
        }
        if (info.pageCount && info.pageSize) {
          info.size = info.pageCount * info.pageSize;
        }
      } catch (error) {
        this.logger.warn("Error getting database info", { error });
      }
    }
    return info;
  }
  /**
   * Run database maintenance operations
   */
  async maintenance() {
    this.checkConnection();
    try {
      this.db.exec("ANALYZE");
      const freeListResult = this.db.pragma("freelist_count");
      const freePages = Array.isArray(freeListResult) ? freeListResult[0] : freeListResult;
      const freePageCount = typeof freePages === "number" ? freePages : Number(freePages);
      if (freePageCount > 100) {
        this.db.exec("VACUUM");
        this.logger.info("Database vacuumed", { freePages: freePageCount });
      }
      this.logger.debug("Database maintenance completed");
    } catch (error) {
      this.logger.error("Error during database maintenance", {
        error: error instanceof Error ? error.message : String(error)
      });
      throw new StorageError("Database maintenance failed", "maintenance", error);
    }
  }
};
export {
  SqliteBackend
};
