// @byterover/cipher/dist/src/core/weaviate-XCM6BZOQ.js
import {
  VectorDimensionError,
  VectorStoreConnectionError,
  VectorStoreError
} from "./chunk-JOPES53R.js";
import {
  ERROR_MESSAGES,
  LOG_PREFIXES
} from "./chunk-4JSJAC4K.js";
import "./chunk-7GLOCKG4.js";
import {
  createLogger
} from "./chunk-GPOSIAVD.js";
import {
  env,
  init_env
} from "./chunk-7UQA6M3D.js";
import "./chunk-NUVPDHVT.js";
import {
  __require,
  init_esm_shims
} from "./chunk-KFOQGTVJ.js";

// src/core/vector_storage/backend/weaviate.ts
init_esm_shims();
import { ApiKey } from "weaviate-ts-client";
init_env();
import { v5 as uuidv5 } from "uuid";
var weaviate = __require("weaviate-ts-client").default;
var WEAVIATE_UUID_NAMESPACE = "6ba7b810-9dad-11d1-80b4-00c04fd430c8";
var WeaviateBackend = class {
  client = void 0;
  config;
  dimension;
  collectionName;
  logger;
  connected = false;
  connectionConfig = {};
  constructor(config) {
    this.config = config;
    this.collectionName = config.collectionName.toLocaleLowerCase();
    this.collectionName = this.collectionName.charAt(0).toUpperCase() + this.collectionName.slice(1);
    this.dimension = config.dimension;
    this.logger = createLogger({
      level: env.CIPHER_LOG_LEVEL || "info"
    });
    this.logger.debug(`${LOG_PREFIXES.WEAVIATE} Backend initialized`, {
      collection: this.collectionName,
      dimension: this.dimension,
      url: config.url,
      host: config.host,
      port: config.port
    });
  }
  /**
   * Ensure the client is available and connected
   */
  ensureClient() {
    if (!this.client) {
      throw new VectorStoreError(ERROR_MESSAGES.NOT_CONNECTED, "operation");
    }
    return this.client;
  }
  /**
   * Convert numeric ID to UUID v5 for Weaviate compatibility
   */
  generateUuidFromId(id) {
    return uuidv5(id.toString(), WEAVIATE_UUID_NAMESPACE);
  }
  validateDimension(vector, operation) {
    if (vector.length !== this.dimension) {
      throw new VectorDimensionError(
        `${ERROR_MESSAGES.INVALID_DIMENSION}: expected ${this.dimension}, got ${vector.length}`,
        this.dimension,
        vector.length
      );
    }
  }
  async connect() {
    if (this.connected) {
      this.logger.debug(`${LOG_PREFIXES.WEAVIATE} Already connected`);
      return;
    }
    this.logger.debug(`${LOG_PREFIXES.WEAVIATE} Connecting to Weaviate`);
    try {
      if (this.config.url) {
        console.log("Connecting to Weaviate Cloud with URL: ", this.config.url);
        let host = this.config.url;
        if (host.startsWith("https://")) {
          host = host.replace("https://", "");
        } else if (host.startsWith("http://")) {
          host = host.replace("http://", "");
        }
        this.connectionConfig = {
          scheme: "https",
          host,
          apiKey: new ApiKey(this.config.apiKey || "")
        };
      } else if (this.config.host) {
        console.log("Connecting to local/custom Weaviate instance: ", this.config.host);
        let scheme = "http";
        let host = this.config.host;
        if (this.config.host.includes("https://")) {
          scheme = "https";
          host = this.config.host.replace("https://", "");
        } else if (this.config.host.includes("http://")) {
          scheme = "http";
          host = this.config.host.replace("http://", "");
        }
        this.connectionConfig = {
          scheme,
          host
        };
        if (this.config.apiKey) {
          this.connectionConfig.apiKey = new ApiKey(this.config.apiKey);
        }
      } else {
        throw new VectorStoreConnectionError(
          "Either url (for Weaviate Cloud) or host (for local instance) must be provided",
          "weaviate",
          new Error("Missing connection configuration")
        );
      }
      console.log("Connection config: ", this.connectionConfig);
      this.client = weaviate.client(this.connectionConfig);
      if (!this.client) {
        throw new VectorStoreConnectionError(
          "Weaviate client not initialized",
          "weaviate",
          new Error("Connection check failed")
        );
      }
      const isReady = await this.client.misc.liveChecker().do();
      if (!isReady) {
        throw new VectorStoreConnectionError(
          "Weaviate instance is not ready",
          "weaviate",
          new Error("Connection check failed")
        );
      }
      await this.ensureCollection();
      this.connected = true;
      this.logger.debug(`${LOG_PREFIXES.WEAVIATE} Successfully connected`);
    } catch (error) {
      this.logger.error(`${LOG_PREFIXES.WEAVIATE} Connection failed`, { error });
      if (error instanceof VectorStoreConnectionError) {
        throw error;
      }
      throw new VectorStoreConnectionError(
        ERROR_MESSAGES.CONNECTION_FAILED,
        "weaviate",
        error
      );
    }
  }
  /**
   * Ensure the collection exists, create if it doesn't
   */
  async ensureCollection() {
    const client = this.ensureClient();
    try {
      const schema = await client.schema.getter().do();
      if (!schema.classes) {
        this.logger.error(`${LOG_PREFIXES.WEAVIATE} Schema has no classes`);
        throw new VectorStoreError("Schema has no classes", "ensureCollection");
      }
      let classExists = false;
      for (const cls of schema.classes) {
        if (!cls.class) {
          this.logger.error(`${LOG_PREFIXES.WEAVIATE} Class has no class name`);
          throw new VectorStoreError("Class has no class name", "ensureCollection");
        }
        if (cls.class === this.collectionName) {
          classExists = true;
          break;
        }
      }
      if (!classExists) {
        this.logger.debug(`${LOG_PREFIXES.WEAVIATE} Creating class ${this.collectionName}`);
        const classObj = {
          class: this.collectionName,
          //benchmark
          description: `Vector store collection for ${this.collectionName}`,
          properties: [
            {
              name: "payload",
              dataType: ["text"],
              description: "JSON string containing all metadata and content"
            }
          ],
          vectorizer: "none"
          // We provide our own vectors
        };
        await client.schema.classCreator().withClass(classObj).do();
        this.logger.debug(`${LOG_PREFIXES.WEAVIATE} Class created: ${this.collectionName}`);
      } else {
        const existingClass = schema.classes?.find((cls) => cls.class === this.collectionName);
        const expectedPayloadProperty = existingClass?.properties?.find(
          (prop) => prop.name === "payload"
        );
        if (!Array.isArray(expectedPayloadProperty?.dataType) || !existingClass || !existingClass?.properties || !(existingClass.properties.length === 1) || !expectedPayloadProperty.dataType.every((v) => typeof v === "string")) {
          this.logger.error(`${LOG_PREFIXES.WEAVIATE} Payload property is not a string[]`);
          throw new VectorStoreError("Payload property is not a string[]", "ensureCollection");
        }
        const schemaMatches = expectedPayloadProperty && expectedPayloadProperty.dataType.includes("text") && existingClass.properties && existingClass.properties.length === 1;
        if (!schemaMatches) {
          this.logger.warn(
            `${LOG_PREFIXES.WEAVIATE} Class ${this.collectionName} schema doesn't match expected structure. Recreating...`
          );
          await client.schema.classDeleter().withClassName(this.collectionName).do();
          this.logger.debug(
            `${LOG_PREFIXES.WEAVIATE} Deleted existing class: ${this.collectionName}`
          );
          const classObj = {
            class: this.collectionName,
            description: `Vector store collection for ${this.collectionName}`,
            properties: [
              {
                name: "payload",
                dataType: ["text"],
                description: "JSON string containing all metadata and content"
              }
            ],
            vectorizer: "none"
            // We provide our own vectors
          };
          await client.schema.classCreator().withClass(classObj).do();
          this.logger.debug(
            `${LOG_PREFIXES.WEAVIATE} Recreated class with correct schema: ${this.collectionName}`
          );
        } else {
          this.logger.debug(
            `${LOG_PREFIXES.WEAVIATE} Class already exists with correct schema: ${this.collectionName}`
          );
        }
      }
    } catch (error) {
      this.logger.error(`${LOG_PREFIXES.WEAVIATE} Failed to ensure class`, { error });
      throw error;
    }
  }
  async insert(vectors, ids, payloads) {
    if (!this.connected) {
      return Promise.reject(new VectorStoreError(ERROR_MESSAGES.NOT_CONNECTED, "insert"));
    }
    if (vectors.length !== ids.length || vectors.length !== payloads.length) {
      return Promise.reject(
        new VectorStoreError("Vectors, IDs, and payloads must have the same length", "insert")
      );
    }
    for (const vector of vectors) {
      this.validateDimension(vector, "insert");
    }
    try {
      const client = this.ensureClient();
      const batcher = client.batch.objectsBatcher();
      for (let i = 0; i < vectors.length; i++) {
        const id = ids[i];
        const vector = vectors[i];
        if (id === void 0) {
          this.logger.error(`${LOG_PREFIXES.WEAVIATE} ID is undefined`);
          throw new VectorStoreError("ID is undefined", "insert");
        }
        if (!vector) {
          this.logger.error(`${LOG_PREFIXES.WEAVIATE} Vector is undefined`);
          throw new VectorStoreError("Vector is undefined", "insert");
        }
        const obj = {
          class: this.collectionName,
          id: this.generateUuidFromId(id),
          properties: {
            payload: JSON.stringify(payloads[i] || {})
          },
          vector
        };
        batcher.withObject(obj);
      }
      const response = await batcher.do();
      if (response && response.some((r) => r.result?.errors)) {
        const errors = response.filter((r) => r.result?.errors).map((r) => r.result.errors.error.map((e) => e.message).join(", ")).join("; ");
        throw new VectorStoreError(`Insert failed: ${errors}`, "insert");
      }
      this.logger.debug(`${LOG_PREFIXES.WEAVIATE} Inserted ${vectors.length} vectors`);
    } catch (error) {
      this.logger.error(`${LOG_PREFIXES.WEAVIATE} Insert failed`, { error });
      throw new VectorStoreError("Failed to insert vectors", "insert", error);
    }
  }
  async search(query, limit = 10, filters) {
    if (!this.connected) {
      throw new VectorStoreError(ERROR_MESSAGES.NOT_CONNECTED, "search");
    }
    this.validateDimension(query, "search");
    try {
      const client = this.ensureClient();
      let searchBuilder = client.graphql.get().withClassName(this.collectionName).withNearVector({ vector: query }).withLimit(limit).withFields("_additional { id certainty distance } payload");
      if (filters) {
        const whereFilter = this.buildWhereFilter(filters);
        if (whereFilter) {
          searchBuilder = searchBuilder.withWhere(whereFilter);
        }
      }
      const response = await searchBuilder.do();
      const results = response.data.Get[this.collectionName] || [];
      return results.map((obj) => ({
        id: parseInt(obj._additional.id) || 0,
        score: obj._additional.certainty || obj._additional.distance || 0,
        payload: obj.payload ? JSON.parse(obj.payload) : {}
      }));
    } catch (error) {
      this.logger.error(`${LOG_PREFIXES.WEAVIATE} Search failed`, { error });
      throw new VectorStoreError(ERROR_MESSAGES.SEARCH_FAILED, "search", error);
    }
  }
  async get(vectorId) {
    if (!this.connected) {
      return Promise.reject(new VectorStoreError(ERROR_MESSAGES.NOT_CONNECTED, "get"));
    }
    try {
      const client = this.ensureClient();
      const response = await client.data.getterById().withClassName(this.collectionName).withId(this.generateUuidFromId(vectorId)).withVector().do();
      if (!response) {
        return null;
      }
      return {
        id: vectorId,
        vector: response.vector || [],
        payload: response.properties?.payload && typeof response.properties.payload === "string" ? JSON.parse(response.properties.payload) : {},
        score: 1
      };
    } catch (error) {
      this.logger.error(`${LOG_PREFIXES.WEAVIATE} Get failed`, { error });
      throw new VectorStoreError("Failed to retrieve vector", "get", error);
    }
  }
  async update(vectorId, vector, payload) {
    if (!this.connected) {
      return Promise.reject(new VectorStoreError(ERROR_MESSAGES.NOT_CONNECTED, "update"));
    }
    this.validateDimension(vector, "update");
    try {
      const client = this.ensureClient();
      await client.data.updater().withClassName(this.collectionName).withId(this.generateUuidFromId(vectorId)).withProperties({
        payload: JSON.stringify(payload)
      }).withVector(vector).do();
      this.logger.debug(`${LOG_PREFIXES.WEAVIATE} Updated vector ${vectorId}`);
    } catch (error) {
      this.logger.error(`${LOG_PREFIXES.WEAVIATE} Update failed`, { error });
      throw new VectorStoreError("Failed to update vector", "update", error);
    }
  }
  async delete(vectorId) {
    if (!this.connected) {
      return Promise.reject(new VectorStoreError(ERROR_MESSAGES.NOT_CONNECTED, "delete"));
    }
    try {
      const client = this.ensureClient();
      await client.data.deleter().withClassName(this.collectionName).withId(this.generateUuidFromId(vectorId)).do();
      this.logger.debug(`${LOG_PREFIXES.WEAVIATE} Deleted vector ${vectorId}`);
    } catch (error) {
      this.logger.error(`${LOG_PREFIXES.WEAVIATE} Delete failed`, { error });
      throw new VectorStoreError("Failed to delete vector", "delete", error);
    }
  }
  async deleteCollection() {
    if (!this.connected) {
      throw new VectorStoreError(ERROR_MESSAGES.NOT_CONNECTED, "deleteCollection");
    }
    try {
      const client = this.ensureClient();
      await client.schema.classDeleter().withClassName(this.collectionName).do();
      this.logger.info(`${LOG_PREFIXES.WEAVIATE} Deleted collection ${this.collectionName}`);
    } catch (error) {
      this.logger.error(`${LOG_PREFIXES.WEAVIATE} Delete collection failed`, { error });
      throw new VectorStoreError("Failed to delete collection", "deleteCollection", error);
    }
  }
  async list(filters, limit = 1e4) {
    if (!this.connected) {
      throw new VectorStoreError(ERROR_MESSAGES.NOT_CONNECTED, "list");
    }
    try {
      const client = this.ensureClient();
      let queryBuilder = client.graphql.get().withClassName(this.collectionName).withLimit(limit).withFields("_additional { id } payload");
      if (filters) {
        const whereFilter = this.buildWhereFilter(filters);
        if (whereFilter) {
          queryBuilder = queryBuilder.withWhere(whereFilter);
        }
      }
      const response = await queryBuilder.do();
      const results = response.data.Get[this.collectionName] || [];
      const transformedResults = results.map((obj) => ({
        id: parseInt(obj._additional.id) || 0,
        vector: [],
        // Vector not included in list operations
        payload: obj.payload ? JSON.parse(obj.payload) : {},
        score: 1
      }));
      return [transformedResults, transformedResults.length];
    } catch (error) {
      this.logger.error(`${LOG_PREFIXES.WEAVIATE} List failed`, { error });
      throw new VectorStoreError("Failed to list vectors", "list", error);
    }
  }
  async disconnect() {
    this.client = void 0;
    this.connected = false;
    this.logger.info(`${LOG_PREFIXES.WEAVIATE} Disconnected`);
  }
  isConnected() {
    return this.connected;
  }
  getBackendType() {
    return "weaviate";
  }
  getDimension() {
    return this.dimension;
  }
  getCollectionName() {
    return this.collectionName;
  }
  async listCollections() {
    if (!this.connected) {
      return Promise.reject(new VectorStoreError(ERROR_MESSAGES.NOT_CONNECTED, "listCollections"));
    }
    try {
      const client = this.ensureClient();
      const schema = await client.schema.getter().do();
      return schema.classes?.map((cls) => cls.class) || [];
    } catch (error) {
      this.logger.error(`${LOG_PREFIXES.WEAVIATE} List collections failed`, { error });
      throw new VectorStoreError("Failed to list collections", "listCollections", error);
    }
  }
  /**
   * Build Weaviate where filter from SearchFilters
   */
  buildWhereFilter(filters) {
    const conditions = [];
    for (const [key, value] of Object.entries(filters)) {
      if (typeof value === "string" || typeof value === "number" || typeof value === "boolean") {
        conditions.push({
          path: ["payload"],
          operator: "Like",
          valueText: `*"${key}":"${value}"*`
        });
      } else if (value && typeof value === "object") {
        if ("any" in value && Array.isArray(value.any)) {
          const orConditions = value.any.map((v) => ({
            path: ["payload"],
            operator: "Like",
            valueText: `*"${key}":"${v}"*`
          }));
          if (orConditions.length > 1) {
            conditions.push({ operator: "Or", operands: orConditions });
          } else if (orConditions.length === 1) {
            conditions.push(orConditions[0]);
          }
        } else {
          if ("gte" in value) {
            conditions.push({
              path: ["payload"],
              operator: "Like",
              valueText: `*"${key}":*`
            });
          }
          if ("lte" in value) {
            conditions.push({
              path: ["payload"],
              operator: "Like",
              valueText: `*"${key}":*`
            });
          }
          if ("gt" in value) {
            conditions.push({
              path: ["payload"],
              operator: "Like",
              valueText: `*"${key}":*`
            });
          }
          if ("lt" in value) {
            conditions.push({
              path: ["payload"],
              operator: "Like",
              valueText: `*"${key}":*`
            });
          }
        }
      }
    }
    if (conditions.length === 0) {
      return void 0;
    } else if (conditions.length === 1) {
      return conditions[0];
    } else {
      return { operator: "And", operands: conditions };
    }
  }
};
export {
  WeaviateBackend
};
